// @splunk/react-ui/Accordion.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = n => {
            /******/ var t = n && n.__esModule ? 
            /******/ () => n["default"]
            /******/ : () => n
            /******/;
            e.d(t, {
                a: t
            });
            /******/            return t;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (n, t) => {
            /******/ for (var r in t) {
                /******/ if (e.o(t, r) && !e.o(n, r)) {
                    /******/ Object.defineProperty(n, r, {
                        enumerable: true,
                        get: t[r]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, n) => Object.prototype.hasOwnProperty.call(e, n)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var n = {};
    // ESM COMPAT FLAG
        e.r(n);
    // EXPORTS
        e.d(n, {
        Panel: () => /* reexport */ x,
        default: () => /* reexport */ K
    });
    // CONCATENATED MODULE: external "react"
    const t = require("react");
    var r =  e.n(t);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var a =  e.n(o);
    // CONCATENATED MODULE: external "lodash/has"
    const i = require("lodash/has");
    var l =  e.n(i);
    // CONCATENATED MODULE: external "lodash/keys"
    const u = require("lodash/keys");
    var c =  e.n(u);
    // CONCATENATED MODULE: external "lodash/omit"
    const f = require("lodash/omit");
    var s =  e.n(f);
    // CONCATENATED MODULE: external "@splunk/react-ui/CollapsiblePanel"
    const p = require("@splunk/react-ui/CollapsiblePanel");
    var d =  e.n(p);
    // CONCATENATED MODULE: external "@splunk/themes"
    const y = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/Accordion/AccordionContext.tsx
    var v =  (0, t.createContext)({});
    v.displayName = "Accordion";
    /* harmony default export */    const b = v;
    // CONCATENATED MODULE: ./src/Accordion/Panel.tsx
    function h() {
        h = Object.assign || function(e) {
            for (var n = 1; n < arguments.length; n++) {
                var t = arguments[n];
                for (var r in t) {
                    if (Object.prototype.hasOwnProperty.call(t, r)) {
                        e[r] = t[r];
                    }
                }
            }
            return e;
        };
        return h.apply(this, arguments);
    }
    function m(e, n) {
        if (e == null) return {};
        var t = g(e, n);
        var r, o;
        if (Object.getOwnPropertySymbols) {
            var a = Object.getOwnPropertySymbols(e);
            for (o = 0; o < a.length; o++) {
                r = a[o];
                if (n.indexOf(r) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, r)) continue;
                t[r] = e[r];
            }
        }
        return t;
    }
    function g(e, n) {
        if (e == null) return {};
        var t = {};
        var r = Object.keys(e);
        var o, a;
        for (a = 0; a < r.length; a++) {
            o = r[a];
            if (n.indexOf(o) >= 0) continue;
            t[o] = e[o];
        }
        return t;
    }
    var O = {
        description: a().string,
        elementRef: a().oneOfType([ a().func, a().object ]),
        inset: a().bool,
        panelId: a().any.isRequired,
        title: a().node.isRequired
    };
    var P = function e(n, t) {
        var r = {
            enterprise: {
                compact: "8px",
                comfortable: "10px"
            },
            prisma: {
                compact: "12px 20px",
                comfortable: "20px"
            }
        };
        return r[n][t];
    };
    var C = {
        flexShrink: 1
    };
    /**
 * `Accordion.Panel` operates as a container component for content in an `Accordion`.
 */    function j(e) {
        var n = e.panelId, o = e.inset, a = m(e, [ "panelId", "inset" ]);
        // @docs-props-type AccordionPanelPropsBase
                var i = (0, y.useSplunkTheme)(), l = i.family, u = i.density;
        var c = (0, t.useContext)(v), f = c.onChange, s = c.openPanelId, p = c.inset;
        var b = s === n;
        var g = o !== undefined ? o : p;
 // Props are different based on controlled/uncontrolled
        // For controlled, onChange required prop for controlled cases and must be defined.
                var O = f ? {
            onChange: f,
            open: b
        } : {
            onChange: undefined
        };
        var j = function e() {
            if (g) {
                var n = P(l, u);
                return {
                    padding: n
                };
            }
            return undefined;
        };
        
        return r().createElement(d(), h({
            panelId: n,
            "data-test": "panel",
            innerBodyStyles: j(),
            "data-test-panel-id": n,
            style: C
        }, a, O));
    }
    j.propTypes = O;
    /* harmony default export */    const x = j;
    // CONCATENATED MODULE: external "styled-components"
    const I = require("styled-components");
    var S =  e.n(I);
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const w = require("@splunk/react-ui/Box");
    var k =  e.n(w);
    // CONCATENATED MODULE: ./src/Accordion/AccordionStyles.ts
    var _ = S()(k()).withConfig({
        displayName: "AccordionStyles__StyledBox",
        componentId: "m39hyh-0"
    })([ "", ";flex-direction:column;margin-bottom:0;", "" ], y.mixins.reset("flex"), (0, 
    y.pick)({
        enterprise: (0, I.css)([ "background:", ";" ], y.variables.backgroundColor)
    }));
    // CONCATENATED MODULE: ./src/Accordion/Accordion.tsx
    function R(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            R = function e(n) {
                return typeof n;
            };
        } else {
            R = function e(n) {
                return n && typeof Symbol === "function" && n.constructor === Symbol && n !== Symbol.prototype ? "symbol" : typeof n;
            };
        }
        return R(e);
    }
    function q() {
        q = Object.assign || function(e) {
            for (var n = 1; n < arguments.length; n++) {
                var t = arguments[n];
                for (var r in t) {
                    if (Object.prototype.hasOwnProperty.call(t, r)) {
                        e[r] = t[r];
                    }
                }
            }
            return e;
        };
        return q.apply(this, arguments);
    }
    function E(e, n) {
        if (!(e instanceof n)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function T(e, n) {
        for (var t = 0; t < n.length; t++) {
            var r = n[t];
            r.enumerable = r.enumerable || false;
            r.configurable = true;
            if ("value" in r) r.writable = true;
            Object.defineProperty(e, r.key, r);
        }
    }
    function A(e, n, t) {
        if (n) T(e.prototype, n);
        if (t) T(e, t);
        return e;
    }
    function B(e, n) {
        if (typeof n !== "function" && n !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(n && n.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (n) D(e, n);
    }
    function D(e, n) {
        D = Object.setPrototypeOf || function e(n, t) {
            n.__proto__ = t;
            return n;
        };
        return D(e, n);
    }
    function M(e) {
        var n = V();
        return function t() {
            var r = z(e), o;
            if (n) {
                var a = z(this).constructor;
                o = Reflect.construct(r, arguments, a);
            } else {
                o = r.apply(this, arguments);
            }
            return N(this, o);
        };
    }
    function N(e, n) {
        if (n && (R(n) === "object" || typeof n === "function")) {
            return n;
        }
        return U(e);
    }
    function U(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function V() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function z(e) {
        z = Object.setPrototypeOf ? Object.getPrototypeOf : function e(n) {
            return n.__proto__ || Object.getPrototypeOf(n);
        };
        return z(e);
    }
    function F(e, n, t) {
        if (n in e) {
            Object.defineProperty(e, n, {
                value: t,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[n] = t;
        }
        return e;
    }
    /** @public */    var G = {
        children: a().node,
        defaultOpenPanelId: a().any,
        elementRef: a().oneOfType([ a().func, a().object ]),
        inset: a().bool,
        onChange: a().func,
        openPanelId: a().any
    };
    var H = {
        inset: false
    };
    var J =  function(e) {
        B(o, e);
        var n = M(o);
        // @docs-props-type AccordionPropsBase
                function o(e) {
            var t;
            E(this, o);
            t = n.call(this, e);
            F(U(t), "controlledExternally", void 0);
            F(U(t), "getCurrentOpenPanel", (function() {
                var e = t.isControlled() ? t.props.openPanelId : t.state.openPanelId;
                return e;
            }));
            F(U(t), "handleChange", (function(e, n) {
                var r, o;
                var a = n.panelId, i = n.reason;
                var l = t.getCurrentOpenPanel();
                var u = l === a ? undefined : a;
                if (!t.isControlled()) {
                    t.setState({
                        openPanelId: u
                    });
                }
                (r = (o = t.props).onChange) === null || r === void 0 ? void 0 : r.call(o, e, {
                    event: e,
                    panelId: u,
                    reason: i
                });
            }));
            t.controlledExternally = l()(e, "openPanelId");
            if (!t.isControlled()) {
                t.state = {
                    openPanelId: e.defaultOpenPanelId
                };
            }
            if (false) {}
            return t;
        }
        A(o, [ {
            key: "componentDidUpdate",
            value: function e(n) {
                if (false) {}
                if (false) {}
            }
        }, {
            key: "isControlled",
            value: function e() {
                return this.controlledExternally;
            }
        }, {
            key: "render",
            value: function e() {
                var n = this.props, a = n.children, i = n.elementRef, l = n.inset;
                var u = this.isControlled() ? this.props.openPanelId : this.state.openPanelId;
                var f = t.Children.toArray(a).filter(t.isValidElement);
 // remove false, null, 0, etc
                                
                return r().createElement(_, q({
                    "data-test": "accordion",
                    "data-test-open-panel-id": u,
                    elementRef: i
                }, s()(this.props, c()(o.propTypes))),  r().createElement(b.Provider, {
                    value: {
                        onChange: this.handleChange,
                        openPanelId: u,
                        inset: l
                    }
                }, f));
            }
        } ]);
        return o;
    }(t.Component);
    F(J, "propTypes", G);
    F(J, "defaultProps", H);
    F(J, "Panel", x);
    /* harmony default export */    const K = J;
    // CONCATENATED MODULE: ./src/Accordion/index.ts
    module.exports = n;
    /******/})();
