// @splunk/react-ui/Anchor.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var r = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(r, {
                a: r
            });
            /******/            return r;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        default: () => /* reexport */ h
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    var n =  e.n(r);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var i =  e.n(o);
    // CONCATENATED MODULE: external "styled-components"
    const a = require("styled-components");
    var c =  e.n(a);
    // CONCATENATED MODULE: external "@splunk/themes"
    const l = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/Anchor/AnchorStyles.ts
    // TODO: name is deprecated, it's only defined here to satisfy ts - remove/replace
    var f = c().a.withConfig({
        displayName: "AnchorStyles__StyledAnchor",
        componentId: "sc-1m9yzet-0"
    })([ "", ";color:inherit;font-size:inherit;font-family:inherit;font-weight:inherit;line-height:inherit;text-decoration:inherit;text-rendering:inherit;text-transform:inherit;cursor:pointer;margin:1px;&:hover::after{content:' #';opacity:0.5;}&:focus,&:active{box-shadow:", ";&::after{content:' #';opacity:0.5;}}" ], l.mixins.reset("inline"), l.variables.focusShadow);
    // CONCATENATED MODULE: ./src/Anchor/Anchor.tsx
    function u() {
        u = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return u.apply(this, arguments);
    }
    function s(e, t) {
        if (e == null) return {};
        var r = p(e, t);
        var n, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                n = i[o];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function p(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var o, i;
        for (i = 0; i < n.length; i++) {
            o = n[i];
            if (t.indexOf(o) >= 0) continue;
            r[o] = e[o];
        }
        return r;
    }
    var d = {
        children: i().node,
        elementRef: i().oneOfType([ i().func, i().object ]),
        name: i().string.isRequired
    };
    /**
 * `Anchor` is a utility component used to add a fragment link to an element.
 */    function y(e) {
        var t = e.children, r = e.elementRef, o = e.name, i = s(e, [ "children", "elementRef", "name" ]);
        // @docs-props-type AnchorPropsBase
                
        return n().createElement(f, u({
            "data-test": "anchor",
            ref: r
        }, i, {
            name: o,
            href: "#".concat(o)
        }), t);
    }
    y.propTypes = d;
    /* harmony default export */    const h = y;
    // CONCATENATED MODULE: ./src/Anchor/index.ts
    module.exports = t;
    /******/})();
