// @splunk/react-ui/AnimationToggle.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var t = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(t, {
                a: t
            });
            /******/            return t;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, t) => {
            /******/ for (var n in t) {
                /******/ if (e.o(t, n) && !e.o(r, n)) {
                    /******/ Object.defineProperty(r, n, {
                        enumerable: true,
                        get: t[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        AnimationToggleContext: () => /* reexport */ u,
        AnimationToggleProvider: () => /* reexport */ h,
        default: () => /* reexport */ b,
        useAnimationToggle: () => /* reexport */ v
    });
    // CONCATENATED MODULE: external "react"
    const t = require("react");
    var n =  e.n(t);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var a =  e.n(o);
    // CONCATENATED MODULE: ./src/AnimationToggle/AnimationToggleContext.tsx
    /**
 * A React context used to control whether animations are enabled or disabled.
 * Prefer `useAnimationToggle` or `AnimationToggle` over consuming this context
 * directly, since they take the user's reduced motion preference into account.
 * This context should only be used if direct access to the animation setting
 * is required. The context type is `Boolean` and defaults to `true`.
 *
 * @public
 */
    var u =  n().createContext(true);
    // CONCATENATED MODULE: ./src/AnimationToggle/useAnimationToggle.tsx
    function i(e, r) {
        return s(e) || l(e, r) || c(e, r) || f();
    }
    function f() {
        throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function c(e, r) {
        if (!e) return;
        if (typeof e === "string") return d(e, r);
        var t = Object.prototype.toString.call(e).slice(8, -1);
        if (t === "Object" && e.constructor) t = e.constructor.name;
        if (t === "Map" || t === "Set") return Array.from(e);
        if (t === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t)) return d(e, r);
    }
    function d(e, r) {
        if (r == null || r > e.length) r = e.length;
        for (var t = 0, n = new Array(r); t < r; t++) {
            n[t] = e[t];
        }
        return n;
    }
    function l(e, r) {
        if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(e))) return;
        var t = [];
        var n = true;
        var o = false;
        var a = undefined;
        try {
            for (var u = e[Symbol.iterator](), i; !(n = (i = u.next()).done); n = true) {
                t.push(i.value);
                if (r && t.length === r) break;
            }
        } catch (e) {
            o = true;
            a = e;
        } finally {
            try {
                if (!n && u["return"] != null) u["return"]();
            } finally {
                if (o) throw a;
            }
        }
        return t;
    }
    function s(e) {
        if (Array.isArray(e)) return e;
    }
    /**
 * A React hook that combines the animation setting (see `AnimationToggleProvider`)
 * and the user's reduced motion preference into one value.
 *
 * @returns {string} `'off'` if animations are disabled. If animations are enabled
 * and the user prefers reduced motion, `'reduced'` is returned, `'on'` otherwise.
 * @public
 */    var v = function e() {
        var r = (0, t.useContext)(u);
        var n = "screen and (prefers-reduced-motion: reduce)";
        var o = typeof window !== "undefined" && !!window.matchMedia;
        var a = (0, t.useState)(o && window.matchMedia(n).matches), f = i(a, 2), c = f[0], d = f[1];
        (0, t.useEffect)((function() {
            if (o) {
                var e = window.matchMedia(n);
                var r = function r() {
                    d(e.matches);
                };
                e.addListener(r);
                return function() {
                    e.removeListener(r);
                };
            }
            return undefined;
        }), [ o, n ]);
        if (r) {
            return c ? "reduced" : "on";
        }
        return "off";
    };
    // CONCATENATED MODULE: ./src/AnimationToggle/AnimationToggle.tsx
    var m = function e(r) {
        return typeof r === "function" ? r() : r;
    };
    var y = {
        on: a().oneOfType([ a().node, a().func ]).isRequired,
        off: a().oneOfType([ a().node, a().func ]).isRequired,
        reduced: a().oneOfType([ a().node, a().func ])
    };
    /**
 * Used to allow animation to be externally controlled through a provider or user preference.
 */    function p(e) {
        var r = e.on, t = e.off, o = e.reduced;
        var a = v();
        if (a === "on") {
            
            return n().createElement(n().Fragment, null, m(r));
        }
        if (a === "reduced" && o) {
            
            return n().createElement(n().Fragment, null, m(o));
        }
        
        return n().createElement(n().Fragment, null, m(t));
 // react-docgen ignores components that don't use JSX/React.createElement.
        // the following line forces detection:
                
        return n().createElement(n().Fragment, null);
 // eslint-disable-line no-unreachable
        }
    p.propTypes = y;
    /* harmony default export */    const b = p;
    // CONCATENATED MODULE: ./src/AnimationToggle/AnimationToggleProvider.tsx
    var g = {
        children: a().node,
        enabled: a().bool
    };
    /**
 * Used to provide a setting to enable or disable animation.
 */    var h = function e(r) {
        var t = r.children, o = r.enabled, a = o === void 0 ? false : o;
        
        return n().createElement(u.Provider, {
            value: a
        }, t);
    };
    h.propTypes = g;
    // CONCATENATED MODULE: ./src/AnimationToggle/index.ts
    module.exports = r;
    /******/})();
