// @splunk/react-ui/Animation.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, t) => {
            /******/ for (var n in t) {
                /******/ if (e.o(t, n) && !e.o(r, n)) {
                    /******/ Object.defineProperty(r, n, {
                        enumerable: true,
                        get: t[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        Spring: () => /* reexport */ l,
        Transition: () => /* reexport */ f
    });
    // CONCATENATED MODULE: external "react-spring"
    const t = require("react-spring");
    // CONCATENATED MODULE: external "@splunk/react-ui/AnimationToggle"
    const n = require("@splunk/react-ui/AnimationToggle");
    // CONCATENATED MODULE: ./src/Animation/Animation.tsx
    function i(e, r) {
        var t = Object.keys(e);
        if (Object.getOwnPropertySymbols) {
            var n = Object.getOwnPropertySymbols(e);
            if (r) n = n.filter((function(r) {
                return Object.getOwnPropertyDescriptor(e, r).enumerable;
            }));
            t.push.apply(t, n);
        }
        return t;
    }
    function o(e) {
        for (var r = 1; r < arguments.length; r++) {
            var t = arguments[r] != null ? arguments[r] : {};
            if (r % 2) {
                i(Object(t), true).forEach((function(r) {
                    u(e, r, t[r]);
                }));
            } else if (Object.getOwnPropertyDescriptors) {
                Object.defineProperties(e, Object.getOwnPropertyDescriptors(t));
            } else {
                i(Object(t)).forEach((function(r) {
                    Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r));
                }));
            }
        }
        return e;
    }
    function u(e, r, t) {
        if (r in e) {
            Object.defineProperty(e, r, {
                value: t,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[r] = t;
        }
        return e;
    }
    function c(e, r) {
        if (e == null) return {};
        var t = a(e, r);
        var n, i;
        if (Object.getOwnPropertySymbols) {
            var o = Object.getOwnPropertySymbols(e);
            for (i = 0; i < o.length; i++) {
                n = o[i];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function a(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var i, o;
        for (o = 0; o < n.length; o++) {
            i = n[o];
            if (r.indexOf(i) >= 0) continue;
            t[i] = e[i];
        }
        return t;
    }
    /**
 * Wrappers to use in place of `Spring` and `Transition` from `react-spring`.
 * Uses `useAnimationToggle` to offer animation control.
 */    function l(e) {
        var r = e.children, i = e.immediate, u = c(e, [ "children", "immediate" ]);
        var a = (0, n.useAnimationToggle)();
        return r((0, t.useSpring)(o({
            immediate: i || a !== "on"
        }, u)));
    }
    function f(e) {
        var r = e.items, i = e.children, u = e.immediate, a = c(e, [ "items", "children", "immediate" ]);
        var l = (0, n.useAnimationToggle)();
        return (0, t.useTransition)(r, o({
            immediate: u || l !== "on"
        }, a))(i);
    }
    // CONCATENATED MODULE: ./src/Animation/index.ts
    module.exports = r;
    /******/})();
