// @splunk/react-ui/Box.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var n = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(n, {
                a: n
            });
            /******/            return n;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, n) => {
            /******/ for (var r in n) {
                /******/ if (e.o(n, r) && !e.o(t, r)) {
                    /******/ Object.defineProperty(t, r, {
                        enumerable: true,
                        get: n[r]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        default: () => /* reexport */ v
    });
    // CONCATENATED MODULE: external "react"
    const n = require("react");
    var r =  e.n(n);
    // CONCATENATED MODULE: external "prop-types"
    const i = require("prop-types");
    var l =  e.n(i);
    // CONCATENATED MODULE: external "styled-components"
    const o = require("styled-components");
    var a =  e.n(o);
    // CONCATENATED MODULE: external "@splunk/themes"
    const f = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/Box/BoxStyles.ts
    var d = a().div.withConfig({
        displayName: "BoxStyles__Styled",
        componentId: "sc-1h4b5f6-0"
    })([ "", ";align-items:stretch;align-content:stretch;flex-flow:row nowrap;flex-grow:0;flex-shrink:0;justify-content:flex-start;width:auto;max-width:100%;@media all and (-ms-high-contrast:none){max-width:calc(100% - 0.01px);}&[data-inline]{display:inline-block;vertical-align:middle;}&[data-flex]{display:flex;}&[data-flex][data-inline]{display:inline-flex;}" ], f.mixins.reset("block"));
    // CONCATENATED MODULE: ./src/Box/Box.tsx
    function c() {
        c = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var n = arguments[t];
                for (var r in n) {
                    if (Object.prototype.hasOwnProperty.call(n, r)) {
                        e[r] = n[r];
                    }
                }
            }
            return e;
        };
        return c.apply(this, arguments);
    }
    function s(e, t) {
        if (e == null) return {};
        var n = u(e, t);
        var r, i;
        if (Object.getOwnPropertySymbols) {
            var l = Object.getOwnPropertySymbols(e);
            for (i = 0; i < l.length; i++) {
                r = l[i];
                if (t.indexOf(r) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, r)) continue;
                n[r] = e[r];
            }
        }
        return n;
    }
    function u(e, t) {
        if (e == null) return {};
        var n = {};
        var r = Object.keys(e);
        var i, l;
        for (l = 0; l < r.length; l++) {
            i = r[l];
            if (t.indexOf(i) >= 0) continue;
            n[i] = e[i];
        }
        return n;
    }
    var p = {
        children: l().node,
        elementRef: l().oneOfType([ l().func, l().object ]),
        flex: l().bool,
        inline: l().bool
    };
    /**
 * Box is a utility component to normalize and abstract common layout styles.
 */    function y(e) {
        var t = e.children, n = e.elementRef, i = e.flex, l = i === void 0 ? false : i, o = e.tag, a = o === void 0 ? "div" : o, f = e.inline, u = f === void 0 ? false : f, p = s(e, [ "children", "elementRef", "flex", "tag", "inline" ]);
        
        return r().createElement(d, c({
            as: a,
            ref: n
        }, p, {
            "data-inline": u || undefined,
            "data-flex": l || undefined
        }), t);
    }
    y.propTypes = p;
    /* harmony default export */    const v = y;
    // CONCATENATED MODULE: ./src/Box/index.ts
    module.exports = t;
    /******/})();
