// @splunk/react-ui/Breadcrumbs.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var t = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(t, {
                a: t
            });
            /******/            return t;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, t) => {
            /******/ for (var n in t) {
                /******/ if (e.o(t, n) && !e.o(r, n)) {
                    /******/ Object.defineProperty(r, n, {
                        enumerable: true,
                        get: t[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        BreadcrumbsContext: () => /* reexport */ E,
        Item: () => /* reexport */ O,
        default: () => /* reexport */ q
    });
    // CONCATENATED MODULE: external "react"
    const t = require("react");
    var n =  e.n(t);
    // CONCATENATED MODULE: external "prop-types"
    const a = require("prop-types");
    var o =  e.n(a);
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const l = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "styled-components"
    const i = require("styled-components");
    var c =  e.n(i);
    // CONCATENATED MODULE: external "@splunk/themes"
    const u = require("@splunk/themes");
    // CONCATENATED MODULE: external "@splunk/react-ui/Link"
    const s = require("@splunk/react-ui/Link");
    var f =  e.n(s);
    // CONCATENATED MODULE: ./src/Breadcrumbs/ItemStyles.ts
    var d = c()(f()).withConfig({
        displayName: "ItemStyles__StyledLink",
        componentId: "sc-1ywtawf-0"
    })([ "display:flex;align-items:center;gap:", ";white-space:nowrap;color:", ";&:where(:hover,:focus):not([disabled],[aria-disabled='true']){color:", ";}" ], u.variables.spacingXSmall, u.variables.contentColorDefault, u.variables.contentColorActive);
    // CONCATENATED MODULE: ./src/Breadcrumbs/Item.tsx
    function p() {
        p = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return p.apply(this, arguments);
    }
    function b(e, r) {
        if (e == null) return {};
        var t = v(e, r);
        var n, a;
        if (Object.getOwnPropertySymbols) {
            var o = Object.getOwnPropertySymbols(e);
            for (a = 0; a < o.length; a++) {
                n = o[a];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function v(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var a, o;
        for (o = 0; o < n.length; o++) {
            a = n[o];
            if (r.indexOf(a) >= 0) continue;
            t[a] = e[a];
        }
        return t;
    }
    var m = Object.freeze({
        allowDisabledLink: true,
        disabled: true,
        to: ""
    });
    var y = {
        elementRef: o().oneOfType([ o().func, o().object ]),
        enableCurrentPage: o().bool,
        endAdornment: o().node,
        isCurrent: o().bool,
        label: o().string.isRequired,
        onClick: o().func,
        startAdornment: o().node,
        to: o().string.isRequired
    };
    var g = {
        enableCurrentPage: false,
        isCurrent: false
    };
    function C(e) {
        var r = e.enableCurrentPage, a = e.endAdornment, o = e.isCurrent, l = e.label, i = e.onClick, c = e.startAdornment, u = e.to, s = b(e, [ "enableCurrentPage", "endAdornment", "isCurrent", "label", "onClick", "startAdornment", "to" ]);
        // @docs-props-type ItemPropsBase
                var f = (0, t.useContext)(E), v = f.onClick;
        var y = {
            to: u
        };
        var g = (0, t.useCallback)((function(e) {
            i === null || i === void 0 ? void 0 : i(e, {
                label: l,
                to: u
            });
            v === null || v === void 0 ? void 0 : v(e, {
                label: l,
                to: u
            });
        }), [ i, v, l, u ]);
        if (o) {
            y["aria-current"] = "page";
            if (!r) {
                Object.assign(y, m);
            }
        }
        var C = i || v ? {
            onClick: g
        } : {};
        
        return n().createElement(d, p({
            "data-test": "item"
        }, C, y, s), c && c, l, a && a);
    }
    C.propTypes = y;
    C.defaultProps = g;
    /* harmony default export */    const O = C;
    // CONCATENATED MODULE: ./src/Breadcrumbs/BreadcrumbsStyles.ts
    var h = c().ol.withConfig({
        displayName: "BreadcrumbsStyles__StyledList",
        componentId: "sc-1maeyfk-0"
    })([ "", " flex-wrap:wrap;" ], u.mixins.reset("flex"));
    var k = c().li.withConfig({
        displayName: "BreadcrumbsStyles__StyledListItem",
        componentId: "sc-1maeyfk-1"
    })([ "display:flex;min-width:max-content;a[aria-current]{font-weight:", ";color:", ";cursor:default;}" ], u.variables.fontWeightSemiBold, u.variables.contentColorActive);
    var P = c().span.withConfig({
        displayName: "BreadcrumbsStyles__StyledSeparator",
        componentId: "sc-1maeyfk-2"
    })([ "cursor:default;padding:0 ", ";color:", ";" ], u.variables.spacingSmall, u.variables.contentColorDefault);
    // CONCATENATED MODULE: ./src/Breadcrumbs/Breadcrumbs.tsx
    function w() {
        w = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return w.apply(this, arguments);
    }
    function S(e, r) {
        if (e == null) return {};
        var t = j(e, r);
        var n, a;
        if (Object.getOwnPropertySymbols) {
            var o = Object.getOwnPropertySymbols(e);
            for (a = 0; a < o.length; a++) {
                n = o[a];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function j(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var a, o;
        for (o = 0; o < n.length; o++) {
            a = n[o];
            if (r.indexOf(a) >= 0) continue;
            t[a] = e[a];
        }
        return t;
    }
    var x = {
        children: o().node.isRequired,
        elementRef: o().oneOfType([ o().func, o().object ]),
        enableCurrentPage: o().bool,
        onClick: o().func
    };
    var _ = {
        enableCurrentPage: false
    };
    var E =  n().createContext({});
    function I(e) {
        var r = e.children, a = e.elementRef, o = e.enableCurrentPage, i = e.onClick, c = S(e, [ "children", "elementRef", "enableCurrentPage", "onClick" ]);
        // @docs-props-type BreadcrumbsPropsBase
                var u;
        var s = t.Children.toArray(r).filter(t.isValidElement).map((function(e, a) {
            if (false) {}
            if (false) {}
 // Set the last child as the current Breadcrumbs.Item
                        var l = a === t.Children.count(r) - 1;
            var i =  (0, t.cloneElement)(e, {
                isCurrent: l,
                enableCurrentPage: o
            });
            
            return n().createElement(k, {
                key: e.props.to
            }, i, !l &&  n().createElement(P, {
                "aria-hidden": "true"
            }, "/"));
        }));
        var f = {
            onClick: i
        };
        
        return n().createElement(E.Provider, {
            value: f
        },  n().createElement("nav", w({
            "data-test": "breadcrumbs",
            "aria-label": (0, l._)("Breadcrumbs"),
            ref: a
        }, c),  n().createElement(h, null, s)));
    }
    I.propTypes = x;
    I.defaultProps = _;
    I.Item = O;
    /* harmony default export */    const q = I;
    // CONCATENATED MODULE: ./src/Breadcrumbs/index.ts
    module.exports = r;
    /******/})();
