// @splunk/react-ui/ButtonGroup.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var t = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(t, {
                a: t
            });
            /******/            return t;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, t) => {
            /******/ for (var n in t) {
                /******/ if (e.o(t, n) && !e.o(r, n)) {
                    /******/ Object.defineProperty(r, n, {
                        enumerable: true,
                        get: t[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        default: () => /* reexport */ m
    });
    // CONCATENATED MODULE: external "react"
    const t = require("react");
    var n =  e.n(t);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var i =  e.n(o);
    // CONCATENATED MODULE: external "styled-components"
    const a = require("styled-components");
    var l =  e.n(a);
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const u = require("@splunk/react-ui/Box");
    var p =  e.n(u);
    // CONCATENATED MODULE: external "@splunk/themes"
    const c = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/ButtonGroup/ButtonGroupStyles.ts
    var s = l()(p()).withConfig({
        displayName: "ButtonGroupStyles__StyledBox",
        componentId: "sc-14fs2xs-0"
    })([ "position:relative;[data-inline] + &{margin-left:", ";}" ], (0, c.pick)({
        enterprise: c.variables.spacingHalf,
        prisma: c.variables.spacingSmall
    }));
    // CONCATENATED MODULE: ./src/ButtonGroup/ButtonGroup.tsx
    function f() {
        f = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return f.apply(this, arguments);
    }
    function d(e, r) {
        if (e == null) return {};
        var t = v(e, r);
        var n, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                n = i[o];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function v(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var o, i;
        for (i = 0; i < n.length; i++) {
            o = n[i];
            if (r.indexOf(o) >= 0) continue;
            t[o] = e[o];
        }
        return t;
    }
    var y = {
        children: i().node,
        elementRef: i().oneOfType([ i().func, i().object ]),
        /** @private */
        flex: i().bool
    };
    function b(e) {
        var r = e.children, o = d(e, [ "children" ]);
        // @docs-props-type ButtonGroupPropsBase
                var i = t.Children.toArray(r).filter(t.isValidElement).map((function(e, r, n) {
            var o = n.length;
            
            return (0, t.cloneElement)(e, {
                prepend: r > 0,
                append: r < o - 1,
                inline: false,
                role: e.props && e.props.role || "menuitem"
            });
        }));
        
        return n().createElement(s, f({
            inline: true,
            flex: true,
            "data-test": "button-group",
            role: "menubar"
        }, o), i);
    }
    b.propTypes = y;
    /* harmony default export */    const m = b;
    // CONCATENATED MODULE: ./src/ButtonGroup/index.ts
    module.exports = r;
    /******/})();
