// @splunk/react-ui/ButtonSimple.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var r = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ r.n = e => {
            /******/ var a = e && e.__esModule ? 
            /******/ () => e["default"]
            /******/ : () => e
            /******/;
            r.d(a, {
                a
            });
            /******/            return a;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ r.d = (e, a) => {
            /******/ for (var o in a) {
                /******/ if (r.o(a, o) && !r.o(e, o)) {
                    /******/ Object.defineProperty(e, o, {
                        enumerable: true,
                        get: a[o]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ r.o = (r, e) => Object.prototype.hasOwnProperty.call(r, e)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ r.r = r => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(r, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(r, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var e = {};
    // ESM COMPAT FLAG
        r.r(e);
    // EXPORTS
        r.d(e, {
        default: () => /* reexport */ vr
    });
    // CONCATENATED MODULE: external "react"
    const a = require("react");
    var o =  r.n(a);
    // CONCATENATED MODULE: external "prop-types"
    const i = require("prop-types");
    var t =  r.n(i);
    // CONCATENATED MODULE: external "@splunk/themes"
    const l = require("@splunk/themes");
    // CONCATENATED MODULE: external "styled-components"
    const n = require("styled-components");
    var s =  r.n(n);
    // CONCATENATED MODULE: external "@splunk/react-ui/Clickable"
    const c = require("@splunk/react-ui/Clickable");
    var d =  r.n(c);
    // CONCATENATED MODULE: external "lodash/merge"
    const v = require("lodash/merge");
    var b =  r.n(v);
    // CONCATENATED MODULE: ./src/ButtonSimple/mixin.ts
    function u(r) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            u = function r(e) {
                return typeof e;
            };
        } else {
            u = function r(e) {
                return e && typeof Symbol === "function" && e.constructor === Symbol && e !== Symbol.prototype ? "symbol" : typeof e;
            };
        }
        return u(r);
    }
    var p = [ "primary", "secondary", "destructive", "subtle" ];
    /**
 * These enterprise... values don't exist in Enterprise @splunk/themes
 * They're provided here for TypeScript compatibility
 */    var f = {
        active: {
            dark: "rgba(0, 0, 0, 0.2)",
            light: "rgba(0, 0, 0, 0.07)"
        },
        hover: {
            dark: "rgba(255, 255, 255, 0.05)",
            light: "rgba(0, 0, 0, 0.03)"
        },
        selected: {
            dark: "rgba(255, 255, 255, 0.1)",
            light: "rgba(0, 0, 0, 0.04)"
        }
    };
    var g = {
        dark: "#272a2f",
        light: "#eeeeee"
    };
    var y = {
        dark: "rgba(255, 255, 255, 0.15)",
        light: "rgba(0, 0, 0, 0.1)"
    };
    function h(r) {
        var e = "transparent";
        var a = "transparent";
        var o = (0, l.pick)({
            enterprise: y,
            prisma: l.variables.interactiveColorBackgroundDisabled
        });
        var i = (0, l.pick)({
            enterprise: {
                dark: l.mixins.overlayColors(y.dark, f.selected.dark),
                light: l.mixins.overlayColors(y.light, f.selected.light)
            },
            prisma: l.mixins.overlayColors(l.variables.interactiveColorBackgroundDisabled, l.variables.interactiveColorOverlaySelected)
        });
        var t = "transparent";
        var n = "transparent";
        var s;
        var c;
        var d = l.variables.contentColorActive;
        var v = l.variables.contentColorDisabled;
        switch (r) {
          case "primary":
            {
                e = l.variables.interactiveColorPrimary;
                a = (0, l.pick)({
                    enterprise: {
                        light: l.mixins.overlayColors(l.variables.interactiveColorPrimary, f.active.light),
                        dark: l.mixins.overlayColors(l.variables.interactiveColorPrimary, f.active.dark)
                    },
                    prisma: l.mixins.overlayColors(l.variables.interactiveColorPrimary, l.variables.interactiveColorOverlayActive)
                });
                t = (0, l.pick)({
                    enterprise: {
                        light: l.mixins.overlayColors(l.variables.interactiveColorPrimary, f.hover.light),
                        dark: l.mixins.overlayColors(l.variables.interactiveColorPrimary, f.hover.dark)
                    },
                    prisma: l.mixins.overlayColors(l.variables.interactiveColorPrimary, l.variables.interactiveColorOverlayHover)
                });
                n = (0, l.pick)({
                    enterprise: {
                        light: l.mixins.overlayColors(l.variables.interactiveColorPrimary, f.selected.light),
                        dark: l.mixins.overlayColors(l.variables.interactiveColorPrimary, f.selected.dark)
                    },
                    prisma: l.mixins.overlayColors(l.variables.interactiveColorPrimary, l.variables.interactiveColorOverlaySelected)
                });
                d = l.variables.contentColorInverted;
                break;
            }

          case "destructive":
            {
                e = l.variables.accentColorNegative;
                a = (0, l.pick)({
                    enterprise: {
                        light: l.mixins.overlayColors(l.variables.accentColorNegative, f.active.light),
                        dark: l.mixins.overlayColors(l.variables.accentColorNegative, f.active.dark)
                    },
                    prisma: l.mixins.overlayColors(l.variables.accentColorNegative, l.variables.interactiveColorOverlayActive)
                });
                t = (0, l.pick)({
                    enterprise: {
                        light: l.mixins.overlayColors(l.variables.accentColorNegative, f.hover.light),
                        dark: l.mixins.overlayColors(l.variables.accentColorNegative, f.hover.dark)
                    },
                    prisma: l.mixins.overlayColors(l.variables.accentColorNegative, l.variables.interactiveColorOverlayHover)
                });
                n = (0, l.pick)({
                    enterprise: {
                        light: l.mixins.overlayColors(l.variables.accentColorNegative, f.selected.light),
                        dark: l.mixins.overlayColors(l.variables.accentColorNegative, f.selected.dark)
                    },
                    prisma: l.mixins.overlayColors(l.variables.accentColorNegative, l.variables.interactiveColorOverlaySelected)
                });
                d = l.variables.contentColorInverted;
                break;
            }

          case "secondary":
            s = l.variables.interactiveColorBorder;
            c = l.variables.interactiveColorBorderDisabled;
            e = (0, l.pick)({
                enterprise: g,
                prisma: l.variables.interactiveColorBackground
            });
            a = (0, l.pick)({
                enterprise: f.active,
                prisma: l.variables.interactiveColorOverlayActive
            });
            t = (0, l.pick)({
                enterprise: f.hover,
                prisma: l.variables.interactiveColorOverlayHover
            });
            n = (0, l.pick)({
                enterprise: f.selected,
                prisma: l.variables.interactiveColorOverlaySelected
            });
            break;

          case "subtle":
            // TODO: SUI-6304 this should be interactiveColorAccent once that variable is introduced in SUI 5
            d = l.variables.interactiveColorPrimary;
            a = (0, l.pick)({
                enterprise: f.active,
                prisma: l.variables.interactiveColorOverlayActive
            });
            t = (0, l.pick)({
                enterprise: f.hover,
                prisma: l.variables.interactiveColorOverlayHover
            });
            n = (0, l.pick)({
                enterprise: f.selected,
                prisma: l.variables.interactiveColorOverlaySelected
            });
            break;

          default:
        }
        var b = {
            background: e,
            backgroundActive: a,
            backgroundDisabled: o,
            backgroundDisabledSelected: i,
            backgroundHover: t,
            backgroundSelected: n,
            borderColor: s,
            borderColorDisabled: c,
            color: d,
            colorDisabled: v
        };
        return b;
    }
    function k(r, e) {
        var a = typeof r === "string" && p.includes(r) ? r : undefined;
        var o = {};
        if (a && e !== undefined) {
            o = e;
        } else if (a === undefined && u(r) === "object" && e === undefined) {
            o = r;
        }
        var i = h(a);
        var t = b()(i, o), s = t.background, c = t.backgroundActive, d = t.backgroundDisabled, v = t.backgroundDisabledSelected, f = t.backgroundHover, g = t.backgroundSelected, y = t.borderColor, k = t.borderColorDisabled, m = t.color, C = t.colorDisabled;
        var x = s != null && s !== "transparent";
        return function() {
            return (0, n.css)([ "border-radius:", ";", ";", " ", " cursor:pointer;position:relative;&:focus,&:hover{z-index:3;}&:not([disabled],[aria-disabled='true']){background-color:", ";color:", ";transition:background-color 0.2s,box-shadow 0.2s,text-decoration 0.2s;&:hover{background-color:", ";}&:active{background-color:", ";transition:none;}", "}&:focus{box-shadow:", ";}&[disabled],&[aria-disabled='true']{border-color:", ";color:", ";cursor:not-allowed;", "}" ], l.variables.borderRadius, (function() {
                return y && (0, n.css)([ "border:1px solid ", ";" ], y);
            }), (function(r) {
                var e = r.$append;
                return e && (0, n.css)([ "border-top-right-radius:0.1px;border-bottom-right-radius:0.1px;border-right:none;" ]);
            }), (function(r) {
                var e = r.$prepend;
                return e && (0, n.css)([ "border-top-left-radius:0.1px;border-bottom-left-radius:0.1px;" ]);
            }), s, m, f, c, (function(r) {
                var e = r.$selected;
                return e && (0, n.css)([ "background-color:", ";" ], g);
            }), l.variables.focusShadow, k, C, (function(r) {
                return x && (0, n.css)([ "background-color:", ";" ], r ? v
                /* TODO: this isn't defined in SUI 5 themes yet */ : d);
            }));
        };
    }
    /* harmony default export */    const m = k;
    // CONCATENATED MODULE: ./src/ButtonSimple/prismaStyles.ts
    var C = (0, n.css)([ "", " &:not([disabled],[aria-disabled='true']){&:focus{background-color:", ";}&:hover{border-color:", ";", "}", " ", "}&[disabled],&[aria-disabled='true']{", "}" ], (0, 
    l.pickVariant)("$variant", {
        secondary: (0, n.css)([ "", " font-weight:", ";" ], m("subtle", {
            color: l.variables.contentColorActive
        }), l.variables.fontWeightSemiBold),
        subtle: (0, n.css)([ "", " font-weight:", ";" ], m("subtle", {
            color: l.variables.contentColorActive
        }), l.variables.fontWeightSemiBold),
        default: (0, n.css)([ "", " font-weight:", ";" ], m("secondary", {
            background: "transparent"
        }), l.variables.fontWeightSemiBold),
        toggle: (0, n.css)([ "", " font-weight:normal;" ], m("secondary", {
            background: "transparent"
        }))
    }), l.variables.interactiveColorOverlayHover, l.variables.interactiveColorBorderHover, (0, 
    l.pickVariant)("$variant", {
        secondary: (0, n.css)([ "box-shadow:", ";" ], l.variables.hoverShadow),
        default: (0, n.css)([ "box-shadow:", ";" ], l.variables.hoverShadow),
        subtle: (0, n.css)([ "box-shadow:", ";" ], l.variables.hoverShadow)
    }), (function(r) {
        var e = r.$selected;
        return e && (0, n.css)([ "border-color:", ";&:hover{background-color:", ";}" ], l.variables.interactiveColorBorderDisabled, l.mixins.overlayColors(l.variables.interactiveColorOverlaySelected, l.variables.interactiveColorOverlayHover));
    }), (function(r) {
        var e = r.$error;
        return e && (0, n.css)([ "border-color:", ";&:hover{border-color:", ";}" ], l.variables.accentColorNegative, l.variables.accentColorNegative);
    }), (function(r) {
        var e = r.$selected;
        return e && (0, n.css)([ "background-color:", ";" ], l.mixins.overlayColors(l.variables.interactiveColorBackgroundDisabled, l.variables.interactiveColorOverlaySelected));
    }));
    var x = (0, n.css)([ "", " font-weight:", ";&:focus{background-color:", ";}&:hover{box-shadow:", ";}" ], (0, 
    l.pickVariant)("$variant", {
        primary: m("primary"),
        destructive: m("destructive"),
        flat: m({
            // this is intentional - the background had always been applying the $selected style
            // since beccfde78f787861217353e9149ce213bee0270 so keeping this for visual consistency
            background: l.mixins.overlayColors(l.variables.interactiveColorBackground, l.variables.interactiveColorOverlaySelected),
            backgroundActive: l.mixins.overlayColors(l.variables.interactiveColorBackground, l.variables.interactiveColorOverlayActive),
            backgroundHover: l.mixins.overlayColors(l.variables.interactiveColorBackground, l.variables.interactiveColorOverlayHover),
            backgroundSelected: l.mixins.overlayColors(l.variables.interactiveColorBackground, l.variables.interactiveColorOverlaySelected)
        })
    }), l.variables.fontWeightSemiBold, (0, l.pickVariant)("$variant", {
        primary: l.mixins.overlayColors(l.variables.interactiveColorPrimary, l.variables.interactiveColorOverlayHover),
        destructive: l.mixins.overlayColors(l.variables.accentColorNegative, l.variables.interactiveColorOverlayHover),
        flat: l.mixins.overlayColors(l.variables.interactiveColorBackground, l.variables.interactiveColorOverlayHover)
    }), l.variables.hoverShadow);
    // CONCATENATED MODULE: ./src/ButtonSimple/enterpriseStyles.ts
    var w = "#1A8929";
 // SUI-2439 to meet WCAG AA compliance
        var S = "#235823";
 // SUI-2439 to meet WCAG AA compliance
        var O = (0, n.css)([ "border-radius:", ";cursor:pointer;position:relative;&:focus{z-index:3;}" ], l.variables.borderRadius);
    var $ = (0, n.css)([ "", " ", "" ], (function(r) {
        var e = r.$append;
        return e && (0, n.css)([ "border-top-right-radius:0;border-bottom-right-radius:0;border-right:none;" ]);
    }), (function(r) {
        var e = r.$prepend;
        return e && (0, n.css)([ "border-top-left-radius:0;border-bottom-left-radius:0;" ]);
    }));
    var D = "inset 0 -1px 0 rgba(0, 0, 0, 0.1)";
    var P = "inset 0 1px 0 rgba(0, 0, 0, 0.1)";
    var j = (0, n.css)([ "", ";border:", ";", ";font-weight:", ";&:not([disabled],[aria-disabled='true']){background-color:", ";color:", ";transition:background-image 0.2s,border 0.2s,box-shadow 0.2s,text-decoration 0.2s;", ";", " &:hover{background-color:", ";box-shadow:", ";}&:active{transition:none;background-color:", ";border-color:", ";", ";}&:focus{", " ", "}}&:not([disabled]){&:focus{box-shadow:", ",", ";}}&[disabled],&[aria-disabled='true']{cursor:not-allowed;color:", ";background-color:", ";border-color:", ";", "}" ], O, (0, 
    l.pick)({
        light: (0, n.css)([ "1px solid ", "" ], l.variables.gray60),
        dark: l.variables.border
    }), $, (0, l.pickVariant)("$variant", {
        default: "normal",
        secondary: l.variables.fontWeightSemiBold
    }), (0, l.pick)({
        light: l.variables.gray98,
        dark: l.variables.gray45
    }), (0, l.pick)({
        light: l.variables.gray45,
        dark: l.variables.white
    }), (0, l.pick)({
        light: (0, n.css)([ "box-shadow:", ";" ], D),
        dark: (0, n.css)([ "box-shadow:inset 0 -1px 0 ", ";" ], l.variables.gray30)
    }), (function(r) {
        var e = r.$selected;
        return e && (0, n.css)([ "box-shadow:", ";background-color:", ";border-color:", ";" ], (0, 
        l.pick)({
            light: P,
            dark: (0, n.css)([ "inset 0 1px 0 ", "" ], l.variables.black)
        }), (0, l.pick)({
            light: l.variables.gray92,
            dark: l.variables.gray22
        }), (0, l.pick)({
            light: l.variables.gray60,
            dark: l.variables.gray20
        }));
    }), (0, l.pick)({
        light: "#ebeeef",
        dark: l.variables.gray30
    }), (0, l.pick)({
        light: D,
        dark: (0, n.css)([ "inset 0 -1px 0 ", "" ], l.variables.gray25)
    }), (0, l.pick)({
        light: l.variables.gray92,
        dark: l.variables.gray22
    }), (0, l.pick)({
        light: l.variables.gray60,
        dark: l.variables.gray20
    }), (0, l.pick)({
        dark: (0, n.css)([ "box-shadow:inset 0 -1px 0 ", ";" ], l.variables.gray22)
    }), (function(r) {
        var e = r.$append;
        return e && (0, n.css)([ "box-shadow:", ",", ",inset -1px 0 0 ", ";" ], (0, l.pick)({
            light: D,
            dark: (0, n.css)([ "inset 0 -1px 0 ", "" ], l.variables.gray30)
        }), l.variables.focusShadow, l.variables.borderColor);
    }), (function(r) {
        var e = r.$append, a = r.$selected;
        return e && a && (0, l.pick)({
            light: (0, n.css)([ "box-shadow:", ",", ",", ",inset -1px 0 0 ", ";" ], D, P, l.variables.focusShadow, l.variables.borderColor),
            dark: (0, n.css)([ "box-shadow:inset 0 1px 0 ", ",", ",inset -1px 0 0 ", ";" ], l.variables.black, l.variables.focusShadow, l.variables.borderColor)
        });
    }), (0, l.pick)({
        light: D,
        dark: (0, n.css)([ "inset 0 -1px 0 ", "" ], l.variables.gray30)
    }), l.variables.focusShadow, l.variables.textDisabledColor, (0, l.pick)({
        light: l.variables.gray96,
        dark: l.variables.gray30
    }), (0, l.pick)({
        light: l.variables.borderLightColor,
        dark: l.variables.gray30
    }), (function(r) {
        var e = r.$selected;
        return e && (0, l.pick)({
            light: (0, n.css)([ "box-shadow:", ";background-color:", ";border-color:", ";" ], P, l.variables.gray92, l.variables.gray80),
            dark: (0, n.css)([ "box-shadow:inset 0 1px 0 ", ";background-color:", ";border-color:", ";" ], l.variables.black, l.variables.gray22, l.variables.gray20)
        });
    }));
    var A = "inset 0 -2px 0 rgba(0, 0, 0, 0.1)";
    var B = "inset 0 2px 0 rgba(0, 0, 0, 0.1)";
    var N = (0, n.css)([ "", ";font-weight:", ";", " &:not([disabled],[aria-disabled='true']){background-color:", ";color:", ";box-shadow:", ";transition:background-image 0.2s,border 0.2s,box-shadow 0.2s,text-decoration 0.2s;", " ", " &:hover{background-color:", ";}&:active{background-color:", ";transition:none;}&:focus{&:active{background-color:", ";transition:none;}}}&:not([disabled]){&:focus{box-shadow:", ",", ";}}&[disabled],&[aria-disabled='true']{color:", ";background-color:", ";cursor:not-allowed;", "}" ], O, l.variables.fontWeightSemiBold, $, w, l.variables.white, A, (function(r) {
        var e = r.$selected;
        return e && (0, n.css)([ "box-shadow:", ";background-color:", ";" ], B, S);
    }), (function(r) {
        var e = r.$prepend;
        return e && (0, n.css)([ "border-left:1px solid ", ";" ], w);
    }), l.variables.brandColorD50, l.variables.brandColorD30, l.variables.brandColorD30, A, l.variables.focusShadow, l.variables.brandColorL30, l.variables.brandColorL10, (function(r) {
        var e = r.$selected;
        return e && (0, n.css)([ "box-shadow:", ";background-color:", ";" ], B, l.variables.brandColorD20);
    }));
    var H = (0, n.css)([ "", ";font-weight:", ";", " &:not([disabled],[aria-disabled='true']){background-color:", ";color:", ";box-shadow:", ";transition:background-image 0.2s,border 0.2s,box-shadow 0.2s,text-decoration 0.2s;", " ", " &:hover{background-color:", ";}&:active{background-color:", ";transition:none;}&:focus{&:active{background-color:", ";transition:none;}}}&:not([disabled]){&:focus{box-shadow:", ",", ";}}&[disabled],&[aria-disabled='true']{color:", ";background-color:", ";cursor:not-allowed;", "}" ], O, l.variables.fontWeightSemiBold, $, l.variables.errorColorD10, l.variables.white, A, (function(r) {
        var e = r.$selected;
        return e && (0, n.css)([ "box-shadow:", ";background-color:", ";" ], B, l.variables.errorColorD30);
    }), (function(r) {
        var e = r.$prepend;
        return e && (0, n.css)([ "border-left:1px solid ", ";" ], l.variables.errorColorD30);
    }), l.variables.errorColorD30, l.variables.errorColorD40, l.variables.errorColorD40, A, l.variables.focusShadow, l.variables.errorColorL10, l.variables.errorColorD10, (function(r) {
        var e = r.$selected;
        return e && (0, n.css)([ "box-shadow:", ";background-color:", ";" ], B, l.variables.errorColorD40);
    }));
    var _ = (0, n.css)([ "", ";color:", ";border:1px solid transparent;&:not([disabled],[aria-disabled='true']){transition:background-color 0.2s,border 0.2s,box-shadow 0.2s,text-decoration 0.2s;", " &:hover{color:", ";background-color:", ";border-color:", ";}", " &:focus{color:", ";}&:active,&[aria-expanded='true']{background-color:", ";transition:none;}}&:not([disabled]){&:focus{box-shadow:", ";}}&[disabled],&[aria-disabled='true']{color:", ";cursor:not-allowed;", "}" ], O, (0, 
    l.pick)({
        light: l.variables.gray45,
        dark: l.variables.white
    }), (function(r) {
        var e = r.$selected;
        return e && (0, n.css)([ "border-color:", ";" ], l.variables.accentColor);
    }), (0, l.pick)({
        light: l.variables.linkColor,
        dark: l.variables.white
    }), l.variables.backgroundColorHover, (0, l.pick)({
        light: l.variables.gray60,
        dark: l.variables.borderColor
    }), (function(r) {
        var e = r.$error;
        return e && (0, n.css)([ "&,&:hover{color:", ";}" ], l.variables.errorColor);
    }), (0, l.pick)({
        light: l.variables.linkColor,
        dark: l.variables.white
    }), (0, l.pick)({
        light: l.variables.gray92,
        dark: l.variables.gray22
    }), l.variables.focusShadow, l.variables.textDisabledColor, (function(r) {
        var e = r.$selected;
        return e && (0, n.css)([ "border-color:", ";" ], l.variables.borderLightColor);
    }));
    // CONCATENATED MODULE: ./src/ButtonSimple/ButtonSimpleStyles.ts
    var T = s()(d()).withConfig({
        displayName: "ButtonSimpleStyles__StyledClickable",
        componentId: "vlarwe-0"
    })([ "", " ", "" ], (function(r) {
        var e = r.to;
        return e ? (0, n.css)([ "", " align-items:center;justify-content:center;" ], l.mixins.reset("flex")) : (0, 
        n.css)([ "", "" ], l.mixins.reset("block"));
    }), (0, l.pick)({
        enterprise: (0, l.pickVariant)("$variant", {
            // the four main styles
            primary: N,
            secondary: j,
            destructive: H,
            subtle: _,
            // same as "secondary"
            default: j,
            // "pill" is an Enterprise-only style
            pill: _,
            // fallbacks for styles unsupported by Enterprise
            toggle: j,
            flat: j
        }),
        prisma: (0, l.pickVariant)("$variant", {
            // the four main styles
            primary: x,
            secondary: C,
            destructive: x,
            subtle: C,
            // same as "secondary"
            default: C,
            // "toggle" and "flat" are Prisma-only styles
            toggle: C,
            flat: x,
            // fallback for "pill" which is not supported by Prisma
            pill: C
        })
    }));
    // CONCATENATED MODULE: ./src/ButtonSimple/ButtonSimple.tsx
    function I(r) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            I = function r(e) {
                return typeof e;
            };
        } else {
            I = function r(e) {
                return e && typeof Symbol === "function" && e.constructor === Symbol && e !== Symbol.prototype ? "symbol" : typeof e;
            };
        }
        return I(r);
    }
    function R() {
        R = Object.assign || function(r) {
            for (var e = 1; e < arguments.length; e++) {
                var a = arguments[e];
                for (var o in a) {
                    if (Object.prototype.hasOwnProperty.call(a, o)) {
                        r[o] = a[o];
                    }
                }
            }
            return r;
        };
        return R.apply(this, arguments);
    }
    function V(r, e) {
        if (r == null) return {};
        var a = W(r, e);
        var o, i;
        if (Object.getOwnPropertySymbols) {
            var t = Object.getOwnPropertySymbols(r);
            for (i = 0; i < t.length; i++) {
                o = t[i];
                if (e.indexOf(o) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(r, o)) continue;
                a[o] = r[o];
            }
        }
        return a;
    }
    function W(r, e) {
        if (r == null) return {};
        var a = {};
        var o = Object.keys(r);
        var i, t;
        for (t = 0; t < o.length; t++) {
            i = o[t];
            if (e.indexOf(i) >= 0) continue;
            a[i] = r[i];
        }
        return a;
    }
    function q(r, e) {
        if (!(r instanceof e)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function E(r, e) {
        for (var a = 0; a < e.length; a++) {
            var o = e[a];
            o.enumerable = o.enumerable || false;
            o.configurable = true;
            if ("value" in o) o.writable = true;
            Object.defineProperty(r, o.key, o);
        }
    }
    function M(r, e, a) {
        if (e) E(r.prototype, e);
        if (a) E(r, a);
        return r;
    }
    function L(r, e) {
        if (typeof e !== "function" && e !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        r.prototype = Object.create(e && e.prototype, {
            constructor: {
                value: r,
                writable: true,
                configurable: true
            }
        });
        if (e) z(r, e);
    }
    function z(r, e) {
        z = Object.setPrototypeOf || function r(e, a) {
            e.__proto__ = a;
            return e;
        };
        return z(r, e);
    }
    function U(r) {
        var e = J();
        return function a() {
            var o = K(r), i;
            if (e) {
                var t = K(this).constructor;
                i = Reflect.construct(o, arguments, t);
            } else {
                i = o.apply(this, arguments);
            }
            return F(this, i);
        };
    }
    function F(r, e) {
        if (e && (I(e) === "object" || typeof e === "function")) {
            return e;
        }
        return G(r);
    }
    function G(r) {
        if (r === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return r;
    }
    function J() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (r) {
            return false;
        }
    }
    function K(r) {
        K = Object.setPrototypeOf ? Object.getPrototypeOf : function r(e) {
            return e.__proto__ || Object.getPrototypeOf(e);
        };
        return K(r);
    }
    function Q(r, e, a) {
        if (e in r) {
            Object.defineProperty(r, e, {
                value: a,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            r[e] = a;
        }
        return r;
    }
    function X(r) {
        return er(r) || rr(r) || Z(r) || Y();
    }
    function Y() {
        throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function Z(r, e) {
        if (!r) return;
        if (typeof r === "string") return ar(r, e);
        var a = Object.prototype.toString.call(r).slice(8, -1);
        if (a === "Object" && r.constructor) a = r.constructor.name;
        if (a === "Map" || a === "Set") return Array.from(r);
        if (a === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(a)) return ar(r, e);
    }
    function rr(r) {
        if (typeof Symbol !== "undefined" && Symbol.iterator in Object(r)) return Array.from(r);
    }
    function er(r) {
        if (Array.isArray(r)) return ar(r);
    }
    function ar(r, e) {
        if (e == null || e > r.length) e = r.length;
        for (var a = 0, o = new Array(e); a < e; a++) {
            o[a] = r[a];
        }
        return o;
    }
    var or = [ "default", "secondary", "primary", "destructive", "subtle" ];
    var ir = [].concat(or, [ "pill" ]);
    var tr = [].concat(or, [ "toggle", "flat" ]);
 // Array.from(new Set... is used to de-duplicate the appearances
        var lr = Array.from(new Set([].concat(X(ir), X(tr))));
    var nr = {
        action: t().string,
        appearance: t().oneOf([ "default", "secondary", "primary", "destructive", "subtle", "pill", "toggle", "flat" ]),
        append: t().bool,
        children: t().node,
        disabled: t().oneOfType([ t().bool, t().oneOf([ "dimmed" ]) ]),
        elementRef: t().oneOfType([ t().func, t().object ]),
        error: t().bool,
        inline: t().bool,
        openInNewContext: t().bool,
        prepend: t().bool,
        selected: t().bool,
        splunkTheme: t().object,
        to: t().string
    };
    var sr = {
        appearance: "default",
        append: false,
        disabled: false,
        error: false,
        inline: true,
        openInNewContext: false,
        prepend: false
    };
    var cr =  function(r) {
        L(a, r);
        var e = U(a);
        function a() {
            var r;
            q(this, a);
            for (var o = arguments.length, i = new Array(o), t = 0; t < o; t++) {
                i[t] = arguments[t];
            }
            r = e.call.apply(e, [ this ].concat(i));
            Q(G(r), "component", null);
            Q(G(r), "handleMount", (function(e) {
                r.component = e;
            }));
            return r;
        }
        M(a, [ {
            key: "focus",
            /**
     * Places focus on the button.
     */
            value: function r() {
                var e;
                (e = this.component) === null || e === void 0 ? void 0 : e.focus();
            }
        }, {
            key: "getVariant",
            value: function r() {
                var e = this.props, a = e.appearance, o = e.error, i = e.splunkTheme;
                var t = a;
                if (i.family === "enterprise") {
                    if (o) {
                        t = "destructive";
                    } else if (!ir.includes(a)) {
                        t = "default";
                    }
                } else if (i.family === "prisma" && !tr.includes(a)) {
                    t = "secondary";
                }
                return t;
            }
        }, {
            key: "render",
            value: function r() {
                var e = this.props, a = e.append, i = e.children, t = e.error, l = e.prepend, n = e.selected, s = V(e, [ "append", "children", "error", "prepend", "selected" ]);
                
                return o().createElement(T, R({
                    "aria-invalid": t,
                    "aria-pressed": n,
                    "data-test": "button-simple",
                    ref: this.handleMount,
                    $variant: this.getVariant(),
                    $append: a,
                    $prepend: l,
                    $selected: n || false,
                    $error: t
                }, s), i);
            }
        } ]);
        return a;
    }(a.Component);
    Q(cr, "defaultProps", sr);
    Q(cr, "propTypes", nr);
    var dr = (0, l.withSplunkTheme)(cr);
    dr.propTypes = cr.propTypes;
    /* harmony default export */    const vr = dr;
    // CONCATENATED MODULE: ./src/ButtonSimple/index.ts
    module.exports = e;
    /******/})();
