// @splunk/react-ui/Button.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var n = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(n, {
                a: n
            });
            /******/            return n;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, n) => {
            /******/ for (var r in n) {
                /******/ if (e.o(n, r) && !e.o(t, r)) {
                    /******/ Object.defineProperty(t, r, {
                        enumerable: true,
                        get: n[r]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        default: () => /* reexport */ U
    });
    // CONCATENATED MODULE: external "react"
    const n = require("react");
    var r =  e.n(n);
    // CONCATENATED MODULE: external "prop-types"
    const i = require("prop-types");
    var a =  e.n(i);
    // CONCATENATED MODULE: external "lodash/isString"
    const o = require("lodash/isString");
    var l =  e.n(o);
    // CONCATENATED MODULE: external "@splunk/react-ui/Dropdown"
    const s = require("@splunk/react-ui/Dropdown");
    // CONCATENATED MODULE: external "@splunk/ui-utils/style"
    const c = require("@splunk/ui-utils/style");
    // CONCATENATED MODULE: external "styled-components"
    const p = require("styled-components");
    var u =  e.n(p);
    // CONCATENATED MODULE: external "@splunk/react-ui/ButtonSimple"
    const f = require("@splunk/react-ui/ButtonSimple");
    var d =  e.n(f);
    // CONCATENATED MODULE: external "@splunk/themes"
    const m = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/Button/ButtonStyles.ts
    var v = u().span.withConfig({
        displayName: "ButtonStyles__StyledIcon",
        componentId: "eqxqs2-0"
    })([ "display:flex;align-items:center;flex-shrink:0;font-size:", ";", "" ], (0, 
    m.pick)({
        enterprise: "inherit",
        prisma: {
            comfortable: "21px",
            compact: "16px"
        }
    }), (0, m.pickVariant)("$isStartIcon", {
        true: {
            enterprise: (0, p.css)([ "padding-right:3px;" ]),
            prisma: (0, p.css)([ "padding-right:8px;" ])
        }
    }));
    var y = u()(d()).withConfig({
        displayName: "ButtonStyles__StyledButtonSimple",
        componentId: "eqxqs2-1"
    })([ "text-decoration:none;white-space:nowrap;flex-grow:1;min-width:0;max-width:100%;line-height:", ";min-height:", ";padding:", ";", " ", " &[data-inline]{:where(:not(a)){display:inline-block;}:where(a){display:inline-flex;}vertical-align:middle;", "}[data-inline] + &&[data-inline]{margin-bottom:0;", "}" ], m.variables.lineHeight, m.variables.inputHeight, (0, 
    m.pick)({
        enterprise: {
            comfortable: "5px 14px",
            compact: "2px 14px"
        },
        prisma: {
            comfortable: function e(t) {
                var n = t.$hasPrismaIcon;
                return n ? "7px 18px" : "9px 18px";
            },
            compact: function e(t) {
                var n = t.$isMenu;
                return n ? "5px 8px 5px 16px" : "5px 16px";
            }
        }
    }), (function(e) {
        var t = e.$inline, n = e.append, r = e.prepend;
        return !t && !n && !r && (0, p.css)([ "width:100%;" ]);
    }), (function(e) {
        var t = e.$isIconOnly;
        return t && (0, p.css)([ "box-sizing:border-box;padding:", ";width:", ";min-width:", ";" ], (0, 
        m.pick)({
            enterprise: {
                comfortable: "0 5px",
                compact: "0 2px"
            },
            prisma: {
                comfortable: (0, p.css)([ "0px calc(", " / 2 - 1px)" ], m.variables.inputHeight),
                compact: (0, p.css)([ "0px calc(", " / 2 - 1px)" ], m.variables.inputHeight)
            }
        }), m.variables.inputHeight, m.variables.inputHeight);
    }), (function(e) {
        var t = e.$isIconOnly;
        return t ? null : (0, p.css)([ "width:auto;" ]);
    }), (function(e) {
        var t = e.prepend;
        return !t && (0, p.css)([ "margin-left:", ";" ], (0, m.pick)({
            enterprise: m.variables.spacingHalf,
            prisma: m.variables.spacingSmall
        }));
    }));
    var b = u().span.withConfig({
        displayName: "ButtonStyles__StyledContentWrapper",
        componentId: "eqxqs2-2"
    })([ "display:flex;flex-direction:row;justify-content:center;align-items:center;flex:1 0 auto;max-width:100%;box-sizing:border-box;padding:", ";" ], (0, 
    m.pick)({
        enterprise: function e(t) {
            var n = t.$error;
            if (n) {
                return "1px";
            }
            return (0, m.pickVariant)("$appearance", {
                default: "0",
                secondary: "0",
                destructive: "1px",
                primary: "1px",
                pill: "0",
                toggle: "0",
                // === default
                flat: "0",
                // === default
                subtle: "0"
            });
        },
        prisma: (0, m.pickVariant)("$appearance", {
            default: 0,
            secondary: "1px 0",
            primary: "1px 0",
            destructive: "1px 0",
            pill: "1px 0",
            // === secondary
            toggle: "0",
            flat: "1px 0",
            subtle: "1px 0"
        })
    }));
    var h = u().span.withConfig({
        displayName: "ButtonStyles__StyledLabel",
        componentId: "eqxqs2-3"
    })([ "text-overflow:ellipsis;overflow:hidden;flex:0 1 auto;&:not(:last-child){padding-right:", ";}", "" ], (0, 
    m.pick)({
        enterprise: "3px",
        prisma: "10px"
    }), (function(e) {
        var t = e.$isMenu;
        return t && (0, p.css)([ "flex:1 1 auto;text-align:left;" ]);
    }));
    var x = u().span.withConfig({
        displayName: "ButtonStyles__StyledCaretWrapper",
        componentId: "eqxqs2-4"
    })([ "display:inline-flex;", "" ], (0, m.pickVariant)("$disabled", {
        false: {
            prisma: (0, m.pickVariant)("$primary", {
                true: (0, p.css)([ "color:", ";" ], m.variables.contentColorInverted),
                false: (0, p.css)([ "color:", ";", ":hover > * > &,", ":active > * > &,{color:", ";}" ], m.variables.contentColorDefault, 
                /* sc-sel */
                y, 
                /* sc-sel */
                y, m.variables.contentColorDefault)
            })
        }
    }));
    // CONCATENATED MODULE: external "@splunk/react-icons/CaretSmallDown"
    const g = require("@splunk/react-icons/CaretSmallDown");
    var w =  e.n(g);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Caret"
    const k = require("@splunk/react-icons/enterprise/Caret");
    var S =  e.n(k);
    // CONCATENATED MODULE: ./src/Button/icons/CaretSmallDown.tsx
    var O = function e() {
        var t = (0, m.useSplunkTheme)(), n = t.isEnterprise, i = t.isCompact;
        var a = i ? "20px" : "24px";
        return n ?  r().createElement(S(), {
            screenReaderText: null,
            hideDefaultTooltip: true,
            size: .5
        }) :  r().createElement(w(), {
            height: a,
            width: a
        });
    };
    /* harmony default export */    const C = O;
    // CONCATENATED MODULE: external "@splunk/react-icons/ArrowSquareTopRightInset"
    const q = require("@splunk/react-icons/ArrowSquareTopRightInset");
    var $ =  e.n(q);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/External"
    const _ = require("@splunk/react-icons/enterprise/External");
    var I =  e.n(_);
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const E = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: ./src/Button/icons/External.tsx
    var P = (0, E._)("Open externally");
    function j() {
        var e = (0, m.useSplunkTheme)(), t = e.isEnterprise;
        return t ?  r().createElement(I(), {
            screenReaderText: P
        }) :  r().createElement($(), {
            "aria-label": P
        },  r().createElement("title", null, P));
    }
    /* harmony default export */    const N = j;
    // CONCATENATED MODULE: ./src/Button/Button.tsx
    function M(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            M = function e(t) {
                return typeof t;
            };
        } else {
            M = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return M(e);
    }
    function T() {
        T = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var n = arguments[t];
                for (var r in n) {
                    if (Object.prototype.hasOwnProperty.call(n, r)) {
                        e[r] = n[r];
                    }
                }
            }
            return e;
        };
        return T.apply(this, arguments);
    }
    function R(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function B(e, t) {
        for (var n = 0; n < t.length; n++) {
            var r = t[n];
            r.enumerable = r.enumerable || false;
            r.configurable = true;
            if ("value" in r) r.writable = true;
            Object.defineProperty(e, r.key, r);
        }
    }
    function D(e, t, n) {
        if (t) B(e.prototype, t);
        if (n) B(e, n);
        return e;
    }
    function H(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) V(e, t);
    }
    function V(e, t) {
        V = Object.setPrototypeOf || function e(t, n) {
            t.__proto__ = n;
            return t;
        };
        return V(e, t);
    }
    function z(e) {
        var t = L();
        return function n() {
            var r = F(e), i;
            if (t) {
                var a = F(this).constructor;
                i = Reflect.construct(r, arguments, a);
            } else {
                i = r.apply(this, arguments);
            }
            return W(this, i);
        };
    }
    function W(e, t) {
        if (t && (M(t) === "object" || typeof t === "function")) {
            return t;
        }
        return A(e);
    }
    function A(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function L() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function F(e) {
        F = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return F(e);
    }
    function G(e, t, n) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: n,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = n;
        }
        return e;
    }
    var J = {
        action: a().string,
        appearance: a().oneOf([ "default", "secondary", "primary", "destructive", "pill", "toggle", "flat", "subtle" ]),
        append: a().bool,
        children: a().node,
        /** @private An additional className to add to the button. */
        classNamePrivate: a().string,
        disabled: a().oneOfType([ a().bool, a().oneOf([ "dimmed" ]) ]),
        elementRef: a().oneOfType([ a().func, a().object ]),
        error: a().bool,
        icon: a().node,
        inline: a().bool,
        isMenu: a().bool,
        label: a().node,
        onClick: a().func,
        openInNewContext: a().bool,
        prepend: a().bool,
        selected: a().bool,
        to: a().string,
        value: a().any
    };
    var K = {
        appearance: "default",
        append: false,
        disabled: false,
        error: false,
        inline: true,
        isMenu: false,
        openInNewContext: false,
        prepend: false
    };
    var Q =  function(e) {
        H(n, e);
        var t = z(n);
        // @docs-props-type ButtonPropsBase
                function n(e) {
            var r;
            R(this, n);
            r = t.call(this, e);
            G(A(r), "component", null);
            G(A(r), "handleClick", (function(e) {
                var t = r.props, n = t.label, i = t.icon, a = t.value, o = t.action, l = t.onClick;
                l === null || l === void 0 ? void 0 : l(e, {
                    label: n,
                    icon: i,
                    value: a,
                    action: o
                });
            }));
            G(A(r), "handleMount", (function(e) {
                r.component = e;
            }));
            if (false) {}
            return r;
        }
        D(n, [ {
            key: "focus",
            /**
     * Places focus on the button.
     */
            value: function e() {
                var t;
                (t = this.component) === null || t === void 0 ? void 0 : t.focus();
            }
        }, {
            key: "render",
            value: function e() {
                var t = this.props, n = t.action, i = t.appearance, a = t.append, o = t.className, s = t.classNamePrivate, p = t.disabled, u = t.error, f = t.icon, d = t.inline, m = t.isMenu, g = t.onClick, w = t.openInNewContext, k = t.prepend, S = t.to, O = t.value;
                var q = this.props, $ = q.children, _ = q.label;
                if (!_ && l()($)) {
                    _ = $;
                    $ = null;
                }
                var I = !!(f && !_ && !m && !$);
                var E = S && w || m || !!f;
                
                return r().createElement(y, T({
                    "aria-haspopup": m || undefined,
                    "aria-invalid": u || undefined,
                    "data-test": "button"
                }, this.props, {
                    className: (0, c.toClassName)(o, s),
                    "data-action": n,
                    "data-inline": d || null,
                    disabled: p,
                    append: a,
                    prepend: k,
                    $inline: d,
                    $hasPrismaIcon: E,
                    $isIconOnly: I,
                    $isMenu: m,
                    value: O,
                    onClick: g ? this.handleClick : undefined,
                    ref: this.handleMount,
                    openInNewContext: w,
                    appearance: i
                }),  r().createElement(b, {
                    $appearance: i,
                    $error: u
                }, f &&  r().createElement(v, {
                    $isStartIcon: !I
                }, f), _ &&  r().createElement(h, {
                    "data-test": "label",
                    $isMenu: m
                }, _), $, m &&  r().createElement(x, {
                    $disabled: p,
                    $primary: i === "primary"
                },  r().createElement(C, null)), w &&  r().createElement(v, {
                    $isStartIcon: false
                },  r().createElement(N, null))));
            }
        } ]);
        return n;
    }(n.Component);
    G(Q, "propTypes", J);
    G(Q, "defaultProps", K);
    G(Q, s.legacyRefMode, true);
    /* harmony default export */    const U = Q;
    // CONCATENATED MODULE: ./src/Button/index.ts
    module.exports = t;
    /******/})();
