// @splunk/react-ui/Calendar.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var r = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(r, {
                a: r
            });
            /******/            return r;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        calendarActiveSelector: () => /* reexport */ qe,
        default: () => /* reexport */ He
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    var n =  e.n(r);
    // CONCATENATED MODULE: external "prop-types"
    const a = require("prop-types");
    var o =  e.n(a);
    // CONCATENATED MODULE: external "lodash/keys"
    const l = require("lodash/keys");
    var i =  e.n(l);
    // CONCATENATED MODULE: external "lodash/omit"
    const u = require("lodash/omit");
    var c =  e.n(u);
    // CONCATENATED MODULE: external "moment"
    const s = require("moment");
    var f =  e.n(s);
    // CONCATENATED MODULE: external "@splunk/ui-utils/id"
    const d = require("@splunk/ui-utils/id");
    // CONCATENATED MODULE: external "lodash/times"
    const p = require("lodash/times");
    var v =  e.n(p);
    // CONCATENATED MODULE: external "@splunk/themes"
    const y = require("@splunk/themes");
    // CONCATENATED MODULE: external "@splunk/ui-utils/keyboard"
    const h = require("@splunk/ui-utils/keyboard");
    // CONCATENATED MODULE: external "@splunk/react-ui/ScreenReaderContent"
    const m = require("@splunk/react-ui/ScreenReaderContent");
    var b =  e.n(m);
    // CONCATENATED MODULE: external "styled-components"
    const Y = require("styled-components");
    var g =  e.n(Y);
    // CONCATENATED MODULE: ./src/Calendar/DayStyles.ts
    /* Element used to prevent vertical shift on firefox when the font-weight is increased with $selected */
    var M = g().div.withConfig({
        displayName: "DayStyles__StyledValue",
        componentId: "sc-1nryhyt-0"
    })([ "left:0;line-height:", ";position:absolute;top:0;width:100%;" ], (0, y.pick)({
        enterprise: "2em",
        prisma: "36px "
    }));
    var D = g().div.withConfig({
        displayName: "DayStyles__StyledInner",
        componentId: "sc-1nryhyt-1"
    })([ "border:", ";border-radius:", ";color:", ";height:", ";position:relative;text-align:center;width:", ";" ], (0, 
    y.pick)({
        enterprise: "1px solid transparent",
        prisma: "none"
    }), (0, y.pick)({
        enterprise: y.variables.borderRadius,
        prisma: "50%"
    }), (0, y.pick)({
        enterprise: {
            light: y.variables.gray45,
            dark: y.variables.gray98
        },
        prisma: y.variables.contentColorActive
    }), (0, y.pick)({
        enterprise: "2em",
        prisma: "36px "
    }), (0, y.pick)({
        enterprise: "calc(2em - 2px)",
        prisma: "36px"
    }));
    var k = g().td.withConfig({
        displayName: "DayStyles__StyledCell",
        componentId: "sc-1nryhyt-2"
    })([ "", ";cursor:pointer;", ";", ";" ], y.mixins.reset("table-cell"), (0, y.pickVariant)("$today", {
        true: {
            enterprise: (0, Y.css)([ "& > ", "{background-color:", ";}&:focus{& > ", "{box-shadow:", ";}}" ], 
            /* sc-sel */
            D, (0, y.pick)({
                light: y.variables.gray92,
                dark: y.variables.gray45
            }), 
            /* sc-sel */
            D, y.variables.focusShadow),
            prisma: (0, Y.css)([ "& > ", "{box-shadow:inset 0 0 0 1px ", ";}&:focus{& > ", "{box-shadow:inset 0 0 0 1px ", ",", ";}}" ], 
            /* sc-sel */
            D, y.variables.interactiveColorBorder, 
            /* sc-sel */
            D, y.variables.interactiveColorBorder, y.variables.focusShadow)
        },
        false: (0, Y.css)([ "&:focus{& > ", "{box-shadow:", ";}}" ], 
        /* sc-sel */
        D, y.variables.focusShadow)
    }), (0, y.pickVariant)("$selected", {
        true: (0, y.pick)({
            enterprise: (0, Y.css)([ "& > ", "{border-color:", ";}" ], 
            /* sc-sel */
            D, y.variables.linkColor),
            prisma: (0, Y.css)([ "& > ", "{background-color:", ";font-weight:", ";}" ], 
            /* sc-sel */
            D, y.variables.interactiveColorOverlaySelected, y.variables.fontWeightBold)
        }),
        false: (0, y.pick)({
            enterprise: (0, Y.css)([ "&:hover{& > ", "{background-color:", ";color:", ";}}" ], 
            /* sc-sel */
            D, y.variables.backgroundColorHover, y.variables.linkColor),
            prisma: (0, Y.css)([ "&:hover{& > ", "{background-color:", ";}}" ], 
            /* sc-sel */
            D, y.variables.interactiveColorOverlayHover)
        })
    }));
    // CONCATENATED MODULE: ./src/Calendar/Day.tsx
    function w() {
        w = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return w.apply(this, arguments);
    }
    function O(e, t) {
        if (e == null) return {};
        var r = C(e, t);
        var n, a;
        if (Object.getOwnPropertySymbols) {
            var o = Object.getOwnPropertySymbols(e);
            for (a = 0; a < o.length; a++) {
                n = o[a];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function C(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var a, o;
        for (o = 0; o < n.length; o++) {
            a = n[o];
            if (t.indexOf(a) >= 0) continue;
            r[a] = e[a];
        }
        return r;
    }
    var S = {
        active: o().bool,
        locale: o().string,
        onClick: o().func,
        onKeydown: o().func,
        selected: o().bool,
        today: o().bool,
        value: o().string
    };
    function x(e) {
        var t = e.active, a = t === void 0 ? false : t, o = e.locale, l = o === void 0 ? "en_US" : o, i = e.onClick, u = e.onKeyDown, c = e.selected, s = c === void 0 ? false : c, d = e.today, p = d === void 0 ? false : d, v = e.value, y = O(e, [ "active", "locale", "onClick", "onKeyDown", "selected", "today", "value" ]);
        var h = (0, r.useCallback)((function(e) {
            e.preventDefault();
            i === null || i === void 0 ? void 0 : i(e, {
                value: v
            });
        }), [ i, v ]);
        var m = (0, r.useCallback)((function(e) {
            e.preventDefault();
            u === null || u === void 0 ? void 0 : u(e, {
                value: v
            });
        }), [ u, v ]);
        var b = v && f()(v, "YYYY-MM-DD", l).format("D");
        
        return n().createElement(k, w({
            tabIndex: a ? 0 : -1,
            "aria-selected": s,
            "data-test-day": b,
            "data-test": "day-of-month",
            "data-active-day": a,
            "data-is-today": p,
            onClick: h,
            onKeyDown: m,
            $today: p,
            $selected: s
        }, y),  n().createElement(D, null,  n().createElement(M, null, b)));
    }
    x.propTypes = S;
    /* harmony default export */    const j = x;
    // CONCATENATED MODULE: ./src/Calendar/DateTableStyles.ts
    var _ = g().table.withConfig({
        displayName: "DateTableStyles__Styled",
        componentId: "sc-1lbrjq0-0"
    })([ "", ";table-layout:fixed;border-spacing:", ";width:", ";margin:", ";" ], y.mixins.reset("table"), (0, 
    y.pick)({
        enterprise: "3px",
        prisma: y.variables.spacingSmall
    }), (0, y.pick)({
        enterprise: "100%",
        prisma: "calc(100% + 16px)"
    }), (0, y.pick)({
        enterprise: "0",
        prisma: "-8px"
    }));
    var T = g().th.withConfig({
        displayName: "DateTableStyles__StyledTableHeader",
        componentId: "sc-1lbrjq0-1"
    })([ "", ";text-align:center;padding-bottom:", ";color:", ";font-size:", ";" ], y.mixins.reset("table-cell"), (0, 
    y.pick)({
        enterprise: "2px",
        prisma: y.variables.spacingXSmall
    }), y.variables.contentColorMuted, (0, y.pick)({
        enterprise: y.variables.fontSizeSmall,
        prisma: y.variables.fontSize
    }));
    // CONCATENATED MODULE: ./src/Calendar/DateTable.tsx
    function E() {
        E = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return E.apply(this, arguments);
    }
    function P(e, t) {
        if (e == null) return {};
        var r = I(e, t);
        var n, a;
        if (Object.getOwnPropertySymbols) {
            var o = Object.getOwnPropertySymbols(e);
            for (a = 0; a < o.length; a++) {
                n = o[a];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function I(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var a, o;
        for (o = 0; o < n.length; o++) {
            a = n[o];
            if (t.indexOf(a) >= 0) continue;
            r[a] = e[a];
        }
        return r;
    }
    var V = {
        displayValue: o().string.isRequired,
        highlightToday: o().bool,
        labelledBy: o().string,
        locale: o().string,
        onChange: o().func,
        onInternalChange: o().func,
        selectedValue: o().string.isRequired,
        todayValue: o().string
    };
    function R(e) {
        return v()(e, (function(e) {
            
            return n().createElement("td", {
                key: "".concat(e)
            });
        }));
    }
    function q(e) {
        var t = e.displayValue, r = e.endDate, a = e.highlightToday, o = e.locale, l = e.onChange, i = e.onKeyDown, u = e.selectedDate, c = e.selectedValue, s = e.startDate, d = e.todayValue;
        var p = f()(s).locale(o).startOf("month");
        var v = [];
        while (p.isSameOrBefore(r)) {
            var y = p.format("YYYY-MM-DD");
            var h = y === d;
            var m = y === c;
            var b = y === t;
            v.push( n().createElement(j, {
                key: y,
                locale: o,
                onClick: l,
                onKeyDown: i,
                active: u ? b : h,
                today: a && h,
                value: y,
                selected: m
            }));
            p.add(1, "day");
        }
        return v;
    }
    function B(e) {
        return e.reduce((function(e, t, r) {
            var n = Math.floor(r / 7);
            e[n].push(t);
            return e;
        }), v()(7, (function() {
            return [];
        }))).map((function(e, t) {
            
            // eslint-disable-next-line react/no-array-index-key
            return n().createElement("tr", {
                key: "$week-".concat(t)
            }, e);
        }));
    }
    function F(e, t) {
        
        return n().createElement("thead", null,  n().createElement("tr", null, v()(7, (function(r) {
            var a = f()().locale(e).weekday(r).format(t);
            var o = f()().locale(e).weekday(r).format("dddd");
            
            return n().createElement(T, {
                key: a
            },  n().createElement(b(), null, o),  n().createElement("span", {
                "aria-hidden": true
            }, a));
        }))));
    }
    function z(e) {
        var t = e.displayValue, r = e.highlightToday, a = r === void 0 ? false : r, o = e.labelledBy, l = e.locale, i = l === void 0 ? "en_US" : l, u = e.onChange, c = e.onInternalChange, s = e.selectedValue, d = e.todayValue, p = P(e, [ "displayValue", "highlightToday", "labelledBy", "locale", "onChange", "onInternalChange", "selectedValue", "todayValue" ]);
        var v = (0, y.useSplunkTheme)(), m = v.family;
        var b = m === "prisma" ? "dd" : "ddd";
        var Y = f()(t, "YYYY-MM-DD", i);
        var g = f()(s, "YYYY-MM-DD", i);
        var M = f()(Y).startOf("month");
        var D = f()(Y).endOf("month").startOf("day");
        var k = function e(t, r) {
            t.preventDefault();
            u === null || u === void 0 ? void 0 : u(t, r);
        };
        var w = function e(t, r) {
            var n = r.value;
            t.preventDefault();
            c === null || c === void 0 ? void 0 : c(t, {
                value: f()(n, "YYYY-MM-DD").add(-1, "d").format("YYYY-MM-DD")
            });
        };
        var O = function e(t, r) {
            var n = r.value;
            t.preventDefault();
            c === null || c === void 0 ? void 0 : c(t, {
                value: f()(n, "YYYY-MM-DD").add(1, "d").format("YYYY-MM-DD")
            });
        };
        var C = function e(t, r) {
            var n = r.value;
            t.preventDefault();
            c === null || c === void 0 ? void 0 : c(t, {
                value: f()(n, "YYYY-MM-DD").add(-1, "w").format("YYYY-MM-DD")
            });
        };
        var S = function e(t, r) {
            var n = r.value;
            t.preventDefault();
            c === null || c === void 0 ? void 0 : c(t, {
                value: f()(n, "YYYY-MM-DD").add(1, "w").format("YYYY-MM-DD")
            });
        };
        var x = function e(t, r) {
            var n = r.value;
            t.preventDefault();
            c === null || c === void 0 ? void 0 : c(t, {
                value: f()(n, "YYYY-MM-DD").add(-1, "M").format("YYYY-MM-DD")
            });
        };
        var j = function e(t, r) {
            var n = r.value;
            t.preventDefault();
            c === null || c === void 0 ? void 0 : c(t, {
                value: f()(n, "YYYY-MM-DD").add(1, "M").format("YYYY-MM-DD")
            });
        };
        var T = function e(t, r) {
            var n = r.value;
            t.preventDefault();
            c === null || c === void 0 ? void 0 : c(t, {
                value: f()(n, "YYYY-MM-DD").add(-1, "y").format("YYYY-MM-DD")
            });
        };
        var I = function e(t, r) {
            var n = r.value;
            t.preventDefault();
            c === null || c === void 0 ? void 0 : c(t, {
                value: f()(n, "YYYY-MM-DD").add(1, "y").format("YYYY-MM-DD")
            });
        };
        var V = function e(t, r) {
            var n = r.value;
            t.preventDefault();
            c === null || c === void 0 ? void 0 : c(t, {
                value: f()(n, "YYYY-MM-DD").startOf("week").format("YYYY-MM-DD")
            });
        };
        var z = function e(t, r) {
            var n = r.value;
            t.preventDefault();
            c === null || c === void 0 ? void 0 : c(t, {
                value: f()(n, "YYYY-MM-DD").endOf("week").format("YYYY-MM-DD")
            });
        };
        var H = function e(t, r) {
            var n = (0, h.keycode)(t.nativeEvent);
            switch (n) {
              case "enter":
              case "space":
                k(t, r);
                break;

              case "up":
                C(t, r);
                break;

              case "down":
                S(t, r);
                break;

              case "left":
                w(t, r);
                break;

              case "right":
                O(t, r);
                break;

              case "page up":
                if (t.shiftKey) {
                    T(t, r);
                } else {
                    x(t, r);
                }
                break;

              case "page down":
                if (t.shiftKey) {
                    I(t, r);
                } else {
                    j(t, r);
                }
                break;

              case "home":
                V(t, r);
                break;

              case "end":
                z(t, r);
                break;

              default:
 // do nothing
                        }
        };
        var N = R(M.format("e")).concat(q({
            displayValue: t,
            endDate: D,
            highlightToday: a,
            locale: i,
            onChange: u,
            onKeyDown: H,
            selectedDate: g,
            selectedValue: s,
            startDate: M,
            todayValue: d
        })).concat(R(6 - D.format("e")));
        
        return n().createElement(_, E({
            role: "grid",
            "aria-labelledby": o
        }, p), F(i, b),  n().createElement("tbody", null, B(N)));
    }
    z.propTypes = V;
    /* harmony default export */    const H = z;
    // CONCATENATED MODULE: external "@splunk/react-icons/ChevronLeft"
    const N = require("@splunk/react-icons/ChevronLeft");
    var K =  e.n(N);
    // CONCATENATED MODULE: external "@splunk/react-icons/ChevronRight"
    const A = require("@splunk/react-icons/ChevronRight");
    var U =  e.n(A);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/ChevronLeft"
    const $ = require("@splunk/react-icons/enterprise/ChevronLeft");
    var L =  e.n($);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/ChevronRight"
    const W = require("@splunk/react-icons/enterprise/ChevronRight");
    var X =  e.n(W);
    // CONCATENATED MODULE: external "@splunk/react-ui/Button"
    const G = require("@splunk/react-ui/Button");
    var J =  e.n(G);
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const Q = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: ./src/Calendar/MonthHeaderStyles.ts
    var Z = g().div.withConfig({
        displayName: "MonthHeaderStyles__StyledBox",
        componentId: "sc-18bmbh7-0"
    })([ "display:flex;position:relative;line-height:", ";padding:0 2px;margin-bottom:10px;" ], y.variables.inputHeight);
    var ee = g().span.withConfig({
        displayName: "MonthHeaderStyles__StyledHeading",
        componentId: "sc-18bmbh7-1"
    })([ "display:block;text-align:center;text-transform:", ";flex:1 0 1px;", ";" ], (0, 
    y.pick)({
        enterprise: "capitalize",
        prisma: "uppercase"
    }), (0, y.pick)({
        prisma: (0, Y.css)([ "font-size:20px;font-weight:bold;color:", ";" ], y.variables.contentColorActive)
    }));
    // CONCATENATED MODULE: ./src/Calendar/MonthHeader.tsx
    function te(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            te = function e(t) {
                return typeof t;
            };
        } else {
            te = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return te(e);
    }
    function re(e, t) {
        if (e == null) return {};
        var r = ne(e, t);
        var n, a;
        if (Object.getOwnPropertySymbols) {
            var o = Object.getOwnPropertySymbols(e);
            for (a = 0; a < o.length; a++) {
                n = o[a];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function ne(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var a, o;
        for (o = 0; o < n.length; o++) {
            a = n[o];
            if (t.indexOf(a) >= 0) continue;
            r[a] = e[a];
        }
        return r;
    }
    function ae(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function oe(e, t) {
        for (var r = 0; r < t.length; r++) {
            var n = t[r];
            n.enumerable = n.enumerable || false;
            n.configurable = true;
            if ("value" in n) n.writable = true;
            Object.defineProperty(e, n.key, n);
        }
    }
    function le(e, t, r) {
        if (t) oe(e.prototype, t);
        if (r) oe(e, r);
        return e;
    }
    function ie(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) ue(e, t);
    }
    function ue(e, t) {
        ue = Object.setPrototypeOf || function e(t, r) {
            t.__proto__ = r;
            return t;
        };
        return ue(e, t);
    }
    function ce(e) {
        var t = de();
        return function r() {
            var n = pe(e), a;
            if (t) {
                var o = pe(this).constructor;
                a = Reflect.construct(n, arguments, o);
            } else {
                a = n.apply(this, arguments);
            }
            return se(this, a);
        };
    }
    function se(e, t) {
        if (t && (te(t) === "object" || typeof t === "function")) {
            return t;
        }
        return fe(e);
    }
    function fe(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function de() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function pe(e) {
        pe = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return pe(e);
    }
    function ve(e, t, r) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: r,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = r;
        }
        return e;
    }
    var ye = {
        headingId: o().string,
        locale: o().string,
        onChange: o().func,
        /** @private */
        splunkTheme: o().object,
        value: o().string.isRequired
    };
    var he = {
        locale: "en_US"
    };
    var me =  function(e) {
        ie(r, e);
        var t = ce(r);
        function r() {
            var e;
            ae(this, r);
            for (var n = arguments.length, a = new Array(n), o = 0; o < n; o++) {
                a[o] = arguments[o];
            }
            e = t.call.apply(t, [ this ].concat(a));
            ve(fe(e), "handlePrevMonthClick", (function(t) {
                e.handleChange(t, -1);
            }));
            ve(fe(e), "handleNextMonthClick", (function(t) {
                e.handleChange(t, 1);
            }));
            return e;
        }
        le(r, [ {
            key: "handleChange",
            value: function e(t, r) {
                var n, a;
                t.stopPropagation();
                var o = f()(this.props.value, "YYYY-MM-DD").add(r, "M").format("YYYY-MM-DD");
                (n = (a = this.props).onChange) === null || n === void 0 ? void 0 : n.call(a, t, {
                    value: o
                });
            }
        }, {
            key: "render",
            value: function e() {
                var t = this.props, a = t.headingId, o = t.locale, l = t.value, i = t.splunkTheme, u = re(t, [ "headingId", "locale", "value", "splunkTheme" ]);
                var s = i.isEnterprise, d = i.isPrisma;
                var p = f()(l, "YYYY-MM-DD", o);
                var v = p.format(r.getMonthYearFormat(o, d));
                var h = {
                    flex: "0 0 auto"
                };
                var m = (0, Q._)("Previous month");
                var b = (0, Q._)("Next month");
                
                return n().createElement(y.SplunkThemeProvider, {
                    density: d ? "comfortable" : i.density
                },  n().createElement(Z, c()(u, "onChange"),  n().createElement(J(), {
                    appearance: "pill",
                    icon: s ?  n().createElement(L(), {
                        hideDefaultTooltip: true,
                        screenReaderText: m,
                        size: 1
                    }) :  n().createElement(K(), {
                        "aria-label": m,
                        height: "24",
                        width: "24"
                    }),
                    "data-test": "previous-month",
                    onClick: this.handlePrevMonthClick,
                    style: h
                }),  n().createElement(ee, {
                    id: a,
                    "data-test": "header-label"
                }, v),  n().createElement(J(), {
                    appearance: "pill",
                    icon: s ?  n().createElement(X(), {
                        hideDefaultTooltip: true,
                        screenReaderText: b,
                        size: 1
                    }) :  n().createElement(U(), {
                        "aria-label": b,
                        height: "24",
                        width: "24"
                    }),
                    "data-test": "next-month",
                    onClick: this.handleNextMonthClick,
                    style: h
                })));
            }
        } ], [ {
            key: "getMonthYearFormat",
            value: function e(t) {
                var n = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
                var a = t.toLowerCase().replace("_", "-");
                return n ? r.monthYearFormatsShortForm[a] || // full locale match
                r.monthYearFormatsShortForm[a.substr(0, 2)] || // language match
                r.monthYearFormatsShortForm["default"] : r.monthYearFormats[a] || // full locale match
                r.monthYearFormats[a.substr(0, 2)] || // language match
                r.monthYearFormats["default"];
 // default
                        }
        } ]);
        return r;
    }(r.Component);
    ve(me, "propTypes", ye);
    ve(me, "defaultProps", he);
    ve(me, "monthYearFormats", {
        default: "MMMM YYYY",
        ja: "YYYY年MMM",
        ko: "YYYY년 MMMM",
        "zh-cn": "YYYY年MMM",
        "zh-tw": "YYYY年MMM"
    });
    ve(me, "monthYearFormatsShortForm", {
        default: "MMM YYYY",
        ja: "YYYY年 MMM",
        ko: "YYYY년 MMMM",
        "zh-cn": "YYYY年 MMM",
        "zh-tw": "YYYY年 MMM"
    });
    var be = (0, y.withSplunkTheme)(me);
    be.propTypes = me.propTypes;
    /* harmony default export */    const Ye = be;
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const ge = require("@splunk/react-ui/Box");
    var Me =  e.n(ge);
    // CONCATENATED MODULE: ./src/Calendar/CalendarStyles.ts
    var De = g()(Me()).withConfig({
        displayName: "CalendarStyles__StyledBox",
        componentId: "sc-18925wt-0"
    })([ "width:", ";padding:", ";" ], (0, y.pick)({
        enterprise: "220px",
        prisma: "332px"
    }), (0, y.pick)({
        enterprise: "8px",
        prisma: y.variables.spacingLarge
    }));
    // CONCATENATED MODULE: ./src/utils/updateReactRef.ts
    /**
 * Updates a React ref. Callback refs and object refs (from `createRef` and `useRef`) are supported.
 *
 * @param ref - The React callback or object ref. Can be `null` or `undefined`.
 * @param current - The new value of the ref.
 */
    function ke(e, t) {
        if (e) {
            if (typeof e === "function") {
                e(t);
            } else {
                // the public signature of this util uses React.Ref<T> to mirror the way React types refs.
                // the intention here is to signal "we will take care of setting 'current', not you".
                e.current = t;
 // eslint-disable-line no-param-reassign
                        }
        }
    }
    // CONCATENATED MODULE: ./src/Calendar/Calendar.tsx
    function we(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            we = function e(t) {
                return typeof t;
            };
        } else {
            we = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return we(e);
    }
    function Oe() {
        Oe = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return Oe.apply(this, arguments);
    }
    function Ce(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function Se(e, t) {
        for (var r = 0; r < t.length; r++) {
            var n = t[r];
            n.enumerable = n.enumerable || false;
            n.configurable = true;
            if ("value" in n) n.writable = true;
            Object.defineProperty(e, n.key, n);
        }
    }
    function xe(e, t, r) {
        if (t) Se(e.prototype, t);
        if (r) Se(e, r);
        return e;
    }
    function je(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) _e(e, t);
    }
    function _e(e, t) {
        _e = Object.setPrototypeOf || function e(t, r) {
            t.__proto__ = r;
            return t;
        };
        return _e(e, t);
    }
    function Te(e) {
        var t = Ie();
        return function r() {
            var n = Ve(e), a;
            if (t) {
                var o = Ve(this).constructor;
                a = Reflect.construct(n, arguments, o);
            } else {
                a = n.apply(this, arguments);
            }
            return Ee(this, a);
        };
    }
    function Ee(e, t) {
        if (t && (we(t) === "object" || typeof t === "function")) {
            return t;
        }
        return Pe(e);
    }
    function Pe(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function Ie() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function Ve(e) {
        Ve = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return Ve(e);
    }
    function Re(e, t, r) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: r,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = r;
        }
        return e;
    }
    var qe = '[data-active-day="true"]';
    var Be = {
        describedBy: o().string,
        highlightToday: o().bool,
        labelledBy: o().string,
        locale: o().string,
        onChange: o().func,
        /** @private. */
        required: o().bool,
        value: o().string,
        elementRef: o().oneOfType([ o().func, o().object ])
    };
    var Fe = {
        locale: "en_US",
        value: f()().format("YYYY-MM-DD")
    };
    var ze =  function(e) {
        je(r, e);
        var t = Te(r);
        function r(e) {
            var a;
            Ce(this, r);
            a = t.call(this, e);
            Re(Pe(a), "monthHeaderId", void 0);
            Re(Pe(a), "todayTimerId", void 0);
            Re(Pe(a), "calendarRef", void 0);
            Re(Pe(a), "handleMonthChange", (function(e, t) {
                var r = t.value;
                a.setState({
                    internalValue: r,
                    source: "month"
                });
            }));
            Re(Pe(a), "handleDayChange", (function(e, t) {
                var r = t.value;
                a.setState({
                    internalValue: r,
                    source: "day"
                });
            }));
            Re(Pe(a), "handleCalendarMount", (function(e) {
                ke(a.calendarRef, e);
                ke(a.props.elementRef, e);
            }));
            a.state = {
                internalValue: e.value,
                todayValue: a.getToday(),
                source: null
            };
            a.monthHeaderId = (0, d.createDOMID)("header");
            a.calendarRef =  n().createRef();
            if (false) {}
            return a;
        }
        xe(r, [ {
            key: "componentDidMount",
            value: function e() {
                var t = this;
                this.todayTimerId = setInterval((function() {
                    var e = t.getToday();
                    if (e !== t.state.todayValue) {
                        t.setState({
                            todayValue: e
                        });
                    }
                }), 1e3);
            }
        }, {
            key: "componentWillUnmount",
            value: function e() {
                if (this.todayTimerId !== undefined) {
                    clearInterval(this.todayTimerId);
                }
            }
        }, {
            key: "componentDidUpdate",
            value: function e(t, r) {
                if (false) {}
                if (this.state.internalValue !== r.internalValue && this.state.source === "day") {
                    this.focusActiveDay();
                }
            }
            /**
     * Place focus on the active day.
     */        }, {
            key: "focusActiveDay",
            value: function e() {
                var t, r;
                var n = (t = this.calendarRef) === null || t === void 0 ? void 0 : (r = t.current) === null || r === void 0 ? void 0 : r.querySelector(qe);
                n === null || n === void 0 ? void 0 : n.focus();
            }
        }, {
            key: "getToday",
            value: function e() {
                var t = f()().locale(this.props.locale).format("YYYY-MM-DD");
                return t;
            }
        }, {
            key: "render",
            value: function e() {
                var t = this.props, a = t.describedBy, o = t.highlightToday, l = t.labelledBy, u = t.locale, s = t.onChange, f = t.required, d = t.value;
                var p = this.state, v = p.internalValue, y = p.todayValue;
                
                return n().createElement(De, Oe({
                    inline: true,
                    "aria-labelledby": l,
                    "aria-describedby": a,
                    "aria-required": f,
                    "data-test": "calendar"
                }, c()(this.props, i()(r.propTypes)), {
                    elementRef: this.handleCalendarMount
                }),  n().createElement(Ye, {
                    headingId: this.monthHeaderId,
                    locale: u,
                    onChange: this.handleMonthChange,
                    value: v
                }),  n().createElement(H, {
                    displayValue: v,
                    highlightToday: o,
                    labelledBy: this.monthHeaderId,
                    locale: u,
                    onChange: s,
                    onInternalChange: this.handleDayChange,
                    selectedValue: d,
                    todayValue: y
                }));
            }
        } ], [ {
            key: "validateValue",
            value: function e(t) {
                if (!f()(t, "YYYY-MM-DD").isValid()) {
                    return new Error("Invalid value prop ".concat(t, " passed to Calendar. value must be a string date formatted 'YYYY-MM-DD'."));
                }
                return null;
            }
        } ]);
        return r;
    }(r.Component);
    Re(ze, "propTypes", Be);
    Re(ze, "defaultProps", Fe);
    /* harmony default export */    const He = ze;
    // CONCATENATED MODULE: ./src/Calendar/index.ts
    module.exports = t;
    /******/})();
