// @splunk/react-ui/CardLayout.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var t = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(t, {
                a: t
            });
            /******/            return t;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, t) => {
            /******/ for (var n in t) {
                /******/ if (e.o(t, n) && !e.o(r, n)) {
                    /******/ Object.defineProperty(r, n, {
                        enumerable: true,
                        get: t[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        default: () => /* reexport */ m
    });
    // CONCATENATED MODULE: external "react"
    const t = require("react");
    var n =  e.n(t);
    // CONCATENATED MODULE: external "prop-types"
    const i = require("prop-types");
    var a =  e.n(i);
    // CONCATENATED MODULE: external "styled-components"
    const o = require("styled-components");
    var c =  e.n(o);
    // CONCATENATED MODULE: external "@splunk/themes"
    const l = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/CardLayout/CardLayoutStyles.ts
    var u = c().div.withConfig({
        displayName: "CardLayoutStyles__Styled",
        componentId: "sc-18kz7zm-0"
    })([ "", ";", " ", "" ], l.mixins.reset("flex"), (function(e) {
        var r = e.$wrapCards;
        return r && (0, o.css)([ "flex-wrap:wrap;" ]);
    }), (0, l.pickVariant)("$alignCards", {
        center: (0, o.css)([ "justify-content:center;" ]),
        right: (0, o.css)([ "justify-content:flex-end;" ])
    }));
    // CONCATENATED MODULE: ./src/CardLayout/CardLayout.tsx
    function d() {
        d = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return d.apply(this, arguments);
    }
    function f(e, r) {
        var t = Object.keys(e);
        if (Object.getOwnPropertySymbols) {
            var n = Object.getOwnPropertySymbols(e);
            if (r) n = n.filter((function(r) {
                return Object.getOwnPropertyDescriptor(e, r).enumerable;
            }));
            t.push.apply(t, n);
        }
        return t;
    }
    function s(e) {
        for (var r = 1; r < arguments.length; r++) {
            var t = arguments[r] != null ? arguments[r] : {};
            if (r % 2) {
                f(Object(t), true).forEach((function(r) {
                    p(e, r, t[r]);
                }));
            } else if (Object.getOwnPropertyDescriptors) {
                Object.defineProperties(e, Object.getOwnPropertyDescriptors(t));
            } else {
                f(Object(t)).forEach((function(r) {
                    Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r));
                }));
            }
        }
        return e;
    }
    function p(e, r, t) {
        if (r in e) {
            Object.defineProperty(e, r, {
                value: t,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[r] = t;
        }
        return e;
    }
    function y(e, r) {
        if (e == null) return {};
        var t = b(e, r);
        var n, i;
        if (Object.getOwnPropertySymbols) {
            var a = Object.getOwnPropertySymbols(e);
            for (i = 0; i < a.length; i++) {
                n = a[i];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function b(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var i, a;
        for (a = 0; a < n.length; a++) {
            i = n[a];
            if (r.indexOf(i) >= 0) continue;
            t[i] = e[i];
        }
        return t;
    }
    var O = {
        alignCards: a().oneOf([ "left", "center", "right" ]),
        cardMaxWidth: a().oneOfType([ a().number, a().string ]),
        cardMinWidth: a().oneOfType([ a().number, a().string ]),
        cardWidth: a().oneOfType([ a().number, a().string ]),
        children: a().node,
        elementRef: a().oneOfType([ a().func, a().object ]),
        gutterSize: a().number,
        wrapCards: a().bool
    };
    function g(e) {
        var r = e.alignCards, i = r === void 0 ? "left" : r, a = e.cardMaxWidth, o = e.cardMinWidth, c = e.cardWidth, l = e.children, f = e.elementRef, p = e.gutterSize, b = p === void 0 ? 10 : p, O = e.style, g = e.wrapCards, m = g === void 0 ? true : g, v = y(e, [ "alignCards", "cardMaxWidth", "cardMinWidth", "cardWidth", "children", "elementRef", "gutterSize", "style", "wrapCards" ]);
        var h = t.Children.toArray(l).filter(t.isValidElement).map((function(e) {
            
            return (0, t.cloneElement)(e, {
                minWidth: c || o || undefined,
                maxWidth: c || a || undefined,
                margin: b / 2
            });
        }));
        
        return n().createElement(u, d({
            "data-test": "card-layout",
            $alignCards: i,
            $wrapCards: m,
            ref: f,
            style: s({
                padding: b / 2
            }, O)
        }, v), h);
    }
    g.propTypes = O;
    /* harmony default export */    const m = g;
    // CONCATENATED MODULE: ./src/CardLayout/index.ts
    module.exports = r;
    /******/})();
