// @splunk/react-ui/Card.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var t = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(t, {
                a: t
            });
            /******/            return t;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, t) => {
            /******/ for (var n in t) {
                /******/ if (e.o(t, n) && !e.o(r, n)) {
                    /******/ Object.defineProperty(r, n, {
                        enumerable: true,
                        get: t[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        Body: () => /* reexport */ m,
        Footer: () => /* reexport */ k,
        Header: () => /* reexport */ Q,
        default: () => /* reexport */ Be
    });
    // CONCATENATED MODULE: external "react"
    const t = require("react");
    var n =  e.n(t);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var i =  e.n(o);
    // CONCATENATED MODULE: external "styled-components"
    const a = require("styled-components");
    var l =  e.n(a);
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const c = require("@splunk/react-ui/Box");
    var s =  e.n(c);
    // CONCATENATED MODULE: external "@splunk/themes"
    const u = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/Card/BodyStyles.ts
    var p = l()(s()).withConfig({
        displayName: "BodyStyles__StyledBox",
        componentId: "sc-1130udp-0"
    })([ "flex:1 1 auto;overflow:auto;height:100%;", ";", "" ], (function(e) {
        var r = e.$inset;
        return r ? (0, a.css)([ "", "" ], (0, u.pick)({
            enterprise: (0, a.css)([ "padding:", ";&:not(:first-child){padding-top:0;}" ], u.variables.spacing),
            prisma: {
                comfortable: (0, a.css)([ "padding:12px 24px;&:first-child{padding-top:28px;}&:last-child{padding-bottom:28px;}" ]),
                compact: (0, a.css)([ "padding:8px 24px;&:first-child{padding-top:20px;}&:last-child{padding-bottom:20px;}" ])
            }
        })) : (0, a.css)([ "padding:0;" ]);
    }), (0, u.pick)({
        prisma: {
            comfortable: (0, a.css)([ "border-radius:", ";color:", ";" ], u.variables.borderRadius, u.variables.contentColorMuted),
            compact: (0, a.css)([ "border-radius:", ";color:", ";" ], u.variables.borderRadius, u.variables.contentColorMuted)
        }
    }));
    // CONCATENATED MODULE: ./src/Card/Body.tsx
    function d() {
        d = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return d.apply(this, arguments);
    }
    function f(e, r) {
        if (e == null) return {};
        var t = b(e, r);
        var n, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                n = i[o];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function b(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var o, i;
        for (i = 0; i < n.length; i++) {
            o = n[i];
            if (r.indexOf(o) >= 0) continue;
            t[o] = e[o];
        }
        return t;
    }
    var v = {
        children: i().node,
        elementRef: i().oneOfType([ i().func, i().object ]),
        inset: i().bool
    };
    /**
 * A styled container for `Card` body content.
 */    function y(e) {
        var r = e.children, t = e.inset, o = t === void 0 ? true : t, i = f(e, [ "children", "inset" ]);
        // @docs-props-type BodyPropsBase
                
        return n().createElement(p, d({
            "data-test": "body",
            $inset: o
        }, i), r);
    }
    y.propTypes = v;
    /* harmony default export */    const m = y;
    // CONCATENATED MODULE: ./src/Card/FooterStyles.ts
    var h = l()(s()).withConfig({
        displayName: "FooterStyles__StyledBox",
        componentId: "sc-1yu3r4s-0"
    })([ "", ";text-align:right;color:", ";", "" ], (0, u.pick)({
        enterprise: (0, a.css)([ "padding:", ";&:first-child{border-top:none;}" ], u.variables.spacing),
        prisma: {
            comfortable: (0, a.css)([ "padding:24px;" ]),
            compact: "padding: 18px 24px"
        }
    }), u.variables.contentColorMuted, (function(e) {
        var r = e.$showTopBorder;
        return r && (0, u.pick)({
            enterprise: (0, a.css)([ "border-top:1px solid ", ";" ], u.variables.gray92)
        });
    }));
    // CONCATENATED MODULE: ./src/Card/Footer.tsx
    function g() {
        g = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return g.apply(this, arguments);
    }
    function w(e, r) {
        if (e == null) return {};
        var t = O(e, r);
        var n, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                n = i[o];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function O(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var o, i;
        for (i = 0; i < n.length; i++) {
            o = n[i];
            if (r.indexOf(o) >= 0) continue;
            t[o] = e[o];
        }
        return t;
    }
    var x = {
        children: i().node,
        elementRef: i().oneOfType([ i().func, i().object ]),
        showBorder: i().bool
    };
    /**
 * A styled container for `Card` footer content.
 */    function S(e) {
        var r = e.children, t = e.showBorder, o = t === void 0 ? true : t, i = w(e, [ "children", "showBorder" ]);
        // @docs-props-type FooterPropsBase
                
        return n().createElement(h, g({
            "data-test": "footer",
            $showTopBorder: o
        }, i), r);
    }
    S.propTypes = x;
    /* harmony default export */    const k = S;
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const j = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "@splunk/react-icons/DotsThreeVertical"
    const C = require("@splunk/react-icons/DotsThreeVertical");
    var P =  e.n(C);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/MoreVertical"
    const _ = require("@splunk/react-icons/enterprise/MoreVertical");
    var T =  e.n(_);
    // CONCATENATED MODULE: external "@splunk/react-ui/Anchor"
    const E = require("@splunk/react-ui/Anchor");
    var B =  e.n(E);
    // CONCATENATED MODULE: external "@splunk/react-ui/Dropdown"
    const R = require("@splunk/react-ui/Dropdown");
    var I =  e.n(R);
    // CONCATENATED MODULE: external "@splunk/react-ui/Tooltip"
    const q = require("@splunk/react-ui/Tooltip");
    var H =  e.n(q);
    // CONCATENATED MODULE: external "@splunk/react-ui/Button"
    const N = require("@splunk/react-ui/Button");
    var $ =  e.n(N);
    // CONCATENATED MODULE: ./src/Card/HeaderStyles.ts
    var A = l().div.withConfig({
        displayName: "HeaderStyles__StyledTitle",
        componentId: "eqi6uk-0"
    })([ "", " ", " margin:0;overflow-wrap:break-word;padding:0;" ], u.mixins.reset("block"), (0, 
    u.pick)({
        enterprise: (0, a.css)([ "font-weight:", ";font-size:", ";line-height:1.2;" ], u.variables.fontWeightSemiBold, u.variables.fontSizeLarge),
        prisma: (0, a.css)([ "font-weight:", ";color:", ";" ], u.variables.fontWeightBold, u.variables.contentColorActive)
    }));
    var W = l().div.withConfig({
        displayName: "HeaderStyles__StyledSubtitle",
        componentId: "eqi6uk-1"
    })([ "", " ", " margin:0;overflow-wrap:break-word;padding:0;font-weight:normal;" ], u.mixins.reset("block"), (0, 
    u.pick)({
        enterprise: (0, a.css)([ "font-size:", ";color:", ";" ], u.variables.fontSizeSmall, u.variables.textGray),
        prisma: (0, a.css)([ "color:", ";" ], u.variables.contentColorDefault)
    }));
    var D = l().div.withConfig({
        displayName: "HeaderStyles__StyledTitleContainer",
        componentId: "eqi6uk-2"
    })([ "", " flex:0 1 auto;", " &:not(:last-child){margin-right:", ";}" ], u.mixins.reset("block"), (function(e) {
        var r = e.$truncateTitle;
        return r && (0, a.css)([ "overflow:hidden;& > ", ",& > ", "{overflow:hidden;text-overflow:ellipsis;white-space:nowrap;}" ], 
        /* sc-sel */
        A, 
        /* sc-sel */
        W);
    }), (0, u.pick)({
        enterprise: u.variables.spacingHalf,
        prisma: u.variables.spacingSmall
    }));
    var M = l().div.withConfig({
        displayName: "HeaderStyles__StyledActionContainer",
        componentId: "eqi6uk-3"
    })([ "", " align-items:center;display:flex;flex:0 0 auto;justify-content:space-between;", ";" ], u.mixins.reset("block"), (0, 
    u.pick)({
        prisma: {
            comfortable: (0, a.css)([ "margin-right:-12px;margin-top:-4px;" ]),
            compact: (0, a.css)([ "margin-right:-12px;margin-top:0;" ])
        }
    }));
    var F = l()($()).withConfig({
        displayName: "HeaderStyles__StyledSecondaryAction",
        componentId: "eqi6uk-4"
    })([ "margin-left:10px;" ]);
    var z = l()(s()).withConfig({
        displayName: "HeaderStyles__StyledBox",
        componentId: "eqi6uk-5"
    })([ "align-items:start;display:flex;flex:0 0 auto;justify-content:space-between;position:relative;min-height:30px;", " ", " >:not(", "):not(", "){flex:1 0 auto;}" ], (0, 
    u.pick)({
        enterprise: (0, a.css)([ "padding:", ";" ], u.variables.spacing),
        prisma: {
            comfortable: (0, a.css)([ "padding:16px 24px;&:last-child{padding-bottom:32px;}" ]),
            compact: (0, a.css)([ "padding:12px 24px;&:last-child{padding-bottom:24px;}" ])
        }
    }), (0, u.pick)({
        prisma: (0, a.css)([ "border-top-left-radius:", ";border-top-right-radius:", ";" ], u.variables.borderRadius, u.variables.borderRadius)
    }), 
    /* sc-sel */
    D, 
    /* sc-sel */
    M);
    // CONCATENATED MODULE: ./src/Card/Header.tsx
    function L(e, r) {
        if (e == null) return {};
        var t = V(e, r);
        var n, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                n = i[o];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function V(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var o, i;
        for (i = 0; i < n.length; i++) {
            o = n[i];
            if (r.indexOf(o) >= 0) continue;
            t[o] = e[o];
        }
        return t;
    }
    function G() {
        G = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return G.apply(this, arguments);
    }
    var U = {
        actionPrimary: i().node,
        actionsSecondary: i().node,
        anchor: i().string,
        children: i().node,
        elementRef: i().oneOfType([ i().func, i().object ]),
        subtitle: i().node,
        title: i().node,
        truncateTitle: i().bool
    };
    // `forwardRef` is used to enable focus on toggle when Dropdown is closing
        var J =  n().forwardRef((function(e, r) {
        var t = (0, u.useSplunkTheme)(), o = t.family, i = t.isCompact;
        var a = i ? "20px" : "24px";
        var l = o === "prisma" ?  n().createElement(P(), {
            height: a,
            width: a
        }) :  n().createElement(T(), null);
        
        return n().createElement(H(), G({
            content: (0, j._)("Actions"),
            contentRelationship: "label"
        }, e),  n().createElement(F, {
            appearance: "secondary",
            "data-test": "actions-secondary-toggle",
            icon: l,
            elementRef: r
        }));
    }));
    /**
 * A styled container for `Card` header content.
 */    function K(e) {
        var r = e.actionPrimary, t = e.actionsSecondary, o = e.anchor, i = e.children, a = e.subtitle, l = e.title, c = e.truncateTitle, s = c === void 0 ? true : c, p = L(e, [ "actionPrimary", "actionsSecondary", "anchor", "children", "subtitle", "title", "truncateTitle" ]);
        // @docs-props-type HeaderPropsBase
                var d = (0, u.useSplunkTheme)(), f = d.family;
        var b = f === "prisma" ? "right" : "below";
        
        return n().createElement(z, G({
            "data-test": "header"
        }, p), (l || a) &&  n().createElement(D, {
            $truncateTitle: s
        }, l &&  n().createElement(A, null, o ?  n().createElement(B(), {
            name: o
        }, l) : l), a &&  n().createElement(W, null, a)), i, (r || t) &&  n().createElement(M, null, r, t &&  n().createElement(I(), {
            defaultPlacement: b,
            toggle:  n().createElement(J, null)
        }, t)));
    }
    K.propTypes = U;
    /* harmony default export */    const Q = K;
    // CONCATENATED MODULE: external "@splunk/react-ui/Clickable"
    const X = require("@splunk/react-ui/Clickable");
    var Y =  e.n(X);
    // CONCATENATED MODULE: ./src/Card/CardStyles.ts
    var Z = "0.2s";
    var ee = l().div.withConfig({
        displayName: "CardStyles__Styled",
        componentId: "ola3x0-0"
    })([ "", ";flex-direction:column;align-items:stretch;min-width:100px;flex:1;vertical-align:top;", " transition:height ", ",width ", ",min-width ", ",max-width ", ",margin ", ",box-shadow ", ",border-color ", ";", " &[data-clickable='true']{cursor:pointer;", " &:hover{", "}&:focus{", "}}" ], u.mixins.reset("inline-flex"), (0, 
    u.pick)({
        enterprise: (0, a.css)([ "background-color:", ";border:1px solid transparent;" ], u.variables.backgroundColor),
        prisma: (0, a.css)([ "background-color:", ";border-radius:", ";box-shadow:", ";" ], u.variables.backgroundColorSection, u.variables.borderRadius, u.variables.embossShadow)
    }), Z, Z, Z, Z, Z, Z, Z, (function(e) {
        var r = e.$cardHasBorder;
        return r && (0, u.pick)({
            enterprise: (0, a.css)([ "border-color:", ";" ], u.variables.borderLightColor)
        });
    }), (function(e) {
        var r = e.$selected;
        return r && (0, u.pick)({
            enterprise: (0, a.css)([ "border-color:", ";" ], u.variables.focusColor),
            prisma: (0, a.css)([ "background-color:", ";" ], u.mixins.overlayColors(u.variables.backgroundColorSection, u.variables.interactiveColorOverlaySelected))
        });
    }), (0, u.pick)({
        enterprise: (0, a.css)([ "box-shadow:", ";" ], u.variables.overlayShadow),
        prisma: (0, a.css)([ "box-shadow:", ";background-color:", ";" ], u.variables.embossShadow, u.mixins.overlayColors(u.variables.backgroundColorSection, u.variables.interactiveColorOverlayHover))
    }), (0, u.pick)({
        enterprise: (0, a.css)([ "box-shadow:", ";" ], u.variables.focusShadow),
        prisma: (0, a.css)([ "box-shadow:", ",", ";background-color:", ";" ], u.variables.embossShadow, u.variables.focusShadow, u.mixins.overlayColors(u.variables.backgroundColorSection, u.variables.interactiveColorOverlayHover))
    }));
    var re = ee.withComponent(Y());
    /* A child element we add to <button>s to make flexbox work.
   Without this wrapper element, child elements won't flex
   in Firefox.
   This unfortunately exposes some fragility in IE (e.g., adding
   a flex: 1 style here will cause the card to collapse) so test
   across the supported browsers if making a change here.
*/    var te = l().div.withConfig({
        displayName: "CardStyles__StyledFirefoxFlexHack",
        componentId: "ola3x0-1"
    })([ "display:flex;flex:1 1 auto;flex-direction:column;align-items:stretch;justify-content:stretch;height:100%;", "" ], (0, 
    u.pick)({
        prisma: (0, a.css)([ "border-radius:", ";" ], u.variables.borderRadius)
    }));
    // CONCATENATED MODULE: ./src/Card/Card.tsx
    function ne(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            ne = function e(r) {
                return typeof r;
            };
        } else {
            ne = function e(r) {
                return r && typeof Symbol === "function" && r.constructor === Symbol && r !== Symbol.prototype ? "symbol" : typeof r;
            };
        }
        return ne(e);
    }
    function oe() {
        oe = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return oe.apply(this, arguments);
    }
    function ie(e, r) {
        return ue(e) || se(e, r) || le(e, r) || ae();
    }
    function ae() {
        throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function le(e, r) {
        if (!e) return;
        if (typeof e === "string") return ce(e, r);
        var t = Object.prototype.toString.call(e).slice(8, -1);
        if (t === "Object" && e.constructor) t = e.constructor.name;
        if (t === "Map" || t === "Set") return Array.from(e);
        if (t === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t)) return ce(e, r);
    }
    function ce(e, r) {
        if (r == null || r > e.length) r = e.length;
        for (var t = 0, n = new Array(r); t < r; t++) {
            n[t] = e[t];
        }
        return n;
    }
    function se(e, r) {
        if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(e))) return;
        var t = [];
        var n = true;
        var o = false;
        var i = undefined;
        try {
            for (var a = e[Symbol.iterator](), l; !(n = (l = a.next()).done); n = true) {
                t.push(l.value);
                if (r && t.length === r) break;
            }
        } catch (e) {
            o = true;
            i = e;
        } finally {
            try {
                if (!n && a["return"] != null) a["return"]();
            } finally {
                if (o) throw i;
            }
        }
        return t;
    }
    function ue(e) {
        if (Array.isArray(e)) return e;
    }
    function pe(e, r) {
        if (!(e instanceof r)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function de(e, r) {
        for (var t = 0; t < r.length; t++) {
            var n = r[t];
            n.enumerable = n.enumerable || false;
            n.configurable = true;
            if ("value" in n) n.writable = true;
            Object.defineProperty(e, n.key, n);
        }
    }
    function fe(e, r, t) {
        if (r) de(e.prototype, r);
        if (t) de(e, t);
        return e;
    }
    function be(e, r) {
        if (typeof r !== "function" && r !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(r && r.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (r) ve(e, r);
    }
    function ve(e, r) {
        ve = Object.setPrototypeOf || function e(r, t) {
            r.__proto__ = t;
            return r;
        };
        return ve(e, r);
    }
    function ye(e) {
        var r = ge();
        return function t() {
            var n = we(e), o;
            if (r) {
                var i = we(this).constructor;
                o = Reflect.construct(n, arguments, i);
            } else {
                o = n.apply(this, arguments);
            }
            return me(this, o);
        };
    }
    function me(e, r) {
        if (r && (ne(r) === "object" || typeof r === "function")) {
            return r;
        }
        return he(e);
    }
    function he(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function ge() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function we(e) {
        we = Object.setPrototypeOf ? Object.getPrototypeOf : function e(r) {
            return r.__proto__ || Object.getPrototypeOf(r);
        };
        return we(e);
    }
    function Oe(e, r) {
        var t = Object.keys(e);
        if (Object.getOwnPropertySymbols) {
            var n = Object.getOwnPropertySymbols(e);
            if (r) n = n.filter((function(r) {
                return Object.getOwnPropertyDescriptor(e, r).enumerable;
            }));
            t.push.apply(t, n);
        }
        return t;
    }
    function xe(e) {
        for (var r = 1; r < arguments.length; r++) {
            var t = arguments[r] != null ? arguments[r] : {};
            if (r % 2) {
                Oe(Object(t), true).forEach((function(r) {
                    Se(e, r, t[r]);
                }));
            } else if (Object.getOwnPropertyDescriptors) {
                Object.defineProperties(e, Object.getOwnPropertyDescriptors(t));
            } else {
                Oe(Object(t)).forEach((function(r) {
                    Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r));
                }));
            }
        }
        return e;
    }
    function Se(e, r, t) {
        if (r in e) {
            Object.defineProperty(e, r, {
                value: t,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[r] = t;
        }
        return e;
    }
    function ke(e, r) {
        if (e == null) return {};
        var t = je(e, r);
        var n, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                n = i[o];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function je(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var o, i;
        for (i = 0; i < n.length; i++) {
            o = n[i];
            if (r.indexOf(o) >= 0) continue;
            t[o] = e[o];
        }
        return t;
    }
    var Ce = {
        children: i().node,
        elementRef: i().oneOfType([ i().func, i().object ]),
        /** @private */
        margin: i().number,
        /** @private */
        maxWidth: i().oneOfType([ i().number, i().string ]),
        /** @private */
        minWidth: i().oneOfType([ i().number, i().string ]),
        onClick: i().func,
        openInNewContext: i().bool,
        selected: i().bool,
        showBorder: i().bool,
        to: i().string,
        value: i().any
    };
    var Pe = {
        openInNewContext: false,
        showBorder: true
    };
    function _e(e) {
        return !!e.onClick || !!e.to;
    }
    function Te(e) {
        var r = e.margin, t = e.maxWidth, n = e.minWidth, o = e.showBorder, i = e.style, a = ke(e, [ "margin", "maxWidth", "minWidth", "showBorder", "style" ]);
        var l = xe({
            minWidth: n,
            maxWidth: t,
            margin: r
        }, i);
        var c = {
            showBorder: o,
            "data-test": "card",
            style: l
        };
        return [ c, a ];
    }
    var Ee =  function(e) {
        be(t, e);
        var r = ye(t);
        function t() {
            var e;
            pe(this, t);
            for (var n = arguments.length, o = new Array(n), i = 0; i < n; i++) {
                o[i] = arguments[i];
            }
            e = r.call.apply(r, [ this ].concat(o));
            Se(he(e), "handleCardClick", (function(r) {
                var t, n;
                var o = e.props, i = o.selected, a = o.value;
                (t = (n = e.props).onClick) === null || t === void 0 ? void 0 : t.call(n, r, {
                    selected: i || false,
                    value: a
                });
            }));
            return e;
        }
        fe(t, [ {
            key: "render",
            value: function e() {
                if (_e(this.props)) {
                    var r = Te(this.props), t = ie(r, 2), o = t[0], i = t[1];
                    var a = o.showBorder;
                    var l = i.children, c = i.elementRef, s = i.onClick, u = i.openInNewContext, p = i.selected, d = i.to, f = ke(i, [ "children", "elementRef", "onClick", "openInNewContext", "selected", "to" ]);
                    
                    return n().createElement(re, oe({
                        "data-clickable": true,
                        $selected: p || false,
                        $cardHasBorder: a,
                        elementRef: c
                    }, o, f, {
                        onClick: s ? this.handleCardClick : undefined,
                        openInNewContext: u,
                        to: d || undefined
                    }),  n().createElement(te, null, l));
                }
                var b = Te(this.props), v = ie(b, 2), y = v[0], m = v[1];
                var h = y.showBorder;
                var g = m.children, w = m.elementRef, O = ke(m, [ "children", "elementRef" ]);
                
                return n().createElement(ee, oe({
                    $selected: false,
                    $cardHasBorder: h,
                    ref: w
                }, y, O), g);
            }
        } ]);
        return t;
    }(t.Component);
    Se(Ee, "propTypes", Ce);
    Se(Ee, "defaultProps", Pe);
    Se(Ee, "Header", Q);
    Se(Ee, "Body", m);
    Se(Ee, "Footer", k);
    /* harmony default export */    const Be = Ee;
    // CONCATENATED MODULE: ./src/Card/index.ts
    module.exports = r;
    /******/})();
