// @splunk/react-ui/Chip.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var a = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(a, {
                a
            });
            /******/            return a;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, a) => {
            /******/ for (var o in a) {
                /******/ if (e.o(a, o) && !e.o(r, o)) {
                    /******/ Object.defineProperty(r, o, {
                        enumerable: true,
                        get: a[o]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        default: () => /* reexport */ G
    });
    // CONCATENATED MODULE: external "react"
    const a = require("react");
    var o =  e.n(a);
    // CONCATENATED MODULE: external "lodash/omit"
    const n = require("lodash/omit");
    var t =  e.n(n);
    // CONCATENATED MODULE: external "prop-types"
    const i = require("prop-types");
    var l =  e.n(i);
    // CONCATENATED MODULE: external "@splunk/react-icons/Cross"
    const s = require("@splunk/react-icons/Cross");
    var c =  e.n(s);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Close"
    const p = require("@splunk/react-icons/enterprise/Close");
    var u =  e.n(p);
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const d = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "@splunk/ui-utils/color"
    const f = require("@splunk/ui-utils/color");
    // CONCATENATED MODULE: external "@splunk/ui-utils/format"
    const v = require("@splunk/ui-utils/format");
    // CONCATENATED MODULE: external "@splunk/themes"
    const b = require("@splunk/themes");
    // CONCATENATED MODULE: external "@splunk/react-ui/ScreenReaderContent"
    const g = require("@splunk/react-ui/ScreenReaderContent");
    var m =  e.n(g);
    // CONCATENATED MODULE: external "styled-components"
    const h = require("styled-components");
    var C =  e.n(h);
    // CONCATENATED MODULE: external "@splunk/react-ui/Clickable"
    const y = require("@splunk/react-ui/Clickable");
    var x =  e.n(y);
    // CONCATENATED MODULE: ./src/Chip/ChipStyles.ts
    var $ = C().div.withConfig({
        displayName: "ChipStyles__StyledBasic",
        componentId: "sc-1sd3tsh-0"
    })([ "", ";background-color:", ";border:1px solid transparent;flex:0 1 auto;max-width:calc(100% - 3px);border-radius:2px;", " ", " ", "" ], b.mixins.reset("inline-flex"), (0, 
    b.pickVariant)("$appearance", {
        default: {
            enterprise: {
                dark: b.variables.gray45,
                light: b.variables.gray92
            },
            prisma: b.variables.neutral100
        },
        info: {
            enterprise: b.variables.infoColorL30
        },
        success: {
            enterprise: b.variables.successColorL30
        },
        warning: {
            enterprise: b.variables.warningColorL30
        },
        error: {
            enterprise: b.variables.errorColorL30,
            prisma: b.variables.accentColorNegative
        },
        outline: {
            prisma: "transparent"
        },
        custom: function e(r) {
            var a = r.$backgroundColor;
            return a;
        }
    }), (0, b.pick)({
        enterprise: (0, h.css)([ "color:", ";line-height:", ";margin:0 2px 2px 0;" ], b.variables.textGray, b.variables.lineHeight),
        prisma: (0, h.css)([ "color:", ";line-height:16px;margin:1px;", "" ], (function(e) {
            var r = e.$appearance;
            return r === "error" ? b.variables.contentColorInverted : b.variables.contentColorDefault;
        }), (function(e) {
            var r = e.$appearance;
            return r === "outline" && (0, h.css)([ "border-color:", ";" ], b.variables.interactiveColorBorder);
        }))
    }), (0, b.pick)({
        enterprise: {
            comfortable: (0, h.css)([ "height:calc(", " - 6px);padding:4px 7px;" ], b.variables.inputHeight),
            compact: (0, h.css)([ "height:calc(", " - 6px);padding:2px 4px;" ], b.variables.inputHeight)
        },
        prisma: {
            comfortable: (0, h.css)([ "height:calc(", " - 8px);padding:7px 7px;font-size:", ";" ], b.variables.inputHeight, b.variables.fontSizeSmall),
            compact: (0, h.css)([ "height:calc(", " - 8px);padding:3px 7px;font-size:", ";" ], b.variables.inputHeight, b.variables.fontSizeSmall)
        }
    }), (function(e) {
        var r = e.$disabled;
        return r && (0, h.css)([ "background-color:", ";" ], (0, b.pickVariant)("$appearance", {
            default: {
                enterprise: {
                    dark: "rgba(0, 0, 0, 0.15)",
                    light: "rgba(0, 0, 0, 0.05)"
                },
                prisma: b.variables.interactiveColorBackgroundDisabled
            },
            info: {
                enterprise: b.mixins.colorWithAlpha(b.variables.infoColorL10, .3)
            },
            success: {
                enterprise: b.mixins.colorWithAlpha(b.variables.successColorL10, .3)
            },
            warning: {
                enterprise: b.mixins.colorWithAlpha(b.variables.warningColorL10, .3)
            },
            error: {
                enterprise: b.mixins.colorWithAlpha(b.variables.errorColorL10, .3),
                prisma: b.variables.interactiveColorBackgroundDisabled
            },
            outline: {
                prisma: b.variables.interactiveColorBackgroundDisabled
            },
            custom: {
                enterprise: {
                    dark: "rgba(0, 0, 0, 0.15)",
                    light: "rgba(0, 0, 0, 0.05)"
                },
                prisma: b.variables.interactiveColorBackgroundDisabled
            }
        }));
    }));
    var k = C()($).withConfig({
        displayName: "ChipStyles__Styled",
        componentId: "sc-1sd3tsh-1"
    })([ "align-items:center;" ]);
    var S = C().div.withConfig({
        displayName: "ChipStyles__StyledInner",
        componentId: "sc-1sd3tsh-2"
    })([ "", ";" ], (0, b.pick)({
        enterprise: (0, h.css)([ "display:flex;max-width:100%;" ]),
        prisma: (0, h.css)([ "display:grid;max-width:100%;height:16px;column-gap:", ";", ";" ], (0, 
        b.pick)({
            prisma: {
                compact: b.variables.spacingXSmall,
                comfortable: b.variables.spacingSmall
            }
        }), (function(e) {
            var r = e.$icon, a = e.$removable;
            var o = (0, h.css)([ "grid-template-columns:1fr;" ]);
            if (r && a) {
                o = (0, h.css)([ "grid-template-columns:0fr 1fr 0fr;" ]);
            } else if (r) {
                o = (0, h.css)([ "grid-template-columns:0fr 1fr;" ]);
            } else if (a) {
                o = (0, h.css)([ "grid-template-columns:1fr 0fr;" ]);
            }
            return o;
        }))
    }));
    var w = C().div.withConfig({
        displayName: "ChipStyles__StyledIcon",
        componentId: "sc-1sd3tsh-3"
    })([ "", ";margin-right:", ";color:", ";" ], (0, b.pick)({
        prisma: {
            compact: (0, h.css)([ "font-size:14px;display:inline-flex;svg{height:16px;vertical-align:baseline;}" ]),
            comfortable: (0, h.css)([ "font-size:18px;display:inline-flex;align-self:center;svg{height:16px;vertical-align:baseline;}" ])
        },
        enterprise: (0, h.css)([ "flex:0 0 auto;" ])
    }), (0, b.pick)({
        enterprise: "3px"
    }), (0, b.pickVariant)("$disabled", {
        true: b.variables.contentColorDisabled,
        false: {
            enterprise: function e(r) {
                var a = r.$foregroundColor;
                return a || b.variables.contentColorDefault;
            },
            prisma: function e(r) {
                var a = r.$foregroundColor, o = r.$appearance;
                return a || (o === "error" ? b.variables.contentColorInverted : b.variables.contentColorDefault);
            }
        }
    }));
    var O = C().div.withConfig({
        displayName: "ChipStyles__StyledLabel",
        componentId: "sc-1sd3tsh-4"
    })([ "", " ", ";color:", ";" ], b.mixins.ellipsis(), (0, b.pick)({
        enterprise: (0, h.css)([ "flex:0 1 auto;" ])
    }), (0, b.pickVariant)("$disabled", {
        true: b.variables.contentColorDisabled,
        false: {
            enterprise: function e(r) {
                var a = r.$appearance, o = r.$foregroundColor;
                return o || (a === "default" || a === "outline" ? b.variables.contentColorDefault : b.variables.gray30);
            },
            prisma: function e(r) {
                var a = r.$appearance, o = r.$foregroundColor;
                return o || (a === "error" ? b.variables.contentColorInverted : b.variables.contentColorActive);
            }
        }
    }));
    var R = C().span.withConfig({
        displayName: "ChipStyles__StyledRemove",
        componentId: "sc-1sd3tsh-5"
    })([ "", ";" ], (0, b.pick)({
        enterprise: (0, h.css)([ "flex:0 0 auto;padding-left:", ";font-size:", ";color:", ";" ], b.variables.spacingQuarter, (0, 
        b.pick)({
            compact: "9px",
            comfortable: "10.5px"
        }), (0, b.pickVariant)("$disabled", {
            true: b.variables.contentColorDisabled,
            false: {
                enterprise: function e(r) {
                    var a = r.$foregroundColor, o = r.$appearance;
                    return a || (o === "default" || o === "outline" ? b.variables.textGray : b.variables.gray30);
                }
            }
        })),
        prisma: (0, h.css)([ "font-size:", ";display:flex;align-items:center;justify-content:center;width:16px;height:16px;border-radius:12px;font-size:9px;color:", ";" ], (0, 
        b.pick)({
            compact: "9px",
            comfortable: "10.5px"
        }), (function(e) {
            var r = e.$disabled, a = e.$foregroundColor;
            return r ? b.variables.contentColorDisabled : a || "inherit";
        }))
    }));
    var _ = C()($.withComponent(x())).withConfig({
        displayName: "ChipStyles__StyledClickable",
        componentId: "sc-1sd3tsh-6"
    })([ "flex:0 1 auto;line-height:16px;cursor:pointer;&:focus{", ";}&:not([disabled]):hover{", ";}", ";" ], (0, 
    b.pick)({
        enterprise: (0, h.css)([ "box-shadow:", ";color:", ";" ], b.variables.focusShadow, b.variables.linkColor),
        prisma: (0, h.css)([ "color:", ";", "{background-color:", ";box-shadow:0 0 0 3px ", ";}" ], b.variables.contentColorActive, 
        /* sc-sel */
        R, b.variables.interactiveColorOverlayHover, b.variables.focusColor)
    }), (0, b.pick)({
        enterprise: (0, h.css)([ "background-color:", ";", "{color:", ";}" ], (0, b.pickVariant)("$appearance", {
            default: {
                dark: b.variables.gray30,
                light: b.variables.gray96
            },
            info: b.variables.infoColorL20,
            success: b.variables.successColorL20,
            warning: b.variables.warningColorL20,
            error: b.variables.errorColorL20,
            custom: {
                dark: b.variables.gray30,
                light: b.variables.gray96
            }
        }), 
        /* sc-sel */
        R, (function(e) {
            var r = e.$foregroundColor;
            return r || b.variables.linkColor;
        })),
        prisma: (0, h.css)([ "color:", ";", "{background-color:", ";", ";}" ], b.variables.contentColorActive, 
        /* sc-sel */
        R, b.variables.interactiveColorOverlayHover, (function(e) {
            var r = e.$appearance;
            return r === "error" && (0, h.css)([ "color:", ";" ], b.variables.contentColorInverted);
        }))
    }), (0, b.pick)({
        prisma: (0, h.css)([ "&:not([disabled]):active ", "{background-color:", ";}" ], 
        /* sc-sel */
        R, b.variables.interactiveColorOverlayActive)
    }));
    // CONCATENATED MODULE: ./src/Chip/Chip.tsx
    function j() {
        j = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var a = arguments[r];
                for (var o in a) {
                    if (Object.prototype.hasOwnProperty.call(a, o)) {
                        e[o] = a[o];
                    }
                }
            }
            return e;
        };
        return j.apply(this, arguments);
    }
    function q(e, r) {
        return L(e) || D(e, r) || I(e, r) || E();
    }
    function E() {
        throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function I(e, r) {
        if (!e) return;
        if (typeof e === "string") return A(e, r);
        var a = Object.prototype.toString.call(e).slice(8, -1);
        if (a === "Object" && e.constructor) a = e.constructor.name;
        if (a === "Map" || a === "Set") return Array.from(e);
        if (a === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(a)) return A(e, r);
    }
    function A(e, r) {
        if (r == null || r > e.length) r = e.length;
        for (var a = 0, o = new Array(r); a < r; a++) {
            o[a] = e[a];
        }
        return o;
    }
    function D(e, r) {
        if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(e))) return;
        var a = [];
        var o = true;
        var n = false;
        var t = undefined;
        try {
            for (var i = e[Symbol.iterator](), l; !(o = (l = i.next()).done); o = true) {
                a.push(l.value);
                if (r && a.length === r) break;
            }
        } catch (e) {
            n = true;
            t = e;
        } finally {
            try {
                if (!o && i["return"] != null) i["return"]();
            } finally {
                if (n) throw t;
            }
        }
        return a;
    }
    function L(e) {
        if (Array.isArray(e)) return e;
    }
    function z(e, r) {
        if (e == null) return {};
        var a = N(e, r);
        var o, n;
        if (Object.getOwnPropertySymbols) {
            var t = Object.getOwnPropertySymbols(e);
            for (n = 0; n < t.length; n++) {
                o = t[n];
                if (r.indexOf(o) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, o)) continue;
                a[o] = e[o];
            }
        }
        return a;
    }
    function N(e, r) {
        if (e == null) return {};
        var a = {};
        var o = Object.keys(e);
        var n, t;
        for (t = 0; t < o.length; t++) {
            n = o[t];
            if (r.indexOf(n) >= 0) continue;
            a[n] = e[n];
        }
        return a;
    }
    var P = {
        appearance: l().oneOf([ "info", "success", "warning", "error", "outline" ]),
        backgroundColor: l().string,
        children: l().node.isRequired,
        disabled: l().bool,
        elementRef: l().oneOfType([ l().func, l().object ]),
        foregroundColor: l().string,
        icon: l().node,
        onRequestRemove: l().func,
        /** Includes this value in `onRequestRemove` callbacks. */
        value: l().any
    };
    var T = {
        disabled: false
    };
    function H(e) {
        return !!e.onRequestRemove;
    }
    function B(e) {
        var r = e.appearance, a = e.backgroundColor, o = e.disabled, n = e.foregroundColor, t = e.icon, i = z(e, [ "appearance", "backgroundColor", "disabled", "foregroundColor", "icon" ]);
 // Validate that the color being rendered into the CSS is restricted to only values that are of CSS type <color>.
                var l = n && (0, f.isCSSColor)(n) ? n : undefined;
        var s = a && (0, f.isCSSColor)(a) ? a : undefined;
        var c = {
            "data-test": "chip",
            $appearance: r || (s ? "custom" : "default"),
            $backgroundColor: s,
            $foregroundColor: l,
            disabled: o,
            icon: t
        };
        return [ c, i ];
    }
    function V(e) {
        var r = B(e), n = q(r, 2), i = n[0], l = n[1];
        var s = i.$appearance, p = i.$foregroundColor, f = i.disabled, g = i.icon, h = z(i, [ "$appearance", "$foregroundColor", "disabled", "icon" ]);
        var C = l.children, y = l.elementRef, x = l.onRequestRemove, $ = l.value, k = z(l, [ "children", "elementRef", "onRequestRemove", "value" ]);
        var E = (0, b.useSplunkTheme)(), I = E.isEnterprise;
        var A = typeof C === "string";
        var D = A ? (0, v.sprintf)((0, d._)("Remove %(children)s"), {
            children: C
        }) : (0, d._)("Remove");
        var L = (0, a.useCallback)((function(e) {
            x(e, {
                value: $
            });
        }), [ x, $ ]);
        
        return o().createElement(_, j({
            $appearance: s,
            $disabled: f,
            $foregroundColor: p,
            "data-test-value": $,
            disabled: f,
            elementRef: y,
            onClick: L
        }, h, t()(k, Object.keys(P))),  o().createElement(S, {
            $icon: !!g,
            $removable: true
        },  o().createElement(m(), null, D), g &&  o().createElement(w, {
            $appearance: s,
            $disabled: f,
            $foregroundColor: p
        }, g),  o().createElement(O, {
            "data-test": "label",
            "aria-hidden": A ? true : undefined,
            $appearance: s,
            $disabled: f,
            $foregroundColor: p
        }, C),  o().createElement(R, {
            $appearance: s,
            $disabled: f,
            $foregroundColor: p
        }, I ?  o().createElement(u(), {
            "data-test": "cross",
            hideDefaultTooltip: true,
            screenReaderText: null,
            size: .85
        }) :  o().createElement(c(), {
            "data-test": "cross",
            height: "20px",
            width: "20px"
        }))));
    }
    function M(e) {
        var r = B(e), a = q(r, 2), n = a[0], i = a[1];
        var l = n.$appearance, s = n.$foregroundColor, c = n.disabled, p = n.icon, u = z(n, [ "$appearance", "$foregroundColor", "disabled", "icon" ]);
        var d = i.children, f = i.elementRef, v = z(i, [ "children", "elementRef" ]);
        
        return o().createElement(k, j({
            $appearance: l,
            $disabled: c,
            ref: f
        }, u, t()(v, Object.keys(P))),  o().createElement(S, {
            $icon: !!p,
            $removable: false
        }, p &&  o().createElement(w, {
            $appearance: l,
            $disabled: c,
            $foregroundColor: s
        }, p),  o().createElement(O, {
            "data-test": "label",
            $appearance: l,
            $disabled: c,
            $foregroundColor: s
        }, d)));
    }
    function W(e) {
        // @docs-props-type ChipPropsBase
        return H(e) ?  o().createElement(V, e) :  o().createElement(M, e);
    }
    W.propTypes = P;
    W.defaultProps = T;
    /* harmony default export */    const G = W;
    // CONCATENATED MODULE: ./src/Chip/index.ts
    module.exports = r;
    /******/})();
