// @splunk/react-ui/CloseButton.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var r = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(r, {
                a: r
            });
            /******/            return r;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        default: () => /* reexport */ j
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    var n =  e.n(r);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var i =  e.n(o);
    // CONCATENATED MODULE: external "@splunk/react-icons/Cross"
    const a = require("@splunk/react-icons/Cross");
    var s =  e.n(a);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Close"
    const l = require("@splunk/react-icons/enterprise/Close");
    var u =  e.n(l);
    // CONCATENATED MODULE: external "@splunk/themes"
    const c = require("@splunk/themes");
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const p = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "styled-components"
    const d = require("styled-components");
    var f =  e.n(d);
    // CONCATENATED MODULE: external "@splunk/react-ui/ButtonSimple"
    const v = require("@splunk/react-ui/ButtonSimple");
    var b =  e.n(v);
    // CONCATENATED MODULE: ./src/CloseButton/CloseButtonStyles.ts
    var y = f()(b()).withConfig({
        displayName: "CloseButtonStyles__StyledButtonSimple",
        componentId: "b3h40f-0"
    })([ "float:right;width:", ";height:", ";text-align:center;" ], c.variables.inputHeight, c.variables.inputHeight);
    // CONCATENATED MODULE: ./src/CloseButton/CloseButton.tsx
    function m() {
        m = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return m.apply(this, arguments);
    }
    function g(e, t) {
        if (e == null) return {};
        var r = h(e, t);
        var n, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                n = i[o];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function h(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var o, i;
        for (i = 0; i < n.length; i++) {
            o = n[i];
            if (t.indexOf(o) >= 0) continue;
            r[o] = e[o];
        }
        return r;
    }
    var O = {
        disabled: i().bool,
        elementRef: i().oneOfType([ i().func, i().object ]),
        screenReaderText: i().string
    };
    function S(e) {
        var t = e.disabled, r = t === void 0 ? false : t, o = e.screenReaderText, i = o === void 0 ? (0, 
        p._)("Close") : o, a = g(e, [ "disabled", "screenReaderText" ]);
        var l = (0, c.useSplunkTheme)(), d = l.isEnterprise, f = l.isCompact;
        var v = f ? "20px" : "24px";
        
        return n().createElement(y, m({
            appearance: "pill",
            disabled: r,
            "aria-label": i
        }, a), d ?  n().createElement(u(), {
            "data-test": "cross",
            inline: false,
            screenReaderText: null,
            size: "13px"
        }) :  n().createElement(s(), {
            "data-test": "cross",
            height: v,
            width: v
        }));
    }
    S.propTypes = O;
    /* harmony default export */    const j = S;
    // CONCATENATED MODULE: ./src/CloseButton/index.ts
    module.exports = t;
    /******/})();
