// @splunk/react-ui/Color.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var r = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(r, {
                a: r
            });
            /******/            return r;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        default: () => /* reexport */ Ue,
        defaultPalette: () => /* reexport */ Ae,
        isValidHEX: () => /* reexport */ ze
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    var n =  e.n(r);
    // CONCATENATED MODULE: external "prop-types"
    const a = require("prop-types");
    var o =  e.n(a);
    // CONCATENATED MODULE: external "lodash/has"
    const l = require("lodash/has");
    var i =  e.n(l);
    // CONCATENATED MODULE: external "lodash/includes"
    const s = require("lodash/includes");
    var u =  e.n(s);
    // CONCATENATED MODULE: external "lodash/keys"
    const c = require("lodash/keys");
    var p =  e.n(c);
    // CONCATENATED MODULE: external "lodash/omit"
    const d = require("lodash/omit");
    var f =  e.n(d);
    // CONCATENATED MODULE: external "lodash/slice"
    const v = require("lodash/slice");
    var h =  e.n(v);
    // CONCATENATED MODULE: external "lodash/startsWith"
    const m = require("lodash/startsWith");
    var y =  e.n(m);
    // CONCATENATED MODULE: external "lodash/toLower"
    const b = require("lodash/toLower");
    var g =  e.n(b);
    // CONCATENATED MODULE: external "lodash/memoize"
    const x = require("lodash/memoize");
    var w =  e.n(x);
    // CONCATENATED MODULE: external "tinycolor2"
    const S = require("tinycolor2");
    var k =  e.n(S);
    // CONCATENATED MODULE: external "@splunk/react-ui/Dropdown"
    const C = require("@splunk/react-ui/Dropdown");
    var E =  e.n(C);
    // CONCATENATED MODULE: external "@splunk/themes"
    const P = require("@splunk/themes");
    // CONCATENATED MODULE: external "@splunk/ui-utils/color"
    const O = require("@splunk/ui-utils/color");
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const _ = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "@splunk/ui-utils/keyboard"
    const q = require("@splunk/ui-utils/keyboard");
    // CONCATENATED MODULE: external "styled-components"
    const I = require("styled-components");
    var j =  e.n(I);
    // CONCATENATED MODULE: external "@splunk/react-ui/Text"
    const T = require("@splunk/react-ui/Text");
    var R =  e.n(T);
    // CONCATENATED MODULE: external "@splunk/react-ui/ScreenReaderContent"
    const B = require("@splunk/react-ui/ScreenReaderContent");
    var V =  e.n(B);
    // CONCATENATED MODULE: external "@splunk/react-icons/Checkmark"
    const A = require("@splunk/react-icons/Checkmark");
    var D =  e.n(A);
    // CONCATENATED MODULE: external "@splunk/react-ui/Clickable"
    const N = require("@splunk/react-ui/Clickable");
    var $ =  e.n(N);
    // CONCATENATED MODULE: ./src/Color/SwatchStyles.ts
    var H = (0, P.pick)({
        enterprise: P.variables.gray80,
        prisma: "rgba(0, 0, 0, 0.2)"
    });
    var z = (0, I.css)([ "background-image:linear-gradient( 45deg,", " 25%,transparent 25%,transparent 75%,", " 75%,", " ),linear-gradient( 45deg,", " 25%,transparent 25%,transparent 75%,", " 75%,", " );background-size:10px 10px;background-position:0 0,5px 5px;background-color:", ";" ], H, H, H, H, H, H, (0, 
    P.pick)({
        enterprise: "transparent",
        prisma: P.variables.white
    }));
    var K = j()(D()).withConfig({
        displayName: "SwatchStyles__StyledCheck",
        componentId: "sc-1wxunhq-0"
    })([ "box-sizing:border-box;color:#d3d3d3;height:100%;width:100%;" ]);
    var F = j()($()).withConfig({
        displayName: "SwatchStyles__StyledClickable",
        componentId: "sc-1wxunhq-1"
    })([ "display:block;flex:0 0 auto;border:", ";", ";", ";", ";", " &:focus{border-color:", ";box-shadow:", ";}", " ", " ", " &[disabled]{background-image:linear-gradient(rgba(255,255,255,0.8),rgba(255,255,255,0.8));}", " ", "" ], (0, 
    P.pick)({
        enterprise: {
            light: (0, I.css)([ "1px solid ", "" ], P.variables.gray60),
            dark: P.variables.border
        },
        prisma: (0, I.css)([ "1px solid ", "" ], P.variables.interactiveColorBorder)
    }), (0, P.pick)({
        prisma: (0, I.css)([ "border-radius:2px;" ])
    }), (function(e) {
        var t = e.$value;
        return t === null && (0, I.css)([ "background-image:linear-gradient( to bottom right,transparent 48%,", ",transparent 52% );", ";" ], (0, 
        P.pick)({
            enterprise: "#ed1e24",
            prisma: P.variables.accentColorNegative
        }), (0, P.pick)({
            prisma: (0, I.css)([ "background-color:", ";" ], P.variables.white)
        }));
    }), (function(e) {
        var t = e.$value;
        return t && (t === "transparent" ? z : "background-color: ".concat(t));
    }), (function(e) {
        var t = e.$noBorder;
        return t && (0, I.css)([ "border:none;" ]);
    }), P.variables.focusColor, P.variables.focusShadow, (0, P.pick)({
        enterprise: {
            comfortable: (0, I.css)([ "width:32px;height:32px;" ]),
            compact: (0, I.css)([ "width:24px;height:24px;" ])
        },
        prisma: (0, I.css)([ "width:24px;height:24px;" ])
    }), (function(e) {
        var t = e.$prepend;
        return t && (0, I.css)([ "border-top-left-radius:0;border-bottom-left-radius:0;" ]);
    }), (function(e) {
        var t = e.$append;
        return t && (0, I.css)([ "margin-right:-1px;border-top-right-radius:0;border-bottom-right-radius:0;border-right:none;" ]);
    }), (function(e) {
        var t = e.$hasError;
        return t && (0, I.css)([ "box-shadow:inset 0 0 0 2px white;border-color:", ";" ], (0, 
        P.pick)({
            enterprise: P.variables.errorColor,
            prisma: P.variables.accentColorNegative
        }));
    }), (function(e) {
        var t = e.$showSelected;
        return t && (0, I.css)([ "display:flex;align-items:center;justify-content:center;" ]);
    }));
    // CONCATENATED MODULE: ./src/Color/Swatch.tsx
    function M() {
        M = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return M.apply(this, arguments);
    }
    function L(e, t) {
        if (e == null) return {};
        var r = U(e, t);
        var n, a;
        if (Object.getOwnPropertySymbols) {
            var o = Object.getOwnPropertySymbols(e);
            for (a = 0; a < o.length; a++) {
                n = o[a];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function U(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var a, o;
        for (o = 0; o < n.length; o++) {
            a = n[o];
            if (t.indexOf(a) >= 0) continue;
            r[a] = e[a];
        }
        return r;
    }
    var W = {
        append: o().bool,
        elementRef: o().oneOfType([ o().func, o().object ]),
        /** @private */
        error: o().bool,
        /** @private */
        noBorder: o().bool,
        /** @private. Call back function when activated. */
        onClick: o().func,
        prepend: o().bool,
        value: o().string,
        /** @private */
        showSelected: o().bool
    };
    var Q =  n().forwardRef((function(e, t) {
        var r = e.append, a = e.error, o = a === void 0 ? false : a, l = e.onClick, i = e.prepend, s = e.value, u = e.noBorder, c = e.showSelected, p = L(e, [ "append", "error", "onClick", "prepend", "value", "noBorder", "showSelected" ]);
        var d = s === null || s === undefined ? (0, _._)("No color") : s;
 // Only valid hexadecimal strings, 'transparent' and `null` are allowed for background colors
        // Invalid values are converted to `undefined`
                var f = s;
        if (!(typeof f === "string" && ((0, O.isValidHexColor)(f) || f === "transparent")) && f !== null && f !== undefined) {
            f = undefined;
        }
        
        return n().createElement(F, M({
            onClick: function e(t) {
                return l === null || l === void 0 ? void 0 : l(t, {
                    value: s
                });
            },
            ref: t,
            $value: f,
            $append: r,
            $prepend: i,
            $hasError: o,
            $noBorder: u,
            $showSelected: c,
            "data-test": "swatch",
            "data-test-value": s
        }, p), c &&  n().createElement(K, {
            "aria-label": (0, _._)("Selected"),
            role: "presentation"
        }),  n().createElement(V(), null, d));
    }));
    Q.propTypes = W;
    /* harmony default export */    const X = Q;
    // CONCATENATED MODULE: ./src/Color/ColorStyles.ts
    var G = j().div.withConfig({
        displayName: "ColorStyles__StyledColor",
        componentId: "jxrost-0"
    })([ "", ";flex-direction:row;" ], P.mixins.reset("flex"));
    var J = j()(X).withConfig({
        displayName: "ColorStyles__StyledInputSwatch",
        componentId: "jxrost-1"
    })([ "", "" ], (0, P.pick)({
        compact: (0, I.css)([ "width:28px;height:28px;" ])
    }));
    var Y = j().div.withConfig({
        displayName: "ColorStyles__StyledInput",
        componentId: "jxrost-2"
    })([ "", ";" ], (0, P.pick)({
        enterprise: (0, I.css)([ "display:flex;width:100%;" ]),
        prisma: (0, I.css)([ "position:relative;", "" ], (function(e) {
            var t = e.$inDropdown;
            return t ? (0, I.css)([ "width:80px;" ]) : (0, I.css)([ "flex:0 1 auto;min-width:0;" ]);
        }))
    }));
    var Z = j()(R()).withConfig({
        displayName: "ColorStyles__StyledInputText",
        componentId: "jxrost-3"
    })([ "", "" ], (function(e) {
        var t = e.$inDropdown;
        return t && (0, I.css)([ "font-size:12px;" ]);
    }));
    var ee = j()(X).withConfig({
        displayName: "ColorStyles__StyledToggleSwatch",
        componentId: "jxrost-4"
    })([ "", "" ], (0, P.pick)({
        prisma: {
            comfortable: (0, I.css)([ "width:40px;height:40px;" ]),
            compact: (0, I.css)([ "width:32px;height:32px;" ])
        },
        enterprise: {
            comfortable: (0, I.css)([ "width:32px;height:32px;" ]),
            compact: (0, I.css)([ "width:28px;height:28px;" ])
        }
    }));
    // CONCATENATED MODULE: external "@splunk/react-icons/ChevronDown"
    const te = require("@splunk/react-icons/ChevronDown");
    var re =  e.n(te);
    // CONCATENATED MODULE: external "@splunk/react-ui/Divider"
    const ne = require("@splunk/react-ui/Divider");
    var ae =  e.n(ne);
    // CONCATENATED MODULE: ./src/Color/PaletteStyles.ts
    var oe = j().div.withConfig({
        displayName: "PaletteStyles__StyledPalette",
        componentId: "qgv9v9-0"
    })([ "width:220px;padding:", ";" ], P.variables.spacingSmall);
    var le = j().ul.withConfig({
        displayName: "PaletteStyles__StyledSwatches",
        componentId: "qgv9v9-1"
    })([ "", ";", "" ], P.mixins.reset("block"), (0, P.pick)({
        enterprise: (0, I.css)([ "margin-bottom:", ";" ], P.variables.spacingQuarter)
    }));
    var ie = j().li.withConfig({
        displayName: "PaletteStyles__StyledSwatch",
        componentId: "qgv9v9-2"
    })([ "", ";margin-right:4px;margin-bottom:4px;vertical-align:top;", "" ], P.mixins.reset("inline-block"), (0, 
    P.pick)({
        enterprise: {
            compact: (0, I.css)([ "&:nth-child(8n){margin-right:0;}" ])
        },
        prisma: (0, I.css)([ "&:nth-child(8n){margin-right:0;}" ])
    }));
    var se = j().div.withConfig({
        displayName: "PaletteStyles__StyledToolBar",
        componentId: "qgv9v9-3"
    })([ "position:relative;display:flex;align-items:center;padding:0;margin-top:", ";margin-bottom:", ";gap:4px;" ], P.variables.spacingSmall, P.variables.spacingHalf);
    var ue = j()($()).withConfig({
        displayName: "PaletteStyles__StyledExpandButton",
        componentId: "qgv9v9-4"
    })([ "position:absolute;bottom:4px;right:4px;padding:2px 4px;border-radius:2px;&:focus{box-shadow:", ";}", "" ], P.variables.focusShadow, (function(e) {
        var t = e.$expanded;
        return t && (0, I.css)([ "transform:rotate(180deg);" ]);
    }));
    var ce = j().div.withConfig({
        displayName: "PaletteStyles__StyledSystemColorPickerWrapper",
        componentId: "qgv9v9-5"
    })([ "position:relative;" ]);
    var pe = j().input.withConfig({
        displayName: "PaletteStyles__StyledSystemColorPicker",
        componentId: "qgv9v9-6"
    })([ "box-sizing:border-box;", "     border:", ";", ";outline:none;appearance:none;padding:1px 10px;", "     margin:0;background:conic-gradient( #ef3434,#ef9a34,#deef34,#78ef34,#34ef56,#34efbc,#34bcef,#3456ef,#7834ef,#de34ef,#ef349a,#ef3434 );background-repeat:no-repeat;cursor:pointer;&::-webkit-color-swatch{border:none;border-radius:2px;}&::-moz-color-swatch{opacity:0;}&:focus{box-shadow:", ";}" ], (0, 
    P.pick)({
        enterprise: {
            comfortable: (0, I.css)([ "width:32px;height:32px;" ]),
            compact: (0, I.css)([ "width:24px;height:24px;" ])
        },
        prisma: (0, I.css)([ "width:24px;height:24px;" ])
    }), (0, P.pick)({
        enterprise: {
            light: (0, I.css)([ "1px solid ", "" ], P.variables.gray60),
            dark: P.variables.border
        },
        prisma: "none"
    }), (0, P.pick)({
        prisma: (0, I.css)([ "border-radius:2px;" ])
    }), (0, P.pick)({
        enterprise: {
            comfortable: (0, I.css)([ "padding:1px 14px;" ])
        }
    }), P.variables.focusShadow);
    var de = j()(D()).withConfig({
        displayName: "PaletteStyles__StyledCheck",
        componentId: "qgv9v9-7"
    })([ "box-sizing:border-box;color:#d3d3d3;height:100%;width:100%;position:absolute;top:0;left:0;pointer-events:none;" ]);
    // CONCATENATED MODULE: ./src/Color/Palette.tsx
    function fe(e) {
        var t = e.expanded, r = e.palette, a = e.value, o = e.swatchToFocusRef, l = e.onSwatchClick, i = e.expandButtonRef, s = e.onPaletteExpand, c = e.onSystemColorPickerChange, p = e.renderInput;
        var d = (0, P.useSplunkTheme)(), f = d.isPrisma, v = d.isLight;
        var m = u()(r.map((function(e) {
            return e && g()(e);
        })), a);
        var y = r.map((function(e, t) {
            var r = e === null ? null : g()(e);
            var i = r === a;
            var s = !m && t === 0 || m && i;
            
            return n().createElement(ie, {
                key: r || "null"
            },  n().createElement(X, {
                value: r,
                ref: s ? o : null,
                onClick: l,
                noBorder: f,
                showSelected: i
            }));
        }));
        var b = function e() {
            var t = !m && a !== "#ffffff" && a !== "#000000" && a !== null && a !== "";
            
            return n().createElement(se, {
                "data-test": "tool-bar"
            }, f &&  n().createElement(P.SplunkThemeProvider, {
                density: "compact"
            }, p && p()),  n().createElement(X, {
                value: null,
                ref: a === null ? o : null,
                onClick: l,
                noBorder: f && !v,
                showSelected: a === null
            }),  n().createElement(X, {
                value: "#ffffff",
                ref: a === "#ffffff" ? o : null,
                onClick: l,
                noBorder: f && !v,
                showSelected: a === "#ffffff"
            }),  n().createElement(X, {
                value: "#000000",
                ref: a === "#000000" ? o : null,
                onClick: l,
                noBorder: f,
                showSelected: a === "#000000"
            }),  n().createElement(ce, null,  n().createElement(pe, {
                type: "color",
                value: a === null ? "" : a,
                onChange: c,
                ref: t ? o : null
            }), t &&  n().createElement(de, {
                "aria-label": (0, _._)("Selected"),
                role: "presentation"
            })));
        };
        var x = y.length === 8 ? 7 : 15;
        var w = h()(y, 0, x);
        var S = h()(y, x);
        if (f) {
            
            return n().createElement(oe, null,  n().createElement("div", {
                style: {
                    position: "relative"
                }
            },  n().createElement(le, null, w, t && S), t && b(),  n().createElement(ue, {
                ref: i,
                $expanded: t,
                onClick: s,
                "data-test": "expand-button"
            },  n().createElement(re(), null))));
        }
        
        return n().createElement(oe, null,  n().createElement(le, null, y),  n().createElement(ae(), null), b(), p && p());
    }
    /* harmony default export */    const ve = fe;
    // CONCATENATED MODULE: ./src/Color/Color.tsx
    function he(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            he = function e(t) {
                return typeof t;
            };
        } else {
            he = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return he(e);
    }
    function me() {
        me = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return me.apply(this, arguments);
    }
    function ye(e) {
        return we(e) || xe(e) || ge(e) || be();
    }
    function be() {
        throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function ge(e, t) {
        if (!e) return;
        if (typeof e === "string") return Se(e, t);
        var r = Object.prototype.toString.call(e).slice(8, -1);
        if (r === "Object" && e.constructor) r = e.constructor.name;
        if (r === "Map" || r === "Set") return Array.from(e);
        if (r === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(r)) return Se(e, t);
    }
    function xe(e) {
        if (typeof Symbol !== "undefined" && Symbol.iterator in Object(e)) return Array.from(e);
    }
    function we(e) {
        if (Array.isArray(e)) return Se(e);
    }
    function Se(e, t) {
        if (t == null || t > e.length) t = e.length;
        for (var r = 0, n = new Array(t); r < t; r++) {
            n[r] = e[r];
        }
        return n;
    }
    function ke(e, t) {
        if (e == null) return {};
        var r = Ce(e, t);
        var n, a;
        if (Object.getOwnPropertySymbols) {
            var o = Object.getOwnPropertySymbols(e);
            for (a = 0; a < o.length; a++) {
                n = o[a];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function Ce(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var a, o;
        for (o = 0; o < n.length; o++) {
            a = n[o];
            if (t.indexOf(a) >= 0) continue;
            r[a] = e[a];
        }
        return r;
    }
    function Ee(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function Pe(e, t) {
        for (var r = 0; r < t.length; r++) {
            var n = t[r];
            n.enumerable = n.enumerable || false;
            n.configurable = true;
            if ("value" in n) n.writable = true;
            Object.defineProperty(e, n.key, n);
        }
    }
    function Oe(e, t, r) {
        if (t) Pe(e.prototype, t);
        if (r) Pe(e, r);
        return e;
    }
    function _e(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) qe(e, t);
    }
    function qe(e, t) {
        qe = Object.setPrototypeOf || function e(t, r) {
            t.__proto__ = r;
            return t;
        };
        return qe(e, t);
    }
    function Ie(e) {
        var t = Re();
        return function r() {
            var n = Be(e), a;
            if (t) {
                var o = Be(this).constructor;
                a = Reflect.construct(n, arguments, o);
            } else {
                a = n.apply(this, arguments);
            }
            return je(this, a);
        };
    }
    function je(e, t) {
        if (t && (he(t) === "object" || typeof t === "function")) {
            return t;
        }
        return Te(e);
    }
    function Te(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function Re() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function Be(e) {
        Be = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return Be(e);
    }
    function Ve(e, t, r) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: r,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = r;
        }
        return e;
    }
    // The default palette in Prisma theme.
        var Ae = [ "#912344", "#D44C20", "#C0891E", "#26AA92", "#2F8AA7", "#2E49AA", "#602CA1", "#BA4ABD" ];
    var De = w()((function(e) {
        var t = [];
        for (var r = 1; r <= e.length * 4; r += 1) {
            var n = e[(r - 1) % e.length];
            if (r > e.length * 3) {
                t.push(n);
            } else {
                var a = 40 - 10 * (r / e.length);
                t.push(k()(n).lighten(a).toString());
            }
        }
        return t;
    }));
 // The tool palette in Prisma theme.
        var Ne = [ "#ffffff", "#000000", null ];
    var $e = {
        append: o().bool,
        defaultValue: o().string,
        describedBy: o().string,
        disabled: o().bool,
        elementRef: o().oneOfType([ o().func, o().object ]),
        error: o().bool,
        hideInput: o().bool,
        labelledBy: o().string,
        name: o().string,
        onChange: o().func,
        palette: o().array,
        prepend: o().bool,
        /** @private */
        required: o().bool,
        /** @private */
        splunkTheme: o().object,
        value: o().string
    };
    var He = {
        append: false,
        disabled: false,
        error: false,
        hideInput: false,
        palette: De(Ae),
        prepend: false
    };
    // TODO: SUI-2551 Deprecate this in favor of ui-utils implementation
        var ze = function e(t) {
        return (0, O.isValidHexColor)(t);
    };
    var Ke = "N/A";
    var Fe = "n/a";
    /*
 * When the dropdown opens:
 * 1. If palette contains the color and the color is selected, the dropdown focus the color.
 * 2. If palette doesn't contain the color:
 *     1. If the current theme `family` is `prisma` the input receives focus.
 *     2. Otherwise the first color swatch receives focus.
 */    var Me =  function(e) {
        _e(r, e);
        var t = Ie(r);
        Oe(r, null, [ {
            key: "hasNull",
            // @docs-props-type ColorPropsBase
            value: function e(t) {
                return t.some((function(e) {
                    return e === null;
                }));
            }
        }, {
            key: "hasTransparent",
            value: function e(t) {
                return t.some((function(e) {
                    return e !== null && g()(e) === "transparent";
                }));
            }
        } ]);
        function r(e) {
            var a;
            Ee(this, r);
            a = t.call(this, e);
            Ve(Te(a), "focusSwatch", void 0);
            Ve(Te(a), "focusExpandButton", void 0);
            Ve(Te(a), "controlledExternally", void 0);
            Ve(Te(a), "dropdown", null);
            Ve(Te(a), "handleSwatchClick", (function(e, t) {
                var r = t.value;
                e.preventDefault();
                var n = a.props, o = n.name, l = n.onChange;
                var i = a.getValue();
                var s = r !== i;
                var u = r === null ? Ke : r;
                if (!a.isControlled()) {
                    a.setState({
                        value: r
                    });
                }
                a.setState({
                    displayValue: u,
                    open: false
                });
                if (s) {
                    l === null || l === void 0 ? void 0 : l({
                        value: r,
                        name: o
                    });
                }
                a.focus();
            }));
            Ve(Te(a), "handlePaletteExpand", (function() {
                a.setState((function(e) {
                    return {
                        expanded: !e.expanded
                    };
                }));
            }));
            Ve(Te(a), "handleTextChange", (function(e, t) {
                var r = t.value;
                a.setState({
                    displayValue: r
                });
            }));
            Ve(Te(a), "handleTextKeyDown", (function(e) {
                if ((0, q.keycode)(e.nativeEvent) === "enter") {
                    e.preventDefault();
                    a.handleRequestClose({
                        reason: "enterKey"
                    });
                    a.focus();
                }
            }));
            Ve(Te(a), "handleButtonClick", (function(e) {
                e.preventDefault();
                a.handleRequestClose({
                    reason: "buttonClick"
                });
                a.focus();
            }));
            Ve(Te(a), "handleRequestClose", (function(e) {
                var t = e.reason;
                var r = a.state.displayValue;
                var n = a.state.value;
                if (g()(r) === Fe) {
                    r = Ke;
                    a.setState({
                        displayValue: r
                    });
                } else if (g()(r) === "transparent") {
                    r = "transparent";
                    a.setState({
                        displayValue: r
                    });
                }
 // If the user types in the hexadecimal number without # in front,
                // we add # automatically.
                                if (r !== Ke && g()(r) !== "transparent" && !y()(r, "#")) {
                    r = "#".concat(r);
                    if (t === "escapeKey") {
                        if (n == null) {
                            a.setState({
                                displayValue: Ke
                            });
                        } else {
                            a.setState({
                                displayValue: n
                            });
                        }
                    } else {
                        a.setState({
                            displayValue: r
                        });
                    }
                }
                if (t !== "contentClick") {
                    a.setState({
                        open: false
                    });
                }
                if (t !== "escapeKey") {
                    a.submitValue(r);
                }
            }));
            Ve(Te(a), "handleRequestOpen", (function() {
                var e = a.props, t = e.splunkTheme, r = e.palette;
                var n = t.isPrisma;
                var o = a.getValue();
                var l = o === null ? Ke : o;
                a.setState({
                    displayValue: l,
                    open: true
                }, (function() {
                    if (a.focusSwatch.current && a.focusSwatch.current.focus) {
                        a.focusSwatch.current.focus();
                    }
                }));
                var i = n ? h()(r, 15) : r;
                var s = !u()(r.map((function(e) {
                    return e && g()(e);
                })), o);
                var c = s || u()(i.map((function(e) {
                    return e && g()(e);
                })), o) || u()(Ne, o);
                if (c && !a.state.expanded) {
                    a.setState({
                        expanded: true
                    });
                }
            }));
            Ve(Te(a), "handleSystemColorPickerChange", (function(e) {
                var t = a.getValue();
                var r = e !== t;
                var n = a.props.name;
                var o = e === null ? Ke : e;
                if (!a.isControlled()) {
                    a.setState({
                        value: e
                    });
                }
                a.setState({
                    displayValue: o
                });
                if (r) {
                    var l, i;
                    (l = (i = a.props).onChange) === null || l === void 0 ? void 0 : l.call(i, {
                        value: e,
                        name: n
                    });
                }
            }));
            a.focusSwatch =  n().createRef();
            a.focusExpandButton =  n().createRef();
            a.controlledExternally = i()(e, "value");
 // value can be hexadecimal color, 'transparent' or null
                        var o = a.controlledExternally ? (0, O.expandShortHandHex)(e.value) : (0, 
            O.expandShortHandHex)(e.defaultValue);
            if (o !== null) {
                o = g()(o);
            }
 // displayValue can be hexadecimal color or 'transparent' to represent 'transparent' or 'N/A' to represent null
                        var l = o === null ? Ke : o;
            a.state = {
                displayValue: l,
                open: false,
                value: o,
                expanded: false
            };
            if (false) {}
            if (false) {
                var s;
            }
            return a;
        }
        Oe(r, [ {
            key: "componentDidUpdate",
            value: function e(t) {
                if (false) {}
                if (false) {}
            }
            /**
     * Standardizes the current value of the Color:
     * returns based on the input being controlled or uncontrolled
     * returns controlled component's value as formatted.
     */        }, {
            key: "getValue",
            value: function e() {
                var t = function e(t) {
                    return t === null ? null : g()(t);
                };
                var r = (0, O.expandShortHandHex)(this.props.value);
                if (this.isControlled()) {
                    return t(r);
                }
                return this.state.value;
            }
        }, {
            key: "submitValue",
            value: function e(t) {
                if (t) {
                    var n = ze(t);
                    if (!n && t !== "transparent" && t !== Ke) {
                        // eslint-disable-next-line no-console
                        console.warn('"'.concat(t, '" is not a valid hex color.'));
                    }
                    if (n || r.hasTransparent(this.props.palette) && t === "transparent" || t === Ke) {
                        var a = (0, O.expandShortHandHex)(t);
                        this.setState({
                            displayValue: a
                        });
                        var o = t !== (this.props.value && g()(this.props.value));
                        var l = this.props.name;
                        var i = t === Ke ? null : t;
                        var s = (0, O.expandShortHandHex)(i);
                        if (t && !this.isControlled()) {
                            this.setState({
                                value: s
                            });
                        }
                        if (o) {
                            var u, c;
                            (u = (c = this.props).onChange) === null || u === void 0 ? void 0 : u.call(c, {
                                value: s,
                                name: l
                            });
                        }
                    }
                }
            }
            /**
     * Place focus on the input.
     */        }, {
            key: "focus",
            value: function e() {
                var t;
                (t = this.dropdown) === null || t === void 0 ? void 0 : t.focus();
            }
        }, {
            key: "isControlled",
            value: function e() {
                return this.controlledExternally;
            }
        }, {
            key: "renderInput",
            value: function e(t) {
                var a = t.textAppend, o = t.textPrepend, l = t.textInline, i = t.inDropdown;
                var s = this.props, u = s.splunkTheme, c = s.required;
                var p = this.state.displayValue;
                var d = u.isPrisma;
                var f = r.hasTransparent(this.props.palette) && g()(p) === "transparent";
                var v = g()(p) === Fe;
                var h = p;
                if (!y()(p, "#")) {
                    h = "#".concat(p);
                }
                h = f ? "transparent" : h;
                h = v ? null : h;
                
                return n().createElement(Y, {
                    $inDropdown: i
                },  n().createElement(Z, {
                    $inDropdown: i,
                    append: a,
                    autoCapitalize: "off",
                    autoComplete: "off",
                    autoCorrect: "off",
                    "aria-label": (0, _._)("Hexadecimal color value"),
                    inline: l,
                    prepend: o,
                    onKeyDown: this.handleTextKeyDown,
                    onChange: this.handleTextChange,
                    required: c,
                    spellCheck: false,
                    value: p
                }), !d &&  n().createElement(J, {
                    "data-test": "textbox-swatch",
                    onClick: this.handleButtonClick,
                    value: h,
                    tabIndex: -1,
                    style: {
                        marginLeft: "1px"
                    }
                }));
            }
        }, {
            key: "renderPalette",
            value: function e() {
                var t = this;
                var r = this.props, a = r.splunkTheme, o = r.palette;
                var l = this.state.expanded;
                var i = a.isPrisma;
                
                return n().createElement(ve, {
                    palette: o,
                    value: this.getValue(),
                    swatchToFocusRef: this.focusSwatch,
                    renderInput: function e() {
                        return i ? t.renderInput({
                            textAppend: false,
                            textPrepend: false,
                            textInline: true,
                            inDropdown: true
                        }) : t.renderInput({
                            textAppend: true,
                            textPrepend: true,
                            textInline: true,
                            inDropdown: false
                        });
                    },
                    expandButtonRef: this.focusExpandButton,
                    onSwatchClick: this.handleSwatchClick,
                    expanded: l,
                    onPaletteExpand: this.handlePaletteExpand,
                    onSystemColorPickerChange: function e(r) {
                        return t.handleSystemColorPickerChange(r.target.value);
                    }
                });
            }
        }, {
            key: "render",
            value: function e() {
                var t = this;
                var a = this.props, o = a.append, l = a.disabled, i = a.describedBy, s = a.elementRef, u = a.error, c = a.hideInput, d = a.labelledBy, v = a.name, h = a.prepend, m = a.splunkTheme, y = ke(a, [ "append", "disabled", "describedBy", "elementRef", "error", "hideInput", "labelledBy", "name", "prepend", "splunkTheme" ]);
                var b = m.isPrisma;
                var g = this.getValue();
                var x = this.state.displayValue;
                var w =  n().createElement(ee, me({
                    append: b && !c ? true : o,
                    "data-test": "color",
                    "data-test-value": x,
                    "aria-describedby": i,
                    "aria-labelledby": d,
                    "aria-invalid": u || undefined,
                    disabled: l,
                    error: u,
                    elementRef: s,
                    name: v,
                    prepend: h,
                    value: g
                }, f()(y, [].concat(ye(p()(r.propTypes)), [ // TS: trick omit
                "onChange" ]))));
                
                return n().createElement(G, null,  n().createElement(E(), {
                    closeReasons: [ "clickAway", "escapeKey", "offScreen", "toggleClick" ],
                    onRequestClose: this.handleRequestClose,
                    onRequestOpen: this.handleRequestOpen,
                    open: this.state.open,
                    ref: function e(r) {
                        t.dropdown = r;
                    },
                    retainFocus: true,
                    takeFocus: false,
                    toggle: w
                }, this.renderPalette()), b && !c && this.renderInput({
                    textAppend: false,
                    textPrepend: true,
                    textInline: true,
                    inDropdown: false
                }));
            }
        } ]);
        return r;
    }(r.Component);
    Ve(Me, "propTypes", $e);
    Ve(Me, "defaultProps", He);
    var Le = (0, P.withSplunkTheme)(Me);
    Le.propTypes = Me.propTypes;
    // TODO: SUI-2551 Deprecate this in favor of ui-utils implementation
    /* harmony default export */    const Ue = Le;
    // CONCATENATED MODULE: ./src/Color/index.ts
    module.exports = t;
    /******/})();
