// @splunk/react-ui/ComboBox.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var n = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(n, {
                a: n
            });
            /******/            return n;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, n) => {
            /******/ for (var o in n) {
                /******/ if (e.o(n, o) && !e.o(t, o)) {
                    /******/ Object.defineProperty(t, o, {
                        enumerable: true,
                        get: n[o]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        Divider: () => /* reexport */ I.Divider,
        Heading: () => /* reexport */ I.Heading,
        Option: () => /* reexport */ X,
        default: () => /* reexport */ he
    });
    // CONCATENATED MODULE: external "react"
    const n = require("react");
    var o =  e.n(n);
    // CONCATENATED MODULE: external "prop-types"
    const r = require("prop-types");
    var i =  e.n(r);
    // CONCATENATED MODULE: external "lodash/extend"
    const a = require("lodash/extend");
    var l =  e.n(a);
    // CONCATENATED MODULE: external "lodash/has"
    const u = require("lodash/has");
    var c =  e.n(u);
    // CONCATENATED MODULE: external "lodash/memoize"
    const s = require("lodash/memoize");
    var f =  e.n(s);
    // CONCATENATED MODULE: external "lodash/omit"
    const p = require("lodash/omit");
    var d =  e.n(p);
    // CONCATENATED MODULE: external "lodash/pick"
    const v = require("lodash/pick");
    var h =  e.n(v);
    // CONCATENATED MODULE: external "lodash/some"
    const y = require("lodash/some");
    var b =  e.n(y);
    // CONCATENATED MODULE: external "@splunk/ui-utils/filter"
    const m = require("@splunk/ui-utils/filter");
    // CONCATENATED MODULE: external "@splunk/ui-utils/id"
    const g = require("@splunk/ui-utils/id");
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const O = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "@splunk/ui-utils/keyboard"
    const S = require("@splunk/ui-utils/keyboard");
    // CONCATENATED MODULE: external "@splunk/react-ui/Menu"
    const I = require("@splunk/react-ui/Menu");
    // CONCATENATED MODULE: external "@splunk/react-ui/Popover"
    const C = require("@splunk/react-ui/Popover");
    var w =  e.n(C);
    // CONCATENATED MODULE: external "@splunk/react-ui/ResultsMenu"
    const k = require("@splunk/react-ui/ResultsMenu");
    var P =  e.n(k);
    // CONCATENATED MODULE: external "@splunk/react-ui/Text"
    const j = require("@splunk/react-ui/Text");
    var x =  e.n(j);
    // CONCATENATED MODULE: ./src/ComboBox/Option.tsx
    function R(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            R = function e(t) {
                return typeof t;
            };
        } else {
            R = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return R(e);
    }
    function M() {
        M = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var n = arguments[t];
                for (var o in n) {
                    if (Object.prototype.hasOwnProperty.call(n, o)) {
                        e[o] = n[o];
                    }
                }
            }
            return e;
        };
        return M.apply(this, arguments);
    }
    function V(e, t) {
        if (e == null) return {};
        var n = D(e, t);
        var o, r;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (r = 0; r < i.length; r++) {
                o = i[r];
                if (t.indexOf(o) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, o)) continue;
                n[o] = e[o];
            }
        }
        return n;
    }
    function D(e, t) {
        if (e == null) return {};
        var n = {};
        var o = Object.keys(e);
        var r, i;
        for (i = 0; i < o.length; i++) {
            r = o[i];
            if (t.indexOf(r) >= 0) continue;
            n[r] = e[r];
        }
        return n;
    }
    function E(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function _(e, t) {
        for (var n = 0; n < t.length; n++) {
            var o = t[n];
            o.enumerable = o.enumerable || false;
            o.configurable = true;
            if ("value" in o) o.writable = true;
            Object.defineProperty(e, o.key, o);
        }
    }
    function q(e, t, n) {
        if (t) _(e.prototype, t);
        if (n) _(e, n);
        return e;
    }
    function A(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) B(e, t);
    }
    function B(e, t) {
        B = Object.setPrototypeOf || function e(t, n) {
            t.__proto__ = n;
            return t;
        };
        return B(e, t);
    }
    function T(e) {
        var t = H();
        return function n() {
            var o = z(e), r;
            if (t) {
                var i = z(this).constructor;
                r = Reflect.construct(o, arguments, i);
            } else {
                r = o.apply(this, arguments);
            }
            return L(this, r);
        };
    }
    function L(e, t) {
        if (t && (R(t) === "object" || typeof t === "function")) {
            return t;
        }
        return F(e);
    }
    function F(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function H() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function z(e) {
        z = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return z(e);
    }
    function K(e, t, n) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: n,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = n;
        }
        return e;
    }
    var N = {
        /** @private */
        active: i().bool,
        description: i().string,
        descriptionPosition: i().oneOf([ "right", "bottom" ]),
        disabled: i().bool,
        icon: i().node,
        label: i().string,
        matchRanges: i().arrayOf(i().shape({
            start: i().number.isRequired,
            end: i().number.isRequired
        })),
        /** @private */
        onClick: i().func,
        truncate: i().bool,
        value: i().string.isRequired
    };
    var W = {
        active: false,
        descriptionPosition: "bottom",
        disabled: false,
        truncate: false
    };
    /**
 * An option within a `ComboBox`. This inherits from
 * [PureComponent](https://reactjs.org/docs/react-api.html#reactpurecomponent)
 * so any elements passed to it must also be pure.
 */    var U =  function(e) {
        A(n, e);
        var t = T(n);
        function n() {
            var e;
            E(this, n);
            for (var o = arguments.length, r = new Array(o), i = 0; i < o; i++) {
                r[i] = arguments[i];
            }
            e = t.call.apply(t, [ this ].concat(r));
            K(F(e), "item", null);
            K(F(e), "handleClick", (function(t) {
                var n = e.props, o = n.disabled, r = n.onClick, i = n.value;
                if (!o) {
                    r === null || r === void 0 ? void 0 : r(t, {
                        value: i
                    });
                }
            }));
            return e;
        }
        q(n, [ {
            key: "scrollIntoViewIfNeeded",
            value: function e() {
                var t;
                (t = this.item) === null || t === void 0 ? void 0 : t.scrollIntoViewIfNeeded();
            }
        }, {
            key: "render",
            value: function e() {
                var t = this;
                // eslint-disable-next-line no-unused-vars
                                var n = this.props, r = n.value, i = n.label, a = V(n, [ "value", "label" ]);
                var l = i === undefined ? r : i;
                
                return o().createElement(I.Item, M({
                    ref: function e(n) {
                        t.item = n;
                    },
                    "data-test": "option",
                    "data-test-value": r
                }, a, {
                    onClick: this.handleClick,
                    role: "option",
                    "aria-selected": false
                }), l);
            }
        } ]);
        return n;
    }(n.PureComponent);
    K(U, "propTypes", N);
    K(U, "defaultProps", W);
    /* harmony default export */    const X = U;
    // CONCATENATED MODULE: ./src/utils/updateReactRef.ts
    /**
 * Updates a React ref. Callback refs and object refs (from `createRef` and `useRef`) are supported.
 *
 * @param ref - The React callback or object ref. Can be `null` or `undefined`.
 * @param current - The new value of the ref.
 */
    function Y(e, t) {
        if (e) {
            if (typeof e === "function") {
                e(t);
            } else {
                // the public signature of this util uses React.Ref<T> to mirror the way React types refs.
                // the intention here is to signal "we will take care of setting 'current', not you".
                e.current = t;
 // eslint-disable-line no-param-reassign
                        }
        }
    }
    // CONCATENATED MODULE: ./src/ComboBox/ComboBox.tsx
    function G(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            G = function e(t) {
                return typeof t;
            };
        } else {
            G = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return G(e);
    }
    function J() {
        J = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var n = arguments[t];
                for (var o in n) {
                    if (Object.prototype.hasOwnProperty.call(n, o)) {
                        e[o] = n[o];
                    }
                }
            }
            return e;
        };
        return J.apply(this, arguments);
    }
    function Q(e, t) {
        var n = Object.keys(e);
        if (Object.getOwnPropertySymbols) {
            var o = Object.getOwnPropertySymbols(e);
            if (t) o = o.filter((function(t) {
                return Object.getOwnPropertyDescriptor(e, t).enumerable;
            }));
            n.push.apply(n, o);
        }
        return n;
    }
    function Z(e) {
        for (var t = 1; t < arguments.length; t++) {
            var n = arguments[t] != null ? arguments[t] : {};
            if (t % 2) {
                Q(Object(n), true).forEach((function(t) {
                    ce(e, t, n[t]);
                }));
            } else if (Object.getOwnPropertyDescriptors) {
                Object.defineProperties(e, Object.getOwnPropertyDescriptors(n));
            } else {
                Q(Object(n)).forEach((function(t) {
                    Object.defineProperty(e, t, Object.getOwnPropertyDescriptor(n, t));
                }));
            }
        }
        return e;
    }
    function $(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function ee(e, t) {
        for (var n = 0; n < t.length; n++) {
            var o = t[n];
            o.enumerable = o.enumerable || false;
            o.configurable = true;
            if ("value" in o) o.writable = true;
            Object.defineProperty(e, o.key, o);
        }
    }
    function te(e, t, n) {
        if (t) ee(e.prototype, t);
        if (n) ee(e, n);
        return e;
    }
    function ne(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) oe(e, t);
    }
    function oe(e, t) {
        oe = Object.setPrototypeOf || function e(t, n) {
            t.__proto__ = n;
            return t;
        };
        return oe(e, t);
    }
    function re(e) {
        var t = le();
        return function n() {
            var o = ue(e), r;
            if (t) {
                var i = ue(this).constructor;
                r = Reflect.construct(o, arguments, i);
            } else {
                r = o.apply(this, arguments);
            }
            return ie(this, r);
        };
    }
    function ie(e, t) {
        if (t && (G(t) === "object" || typeof t === "function")) {
            return t;
        }
        return ae(e);
    }
    function ae(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function le() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function ue(e) {
        ue = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return ue(e);
    }
    function ce(e, t, n) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: n,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = n;
        }
        return e;
    }
    var se = {
        animateLoading: i().bool,
        append: i().bool,
        children: i().node,
        controlledFilter: i().bool,
        defaultPlacement: i().oneOf([ "above", "below", "vertical" ]),
        defaultValue: i().string,
        describedBy: i().string,
        disabled: i().bool,
        elementRef: i().oneOfType([ i().func, i().object ]),
        error: i().bool,
        footerMessage: i().node,
        inline: i().bool,
        inputRef: i().oneOfType([ i().func, i().object ]),
        isLoadingOptions: i().bool,
        labelledBy: i().string,
        loadingMessage: i().node,
        menuStyle: i().object,
        name: i().string,
        noOptionsMessage: i().node,
        onBlur: i().func,
        onChange: i().func,
        onClose: i().func,
        onFocus: i().func,
        onKeyDown: i().func,
        onOpen: i().func,
        onScroll: i().func,
        onScrollBottom: i().func,
        onSelect: i().func,
        placeholder: i().string,
        prepend: i().bool,
        size: i().oneOf([ "small", "medium", "large" ]),
        value: i().string,
        /** @private. */
        virtualization: i().number
    };
    var fe = {
        animateLoading: false,
        controlledFilter: false,
        defaultPlacement: "vertical",
        disabled: false,
        error: false,
        inline: false,
        isLoadingOptions: false,
        menuStyle: {},
        placeholder: (0, O._)("Select..."),
        size: "medium"
    };
    function pe(e, t) {
        var n = t.clientX, o = t.clientY;
        var r = e.getBoundingClientRect(), i = r.top, a = r.left, l = r.bottom, u = r.right;
        return n > a && n < u && o > i && o < l;
    }
    var de = f()((function(e) {
        var t = e.anchorWidth, n = e.maxHeight, o = e.menuStyle;
        return l()({
            maxHeight: n,
            overflow: "auto",
            width: Math.max(t !== null && t !== void 0 ? t : 0, 200)
        }, o);
    }));
    /**
 * `ComboBox` allows the user to select a predefined string or enter a new value. Unlike `Select`
 * and `Multiselect`, `Option` value must always be a string and `Option` does not have a label
 * property.
 */    var ve =  function(e) {
        ne(r, e);
        var t = re(r);
        // @docs-props-type ComboBoxPropsBase
                function r(e) {
            var i;
            $(this, r);
            i = t.call(this, e);
            ce(ae(i), "activeItemId", void 0);
            ce(ae(i), "activeValue", void 0);
            ce(ae(i), "availableOptionCount", 0);
            ce(ae(i), "controlledExternally", void 0);
            ce(ae(i), "focusCalledInternally", false);
            ce(ae(i), "options", void 0);
            ce(ae(i), "popoverId", void 0);
            ce(ae(i), "previousActiveIndex", null);
            ce(ae(i), "textInput", null);
            ce(ae(i), "handleInputMount", (function(e) {
                i.textInput = e;
                Y(i.props.inputRef, e);
            }));
            ce(ae(i), "handleAnchorMount", (function(e) {
                i.setState({
                    anchor: e
                });
                Y(i.props.elementRef, e);
            }));
            ce(ae(i), "handleActiveOptionMount", (function(e) {
                if (i.previousActiveIndex !== i.state.activeIndex) {
                    e === null || e === void 0 ? void 0 : e.scrollIntoViewIfNeeded();
                }
            }));
            ce(ae(i), "handleInputFocus", (function(e, t) {
                var n, o;
                /* SUI-930 On IE 11 this handler is essentially deferred after calling
       * this.textInput.focus(). this.focusCalledInternally enables the focus event to be ignored
       * when the menu closes. */                if (i.focusCalledInternally) {
                    i.focusCalledInternally = false;
                } else {
                    i.open();
                }
                (n = (o = i.props).onFocus) === null || n === void 0 ? void 0 : n.call(o, e, t);
            }));
            ce(ae(i), "handleChange", (function(e, t) {
                var n, o;
                var r = t.value, a = t.selectedOption;
                e.stopPropagation();
                var l = i.props.name;
                var u = a ? undefined : r;
                if (i.isControlled()) {
                    i.setState({
                        activeIndex: 0,
                        typedValue: u
                    });
                } else {
                    i.setState({
                        value: r,
                        activeIndex: 0,
                        typedValue: u
                    });
                }
                (n = (o = i.props).onChange) === null || n === void 0 ? void 0 : n.call(o, e, {
                    value: r,
                    name: l
                });
            }));
            ce(ae(i), "handleSelectOption", (function(e, t) {
                i.handleChange(e, Z(Z({}, t), {}, {
                    selectedOption: true
                }));
                i.focusCalledInternally = true;
                i.focus();
                i.close();
            }));
            ce(ae(i), "handleInputKeyDown", (function(e) {
                var t = i.props, o = t.children, r = t.onKeyDown, a = t.onScrollBottom;
                var l = i.availableOptionCount;
                var u = i.activeValue;
                var c = (0, S.keycode)(e.nativeEvent);
                if (i.state.open) {
                    switch (c) {
                      case "enter":
                        {
                            if (u) {
                                i.handleSelectOption(e, {
                                    value: u
                                });
                            }
                            break;
                        }

                      case "tab":
                        i.close();
                        break;

                      case "down":
                        i.setState((function(e) {
                            return {
                                activeIndex: Math.min(e.activeIndex + 1, l - 1)
                            };
                        }));
                        if (o && a) {
                            var s = n.Children.count(o) - 2;
                            if (i.state.activeIndex === s) {
                                i.handleScrollBottom(e);
                            }
                        }
                        break;

                      case "up":
                        i.setState((function(e) {
                            return {
                                activeIndex: Math.max(e.activeIndex - 1, 0)
                            };
                        }));
                        break;

                      default:
 // do nothing
                                        }
                } else if ((0, S.addsCharacter)(e.nativeEvent) !== false || // Safari 9.0 returns undefined
                c === "enter" || c === "backspace" || c === "down" || c === "up") {
                    i.open();
                }
                r === null || r === void 0 ? void 0 : r(e);
            }));
            ce(ae(i), "handleInputClick", (function() {
                if (!i.state.open && !i.props.disabled) {
                    i.open();
                }
            }));
            ce(ae(i), "handleRequestClose", (function(e) {
                var t = e.event, n = e.reason;
                var o = n === "offScreen" || n === "escapeKey" || n === "clickAway" && i.state.anchor && !pe(i.state.anchor, t);
                if (o) {
                    i.close();
                }
            }));
            ce(ae(i), "handleScrollBottom", (function(e) {
                if (i.state.open && !i.props.isLoadingOptions) {
                    var t, n;
                    (t = (n = i.props).onScrollBottom) === null || t === void 0 ? void 0 : t.call(n, e);
                }
            }));
            ce(ae(i), "prepareOptions", (function() {
                var e = i.props, t = e.children, r = e.controlledFilter;
                var a = i.state, l = a.activeIndex, u = a.typedValue;
                var s = i.getValue();
                var f = n.Children.toArray(t).filter(n.isValidElement);
                var p = b()(f, (function(e) {
                    if (e.type === I.Heading) {
                        return false;
                    }
                    if (e.props.label === undefined) {
                        return e.props.value === s;
                    }
                    if (e.props.label !== undefined && u !== undefined) {
                        return e.props.label === s;
                    }
                    return e.props.value === s;
                }));
                if (!p && s) {
                    f.unshift( o().createElement(X, {
                        description: (0, O._)("(New value)"),
                        descriptionPosition: "right",
                        key: "currentValueOption",
                        value: s,
                        "data-test-current-value-option": s
                    }));
                }
 // Highlight Active
                                i.availableOptionCount = 0;
                i.activeValue = undefined;
                var d = (0, m.stringToKeywords)(u ? s : i.getDisplayValue());
                i.options = (r ? f : f.filter((function(e) {
                    if (e.props.label !== undefined) {
                        return (0, m.testPhrase)(e.props.label, d);
                    }
                    if (e.props.value !== undefined) {
                        return (0, m.testPhrase)(e.props.value, d);
                    }
                    return true;
 // Keep all headers and non-interactive options
                                }))).map((function(e, t) {
                    if (!c()(e.props, "active")) {
                        // ignore Headings and Dividers
                        return e;
                    }
                    var o = i.availableOptionCount === l;
                    i.availableOptionCount += 1;
                    var a = e.props.label !== undefined ? e.props.label : e.props.value;
                    var u = e.props.matchRanges;
                    var f = !r && s && (p || t > 0) ? (0, m.keywordLocations)(a, d) || undefined : undefined;
                    if (o) {
                        if (!e.props.disabled) {
                            i.activeValue = e.props.value;
                        }
                        
                        return (0, n.cloneElement)(e, {
                            ref: i.handleActiveOptionMount,
                            id: i.activeItemId,
                            onClick: i.handleSelectOption,
                            matchRanges: u || f,
                            active: true
                        });
                    }
                    
                    return (0, n.cloneElement)(e, {
                        onClick: i.handleSelectOption,
                        matchRanges: u || f
                    });
                }));
            }));
            ce(ae(i), "renderMenu", (function(e) {
                var t = e.anchorWidth, n = e.maxHeight;
                var r = i.props, a = r.isLoadingOptions, l = r.menuStyle, u = r.onScrollBottom, c = r.virtualization;
                var s = de({
                    anchorWidth: t,
                    maxHeight: n,
                    menuStyle: l
                });
                var f = Z({
                    style: s,
                    controlledExternally: true,
                    onScrollBottom: u ? i.handleScrollBottom : undefined,
                    isLoading: a
                }, h()(i.props, "className", "noOptionsMessage", "footerMessage", "animateLoading", "loadingMessage", "onScroll"));
                if (c) {
                    
                    return o().createElement(k.VirtualizedResultsMenu, J({
                        virtualization: c
                    }, f), i.options);
                }
                
                return o().createElement(P(), f, i.options);
            }));
            i.state = {
                activeIndex: 0,
                anchor: null,
                typedValue: undefined,
                open: false,
                value: e.defaultValue || ""
            };
            i.controlledExternally = c()(e, "value");
            i.popoverId = (0, g.createDOMID)("popover");
            i.activeItemId = (0, g.createDOMID)("active-item");
            if (false) {}
            if (false) {}
            return i;
        }
        te(r, [ {
            key: "componentDidUpdate",
            value: function e(t, n) {
                if (false) {}
                this.previousActiveIndex = n.activeIndex;
            }
        }, {
            key: "getValue",
            value: function e() {
                return this.isControlled() ? this.props.value : this.state.value;
            }
        }, {
            key: "getDisplayValue",
            value: function e() {
                var t = this.getValue();
                var o = this.state.typedValue;
                var r = n.Children.toArray(this.props.children).filter(n.isValidElement);
                var i = !o && r.find((function(e) {
                    return e.props.label !== undefined && e.props.value === t;
                }));
                if (i) {
                    return i === null || i === void 0 ? void 0 : i.props.label;
                }
                return t;
            }
        }, {
            key: "open",
            value: function e() {
                var t = this;
                this.setState({
                    open: true,
                    activeIndex: 0
                }, (function() {
                    var e, n;
                    (e = (n = t.props).onOpen) === null || e === void 0 ? void 0 : e.call(n);
                }));
            }
        }, {
            key: "close",
            value: function e() {
                var t = this;
                this.setState({
                    open: false
                }, (function() {
                    var e, n;
                    t.previousActiveIndex = null;
                    (e = (n = t.props).onClose) === null || e === void 0 ? void 0 : e.call(n);
                }));
            }
            /**
     * Focus the `ComboBox`.
     */        }, {
            key: "focus",
            value: function e() {
                var t;
                (t = this.textInput) === null || t === void 0 ? void 0 : t.focus();
            }
        }, {
            key: "isControlled",
            value: function e() {
                return this.controlledExternally;
            }
        }, {
            key: "render",
            value: function e() {
                var t = this.props.defaultPlacement;
                var n = this.state, r = n.anchor, i = n.open;
                var a = this.getValue();
                var l = this.getDisplayValue();
                this.prepareOptions();
                
                return o().createElement(x(), J({
                    autoCapitalize: "off",
                    autoComplete: "off",
                    autoCorrect: "off",
                    spellCheck: false,
                    canClear: true,
                    "data-test": "combo-box"
                }, d()(this.props, "animateLoading", "className", "controlledFilter", "defaultValue", "footerMessage", "isLoadingOptions", "loadingMessage", "menuStyle", "noOptionsMessage", "onClose", "onOpen", "onScroll", "onScrollBottom", "spellCheck"), {
                    "data-test-popover-id": this.popoverId,
                    "data-test-label": l,
                    "data-test-value": a,
                    "data-test-open": i && !!r,
                    onFocus: this.handleInputFocus,
                    onClick: this.handleInputClick,
                    onChange: this.handleChange,
                    onKeyDown: this.handleInputKeyDown,
                    inputRef: this.handleInputMount,
                    elementRef: this.handleAnchorMount,
                    role: "combobox",
                    value: l,
                    "aria-activedescendant": this.state.open && Array.isArray(this.options) && this.options.length > 0 ? this.activeItemId : undefined,
                    "aria-expanded": i,
                    "aria-label": (0, O._)("Value input"),
                    "aria-controls": i ? this.popoverId : undefined
                }),  o().createElement(w(), {
                    anchor: r,
                    autoCloseWhenOffScreen: true,
                    canCoverAnchor: false,
                    defaultPlacement: t,
                    id: this.popoverId,
                    onRequestClose: this.handleRequestClose,
                    open: i && !!r,
                    repositionMode: "flip"
                }, this.renderMenu));
            }
        } ]);
        return r;
    }(n.Component);
    ce(ve, "propTypes", se);
    ce(ve, "defaultProps", fe);
    ce(ve, "Option", X);
    ce(ve, "Divider", I.Divider);
    ce(ve, "Heading", I.Heading);
    /* harmony default export */    const he = ve;
    // CONCATENATED MODULE: ./src/ComboBox/index.ts
    module.exports = t;
    /******/})();
