// @splunk/react-ui/Concertina.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var n = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(n, {
                a: n
            });
            /******/            return n;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, n) => {
            /******/ for (var r in n) {
                /******/ if (e.o(n, r) && !e.o(t, r)) {
                    /******/ Object.defineProperty(t, r, {
                        enumerable: true,
                        get: n[r]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        Panel: () => /* reexport */ Pe,
        default: () => /* reexport */ Ve
    });
    // CONCATENATED MODULE: external "react"
    const n = require("react");
    var r =  e.n(n);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var i =  e.n(o);
    // CONCATENATED MODULE: external "lodash/debounce"
    const a = require("lodash/debounce");
    var l =  e.n(a);
    // CONCATENATED MODULE: external "lodash/take"
    const s = require("lodash/take");
    var c =  e.n(s);
    // CONCATENATED MODULE: external "lodash/takeRight"
    const u = require("lodash/takeRight");
    var f =  e.n(u);
    // CONCATENATED MODULE: external "@splunk/react-ui/EventListener"
    const p = require("@splunk/react-ui/EventListener");
    var d =  e.n(p);
    // CONCATENATED MODULE: external "lodash/keys"
    const h = require("lodash/keys");
    var y =  e.n(h);
    // CONCATENATED MODULE: external "lodash/omit"
    const b = require("lodash/omit");
    var v =  e.n(b);
    // CONCATENATED MODULE: external "styled-components"
    const g = require("styled-components");
    var m =  e.n(g);
    // CONCATENATED MODULE: external "@splunk/react-ui/Clickable"
    const k = require("@splunk/react-ui/Clickable");
    var S =  e.n(k);
    // CONCATENATED MODULE: external "@splunk/themes"
    const w = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/Concertina/HeadingStyles.ts
    var C = m()(S()).withConfig({
        displayName: "HeadingStyles__StyledClickable",
        componentId: "sc-1h92oaj-0"
    })([ "display:block;width:100%;cursor:pointer;position:relative;padding:6px ", ";line-height:", ";border-radius:0;color:", ";", " &:focus:not([disabled]){box-shadow:", ";}", "" ], w.variables.spacing, w.variables.lineHeight, w.variables.textColor, (0, 
    w.pickVariant)("$inner", {
        true: (0, g.css)([ "*:not(:first-child) > &&{border-top:2px solid ", ";}*:not(:last-child) > &&{border-bottom:2px solid ", ";}" ], w.variables.backgroundColor, w.variables.backgroundColor),
        false: (0, g.css)([ "& + &&{border-top:2px solid ", ";}" ], w.variables.backgroundColor)
    }), w.variables.focusShadowInset, (0, w.pickVariant)("$status", {
        normal: (0, g.css)([ "background-color:", ";&:hover:not([disabled]){background-color:", ";}" ], (0, 
        w.pick)({
            dark: w.variables.gray45,
            light: w.variables.gray96
        }), (0, w.pick)({
            dark: w.variables.gray30,
            light: w.variables.gray92
        })),
        disabled: (0, g.css)([ "background-color:", ";" ], (0, w.pick)({
            dark: w.variables.gray80,
            light: w.variables.gray96
        })),
        error: (0, g.css)([ "background-color:", ";color:", ";&:hover:not([disabled]){background-color:", ";}" ], w.variables.errorColor, w.variables.white, w.variables.errorColorD10),
        warning: (0, g.css)([ "background-color:", ";color:", ";&:hover:not([disabled]){background-color:", ";box-shadow:none;}" ], w.variables.warningColor, w.variables.white, w.variables.warningColorD10)
    }));
    var O = m().span.withConfig({
        displayName: "HeadingStyles__StyledHeadingContent",
        componentId: "sc-1h92oaj-1"
    })([ "display:flex;justify-content:space-between;" ]);
    var P = m().span.withConfig({
        displayName: "HeadingStyles__StyledDescription",
        componentId: "sc-1h92oaj-2"
    })([ "font-size:", ";" ], w.variables.fontSizeSmall);
    // CONCATENATED MODULE: ./src/Concertina/Heading.tsx
    function j(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            j = function e(t) {
                return typeof t;
            };
        } else {
            j = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return j(e);
    }
    function E() {
        E = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var n = arguments[t];
                for (var r in n) {
                    if (Object.prototype.hasOwnProperty.call(n, r)) {
                        e[r] = n[r];
                    }
                }
            }
            return e;
        };
        return E.apply(this, arguments);
    }
    function x(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function _(e, t) {
        for (var n = 0; n < t.length; n++) {
            var r = t[n];
            r.enumerable = r.enumerable || false;
            r.configurable = true;
            if ("value" in r) r.writable = true;
            Object.defineProperty(e, r.key, r);
        }
    }
    function H(e, t, n) {
        if (t) _(e.prototype, t);
        if (n) _(e, n);
        return e;
    }
    function T(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) R(e, t);
    }
    function R(e, t) {
        R = Object.setPrototypeOf || function e(t, n) {
            t.__proto__ = n;
            return t;
        };
        return R(e, t);
    }
    function I(e) {
        var t = M();
        return function n() {
            var r = A(e), o;
            if (t) {
                var i = A(this).constructor;
                o = Reflect.construct(r, arguments, i);
            } else {
                o = r.apply(this, arguments);
            }
            return q(this, o);
        };
    }
    function q(e, t) {
        if (t && (j(t) === "object" || typeof t === "function")) {
            return t;
        }
        return z(e);
    }
    function z(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function M() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function A(e) {
        A = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return A(e);
    }
    function W(e, t, n) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: n,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = n;
        }
        return e;
    }
    var B = {
        children: i().node.isRequired,
        description: i().node,
        /** @private */
        index: i().number.isRequired,
        /** @private */
        onClick: i().func,
        /** @private */
        panelId: i().string,
        /** @private */
        position: i().oneOf([ "top", "inner", "bottom" ]),
        status: i().oneOf([ "normal", "warning", "error", "disabled" ])
    };
    var D = {
        position: "inner"
    };
    var N =  function(e) {
        T(n, e);
        var t = I(n);
        function n() {
            var e;
            x(this, n);
            for (var r = arguments.length, o = new Array(r), i = 0; i < r; i++) {
                o[i] = arguments[i];
            }
            e = t.call.apply(t, [ this ].concat(o));
            W(z(e), "container", null);
            W(z(e), "handleClick", (function(t) {
                var n, r;
                (n = (r = e.props).onClick) === null || n === void 0 ? void 0 : n.call(r, t, {
                    index: e.props.index,
                    panelId: e.props.panelId,
                    position: e.props.position
                });
            }));
            W(z(e), "handleMount", (function(t) {
                e.container = t;
            }));
            return e;
        }
        H(n, [ {
            key: "focus",
            value: function e() {
                var t;
                (t = this.container) === null || t === void 0 ? void 0 : t.focus();
            }
        }, {
            key: "render",
            value: function e() {
                var t = this.props, o = t.children, i = t.description, a = t.panelId, l = t.position, s = t.status;
                
                return r().createElement(C, E({
                    disabled: s === "disabled",
                    $inner: l === "inner",
                    $status: s,
                    onClick: this.handleClick,
                    ref: this.handleMount,
                    "data-test": "heading",
                    "data-test-panel-id": a
                }, v()(this.props, y()(n.propTypes))),  r().createElement(O, null,  r().createElement("span", {
                    "data-concertina-role": "title"
                }, o), i &&  r().createElement(P, {
                    "data-concertina-role": "description"
                }, i)));
            }
        } ]);
        return n;
    }(n.Component);
    W(N, "propTypes", B);
    W(N, "defaultProps", D);
    /* harmony default export */    const $ = N;
    // CONCATENATED MODULE: external "lodash/throttle"
    const U = require("lodash/throttle");
    var V =  e.n(U);
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const L = require("@splunk/react-ui/Box");
    var F =  e.n(L);
    // CONCATENATED MODULE: external "@splunk/react-ui/Scroll"
    const G = require("@splunk/react-ui/Scroll");
    var J =  e.n(G);
    // CONCATENATED MODULE: ./src/Concertina/ConcertinaStyles.ts
    var K = m()(J()).withConfig({
        displayName: "ConcertinaStyles__StyledScroll",
        componentId: "sc-1fzuh39-0"
    })([ "position:absolute;left:0;top:0;right:0;bottom:0;" ]);
    var Q = m()(K.withComponent(F())).withConfig({
        displayName: "ConcertinaStyles__StyledBox",
        componentId: "sc-1fzuh39-1"
    })([ "overflow:hidden;" ]);
    var X = m().div.withConfig({
        displayName: "ConcertinaStyles__StyledTop",
        componentId: "sc-1fzuh39-2"
    })([ "position:absolute;left:0;top:0;right:0;z-index:1;background-color:", ";&::after{content:'';position:absolute;bottom:-10px;height:10px;left:0;right:0;background-image:linear-gradient(180deg,rgba(0,0,0,0.15),rgba(0,0,0,0));}" ], w.variables.backgroundColor);
    var Y = m().div.withConfig({
        displayName: "ConcertinaStyles__StyledBottom",
        componentId: "sc-1fzuh39-3"
    })([ "position:absolute;left:0;bottom:0;right:0;background-color:", ";&::before{content:'';position:absolute;top:-5px;height:5px;left:0;right:0;background-image:linear-gradient(0deg,rgba(0,0,0,0.15),rgba(0,0,0,0));}" ], w.variables.backgroundColor);
    var Z = m().div.withConfig({
        displayName: "ConcertinaStyles__StyledPanelBody",
        componentId: "sc-1fzuh39-4"
    })([ "", "" ], (0, w.pickVariant)("$status", {
        error: {
            dark: (0, g.css)([ "background-color:", ";" ], w.mixins.colorWithAlpha(w.variables.errorColor, .5)),
            light: (0, g.css)([ "background-color:", ";" ], w.variables.errorColorL50)
        },
        warning: {
            dark: (0, g.css)([ "background-color:", ";" ], w.mixins.colorWithAlpha(w.variables.warningColor, .5)),
            light: (0, g.css)([ "background-color:", ";" ], w.variables.warningColorL50)
        },
        disabled: (0, g.css)([ "display:none;" ])
    }));
    // CONCATENATED MODULE: ./src/utils/updateReactRef.ts
    /**
 * Updates a React ref. Callback refs and object refs (from `createRef` and `useRef`) are supported.
 *
 * @param ref - The React callback or object ref. Can be `null` or `undefined`.
 * @param current - The new value of the ref.
 */
    function ee(e, t) {
        if (e) {
            if (typeof e === "function") {
                e(t);
            } else {
                // the public signature of this util uses React.Ref<T> to mirror the way React types refs.
                // the intention here is to signal "we will take care of setting 'current', not you".
                e.current = t;
 // eslint-disable-line no-param-reassign
                        }
        }
    }
    // CONCATENATED MODULE: ./src/Concertina/ConcertinaContext.tsx
    var te =  (0, n.createContext)({});
    te.displayName = "Concertina";
    /* harmony default export */    const ne = te;
    // CONCATENATED MODULE: ./src/Concertina/Panel.tsx
    function re(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            re = function e(t) {
                return typeof t;
            };
        } else {
            re = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return re(e);
    }
    function oe() {
        oe = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var n = arguments[t];
                for (var r in n) {
                    if (Object.prototype.hasOwnProperty.call(n, r)) {
                        e[r] = n[r];
                    }
                }
            }
            return e;
        };
        return oe.apply(this, arguments);
    }
    function ie(e) {
        return ce(e) || se(e) || le(e) || ae();
    }
    function ae() {
        throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function le(e, t) {
        if (!e) return;
        if (typeof e === "string") return ue(e, t);
        var n = Object.prototype.toString.call(e).slice(8, -1);
        if (n === "Object" && e.constructor) n = e.constructor.name;
        if (n === "Map" || n === "Set") return Array.from(e);
        if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return ue(e, t);
    }
    function se(e) {
        if (typeof Symbol !== "undefined" && Symbol.iterator in Object(e)) return Array.from(e);
    }
    function ce(e) {
        if (Array.isArray(e)) return ue(e);
    }
    function ue(e, t) {
        if (t == null || t > e.length) t = e.length;
        for (var n = 0, r = new Array(t); n < t; n++) {
            r[n] = e[n];
        }
        return r;
    }
    function fe(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function pe(e, t) {
        for (var n = 0; n < t.length; n++) {
            var r = t[n];
            r.enumerable = r.enumerable || false;
            r.configurable = true;
            if ("value" in r) r.writable = true;
            Object.defineProperty(e, r.key, r);
        }
    }
    function de(e, t, n) {
        if (t) pe(e.prototype, t);
        if (n) pe(e, n);
        return e;
    }
    function he(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) ye(e, t);
    }
    function ye(e, t) {
        ye = Object.setPrototypeOf || function e(t, n) {
            t.__proto__ = n;
            return t;
        };
        return ye(e, t);
    }
    function be(e) {
        var t = me();
        return function n() {
            var r = ke(e), o;
            if (t) {
                var i = ke(this).constructor;
                o = Reflect.construct(r, arguments, i);
            } else {
                o = r.apply(this, arguments);
            }
            return ve(this, o);
        };
    }
    function ve(e, t) {
        if (t && (re(t) === "object" || typeof t === "function")) {
            return t;
        }
        return ge(e);
    }
    function ge(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function me() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function ke(e) {
        ke = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return ke(e);
    }
    function Se(e, t, n) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: n,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = n;
        }
        return e;
    }
    var we = {
        children: i().node,
        description: i().node,
        elementRef: i().oneOfType([ i().func, i().object ]),
        /** @private */
        index: i().number,
        panelId: i().string,
        status: i().oneOf([ "normal", "warning", "error", "disabled" ]),
        title: i().node.isRequired
    };
    var Ce = {
        status: "normal"
    };
    var Oe =  function(e) {
        he(n, e);
        var t = be(n);
        // @docs-props-type PanelPropsBase
                function n(e) {
            var r;
            fe(this, n);
            r = t.call(this, e);
            Se(ge(r), "lastHeight", void 0);
            Se(ge(r), "measureHeight", (function(e) {
                var t, n, o;
                if (!r.state.containerEl || !r.state.bodyEl) {
                    return;
                }
                var i = r.state.containerEl.clientHeight;
                if (i === r.lastHeight) {
                    return;
                }
                var a = {
                    headingHeight: r.state.containerEl.offsetHeight - r.state.bodyEl.offsetHeight,
                    height: r.state.containerEl.offsetHeight,
                    index: (t = r.props.index) !== null && t !== void 0 ? t : 0,
                    panelId: r.props.panelId
                };
                r.lastHeight = i;
                (n = (o = r.context).onChange) === null || n === void 0 ? void 0 : n.call(o, e, a);
            }));
            Se(ge(r), "handleResize", V()(r.measureHeight, 300));
            Se(ge(r), "handleMount", (function(e) {
                r.setState({
                    containerEl: e
                });
                ee(r.props.elementRef, e);
            }));
            Se(ge(r), "handleHeadingMount", (function(e) {
                r.setState({
                    heading: e
                });
            }));
            Se(ge(r), "handleBodyMount", (function(e) {
                r.setState({
                    bodyEl: e
                });
            }));
            r.state = {
                bodyEl: null,
                containerEl: null,
                heading: null
            };
            return r;
        }
        de(n, [ {
            key: "componentDidUpdate",
            value: function e(t, n) {
                if (this.props.children !== t.children || !n.containerEl) {
                    this.measureHeight();
                }
            }
        }, {
            key: "componentWillUnmount",
            value: function e() {
                this.handleResize.cancel();
            }
        }, {
            key: "focus",
            value: function e() {
                var t;
                (t = this.state.heading) === null || t === void 0 ? void 0 : t.focus();
            }
        }, {
            key: "render",
            value: function e() {
                var t = this.props, o = t.children, i = t.description, a = t.index, l = t.panelId, s = t.status, c = t.style, u = t.title;
                var f = this.context.onHeadingClick;
                
                return r().createElement("div", oe({
                    ref: this.handleMount,
                    "data-test": "panel",
                    "data-test-panel-id": l || "".concat(a)
                }, v()(this.props, [].concat(ie(y()(n.propTypes)), [ "style" ]))),  r().createElement($, {
                    onClick: f,
                    ref: this.handleHeadingMount,
                    description: i,
                    index: a !== null && a !== void 0 ? a : 0,
                    panelId: l || "".concat(a),
                    position: "inner",
                    status: s
                }, u),  r().createElement(d(), {
                    target: window,
                    eventType: "resize",
                    listener: this.handleResize
                }),  r().createElement(Z, {
                    $status: s,
                    "data-test": "body",
                    style: c,
                    ref: this.handleBodyMount
                }, o));
            }
        } ]);
        return n;
    }(n.Component);
    Se(Oe, "propTypes", we);
    Se(Oe, "defaultProps", Ce);
    Se(Oe, "contextType", ne);
    /* harmony default export */    const Pe = Oe;
    // CONCATENATED MODULE: ./src/Concertina/Concertina.tsx
    function je(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            je = function e(t) {
                return typeof t;
            };
        } else {
            je = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return je(e);
    }
    function Ee() {
        Ee = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var n = arguments[t];
                for (var r in n) {
                    if (Object.prototype.hasOwnProperty.call(n, r)) {
                        e[r] = n[r];
                    }
                }
            }
            return e;
        };
        return Ee.apply(this, arguments);
    }
    function xe(e, t) {
        if (e == null) return {};
        var n = _e(e, t);
        var r, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                r = i[o];
                if (t.indexOf(r) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, r)) continue;
                n[r] = e[r];
            }
        }
        return n;
    }
    function _e(e, t) {
        if (e == null) return {};
        var n = {};
        var r = Object.keys(e);
        var o, i;
        for (i = 0; i < r.length; i++) {
            o = r[i];
            if (t.indexOf(o) >= 0) continue;
            n[o] = e[o];
        }
        return n;
    }
    function He(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function Te(e, t) {
        for (var n = 0; n < t.length; n++) {
            var r = t[n];
            r.enumerable = r.enumerable || false;
            r.configurable = true;
            if ("value" in r) r.writable = true;
            Object.defineProperty(e, r.key, r);
        }
    }
    function Re(e, t, n) {
        if (t) Te(e.prototype, t);
        if (n) Te(e, n);
        return e;
    }
    function Ie(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) qe(e, t);
    }
    function qe(e, t) {
        qe = Object.setPrototypeOf || function e(t, n) {
            t.__proto__ = n;
            return t;
        };
        return qe(e, t);
    }
    function ze(e) {
        var t = We();
        return function n() {
            var r = Be(e), o;
            if (t) {
                var i = Be(this).constructor;
                o = Reflect.construct(r, arguments, i);
            } else {
                o = r.apply(this, arguments);
            }
            return Me(this, o);
        };
    }
    function Me(e, t) {
        if (t && (je(t) === "object" || typeof t === "function")) {
            return t;
        }
        return Ae(e);
    }
    function Ae(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function We() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function Be(e) {
        Be = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return Be(e);
    }
    function De(e, t, n) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: n,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = n;
        }
        return e;
    }
    var Ne = {
        children: i().node,
        elementRef: i().oneOfType([ i().func, i().object ])
    };
    var $e = {};
    /**
 * @deprecated
 * Concertina has been deprecated and will be removed in a future major version.
 */    var Ue =  function(e) {
        Ie(o, e);
        var t = ze(o);
        // @docs-props-type ConcertinaPropsBase
                function o(e) {
            var n;
            He(this, o);
            n = t.call(this, e);
            De(Ae(n), "panelPositions", void 0);
            De(Ae(n), "handleUpdate", (function(e, t) {
                n.panelPositions[t.index] = t;
                var r = 0;
                n.panelPositions.forEach((function(e) {
                    // eslint-disable-next-line no-param-reassign
                    e.offsetTop = r;
                    r += e.height || 0;
                }));
                n.updateHeadings();
            }));
            De(Ae(n), "handleElementMount", (function(e) {
                n.setState({
                    scrollEl: e
                });
 // The scrollEl gets unmounted during the animation, it's necessary to cache the
                // recorded height until it is remounted.
                                if (e) {
                    n.setState({
                        innerWidth: e.clientWidth
                    });
                }
            }));
            De(Ae(n), "handleResize", (function() {
                n.setState((function(e) {
                    var t;
                    return {
                        innerWidth: (t = e.scrollEl) === null || t === void 0 ? void 0 : t.clientWidth
                    };
                }));
            }));
            De(Ae(n), "handleScroll", (function() {
                n.updateHeadings();
            }));
            De(Ae(n), "handleScrollComplete", (function() {
                n.setState({
                    targetTop: undefined
                });
            }));
            De(Ae(n), "handleClick", (function(e, t) {
                var r;
                var o = t.index, i = t.position;
                var a = n.state.scrollEl;
                if (!a) {
                    return;
                }
                var l = c()(n.panelPositions, o).map((function(e) {
                    return e.headingHeight;
                })).reduce((function(e, t) {
                    return e + t;
                }), 0);
 // sum the heights
                                var s = f()(n.panelPositions, n.panelPositions.length - o - 1).map((function(e) {
                    return e.headingHeight;
                })).reduce((function(e, t) {
                    return e + t;
                }), 0);
 // sum the heights
                                var u = a.clientHeight - l - s;
                var p = n.panelPositions[o].height;
                var d = ((r = n.panelPositions[o].offsetTop) !== null && r !== void 0 ? r : 0) - l;
                var h = d - (u - p);
                var y = u > p && (i === "bottom" || i === "inner");
 // if the entire panel is in view, do nothing
                                if (y && h < a.scrollTop) {
                    return;
                }
                var b = y ? h : d;
 // handleRest will shift focus;
                                if (i !== "inner") {
                    e.currentTarget.blur();
                }
                n.setState({
                    targetTop: b
                });
            }));
            De(Ae(n), "updateHeadingsImpl", (function() {
                var e = n.state.scrollEl;
                if (!e) {
                    return;
                }
                var t = e.scrollTop;
                function r(e, n) {
                    if (n.offsetTop !== undefined && n.offsetTop < t) {
                        t += n.headingHeight;
                        return e + 1;
                    }
                    return e;
                }
                var o = t + e.clientHeight;
                function i(e, t) {
                    if (t.offsetTop !== undefined && t.offsetTop + t.headingHeight > o) {
                        o -= t.headingHeight;
                        return e + 1;
                    }
                    return e;
                }
                n.setState({
                    panelsTopCount: n.panelPositions.reduce(r, 0),
                    panelsBottomCount: n.panelPositions.reduceRight(i, 0)
                });
            }));
            De(Ae(n), "updateHeadings", l()(n.updateHeadingsImpl, 0));
            n.state = {
                innerWidth: "100%",
                panelsBottomCount: 0,
                panelsTopCount: 0,
                scrollEl: null
            };
            n.panelPositions = [];
            return n;
        }
        Re(o, [ {
            key: "componentWillUnmount",
            value: function e() {
                this.updateHeadings.cancel();
            }
        }, {
            key: "renderHeadings",
            value: function e(t, n) {
                var o = this;
                var i = t.map((function(e) {
                    var t;
                    
                    return r().createElement($, {
                        onClick: o.handleClick,
                        description: e.props.description,
                        panelId: e.props.panelId,
                        index: (t = e.props.index) !== null && t !== void 0 ? t : 0,
                        position: n,
                        status: e.props.status,
                        key: e.props.panelId,
                        tabIndex: -1
                    }, e.props.title);
                }));
                return i;
            }
        }, {
            key: "render",
            value: function e() {
                if (false) {}
                var t = this.props, o = t.children, i = xe(t, [ "children" ]);
                var a = n.Children.toArray(o).filter(n.isValidElement).map((function(e, t) {
                    
                    return (0, n.cloneElement)(e, {
                        index: t
                    });
                }));
                var l = c()(a, this.state.panelsTopCount);
                var s = f()(a, this.state.panelsBottomCount);
                
                return r().createElement(Q, Ee({
                    "data-test": "concertina"
                }, i),  r().createElement(d(), {
                    target: window,
                    eventType: "resize",
                    listener: this.handleResize
                }), this.state.panelsTopCount > 0 &&  r().createElement(X, {
                    style: {
                        width: this.state.innerWidth
                    },
                    "data-test": "dock-top"
                }, this.renderHeadings(l, "top")),  r().createElement(K, {
                    onScroll: this.handleScroll,
                    stopScrollPropagation: true,
                    key: "scroll-container",
                    elementRef: this.handleElementMount,
                    "data-test": "scroll",
                    top: this.state.targetTop,
                    onScrollComplete: this.handleScrollComplete
                },  r().createElement(ne.Provider, {
                    value: {
                        onHeadingClick: this.handleClick,
                        onChange: this.handleUpdate
                    }
                }, a)), this.state.panelsBottomCount > 0 &&  r().createElement(Y, {
                    style: {
                        width: this.state.innerWidth
                    },
                    "data-test": "dock-bottom"
                }, this.renderHeadings(s, "bottom")));
            }
        } ]);
        return o;
    }(n.Component);
    De(Ue, "propTypes", Ne);
    De(Ue, "defaultProps", $e);
    De(Ue, "Panel", Pe);
    /* harmony default export */    const Ve = Ue;
    // CONCATENATED MODULE: ./src/Concertina/index.ts
    module.exports = t;
    /******/})();
