// @splunk/react-ui/ControlGroup.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var r = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(r, {
                a: r
            });
            /******/            return r;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        ControlGroupContext: () => /* reexport */ T,
        default: () => /* reexport */ se
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    var n =  e.n(r);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var i =  e.n(o);
    // CONCATENATED MODULE: external "lodash/find"
    const l = require("lodash/find");
    var a =  e.n(l);
    // CONCATENATED MODULE: external "lodash/isFinite"
    const s = require("lodash/isFinite");
    var p =  e.n(s);
    // CONCATENATED MODULE: external "@splunk/themes"
    const c = require("@splunk/themes");
    // CONCATENATED MODULE: external "@splunk/ui-utils/id"
    const u = require("@splunk/ui-utils/id");
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const f = require("@splunk/react-ui/Box");
    var d =  e.n(f);
    // CONCATENATED MODULE: external "@splunk/react-ui/ScreenReaderContent"
    const y = require("@splunk/react-ui/ScreenReaderContent");
    var b =  e.n(y);
    // CONCATENATED MODULE: external "styled-components"
    const h = require("styled-components");
    var m =  e.n(h);
    // CONCATENATED MODULE: external "@splunk/react-ui/Tooltip"
    const v = require("@splunk/react-ui/Tooltip");
    var g =  e.n(v);
    // CONCATENATED MODULE: ./src/ControlGroup/ControlGroupStyles.ts
    var O = m()(d()).withConfig({
        displayName: "ControlGroupStyles__StyledBox",
        componentId: "wjnyif-0"
    })([ "", " ", " max-width:600px;margin-bottom:", ";", " color:", ";", "" ], c.mixins.reset("block"), c.mixins.clearfix(), (0, 
    c.pick)({
        enterprise: (0, h.css)([ "calc(", " * 0.75)" ], c.variables.spacing),
        prisma: {
            comfortable: "16px",
            compact: "12px"
        }
    }), (0, c.pick)({
        prisma: {
            comfortable: (0, h.css)([ "&:first-child{margin-top:", ";}&:last-child{margin-bottom:", ";}" ], c.variables.spacingMedium, c.variables.spacingMedium),
            compact: (0, h.css)([ "&:first-child{margin-top:", ";}&:last-child{margin-bottom:", ";}" ], c.variables.spacingSmall, c.variables.spacingSmall)
        }
    }), (0, c.pick)({
        enterprise: c.variables.textColor,
        prisma: c.variables.contentColorMuted
    }), (function(e) {
        var t = e.$error;
        return t && (0, h.css)([ "color:", ";" ], (0, c.pick)({
            enterprise: {
                light: c.variables.errorColorD10,
                dark: c.variables.errorColorL20
            },
            prisma: c.variables.accentColorNegative
        }));
    }));
    var x = m()(d()).withConfig({
        displayName: "ControlGroupStyles__StyledControlsStackBox",
        componentId: "wjnyif-1"
    })([ "flex-direction:column;" ]);
    var S = m().div.withConfig({
        displayName: "ControlGroupStyles__StyledLabelWrapper",
        componentId: "wjnyif-2"
    })([ "display:inline-flex;align-items:center;", " ", "" ], (0, c.pick)({
        enterprise: (0, h.css)([ "justify-content:flex-end;" ])
    }), (function(e) {
        var t = e.$labelPosition;
        return t === "top" ? (0, h.css)([ "", "" ], (0, c.pick)({
            prisma: (0, h.css)([ "height:24px;padding-bottom:", ";" ], c.variables.spacingXSmall)
        })) : (0, h.css)([ "", "" ], (0, c.pick)({
            prisma: (0, h.css)([ "min-height:", ";" ], c.variables.inputHeight)
        }));
    }));
    var I = m()(S).withConfig({
        displayName: "ControlGroupStyles__StyledLabelWrapperLeft",
        componentId: "wjnyif-3"
    })([ "float:left;" ]);
    var C = m().label.withConfig({
        displayName: "ControlGroupStyles__StyledLabel",
        componentId: "wjnyif-4"
    })([ "padding:", ";word-wrap:break-word;color:inherit;text-align:", ";" ], (0, c.pick)({
        enterprise: {
            comfortable: "6px 0",
            compact: "4px 0"
        }
    }), (0, c.pick)({
        enterprise: "right",
        prisma: "left"
    }));
    var k = m()(g()).withConfig({
        displayName: "ControlGroupStyles__StyledTooltip",
        componentId: "wjnyif-5"
    })([ "", "" ], (function(e) {
        var t = e.$labelPosition;
        return t === "top" && (0, h.css)([ "button{", "}" ], (0, c.pick)({
            prisma: (0, h.css)([ "svg{width:20px;height:20px;}padding:4px;" ])
        }));
    }));
    var w = m().div.withConfig({
        displayName: "ControlGroupStyles__StyledHelp",
        componentId: "wjnyif-6"
    })([ "", ";font-size:", ";color:", ";margin-top:", ";", "" ], c.mixins.reset("block"), c.variables.fontSizeSmall, c.variables.contentColorMuted, (0, 
    c.pick)({
        enterprise: "2px",
        prisma: "4px"
    }), (function(e) {
        var t = e.$error;
        return t && (0, h.css)([ "color:inherit;" ]);
    }));
    var j = m().div.withConfig({
        displayName: "ControlGroupStyles__StyledErrorText",
        componentId: "wjnyif-7"
    })([ "", ";font-size:", ";[aria-invalid] > &{color:inherit;}", "" ], c.mixins.reset("block"), c.variables.fontSizeSmall, (function(e) {
        var t = e.$help;
        return !t && (0, h.css)([ "margin-top:", ";" ], (0, c.pick)({
            enterprise: "2px",
            prisma: "4px"
        }));
    }));
    var P = m().span.withConfig({
        displayName: "ControlGroupStyles__StyledAsterisk",
        componentId: "wjnyif-8"
    })([ "color:inherit;margin-right:2px;" ]);
    // CONCATENATED MODULE: ./src/ControlGroup/ControlGroupContext.tsx
    /**
 * A React context used to inform subcomponets of ControlGroup of parent component values.
 * The context interface is `ControlGroupContextValue`.
 * Defaults to `'{}'`.
 * @public
 */
    var _ =  n().createContext({});
    /* harmony default export */    const T = _;
    // CONCATENATED MODULE: ./src/ControlGroup/ControlGroup.tsx
    function L(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            L = function e(t) {
                return typeof t;
            };
        } else {
            L = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return L(e);
    }
    function E() {
        E = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return E.apply(this, arguments);
    }
    function D(e, t) {
        var r = Object.keys(e);
        if (Object.getOwnPropertySymbols) {
            var n = Object.getOwnPropertySymbols(e);
            if (t) n = n.filter((function(t) {
                return Object.getOwnPropertyDescriptor(e, t).enumerable;
            }));
            r.push.apply(r, n);
        }
        return r;
    }
    function q(e) {
        for (var t = 1; t < arguments.length; t++) {
            var r = arguments[t] != null ? arguments[t] : {};
            if (t % 2) {
                D(Object(r), true).forEach((function(t) {
                    Z(e, t, r[t]);
                }));
            } else if (Object.getOwnPropertyDescriptors) {
                Object.defineProperties(e, Object.getOwnPropertyDescriptors(r));
            } else {
                D(Object(r)).forEach((function(t) {
                    Object.defineProperty(e, t, Object.getOwnPropertyDescriptor(r, t));
                }));
            }
        }
        return e;
    }
    function M(e, t) {
        if (e == null) return {};
        var r = V(e, t);
        var n, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                n = i[o];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function V(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var o, i;
        for (i = 0; i < n.length; i++) {
            o = n[i];
            if (t.indexOf(o) >= 0) continue;
            r[o] = e[o];
        }
        return r;
    }
    function $(e, t) {
        return z(e) || N(e, t) || G(e, t) || A();
    }
    function A() {
        throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function G(e, t) {
        if (!e) return;
        if (typeof e === "string") return B(e, t);
        var r = Object.prototype.toString.call(e).slice(8, -1);
        if (r === "Object" && e.constructor) r = e.constructor.name;
        if (r === "Map" || r === "Set") return Array.from(e);
        if (r === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(r)) return B(e, t);
    }
    function B(e, t) {
        if (t == null || t > e.length) t = e.length;
        for (var r = 0, n = new Array(t); r < t; r++) {
            n[r] = e[r];
        }
        return n;
    }
    function N(e, t) {
        if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(e))) return;
        var r = [];
        var n = true;
        var o = false;
        var i = undefined;
        try {
            for (var l = e[Symbol.iterator](), a; !(n = (a = l.next()).done); n = true) {
                r.push(a.value);
                if (t && r.length === t) break;
            }
        } catch (e) {
            o = true;
            i = e;
        } finally {
            try {
                if (!n && l["return"] != null) l["return"]();
            } finally {
                if (o) throw i;
            }
        }
        return r;
    }
    function z(e) {
        if (Array.isArray(e)) return e;
    }
    function R(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function W(e, t) {
        for (var r = 0; r < t.length; r++) {
            var n = t[r];
            n.enumerable = n.enumerable || false;
            n.configurable = true;
            if ("value" in n) n.writable = true;
            Object.defineProperty(e, n.key, n);
        }
    }
    function F(e, t, r) {
        if (t) W(e.prototype, t);
        if (r) W(e, r);
        return e;
    }
    function J(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) H(e, t);
    }
    function H(e, t) {
        H = Object.setPrototypeOf || function e(t, r) {
            t.__proto__ = r;
            return t;
        };
        return H(e, t);
    }
    function U(e) {
        var t = Q();
        return function r() {
            var n = Y(e), o;
            if (t) {
                var i = Y(this).constructor;
                o = Reflect.construct(n, arguments, i);
            } else {
                o = n.apply(this, arguments);
            }
            return X(this, o);
        };
    }
    function X(e, t) {
        if (t && (L(t) === "object" || typeof t === "function")) {
            return t;
        }
        return K(e);
    }
    function K(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function Q() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function Y(e) {
        Y = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return Y(e);
    }
    function Z(e, t, r) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: r,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = r;
        }
        return e;
    }
    // props ControlGroup may access and/or override on children
        var ee = {
        children: i().node,
        controlsLayout: i().oneOf([ "fill", "fillJoin", "none", "stack" ]),
        elementRef: i().oneOfType([ i().func, i().object ]),
        error: i().oneOfType([ i().bool, i().string ]),
        help: i().node,
        hideLabel: i().bool,
        label: i().string.isRequired,
        labelFor: i().string,
        labelPosition: i().oneOf([ "left", "top" ]),
        labelWidth: i().oneOfType([ i().number, i().string ]),
        required: i().bool,
        size: i().oneOf([ "small", "medium" ]),
        /** @private */
        splunkTheme: i().object,
        tooltip: i().node,
        tooltipDefaultPlacement: i().oneOf([ "above", "below", "left", "right", "theme" ])
    };
    var te = {
        controlsLayout: "fill",
        error: false,
        hideLabel: false,
        labelPosition: "left",
        labelWidth: 120,
        required: false,
        size: "medium"
    };
    // Style cloned onto child elements when fill behavior set.
        var re = {
        flexGrow: 1
    };
 // Margin bootom values cloned onto child elements when the stacking behavior is needed.
        var ne = "0";
    var oe = "16px";
    var ie = "12px";
    /**
 * `ControlGroup` places a label and optional help text around one or more controls. The `ControlGroup`
 * will automatically add aria attributes to associate the controls with the labels and help text to
 * address accessibility requirements.
 *
 * `ControlGroup` provides layouts to assist in aligning and laying out controls, but the defaults are
 * not helpful in all cases, nor will the layout options address all cases. Consider setting
 * `controlsLayout` to none and manually positioning the controls as required.
 *
 * `ControlGroup` uses the HTML `label` tag. The rules for determining which child component is linked
 * to the label's `for` attribute are:
 * 1. If one or more `children` are `Text` components, the first one is used.
 * 2. If there aren't any `Text` components, the first child is used.
 *
 * If the linked child supports an `inputId` prop and it's set, its value is used for the label's `for`
 * attribute. If `inputId` is supported but not set a generated id is used instead. If `inputId` isn't
 * supported `id` is used. The `labelFor` prop may be used to override the `for` attribute.
 */    var le =  function(e) {
        J(o, e);
        var t = U(o);
        // @docs-props-type ControlGroupPropsBase
                function o(e) {
            var r;
            R(this, o);
            r = t.call(this, e);
            Z(K(r), "errorId", void 0);
            Z(K(r), "helpId", void 0);
            Z(K(r), "labelId", void 0);
            Z(K(r), "contextValue", void 0);
            Z(K(r), "childIds", void 0);
            Z(K(r), "stackLayoutChildStyle", void 0);
            Z(K(r), "getChildID", (function(e, t) {
                var n = "".concat(e, "-").concat(t);
                var o = r.childIds[n];
                if (!o || !o.length) {
                    o = (0, u.createDOMID)(e);
                    r.childIds[n] = o;
                }
                return o;
            }));
            Z(K(r), "getStackLayoutChildStyle", (function(e, t) {
                if (e) {
                    if (t) {
                        r.stackLayoutChildStyle.marginBottom = ie;
                    }
                    r.stackLayoutChildStyle.marginBottom = oe;
                } else {
                    r.stackLayoutChildStyle.marginBottom = ne;
                }
                return r.stackLayoutChildStyle;
            }));
            Z(K(r), "hasInputId", (function(e) {
                return e.type && e.type.propTypes && Object.prototype.hasOwnProperty.call(e.type.propTypes, "inputId");
            }));
            Z(K(r), "getLinkedId", (function(e) {
                if (e.length === 0) {
                    return undefined;
                }
                var t = $(e, 1), n = t[0];
                var o = a()(e, (function(e) {
                    return e.type && e.type.componentType === "Text";
                }));
                var i = r.hasInputId(n);
 // see class doc block for details
                                if (e.length === 1 && n && n.props) {
                    return i ? n.props.inputId : n.props.id;
                }
                if (o && o.props) {
                    return o.props.inputId;
                }
                if (n.props) {
                    return i ? n.props.inputId : n.props.id;
                }
                return undefined;
            }));
            r.errorId = (0, u.createDOMID)("error");
            r.helpId = (0, u.createDOMID)("help");
            r.labelId = (0, u.createDOMID)("label");
            r.contextValue = r.createContextValue();
            r.childIds = {};
            r.stackLayoutChildStyle = {};
            if (false) {}
            return r;
        }
        F(o, [ {
            key: "shouldComponentUpdate",
            value: function e(t) {
                if (n().Children.count(t.children) !== n().Children.count(this.props.children)) {
                    this.childIds = {};
                }
                return true;
            }
        }, {
            key: "createContextValue",
            value: function e() {
                return {
                    labelAttrs: {
                        text: this.props.label,
                        id: this.labelId
                    }
                };
            }
        }, {
            key: "getContextValue",
            value: function e() {
                var t, r;
                if (this.props.label !== ((t = this.contextValue.labelAttrs) === null || t === void 0 ? void 0 : t.text) || this.labelId !== ((r = this.contextValue.labelAttrs) === null || r === void 0 ? void 0 : r.id)) {
                    this.contextValue = this.createContextValue();
                }
                return this.contextValue;
            }
        }, {
            key: "render",
            value: function e() {
                var t = this;
                var o = this.props, i = o.children, l = o.controlsLayout, a = o.error, s = o.help, c = o.hideLabel, u = o.label, f = o.labelFor, y = o.labelPosition, h = o.labelWidth, m = o.required, v = o.size, g = o.splunkTheme, _ = o.tooltip, L = o.tooltipDefaultPlacement, D = M(o, [ "children", "controlsLayout", "error", "help", "hideLabel", "label", "labelFor", "labelPosition", "labelWidth", "required", "size", "splunkTheme", "tooltip", "tooltipDefaultPlacement" ]);
                var V = g.isPrisma, $ = g.isCompact;
                var A = r.Children.toArray(i).filter(r.isValidElement);
                var G = A.length;
                var B = D;
 // Clean the Children
                                var N = function e(n, o) {
                    var i = {
                        key: n.key || o
                    };
                    if (l === "fillJoin") {
                        if (o > 0) {
                            i.prepend = true;
                        }
                        if (o < G - 1) {
                            i.append = true;
                        }
                        if (i.prepend || i.append) {
                            i.inline = false;
                        }
                    }
                    if (l === "fill" && G > 1) {
                        i.inline = true;
                    }
                    if (l === "stack") {
                        i.inline = false;
                        if (o < G - 1) {
                            // The spread operator is used as css styles are not nested, so there is no need for a deep copy.
                            var a = t.getStackLayoutChildStyle(V, $);
                            i.style = n.props.style ? q(q({}, n.props.style), a) : a;
                        }
                    }
                    i.labelledBy = t.labelId;
                    if (t.props.help && typeof t.props.error === "boolean") {
                        i.describedBy = t.helpId;
                    } else if (t.props.help && typeof t.props.error === "string") {
                        i.describedBy = "".concat(t.helpId, " ").concat(t.errorId);
                    } else if (!t.props.help && typeof t.props.error === "string") {
                        i.describedBy = t.errorId;
                    }
                    if (t.props.required) {
                        i.required = true;
                    }
 // TODO: SUI-3402 refactor so that parent component doesn't have to know what it's child components are
                                        if (n.type.componentType === "Multiselect") {
                        var s = n;
                        if (s.props.compact === false) {
                            i.labelText = u;
                        }
                    } else {
                        i.labelText = u;
                    }
                    if (G === 1 && (l === "fillJoin" || l === "fill")) {
                        // Some controls like Select do not grow by default, so flexGrow is applied
                        // The spread operator is used as css styles are not nested, so there is no need for a deep copy.
                        i.style = n.props.style ? q(q({}, n.props.style), re) : re;
                    }
                    if (!f && n) {
                        if (t.hasInputId(n)) {
                            i.inputId = n.props.inputId || t.getChildID("input-id", o);
                        } else {
                            i.id = n.props.id || t.getChildID("id", o);
                        }
                    }
                    
                    return (0, r.cloneElement)(n, i);
                };
                var z = A.map(N);
                var R = this.getLinkedId(z);
                if (a) {
                    B["aria-invalid"] = true;
                }
                var W = y === "left" ? {
                    width: h
                } : undefined;
                var F = p()(h) ? "".concat(h, "px") : h;
                var J = V ? "16px" : "20px";
                var H = y === "left" ? {
                    marginLeft: "calc(".concat(F, " + ").concat(J, ")")
                } : undefined;
                var U = l === "stack" ? x : d();
                var X = y === "left" ? I : S;
                var K =  n().createElement(X, {
                    style: W,
                    $labelPosition: y
                },  n().createElement(C, {
                    "data-size": v,
                    "data-test": "label",
                    id: this.labelId,
                    htmlFor: f || R,
                    $tooltip: !!_
                }, m &&  n().createElement(P, {
                    "aria-hidden": "true"
                }, "*"), u, !V && !c && _ &&  n().createElement("span", null, " "), !c && _ &&  n().createElement(b(), null, _)), !c && _ &&  n().createElement(k, {
                    $labelPosition: y,
                    closeWhen: "notOnClick",
                    content: _,
                    defaultPlacement: L,
                    "aria-hidden": "true"
                }));
                var Q = typeof a === "string" || a === true;
                var Y = typeof a === "boolean";
                
                return n().createElement(O, E({
                    "data-test": "control-group",
                    "data-test-required": m,
                    $error: Q
                }, B),  n().createElement(T.Provider, {
                    value: this.getContextValue()
                }, c ?  n().createElement(b(), null, K) : K,  n().createElement(U, {
                    "data-test": "controls",
                    flex: l !== "none",
                    style: H
                }, z), s &&  n().createElement(w, {
                    "data-test": "help",
                    id: this.helpId,
                    style: H,
                    $error: Q && Y
                }, s), Q && !Y &&  n().createElement(j, {
                    "data-test": "error",
                    id: this.errorId,
                    style: H,
                    $help: !!s
                }, a)));
            }
        } ]);
        return o;
    }(r.Component);
    Z(le, "propTypes", ee);
    Z(le, "defaultProps", te);
    var ae = (0, c.withSplunkTheme)(le);
    ae.propTypes = le.propTypes;
    /* harmony default export */    const se = ae;
    // CONCATENATED MODULE: ./src/ControlGroup/index.ts
    module.exports = t;
    /******/})();
