// @splunk/react-ui/DefinitionList.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var r = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(r, {
                a: r
            });
            /******/            return r;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        Description: () => /* reexport */ m,
        Term: () => /* reexport */ x,
        default: () => /* reexport */ R
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    var n =  e.n(r);
    // CONCATENATED MODULE: external "prop-types"
    const i = require("prop-types");
    var o =  e.n(i);
    // CONCATENATED MODULE: external "styled-components"
    const l = require("styled-components");
    var a =  e.n(l);
    // CONCATENATED MODULE: external "@splunk/themes"
    const c = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/DefinitionList/DescriptionStyles.ts
    var f = a().dd.withConfig({
        displayName: "DescriptionStyles__Styled",
        componentId: "sc-1jglixb-0"
    })([ "", ";", ";min-height:", ";padding-left:5px;" ], c.mixins.reset("block"), (0, 
    c.pick)({
        prisma: c.mixins.typography("body", {
            color: "active"
        }),
        enterprise: c.mixins.typography("body")
    }), c.variables.lineHeight);
    // CONCATENATED MODULE: ./src/DefinitionList/Description.tsx
    function p() {
        p = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return p.apply(this, arguments);
    }
    function s(e, t) {
        if (e == null) return {};
        var r = u(e, t);
        var n, i;
        if (Object.getOwnPropertySymbols) {
            var o = Object.getOwnPropertySymbols(e);
            for (i = 0; i < o.length; i++) {
                n = o[i];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function u(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var i, o;
        for (o = 0; o < n.length; o++) {
            i = n[o];
            if (t.indexOf(i) >= 0) continue;
            r[i] = e[i];
        }
        return r;
    }
    var y = {
        children: o().node,
        elementRef: o().oneOfType([ o().func, o().object ])
    };
    /**
 * Container component for a `DefinitionList` description.
 */    function d(e) {
        var t = e.children, r = e.elementRef, i = s(e, [ "children", "elementRef" ]);
        // @docs-props-type DescriptionPropsBase
                
        return n().createElement(f, p({
            "data-test": "description",
            ref: r
        }, i), t);
    }
    d.propTypes = y;
    /* harmony default export */    const m = d;
    // CONCATENATED MODULE: ./src/DefinitionList/DefinitionListStyles.ts
    var v = a().dl.withConfig({
        displayName: "DefinitionListStyles__Styled",
        componentId: "sc-1pqfrgm-0"
    })([ "", ";", ";" ], c.mixins.reset("block"), c.mixins.typography("body"));
    // CONCATENATED MODULE: ./src/DefinitionList/TermStyles.ts
    var h = a().dt.withConfig({
        displayName: "TermStyles__Styled",
        componentId: "sc-1ao97j9-0"
    })([ "", ";", " float:left;width:120px;overflow:hidden;overflow-x:hidden;white-space:nowrap;word-wrap:normal;", "" ], c.mixins.reset("block"), (0, 
    c.pick)({
        prisma: c.mixins.typography("body", {
            color: "muted"
        }),
        enterprise: c.mixins.typography("body")
    }), (0, c.pick)({
        enterprise: (0, l.css)([ "&::after{content:'", "';}" ], " ".concat(".".repeat(100)))
    }));
    // CONCATENATED MODULE: ./src/DefinitionList/Term.tsx
    function b() {
        b = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return b.apply(this, arguments);
    }
    function O(e, t) {
        if (e == null) return {};
        var r = g(e, t);
        var n, i;
        if (Object.getOwnPropertySymbols) {
            var o = Object.getOwnPropertySymbols(e);
            for (i = 0; i < o.length; i++) {
                n = o[i];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function g(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var i, o;
        for (o = 0; o < n.length; o++) {
            i = n[o];
            if (t.indexOf(i) >= 0) continue;
            r[i] = e[i];
        }
        return r;
    }
    var j = {
        children: o().node,
        elementRef: o().oneOfType([ o().func, o().object ])
    };
    /**
 * Container component for a `DefinitionList` term.
 */    function w(e) {
        var t = e.children, r = e.elementRef, i = O(e, [ "children", "elementRef" ]);
        // @docs-props-type TermPropsBase
                
        return n().createElement(h, b({
            "data-test": "term",
            ref: r
        }, i), t);
    }
    w.propTypes = j;
    /* harmony default export */    const x = w;
    // CONCATENATED MODULE: ./src/DefinitionList/DefinitionList.tsx
    function S() {
        S = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return S.apply(this, arguments);
    }
    function P(e, t) {
        if (e == null) return {};
        var r = T(e, t);
        var n, i;
        if (Object.getOwnPropertySymbols) {
            var o = Object.getOwnPropertySymbols(e);
            for (i = 0; i < o.length; i++) {
                n = o[i];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function T(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var i, o;
        for (o = 0; o < n.length; o++) {
            i = n[o];
            if (t.indexOf(i) >= 0) continue;
            r[i] = e[i];
        }
        return r;
    }
    var k = {
        children: o().node,
        elementRef: o().oneOfType([ o().func, o().object ]),
        termWidth: o().oneOfType([ o().number, o().string ])
    };
    function _(e) {
        var t = e.children, n = e.termWidth;
        var i = r.Children.toArray(t).filter(r.isValidElement);
        return i.map((function(e) {
            if (e.type === x) {
                
                return (0, r.cloneElement)(e, {
                    style: {
                        width: n
                    }
                });
            }
            if (e.type === m) {
                
                return (0, r.cloneElement)(e, {
                    style: {
                        marginLeft: n
                    }
                });
            }
            return e;
        }));
    }
    function E(e) {
        var t = e.children, r = e.elementRef, i = e.termWidth, o = i === void 0 ? 120 : i, l = P(e, [ "children", "elementRef", "termWidth" ]);
        // @docs-props-type DefinitionListPropsBase
                
        return n().createElement(v, S({
            "data-test": "definition-list",
            ref: r
        }, l), _({
            children: t,
            termWidth: o
        }));
    }
    E.propTypes = k;
    E.Description = m;
    E.Term = x;
    /* harmony default export */    const R = E;
    // CONCATENATED MODULE: ./src/DefinitionList/index.ts
    module.exports = t;
    /******/})();
