// @splunk/react-ui/File.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var t = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(t, {
                a: t
            });
            /******/            return t;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, t) => {
            /******/ for (var a in t) {
                /******/ if (e.o(t, a) && !e.o(r, a)) {
                    /******/ Object.defineProperty(r, a, {
                        enumerable: true,
                        get: t[a]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        Item: () => /* reexport */ ne,
        default: () => /* reexport */ Fe
    });
    // CONCATENATED MODULE: external "react"
    const t = require("react");
    var a =  e.n(t);
    // CONCATENATED MODULE: external "prop-types"
    const n = require("prop-types");
    var i =  e.n(n);
    // CONCATENATED MODULE: external "lodash/debounce"
    const o = require("lodash/debounce");
    var l =  e.n(o);
    // CONCATENATED MODULE: external "lodash/omit"
    const s = require("lodash/omit");
    var c =  e.n(s);
    // CONCATENATED MODULE: external "lodash/toUpper"
    const p = require("lodash/toUpper");
    var d =  e.n(p);
    // CONCATENATED MODULE: external "lodash/isString"
    const u = require("lodash/isString");
    var v =  e.n(u);
    // CONCATENATED MODULE: external "@splunk/react-ui/EventListener"
    const f = require("@splunk/react-ui/EventListener");
    var m =  e.n(f);
    // CONCATENATED MODULE: external "@splunk/themes"
    const b = require("@splunk/themes");
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const h = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "@splunk/ui-utils/id"
    const y = require("@splunk/ui-utils/id");
    // CONCATENATED MODULE: external "lodash/isUndefined"
    const g = require("lodash/isUndefined");
    var C =  e.n(g);
    // CONCATENATED MODULE: external "@splunk/react-icons/Cross"
    const x = require("@splunk/react-icons/Cross");
    var k =  e.n(x);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Close"
    const w = require("@splunk/react-icons/enterprise/Close");
    var L =  e.n(w);
    // CONCATENATED MODULE: external "@splunk/react-ui/Progress"
    const O = require("@splunk/react-ui/Progress");
    var S =  e.n(O);
    // CONCATENATED MODULE: external "@splunk/ui-utils/format"
    const I = require("@splunk/ui-utils/format");
    // CONCATENATED MODULE: ./src/File/FileContext.tsx
    var _ =  (0, t.createContext)({
        disabled: false
    });
    _.displayName = "File";
    /* harmony default export */    const E = _;
    // CONCATENATED MODULE: external "@splunk/react-icons/SVGEnterprise"
    const R = require("@splunk/react-icons/SVGEnterprise");
    var D =  e.n(R);
    // CONCATENATED MODULE: ./src/File/PaperClip.tsx
    function q() {
        q = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var a in t) {
                    if (Object.prototype.hasOwnProperty.call(t, a)) {
                        e[a] = t[a];
                    }
                }
            }
            return e;
        };
        return q.apply(this, arguments);
    }
    function j(e) {
        
        return a().createElement(D(), q({
            hideDefaultTooltip: true,
            viewBox: "0 0 16 16",
            width: "16px",
            height: "16px"
        }, e),  a().createElement("path", {
            d: "M8.624 14.7021C8.23332 15.0928 7.59989 15.0928 7.2092 14.7021L2.33962 9.83209C0.546941 8.03927 0.54695 5.1325 2.33964 3.33968C4.13162 1.54758 7.03672 1.54684 8.82962 3.33802L14.1746 8.67787C15.2682 9.77043 15.2687 11.5429 14.1756 12.636C13.083 13.7288 11.3115 13.7288 10.2189 12.636L4.87676 7.29338C4.48607 6.90266 4.48607 6.26917 4.87676 5.87845C5.26744 5.48773 5.90087 5.48773 6.29155 5.87845L11.6337 11.2211C11.9449 11.5324 12.4496 11.5324 12.7608 11.2211C13.0722 10.9097 13.0721 10.4048 12.7606 10.0936L7.41557 4.75369C6.40415 3.74325 4.76532 3.74367 3.75442 4.75463C2.74312 5.766 2.74312 7.40578 3.75441 8.41715L8.62399 13.2871C9.01468 13.6779 9.01468 14.3113 8.624 14.7021Z",
            fill: "currentColor"
        }));
    }
    // CONCATENATED MODULE: ./src/File/Retry.tsx
    function P() {
        P = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var a in t) {
                    if (Object.prototype.hasOwnProperty.call(t, a)) {
                        e[a] = t[a];
                    }
                }
            }
            return e;
        };
        return P.apply(this, arguments);
    }
    function T(e) {
        var r = (0, b.useSplunkTheme)(), t = r.density;
        var n = t === "compact" ? "20px" : "24px";
        
        return a().createElement(D(), P({
            hideDefaultTooltip: true,
            viewBox: "0 0 24 24",
            width: n,
            height: n
        }, e),  a().createElement("path", {
            d: "M18.2685 14.3175L18.4349 14.9384C18.6078 14.558 18.7453 14.162 18.8452 13.7556C18.9449 13.3507 19.0068 12.9373 19.0303 12.521L19.0362 12.4167C19.0663 11.883 19.0409 11.3476 18.9604 10.8191L18.9462 10.7261C18.8794 10.287 18.7695 9.85559 18.6181 9.43803C18.4827 9.06459 18.3148 8.70378 18.1162 8.35979L18.0868 8.30889C17.8331 7.86964 17.5281 7.46215 17.1781 7.09503C16.9596 6.8658 16.7243 6.65314 16.4743 6.4588L16.318 6.33739C16.0132 6.10052 15.6825 5.88712 15.34 5.70564C14.95 5.49901 14.5348 5.32943 14.1132 5.20545C13.7201 5.08986 13.317 5.01136 12.9093 4.97099L12.8406 4.96419C12.4182 4.92237 11.993 4.91829 11.5699 4.95201C11.1331 4.98682 10.7005 5.06176 10.2774 5.17589L9.70069 5.33148C9.16747 5.47533 8.61859 5.15969 8.47474 4.62647C8.33089 4.09324 8.64653 3.54437 9.17976 3.40051L9.7565 3.24492C10.298 3.09883 10.8519 3.00289 11.411 2.95833C11.9526 2.91517 12.497 2.92038 13.0377 2.97392L13.1063 2.98072C13.6384 3.03341 14.1644 3.13584 14.6774 3.28668C15.2296 3.44906 15.7694 3.66976 16.2764 3.9384C16.7197 4.17329 17.1479 4.44938 17.5453 4.75819L17.7015 4.8796C18.0299 5.13478 18.3388 5.41401 18.6257 5.715C19.0853 6.19704 19.4858 6.73209 19.8188 7.30885L19.8482 7.35975C20.1055 7.80535 20.323 8.27272 20.4984 8.75646C20.6944 9.29736 20.8368 9.85624 20.9234 10.425L20.9376 10.518C21.0389 11.1833 21.0709 11.8573 21.033 12.5293L21.0271 12.6335C20.9967 13.1731 20.9164 13.7087 20.7873 14.2334C20.6488 14.7966 20.4548 15.3442 20.2083 15.8683L20.718 15.7317C21.2515 15.5888 21.7998 15.9054 21.9428 16.4389C22.0857 16.9723 21.7691 17.5207 21.2356 17.6636L18.8208 18.3106C18.0206 18.525 17.1981 18.0501 16.9837 17.2499L16.3367 14.8352C16.1937 14.3017 16.5103 13.7534 17.0438 13.6104C17.5773 13.4675 18.1256 13.7841 18.2685 14.3175ZM3.92363 7.90992L3.33327 8.06809C2.7998 8.21103 2.25147 7.89443 2.10854 7.36097C1.9656 6.8275 2.2822 6.27917 2.81566 6.13623L5.2305 5.48922C6.03069 5.27483 6.85318 5.7497 7.06759 6.54989L7.71463 8.96466C7.85757 9.49813 7.54099 10.0465 7.00752 10.1894C6.47405 10.3324 5.92572 10.0158 5.78278 9.4823L5.64111 8.9536C5.46988 9.32312 5.33225 9.70767 5.23007 10.1026C5.1256 10.5063 5.05872 10.9189 5.03027 11.3349L5.02314 11.4391C4.98667 11.9725 5.00567 12.5081 5.07982 13.0376L5.09286 13.1307C5.15447 13.5705 5.25921 14.0032 5.40556 14.4226C5.53644 14.7976 5.70008 15.1604 5.89456 15.5068L5.92334 15.558C6.17169 16.0003 6.47181 16.4114 6.81738 16.7826C7.03315 17.0145 7.26586 17.2299 7.51359 17.4273L7.66835 17.5505C7.97029 17.791 8.29849 18.0084 8.63879 18.1939C9.02626 18.4052 9.43938 18.5797 9.8595 18.7088C10.2512 18.829 10.6533 18.9124 11.0605 18.9576L11.1291 18.9652C11.551 19.0121 11.9761 19.0213 12.3996 18.9926C12.8368 18.963 13.2703 18.8933 13.6946 18.7842L14.2732 18.6355C14.8081 18.4981 15.3532 18.8202 15.4906 19.3552C15.6281 19.8901 15.3059 20.4351 14.771 20.5726L14.1924 20.7213C13.6492 20.8609 13.0943 20.9502 12.5346 20.988C11.9925 21.0247 11.4483 21.013 10.9083 20.953L10.8397 20.9454C10.3082 20.8863 9.78352 20.7776 9.27238 20.6206C8.72212 20.4517 8.18505 20.2245 7.68131 19.9498C7.24086 19.7097 6.81595 19.4285 6.42229 19.1149L6.26753 18.9917C5.94226 18.7326 5.6367 18.4497 5.35338 18.1453C4.89963 17.6578 4.50556 17.118 4.17948 16.5373L4.1507 16.486C3.89877 16.0374 3.6868 15.5674 3.51725 15.0816C3.32768 14.5384 3.192 13.9779 3.1122 13.4081L3.09915 13.315C3.0058 12.6485 2.98189 11.9741 3.0278 11.3027L3.03493 11.1985C3.07179 10.6594 3.15847 10.1248 3.29383 9.60161C3.44505 9.01718 3.65617 8.4503 3.92363 7.90992Z",
            fill: "currentColor"
        }));
    }
    // CONCATENATED MODULE: ./src/File/Trash.tsx
    function $() {
        $ = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var a in t) {
                    if (Object.prototype.hasOwnProperty.call(t, a)) {
                        e[a] = t[a];
                    }
                }
            }
            return e;
        };
        return $.apply(this, arguments);
    }
    function N(e) {
        var r = (0, b.useSplunkTheme)(), t = r.density;
        var n = t === "compact" ? "20px" : "24px";
        
        return a().createElement(D(), $({
            hideDefaultTooltip: true,
            viewBox: "0 0 24 24",
            width: n,
            height: n
        }, e),  a().createElement("path", {
            d: "M10 4V6H14V4H10ZM16 6H17H19H20C20.5523 6 21 6.44772 21 7C21 7.55228 20.5523 8 20 8H19V20.5C19 21.3284 18.3284 22 17.5 22H6.5C5.67157 22 5 21.3284 5 20.5V8H4C3.44772 8 3 7.55228 3 7C3 6.44772 3.44772 6 4 6H5H7H8V3.5C8 2.67157 8.67157 2 9.5 2H14.5C15.3284 2 16 2.67157 16 3.5V6ZM7 8V20H17V8H7ZM10.5858 14L9.2929 15.2929C8.90237 15.6834 8.90237 16.3166 9.2929 16.7071C9.68342 17.0976 10.3166 17.0976 10.7071 16.7071L12 15.4142L13.2929 16.7071C13.6834 17.0976 14.3166 17.0976 14.7071 16.7071C15.0976 16.3166 15.0976 15.6834 14.7071 15.2929L13.4142 14L14.7071 12.7071C15.0976 12.3166 15.0976 11.6834 14.7071 11.2929C14.3166 10.9024 13.6834 10.9024 13.2929 11.2929L12 12.5858L10.7071 11.2929C10.3166 10.9024 9.68342 10.9024 9.29289 11.2929C8.90237 11.6834 8.90237 12.3166 9.29289 12.7071L10.5858 14Z",
            fill: "currentColor"
        }));
    }
    // CONCATENATED MODULE: external "styled-components"
    const M = require("styled-components");
    var F =  e.n(M);
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const H = require("@splunk/react-ui/Box");
    var A =  e.n(H);
    // CONCATENATED MODULE: external "@splunk/react-ui/Clickable"
    const B = require("@splunk/react-ui/Clickable");
    var z =  e.n(B);
    // CONCATENATED MODULE: external "@splunk/themes/mixins"
    const V = require("@splunk/themes/mixins");
    // CONCATENATED MODULE: ./src/File/ItemStyles.ts
    var Z = F()(A()).withConfig({
        displayName: "ItemStyles__StyledBox",
        componentId: "sc-1fsk3kr-0"
    })([ "position:relative;width:100%;max-width:400px;margin:", ";background-color:", ";border-radius:", ";min-height:32px;color:", ";", " ", " ", " ", ";" ], (0, 
    b.pick)({
        enterprise: "5px auto 0",
        prisma: {
            comfortable: "8px auto 0",
            compact: "4px auto 0"
        }
    }), (0, b.pick)({
        enterprise: {
            light: b.variables.gray92,
            dark: b.variables.gray45
        },
        prisma: b.variables.interactiveColorBackground
    }), (0, b.pick)({
        enterprise: "2px",
        prisma: b.variables.borderRadius
    }), (0, b.pick)({
        enterprise: b.variables.textGray,
        prisma: b.variables.contentColorDefault
    }), (0, b.pick)({
        prisma: (0, M.css)([ "border:1px solid ", ";padding-left:12px;" ], b.variables.neutral200),
        enterprise: (0, M.css)([ "line-height:24px;" ])
    }), (0, b.pick)({
        enterprise: {
            compact: (0, M.css)([ "min-height:24px;" ])
        }
    }), (function(e) {
        var r = e.$error;
        return r && (0, b.pick)({
            enterprise: (0, M.css)([ "box-shadow:inset 0 0 0 1px ", ";" ], b.variables.errorColor),
            prisma: (0, M.css)([ "box-shadow:inset 0 0 0 1px ", ";background-color:", ";" ], b.variables.accentColorNegative, b.variables.interactiveColorBackground)
        });
    }), (function(e) {
        var r = e.$disabled;
        return r && (0, M.css)([ "cursor:not-allowed;background-color:", ";" ], (0, b.pick)({
            enterprise: {
                dark: b.variables.gray60,
                light: b.variables.gray96
            },
            prisma: (0, V.overlayColors)(b.variables.interactiveColorBackground, b.variables.interactiveColorBackgroundDisabled)
        }));
    }));
    var G = F().div.withConfig({
        displayName: "ItemStyles__StyledErrorText",
        componentId: "sc-1fsk3kr-1"
    })([ "font-size:", ";color:", ";line-height:16px;margin-top:4px;text-align:left;" ], b.variables.fontSizeSmall, b.variables.accentColorNegative);
    var W = F()(z()).withConfig({
        displayName: "ItemStyles__StyledIconClickable",
        componentId: "sc-1fsk3kr-2"
    })([ "color:", ";border-radius:", ";padding:", ";text-align:center;", " ", " &:focus{color:", ";", "}&:hover{background-color:", ";color:", ";}" ], (0, 
    b.pick)({
        enterprise: "inherit",
        prisma: b.variables.contentColorDefault
    }), (0, b.pick)({
        enterprise: "0 2px 2px 0",
        prisma: "50%"
    }), (0, b.pick)({
        enterprise: "3px 0",
        prisma: {
            comfortable: "0 8px",
            compact: "0 6px"
        }
    }), (0, b.pick)({
        enterprise: (0, M.css)([ "flex-basis:", ";" ], b.variables.inputHeight)
    }), (function(e) {
        var r = e.$error;
        return r && (0, M.css)([ "", "" ], (0, b.pick)({
            enterprise: (0, M.css)([ "border:1px solid ", ";border-left:none;" ], b.variables.errorColor)
        }));
    }), (0, b.pick)({
        enterprise: b.variables.linkColor,
        prisma: b.variables.contentColorActive
    }), (0, b.pick)({
        enterprise: (0, M.css)([ "box-shadow:", ";" ], b.variables.focusShadow),
        prisma: (0, M.css)([ "background-color:", ";box-shadow:0 0 0 3px ", ";" ], b.variables.interactiveColorOverlayHover, b.variables.focusColor)
    }), (0, b.pick)({
        enterprise: {
            light: b.variables.gray96,
            dark: b.variables.gray30
        },
        prisma: b.variables.interactiveColorOverlayHover
    }), (0, b.pick)({
        enterprise: b.variables.linkColor,
        prisma: b.variables.contentColorActive
    }));
 // for Prisma theme only
        var U = F()(z()).withConfig({
        displayName: "ItemStyles__StyledTrashIcon",
        componentId: "sc-1fsk3kr-3"
    })([ "color:", ";border-radius:50%;padding:", ";margin-right:", ";text-align:center;&:focus{color:", ";background-color:", ";box-shadow:0 0 0 3px ", ";}&:hover{background-color:", ";color:", ";}" ], b.variables.contentColorDefault, (0, 
    b.pick)({
        comfortable: "0 8px",
        compact: "0 6px"
    }), (0, b.pick)({
        comfortable: "0 4px",
        compact: "0 6px"
    }), b.variables.contentColorActive, b.variables.interactiveColorOverlayHover, b.variables.focusColor, b.variables.interactiveColorOverlayHover, b.variables.contentColorActive);
    var X = F().div.withConfig({
        displayName: "ItemStyles__StyledLabel",
        componentId: "sc-1fsk3kr-4"
    })([ "color:", ";overflow:hidden;white-space:nowrap;text-overflow:ellipsis;flex:1 0 0px;padding:", ";border-radius:", ";", "" ], b.variables.contentColorActive, (0, 
    b.pick)({
        enterprise: "3px 0 3px 10px",
        prisma: {
            comfortable: "10px 0 10px 0",
            compact: "6px 0 6px 0"
        }
    }), (0, b.pick)({
        enterprise: "2px",
        prisma: b.variables.borderRadius
    }), (function(e) {
        var r = e.$disabled;
        return r && (0, b.pick)({
            enterprise: {
                light: (0, M.css)([ "color:", ";" ], b.variables.textGray),
                dark: (0, M.css)([ "color:", ";" ], b.variables.textGray)
            },
            prisma: (0, M.css)([ "color:", ";" ], b.variables.contentColorDisabled)
        });
    }));
 // for Prisma theme only
        var Q = F().div.withConfig({
        displayName: "ItemStyles__StyledPaperClipIcon",
        componentId: "sc-1fsk3kr-5"
    })([ "color:", ";padding:", ";", "" ], b.variables.neutral400, (0, b.pick)({
        comfortable: "8px 4px 8px 0",
        compact: "4px 4px 4px 0"
    }), (function(e) {
        var r = e.$disabled;
        return r && (0, M.css)([ "color:", ";cursor:not-allowed;" ], b.variables.contentColorDisabled);
    }));
 // for Prisma theme only
        var J = F().span.withConfig({
        displayName: "ItemStyles__StyledPercentage",
        componentId: "sc-1fsk3kr-6"
    })([ "color:", ";" ], b.variables.contentColorMuted);
 // for Prisma theme only
        var K = F().div.withConfig({
        displayName: "ItemStyles__StyledProgress",
        componentId: "sc-1fsk3kr-7"
    })([ "position:absolute;left:0;top:0;width:", "%;height:100%;background:", ";transition:width 300ms;border-radius:", ";" ], (function(e) {
        var r = e.$uploadPercentage;
        return r;
    }), b.variables.interactiveColorOverlaySelected, b.variables.borderRadius);
    // CONCATENATED MODULE: ./src/File/Item.tsx
    function Y() {
        Y = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var a in t) {
                    if (Object.prototype.hasOwnProperty.call(t, a)) {
                        e[a] = t[a];
                    }
                }
            }
            return e;
        };
        return Y.apply(this, arguments);
    }
    function ee(e, r) {
        if (e == null) return {};
        var t = re(e, r);
        var a, n;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (n = 0; n < i.length; n++) {
                a = i[n];
                if (r.indexOf(a) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, a)) continue;
                t[a] = e[a];
            }
        }
        return t;
    }
    function re(e, r) {
        if (e == null) return {};
        var t = {};
        var a = Object.keys(e);
        var n, i;
        for (i = 0; i < a.length; i++) {
            n = a[i];
            if (r.indexOf(n) >= 0) continue;
            t[n] = e[n];
        }
        return t;
    }
    var te = {
        disabled: i().bool,
        elementRef: i().oneOfType([ i().func, i().object ]),
        error: i().bool,
        /** @private */
        index: i().number,
        itemId: i().any,
        name: i().string.isRequired,
        /** If the uploadPercentage is 0, the item is assumed to be queued. If the upload is complete or
   * not applicable, uploadPercentage must be undefined. */
        uploadPercentage: i().number
    };
    function ae(e) {
        var r = e.disabled, n = e.error, i = n === void 0 ? false : n, o = e.itemId, l = e.name, s = e.index, c = s === void 0 ? 0 : s, p = e.uploadPercentage, d = ee(e, [ "disabled", "error", "itemId", "name", "index", "uploadPercentage" ]);
        // @docs-props-type ItemPropsBase
                var u = (0, b.useSplunkTheme)(), v = u.isPrisma, f = u.isEnterprise, m = u.isCompact;
        var y = m ? "20px" : "24px";
        var g = (0, t.useContext)(_), x = g.onRequestRemove;
        function w(e) {
            e.preventDefault();
            x === null || x === void 0 ? void 0 : x({
                itemId: o,
                name: l,
                index: c
            });
        }
        var O = (0, t.useContext)(_), E = O.onRequestRetry;
        function R(e) {
            e.preventDefault();
            E === null || E === void 0 ? void 0 : E({
                itemId: o,
                name: l,
                index: c
            });
        }
        var D = (0, t.useContext)(_), q = D.disabled;
        var P = (0, I.sprintf)((0, h._)('Remove "%(filename)s"'), {
            filename: l
        });
        var $ = (0, I.sprintf)((0, h._)('Retry "%(filename)s"'), {
            filename: l
        });
        var M = (0, I.sprintf)((0, h._)('Uploading "%(filename)s"'), {
            filename: l
        });
        var F = (0, h._)("Remove");
        var H =  a().createElement("span", null,  a().createElement("span", {
            style: {
                paddingRight: "4px"
            }
        }, (0, h._)("Uploading ")),  a().createElement(J, null, p, "%"));
        var A = v && !C()(p) && p > 0 ? H : l;
        var B = r !== undefined ? r : q;
        
        return a().createElement(a().Fragment, null,  a().createElement(Z, Y({
            "data-test": "item",
            "data-test-disabled": B,
            "data-test-error": i
        }, d, {
            flex: true,
            $error: i,
            $disabled: B
        }), v && C()(p) &&  a().createElement(Q, {
            "data-test": "paper-clip",
            $disabled: B
        },  a().createElement(j, {
            "aria-hidden": "true"
        })),  a().createElement(X, {
            "data-test": "label",
            $error: i,
            $disabled: B
        }, A), v && E && i &&  a().createElement(W, {
            "data-test": "retry",
            onClick: R,
            "aria-label": $
        },  a().createElement(T, {
            screenReaderText: (0, h._)("Retry")
        })), !B && (v && C()(p) && !i ?  a().createElement(U, {
            "data-test": "remove",
            onClick: w,
            "aria-label": P
        },  a().createElement(N, {
            screenReaderText: (0, h._)("Remove")
        })) :  a().createElement(W, {
            "data-test": "remove",
            $error: i,
            onClick: w,
            "aria-label": P
        }, f ?  a().createElement(L(), {
            hideDefaultTooltip: true,
            screenReaderText: F
        }) :  a().createElement(k(), {
            "aria-label": F,
            height: y,
            width: y
        }))), !C()(p) && p > 0 && (v ?  a().createElement(K, {
            $uploadPercentage: p,
            "aria-label": M,
            role: "progressbar",
            "aria-valuenow": p,
            "aria-valuemin": 0,
            "aria-valuemax": 100
        }) :  a().createElement(S(), {
            style: {
                position: "absolute",
                left: 0,
                top: 0,
                right: 0,
                zIndex: 1
            },
            percentage: p,
            "aria-label": M
        }))), v && i &&  a().createElement(G, null, (0, h._)("Something went wrong.")));
    }
    ae.propTypes = te;
    /* harmony default export */    const ne = ae;
    // CONCATENATED MODULE: ./src/File/Icon.tsx
    function ie() {
        ie = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var a in t) {
                    if (Object.prototype.hasOwnProperty.call(t, a)) {
                        e[a] = t[a];
                    }
                }
            }
            return e;
        };
        return ie.apply(this, arguments);
    }
    function oe(e) {
        
        return a().createElement(D(), ie({
            hideDefaultTooltip: true,
            screenReaderText: (0, h._)("File upload"),
            viewBox: "0 0 72 88"
        }, e),  a().createElement("path", {
            d: "M50,27 L68.0005854,27 C70.2022516,27 72,28.7919267 72,31.0023804 L72,83.9976196 C72,86.2074215 70.2094011,88 68.0005854,88 L3.99941455,88 C1.79774843,88 0,86.2080733 0,83.9976196 L0,31.0023804 C0,28.7925785 1.79059889,27 3.99941455,27 L21,27 L21,32 L5.99898406,32 C5.4472604,32 5,32.4408979 5,32.9958767 L5,82.0041233 C5,82.5541308 5.44605521,83 5.99898406,83 L66.0010159,83 C66.5527396,83 67,82.5591021 67,82.0041233 L67,32.9958767 C67,32.4458692 66.5539448,32 66.0010159,32 L50,32 L50,27 Z"
        }),  a().createElement("path", {
            d: "M41.9634682,10 L41.9634682,28 L46.9634682,28 L46.9634682,5 L44.4634682,5 L23.9634682,5 L23.9634682,10 L41.9634682,10 Z",
            transform: "translate(35.463468, 16.500000) rotate(-45.000000) translate(-35.463468, -16.500000) "
        }),  a().createElement("rect", {
            x: "33",
            y: "3",
            width: "5",
            height: "51"
        }));
    }
    // CONCATENATED MODULE: ./src/File/FileStyles.ts
    var le = F().input.withConfig({
        displayName: "FileStyles__StyledInput",
        componentId: "sc-1wyh4cf-0"
    })([ "&[type='file']{width:0.1px;height:0.1px;opacity:0;overflow:hidden;position:absolute;z-index:-1;}" ]);
    var se = (0, M.css)([ "fill:", ";" ], (0, b.pick)({
        enterprise: {
            light: b.variables.gray60,
            dark: b.variables.gray80
        }
    }));
    var ce = F()(oe).withConfig({
        displayName: "FileStyles__StyledNormalIcon",
        componentId: "sc-1wyh4cf-1"
    })([ "", ";height:1.4em;width:1.4em;display:inline-block;vertical-align:middle;padding-bottom:3px;" ], se);
    var pe = F()(oe).withConfig({
        displayName: "FileStyles__StyledWindowIcon",
        componentId: "sc-1wyh4cf-2"
    })([ "", ";height:48px;width:48px;position:absolute;top:30px;left:50%;transform:translateX(-50%);" ], se);
    var de = F().label.withConfig({
        displayName: "FileStyles__StyledLink",
        componentId: "sc-1wyh4cf-3"
    })([ "", ";color:", ";cursor:pointer;font-size:inherit;font-weight:inherit;&:hover,&[data-focused]{text-decoration:underline;}&[data-focused]{box-shadow:", ";}" ], b.mixins.reset("inline"), b.variables.linkColor, (0, 
    b.pick)({
        enterprise: b.variables.focusShadowInset,
        prisma: b.variables.focusShadow
    }));
    var ue = F().div.withConfig({
        displayName: "FileStyles__StyledNormalText",
        componentId: "sc-1wyh4cf-4"
    })([ "", " ", " ", " ", "" ], (0, b.pick)({
        enterprise: (0, M.css)([ "display:inline-block;" ]),
        prisma: (0, M.css)([ "display:inline-flex;flex-direction:column;justify-content:center;text-align:center;color:", ";border:1px dashed ", ";padding:8px;border-radius:", ";min-height:128px;&:not(:last-child){margin-bottom:4px;}" ], b.variables.contentColorActive, b.variables.contentColorMuted, b.variables.borderRadius)
    }), (function(e) {
        var r = e.$dragOver;
        return r && (0, M.css)([ "", "" ], (0, b.pick)({
            prisma: (0, M.css)([ "border:1px dashed ", ";background-color:", ";" ], b.variables.interactiveColorPrimary, b.mixins.colorWithAlpha(b.variables.interactiveColorPrimary, .1))
        }));
    }), (function(e) {
        var r = e.$error, t = e.$dragOver;
        return r && !t && (0, b.pick)({
            prisma: (0, M.css)([ "border:1px dashed ", ";color:", ";background-color:", ";" ], b.variables.accentColorNegative, b.variables.accentColorNegative, b.mixins.colorWithAlpha(b.variables.accentColorNegative, .1))
        });
    }), (function(e) {
        var r = e.$disabled;
        return r && (0, M.css)([ "border:none;min-height:0;color:", ";cursor:not-allowed;" ], (0, 
        b.pick)({
            prisma: b.variables.contentColorDisabled
        }));
    }));
    var ve = F().div.withConfig({
        displayName: "FileStyles__StyledWindowText",
        componentId: "sc-1wyh4cf-5"
    })([ "margin-top:", ";margin-bottom:", ";font-size:", ";", "" ], (0, b.pick)({
        enterprise: (0, M.css)([ "calc(", " * 4)" ], b.variables.spacing),
        prisma: (0, M.css)([ "calc(", " * 4)" ], b.variables.spacingLarge)
    }), (0, b.pick)({
        enterprise: b.variables.spacingHalf,
        prisma: b.variables.spacingMedium
    }), b.variables.fontSizeXLarge, (0, b.pick)({
        prisma: (0, M.css)([ "color:", ";" ], b.variables.contentColorActive)
    }));
    var fe = F().div.withConfig({
        displayName: "FileStyles__StyledSubtitle",
        componentId: "sc-1wyh4cf-6"
    })([ "color:", ";padding:4px 8px;font-size:12px;line-height:16px;overflow:hidden;text-overflow:ellipsis;white-space:nowrap;" ], b.variables.contentColorMuted);
    var me = F().div.withConfig({
        displayName: "FileStyles__StyledHelp",
        componentId: "sc-1wyh4cf-7"
    })([ "margin-bottom:", ";" ], (0, b.pick)({
        enterprise: b.variables.spacingXLarge,
        prisma: (0, M.css)([ "calc(", " * 1.5)" ], b.variables.spacingLarge)
    }));
    var be = F().div.withConfig({
        displayName: "FileStyles__StyledWindowDrop",
        componentId: "sc-1wyh4cf-8"
    })([ "position:fixed;top:0;left:0;right:0;bottom:0;border:", ";z-index:", " + 10;" ], (0, 
    b.pick)({
        enterprise: "5px solid ".concat(b.variables.accentColorL10),
        prisma: "5px solid ".concat(b.variables.interactiveColorPrimary)
    }), b.variables.zindexModal);
    var he = F()(A()).withConfig({
        displayName: "FileStyles__StyledDropNormalTargetBox",
        componentId: "sc-1wyh4cf-9"
    })([ "", ";flex-direction:column;justify-content:center;text-align:center;border-radius:", ";", " ", " ", " ", " ", " ", "" ], b.mixins.reset("flex"), b.variables.borderRadius, (0, 
    b.pick)({
        enterprise: (0, M.css)([ "line-height:calc(", " - 2px);" ], b.variables.inputHeight)
    }), (0, b.pick)({
        enterprise: {
            comfortable: (0, M.css)([ "min-height:73px;padding:", ";" ], b.variables.spacingQuarter),
            compact: (0, M.css)([ "min-height:63px;padding:3px;" ])
        }
    }), (0, b.pick)({
        enterprise: {
            light: (0, M.css)([ "border:1px dashed ", ";" ], b.variables.borderColor),
            dark: (0, M.css)([ "border:1px dashed ", ";" ], b.variables.textGray)
        }
    }), (function(e) {
        var r = e.$dragOver;
        return r && (0, b.pick)({
            enterprise: (0, M.css)([ "border:1px solid ", ";" ], b.variables.accentColorL10)
        });
    }), (function(e) {
        var r = e.$error, t = e.$dragOver;
        return r && !t && (0, b.pick)({
            enterprise: (0, M.css)([ "border:1px solid ", ";" ], b.variables.errorColor)
        });
    }), (function(e) {
        var r = e.$disabled, t = e.$fileCount;
        return r && (0, M.css)([ "border:", ";color:", ";cursor:not-allowed;", " ", "" ], (0, 
        b.pick)({
            enterprise: "none",
            prisma: "1px solid ".concat(b.variables.contentColorDisabled)
        }), (0, b.pick)({
            enterprise: b.variables.textGray,
            prisma: b.variables.contentColorDisabled
        }), t === 0 && (0, b.pick)({
            enterprise: {
                light: (0, M.css)([ "background-color:", ";" ], b.variables.gray96),
                dark: (0, M.css)([ "background-color:", ";" ], b.variables.gray60)
            },
            prisma: (0, M.css)([ "border:1px dashed ", ";min-height:128px;" ], b.variables.contentColorDisabled)
        }), t !== 0 && (0, b.pick)({
            prisma: (0, M.css)([ "", "{padding:0;}" ], 
            /* sc-sel */
            ue)
        }));
    }));
    var ye = F()(A()).withConfig({
        displayName: "FileStyles__StyledWindowDropTargetBox",
        componentId: "sc-1wyh4cf-10"
    })([ "position:relative;text-align:center;min-height:250px;padding:", ";", "" ], b.variables.spacingLarge, (function(e) {
        var r = e.$disabled;
        return r && (0, M.css)([ "color:", ";" ], (0, b.pick)({
            enterprise: b.variables.textGray,
            prisma: b.variables.contentColorMuted
        }));
    }));
    // CONCATENATED MODULE: ./src/File/File.tsx
    function ge(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            ge = function e(r) {
                return typeof r;
            };
        } else {
            ge = function e(r) {
                return r && typeof Symbol === "function" && r.constructor === Symbol && r !== Symbol.prototype ? "symbol" : typeof r;
            };
        }
        return ge(e);
    }
    function Ce() {
        Ce = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var a in t) {
                    if (Object.prototype.hasOwnProperty.call(t, a)) {
                        e[a] = t[a];
                    }
                }
            }
            return e;
        };
        return Ce.apply(this, arguments);
    }
    function xe(e, r) {
        if (e == null) return {};
        var t = ke(e, r);
        var a, n;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (n = 0; n < i.length; n++) {
                a = i[n];
                if (r.indexOf(a) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, a)) continue;
                t[a] = e[a];
            }
        }
        return t;
    }
    function ke(e, r) {
        if (e == null) return {};
        var t = {};
        var a = Object.keys(e);
        var n, i;
        for (i = 0; i < a.length; i++) {
            n = a[i];
            if (r.indexOf(n) >= 0) continue;
            t[n] = e[n];
        }
        return t;
    }
    function we(e, r) {
        if (!(e instanceof r)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function Le(e, r) {
        for (var t = 0; t < r.length; t++) {
            var a = r[t];
            a.enumerable = a.enumerable || false;
            a.configurable = true;
            if ("value" in a) a.writable = true;
            Object.defineProperty(e, a.key, a);
        }
    }
    function Oe(e, r, t) {
        if (r) Le(e.prototype, r);
        if (t) Le(e, t);
        return e;
    }
    function Se(e, r) {
        if (typeof r !== "function" && r !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(r && r.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (r) Ie(e, r);
    }
    function Ie(e, r) {
        Ie = Object.setPrototypeOf || function e(r, t) {
            r.__proto__ = t;
            return r;
        };
        return Ie(e, r);
    }
    function _e(e) {
        var r = De();
        return function t() {
            var a = qe(e), n;
            if (r) {
                var i = qe(this).constructor;
                n = Reflect.construct(a, arguments, i);
            } else {
                n = a.apply(this, arguments);
            }
            return Ee(this, n);
        };
    }
    function Ee(e, r) {
        if (r && (ge(r) === "object" || typeof r === "function")) {
            return r;
        }
        return Re(e);
    }
    function Re(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function De() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function qe(e) {
        qe = Object.setPrototypeOf ? Object.getPrototypeOf : function e(r) {
            return r.__proto__ || Object.getPrototypeOf(r);
        };
        return qe(e);
    }
    function je(e, r, t) {
        if (r in e) {
            Object.defineProperty(e, r, {
                value: t,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[r] = t;
        }
        return e;
    }
    /** @public */    var Pe = {
        accept: i().string,
        allowMultiple: i().bool,
        children: i().node,
        disabled: i().bool,
        dropAnywhere: i().bool,
        elementRef: i().oneOfType([ i().func, i().object ]),
        fullscreen: i().bool,
        error: i().bool,
        help: i().node,
        name: i().string,
        onRequestAdd: i().func,
        onRequestRemove: i().func,
        onRequestRetry: i().func,
        /** @private. */
        required: i().bool,
        supportsMessage: i().node,
        /** @private */
        splunkTheme: i().object
    };
    var Te = {
        allowMultiple: false,
        disabled: false,
        dropAnywhere: false,
        error: false
    };
    /**
 * File provides the ability to accept files and present uploaded files. It does not provide
 * file readers, only a reference to the file. This can be used to post binary content, or
 * upload using an array buffer.
 */    var $e =  function(e) {
        Se(n, e);
        var r = _e(n);
        // @docs-props-type FilePropsBase
                function n(e) {
            var t;
            we(this, n);
            t = r.call(this, e);
            je(Re(t), "inputId", void 0);
            je(Re(t), "inputCount", void 0);
            je(Re(t), "handleInputChange", (function(e) {
                var r;
                t.addFiles((r = e.currentTarget.files) !== null && r !== void 0 ? r : undefined);
            }));
            je(Re(t), "handleInputFocus", (function() {
                t.setState({
                    focusedInput: true
                });
            }));
            je(Re(t), "handleInputBlur", (function() {
                t.setState({
                    focusedInput: false
                });
            }));
            je(Re(t), "handleDragOver", (function(e) {
                if (!t.state.dragOver) {
                    t.setState({
                        dragOver: true
                    });
                }
                e.preventDefault();
            }));
            je(Re(t), "handleDragLeave", (function() {
                t.setState({
                    dragOver: false
                });
            }));
            je(Re(t), "handleDrop", (function(e) {
                var r;
                e.preventDefault();
                t.handleDragLeave();
                t.addFiles((r = e.dataTransfer) === null || r === void 0 ? void 0 : r.files);
            }));
            t.state = {
                dragOver: false,
                focusedInput: false
            };
            t.inputId = (0, y.createDOMID)();
            /* Each time a file is uploaded this is incremented and used to generate the
     * file input's key. In this way we get a new input without a value. */            t.inputCount = 0;
            t.handleDragLeave = l()(t.handleDragLeave, 300);
            return t;
        }
        Oe(n, [ {
            key: "addFiles",
            value: function e(r) {
                var t = this.props.name;
                if (r !== undefined && r.length > 0) {
                    var a, n;
                    var i = Array.from(r);
                    (a = (n = this.props).onRequestAdd) === null || a === void 0 ? void 0 : a.call(n, this.props.allowMultiple ? i : [ i[0] ], {
                        name: t
                    });
                } else {
                    var o, l;
                    (o = (l = this.props).onRequestAdd) === null || o === void 0 ? void 0 : o.call(l, [], {
                        name: t
                    });
                }
                this.inputCount += 1;
            }
        }, {
            key: "render",
            value: function e() {
                var r = this.props, n = r.accept, i = r.allowMultiple, o = r.children, l = r.disabled, s = r.dropAnywhere, p = r.error, u = r.fullscreen, f = r.help, b = r.name, y = r.onRequestRemove, g = r.onRequestRetry, C = r.required, x = r.splunkTheme, k = xe(r, [ "accept", "allowMultiple", "children", "disabled", "dropAnywhere", "error", "fullscreen", "help", "name", "onRequestRemove", "onRequestRetry", "required", "splunkTheme" ]);
                var w = x.isPrisma;
                var L = 0;
                var O = t.Children.toArray(o).filter(t.isValidElement).map((function(e, r) {
                    L += 1;
                    
                    return (0, t.cloneElement)(e, {
                        index: r,
                        key: e.key || e.props.itemId || "item-".concat(r)
                    });
                }));
                var S = u || s;
                var I = S || l;
                var _ = S && !l;
                var R = u ? ye : he;
                var D = u ? ve : ue;
                var q = u ? pe : ce;
                var j = I ? false : this.state.dragOver || false;
                var P = v()(n) && d()(n.split(",").join(", "));
                var T = function e(r) {
                    g === null || g === void 0 ? void 0 : g({
                        event: r,
                        filename: r.name,
                        index: r.index,
                        itemId: r.itemId,
                        name: b
                    });
                };
                var $ = function e(r) {
                    y === null || y === void 0 ? void 0 : y({
                        event: r,
                        index: r.index,
                        filename: r.name,
                        itemId: r.itemId,
                        name: b
                    });
                };
                var N = null;
                if (this.props.supportsMessage) {
                    N =  a().createElement(fe, {
                        "data-test": "file-supports"
                    }, this.props.supportsMessage);
                } else if (w && !l) {
                    N = n ?  a().createElement(fe, {
                        "data-test": "file-supports"
                    }, (0, h._)("Supports ".concat(P))) :  a().createElement(fe, {
                        "data-test": "file-supports"
                    }, (0, h._)("Supports all file types"));
                }
                
                return a().createElement(R, Ce({
                    onDragOver: I ? undefined : this.handleDragOver,
                    onDragLeave: I ? undefined : this.handleDragLeave,
                    onDrop: I ? undefined : this.handleDrop,
                    $disabled: l,
                    $dragOver: j,
                    $error: p,
                    $fileCount: L,
                    "data-test": "file",
                    "data-test-disabled": l,
                    "data-test-error": p
                }, c()(k, "onRequestAdd", "onRequestRemove")),  a().createElement(D, {
                    $disabled: l,
                    $dragOver: j,
                    $error: p
                }, !l && !w &&  a().createElement(q, {
                    $dragOver: j,
                    $error: p
                }), " ",  a().createElement("span", {
                    "data-test": "file-label"
                }, !S && !l && (0, h._)("Drop your file here or"), _ && (0, h._)("Drop your file anywhere or"), " ", L === 0 && l && (0, 
                h._)("No files selected"),  a().createElement(de, {
                    htmlFor: this.inputId,
                    "data-test": "file-link",
                    "data-focused": this.state.focusedInput || null
                },  a().createElement(le, {
                    accept: n,
                    "data-test": "file-input",
                    disabled: l,
                    id: this.inputId,
                    key: "file-input-".concat(this.inputCount),
                    multiple: i || undefined,
                    onBlur: this.handleInputBlur,
                    onChange: this.handleInputChange,
                    onFocus: this.handleInputFocus,
                    required: C,
                    type: "file"
                }), !l && (0, h._)("upload file…"))), N), !l && f &&  a().createElement(me, {
                    "data-test": "help"
                }, f), _ && this.state.dragOver &&  a().createElement(be, {
                    "data-test": "file-window-drop",
                    onDragLeave: this.handleDragLeave
                }), _ &&  a().createElement(a().Fragment, null,  a().createElement(m(), {
                    target: window,
                    eventType: "dragover",
                    listener: this.handleDragOver
                }),  a().createElement(m(), {
                    target: window,
                    eventType: "drop",
                    listener: this.handleDrop
                })),  a().createElement(E.Provider, {
                    value: {
                        disabled: l,
                        onRequestRemove: y ? $ : undefined,
                        onRequestRetry: g ? T : undefined
                    }
                }, O));
            }
        } ]);
        return n;
    }(t.Component);
    je($e, "propTypes", Pe);
    je($e, "defaultProps", Te);
    je($e, "Item", ne);
    var Ne = (0, b.withSplunkTheme)($e);
    var Me = Ne;
    Me.propTypes = $e.propTypes;
    Me.Item = ne;
    /* harmony default export */    const Fe = Me;
    // CONCATENATED MODULE: ./src/File/index.ts
    module.exports = r;
    /******/})();
