// @splunk/react-ui/Layer.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var n = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(n, {
                a: n
            });
            /******/            return n;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, n) => {
            /******/ for (var r in n) {
                /******/ if (e.o(n, r) && !e.o(t, r)) {
                    /******/ Object.defineProperty(t, r, {
                        enumerable: true,
                        get: n[r]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/global */
    /******/    (() => {
        /******/ e.g = function() {
            /******/ if (typeof globalThis === "object") return globalThis;
            /******/            try {
                /******/ return this || new Function("return this")();
                /******/            } catch (e) {
                /******/ if (typeof window === "object") return window;
                /******/            }
            /******/        }();
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        LayerContext: () => /* reexport */ m,
        LayerStackContext: () => /* reexport */ h,
        LayerStackGlobalProvider: () => /* reexport */ k,
        default: () => /* reexport */ F
    });
    // CONCATENATED MODULE: external "react"
    const n = require("react");
    var r =  e.n(n);
    // CONCATENATED MODULE: external "react-dom"
    const o = require("react-dom");
    // CONCATENATED MODULE: external "prop-types"
    const i = require("prop-types");
    var a =  e.n(i);
    // CONCATENATED MODULE: external "lodash/includes"
    const c = require("lodash/includes");
    var u =  e.n(c);
    // CONCATENATED MODULE: external "lodash/last"
    const l = require("lodash/last");
    var s =  e.n(l);
    // CONCATENATED MODULE: external "lodash/pull"
    const f = require("lodash/pull");
    var p =  e.n(f);
    // CONCATENATED MODULE: external "@splunk/ui-utils/keyboard"
    const y = require("@splunk/ui-utils/keyboard");
    // CONCATENATED MODULE: external "@splunk/react-ui/EventListener"
    const d = require("@splunk/react-ui/EventListener");
    var v =  e.n(d);
    // CONCATENATED MODULE: ./src/Layer/LayerStack.tsx
    /**
 * `LayerStackContext` is used to manage the array that `Layer` (and components that
 * depend on it, such as `Modal`) uses to determine the order of open layers.
 * @public
 */
    var h =  r().createContext([]);
    var m =  r().createContext({});
    var b = {
        children: a().node,
        name: a().string,
        scope: a().object,
        separateStackingContexts: a().bool
    };
    /* global global */
    /**
 * A `LayerStackContext` provider that stores a shared layer stack using a global variable.
 * Applications should only use this provider if there's a known need to support multiple
 * instances of this library on the same page.
 */    function k(t) {
        var n = t.children, o = t.name, i = o === void 0 ? "__splunkui_layer_instances__" : o, a = t.scope, c = a === void 0 ? typeof window !== "undefined" ? window : e.g : a, u = t.separateStackingContexts, l = u === void 0 ? false : u;
        if (!c[i]) {
            c[i] = [];
 // eslint-disable-line no-param-reassign
                }
        
        return r().createElement(h.Provider, {
            value: c[i]
        },  r().createElement(m.Provider, {
            value: {
                separateStackingContexts: l
            }
        }, n));
    }
    k.propTypes = b;
    // CONCATENATED MODULE: external "styled-components"
    const w = require("styled-components");
    var g =  e.n(w);
    // CONCATENATED MODULE: external "@splunk/themes"
    const C = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/Layer/LayerStyles.ts
    var S = g().div.withConfig({
        displayName: "LayerStyles__StyledLayer",
        componentId: "ii6psl-0"
    })([ "", "" ], (function(e) {
        var t = e.$separateStackingContexts;
        return t && (0, w.css)([ "isolation:isolate;z-index:", ";" ], C.variables.zindexLayer);
    }));
    // CONCATENATED MODULE: ./src/utils/ssrDocument.ts
    /* eslint-disable @typescript-eslint/no-empty-function */
    var E = {
        body: {
            appendChild: function e() {
                return [];
            }
        },
        addEventListener: function e() {},
        removeEventListener: function e() {},
        activeElement: {
            blur: function e() {},
            nodeName: ""
        },
        querySelector: function e() {
            return null;
        },
        querySelectorAll: function e() {
            return [];
        },
        getElementById: function e() {
            return null;
        },
        createEvent: function e() {
            return {
                initEvent: function e() {}
            };
        },
        createElement: function e() {
            return {
                children: [],
                childNodes: [],
                style: {},
                setAttribute: function e() {},
                getElementsByTagName: function e() {
                    return [];
                }
            };
        },
        createElementNS: function e() {
            return {};
        },
        importNode: function e() {
            return null;
        },
        location: {
            hash: "",
            host: "",
            hostname: "",
            href: "",
            origin: "",
            pathname: "",
            protocol: "",
            search: ""
        }
    };
    function O() {
        var e = typeof document !== "undefined" ? document : E;
        return e;
    }
    // CONCATENATED MODULE: ./src/Layer/Layer.tsx
    function L(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            L = function e(t) {
                return typeof t;
            };
        } else {
            L = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return L(e);
    }
    function _(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function x(e, t) {
        for (var n = 0; n < t.length; n++) {
            var r = t[n];
            r.enumerable = r.enumerable || false;
            r.configurable = true;
            if ("value" in r) r.writable = true;
            Object.defineProperty(e, r.key, r);
        }
    }
    function P(e, t, n) {
        if (t) x(e.prototype, t);
        if (n) x(e, n);
        return e;
    }
    function j(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) q(e, t);
    }
    function q(e, t) {
        q = Object.setPrototypeOf || function e(t, n) {
            t.__proto__ = n;
            return t;
        };
        return q(e, t);
    }
    function T(e) {
        var t = A();
        return function n() {
            var r = K(e), o;
            if (t) {
                var i = K(this).constructor;
                o = Reflect.construct(r, arguments, i);
            } else {
                o = r.apply(this, arguments);
            }
            return R(this, o);
        };
    }
    function R(e, t) {
        if (t && (L(t) === "object" || typeof t === "function")) {
            return t;
        }
        return D(e);
    }
    function D(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function A() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function K(e) {
        K = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return K(e);
    }
    function M(e, t, n) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: n,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = n;
        }
        return e;
    }
    /** @public */    var N = [ "clickAway", "escapeKey" ];
    var W = {
        children: a().node,
        closeReasons: a().arrayOf(a().oneOf(N)),
        onRequestClose: a().func,
        open: a().bool
    };
    var z = {
        closeReasons: N,
        open: false
    };
    var B =  function(e) {
        j(n, e);
        var t = T(n);
        // @docs-props-type LayerPropsBase
        // eslint-disable-next-line react/sort-comp
        // LayerStackContext (an array) keeps track of the current instances of Layer. This is
        // used by Layer#handleKeyDown to determine if the escapeKey event should be handled by
        // the current instance. Only the topmost Layer instance should honor the escapeKey.
        // TODO: enable once this is sorted out within the babel/ts ecosystem
        // declare context: React.ContextType<typeof LayerStackContext>;
                function n(e) {
            var r;
            _(this, n);
            r = t.call(this, e);
            M(D(r), "layerClickEvent", null);
            M(D(r), "handleClickOnLayer", (function(e) {
                var t = e.nativeEvent;
                r.layerClickEvent = t;
            }));
            M(D(r), "handleClickOnWindow", (function(e) {
                var t, n;
                // clicks inside the layer should not be considered clickAways
                                if (!r.props.open || !u()(r.props.closeReasons, "clickAway") || r.layerClickEvent === e) {
                    return;
                }
                (t = (n = r.props).onRequestClose) === null || t === void 0 ? void 0 : t.call(n, {
                    event: e,
                    reason: "clickAway"
                });
            }));
            M(D(r), "handleKeyDownOnWindow", (function(e) {
                if (r.props.open && (0, y.keycode)(e) === "esc" && s()(r.getLayerStack()) === D(r) && u()(r.props.closeReasons, "escapeKey")) {
                    var t, n;
                    (t = (n = r.props).onRequestClose) === null || t === void 0 ? void 0 : t.call(n, {
                        event: e,
                        reason: "escapeKey"
                    });
                }
            }));
            var o = O();
            if (!n.layerContainer) {
                n.layerContainer = o.createElement("div");
                n.layerContainer.setAttribute("data-test", "layer-container");
                o.body.appendChild(n.layerContainer);
            }
            return r;
        }
        P(n, [ {
            key: "componentDidMount",
            value: function e() {
                if (this.props.open) {
                    this.getLayerStack().push(this);
                }
            }
        }, {
            key: "componentDidUpdate",
            value: function e(t) {
                if (!t.open && this.props.open) {
                    this.getLayerStack().push(this);
                } else if (t.open && !this.props.open) {
                    p()(this.getLayerStack(), this);
                }
            }
        }, {
            key: "componentWillUnmount",
            value: function e() {
                p()(this.getLayerStack(), this);
            }
        }, {
            key: "getLayerStack",
            value: function e() {
                return this.context;
            }
        }, {
            key: "render",
            value: function e() {
                var t = this;
                var i = this.props, a = i.children, c = i.open;
                if (c) {
                    var u =  (0, o.createPortal)( r().createElement(m.Consumer, null, (function(e) {
                        var n = e.separateStackingContexts, o = n === void 0 ? false : n;
                        
                        return r().createElement(S, {
                            $separateStackingContexts: o,
                            "data-test": "layer",
                            onMouseDown: t.handleClickOnLayer,
                            onTouchStart: t.handleClickOnLayer
                        }, a);
                    })), n.layerContainer);
                    
                    return r().createElement(r().Fragment, null,  r().createElement(v(), {
                        target: window,
                        eventType: "keydown",
                        listener: this.handleKeyDownOnWindow,
                        key: "eventListenerKeydown"
                    }),  r().createElement(v(), {
                        target: window,
                        eventType: "mousedown",
                        listener: this.handleClickOnWindow,
                        key: "eventListenerMouseDown"
                    }),  r().createElement(v(), {
                        target: window,
                        eventType: "touchstart",
                        listener: this.handleClickOnWindow,
                        key: "eventListenerTouchStart",
                        options: {
                            passive: true
                        }
                    }), u);
                }
                return null;
            }
        } ]);
        return n;
    }(n.Component);
    M(B, "layerContainer", null);
    M(B, "possibleCloseReasons", N);
    M(B, "propTypes", W);
    M(B, "defaultProps", z);
    M(B, "contextType", h);
    /* harmony default export */    const F = B;
    // CONCATENATED MODULE: ./src/Layer/index.ts
    module.exports = t;
    /******/})();
