// @splunk/react-ui/Link.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var n = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(n, {
                a: n
            });
            /******/            return n;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, n) => {
            /******/ for (var r in n) {
                /******/ if (e.o(n, r) && !e.o(t, r)) {
                    /******/ Object.defineProperty(t, r, {
                        enumerable: true,
                        get: n[r]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        default: () => /* reexport */ J
    });
    // CONCATENATED MODULE: external "react"
    const n = require("react");
    var r =  e.n(n);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var i =  e.n(o);
    // CONCATENATED MODULE: external "@splunk/react-ui/ScreenReaderContent"
    const a = require("@splunk/react-ui/ScreenReaderContent");
    var l =  e.n(a);
    // CONCATENATED MODULE: external "@splunk/react-ui/Dropdown"
    const u = require("@splunk/react-ui/Dropdown");
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const c = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "styled-components"
    const s = require("styled-components");
    var f =  e.n(s);
    // CONCATENATED MODULE: external "@splunk/react-icons/ArrowSquareTopRight"
    const p = require("@splunk/react-icons/ArrowSquareTopRight");
    var d =  e.n(p);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/External"
    const y = require("@splunk/react-icons/enterprise/External");
    var b =  e.n(y);
    // CONCATENATED MODULE: external "@splunk/themes"
    const v = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/Link/icons/External.tsx
    var h = (0, s.css)([ "width:0.8em;height:0.8em;vertical-align:baseline;" ]);
    var m = f()(b()).withConfig({
        displayName: "External__StyledEnterprise",
        componentId: "sc-13bdx45-0"
    })([ "", " margin:0 0 0 3px;" ], h);
    var g = f()(d()).withConfig({
        displayName: "External__StyledPrisma",
        componentId: "sc-13bdx45-1"
    })([ "", " margin:0 0 0 4px;" ], h);
    function w() {
        var e = (0, v.useSplunkTheme)(), t = e.isEnterprise;
        return t ?  r().createElement(m, {
            screenReaderText: null
        }) :  r().createElement(g, null);
    }
    /* harmony default export */    const O = w;
    // CONCATENATED MODULE: external "@splunk/react-ui/Clickable"
    const x = require("@splunk/react-ui/Clickable");
    var S =  e.n(x);
    // CONCATENATED MODULE: ./src/Link/LinkStyles.ts
    var k = f()(S()).withConfig({
        displayName: "LinkStyles__StyledClickable",
        componentId: "sc-1hhltcf-0"
    })([ "", ";color:", ";text-decoration:none;font-size:inherit;font-weight:inherit;line-height:inherit;border-radius:", ";&:not([disabled],[aria-disabled='true']){cursor:pointer;", ";&:focus{box-shadow:", ";outline:0;&:active{box-shadow:none;}}}&[disabled],&[aria-disabled='true']{color:", ";}" ], v.mixins.reset("inline"), (0, 
    v.pick)({
        enterprise: v.variables.linkColor,
        prisma: v.variables.interactiveColorPrimary
    }), (0, v.pick)({
        enterprise: 0,
        prisma: "1px"
    }), (0, v.pick)({
        enterprise: (0, s.css)([ "&:hover{text-decoration:underline;}" ]),
        prisma: (0, s.css)([ "&:hover:not(:focus),&:active{text-decoration:underline;text-underline-position:under;}" ])
    }), v.variables.focusShadow, v.variables.contentColorDisabled);
    // CONCATENATED MODULE: ./src/Link/Link.tsx
    function j(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            j = function e(t) {
                return typeof t;
            };
        } else {
            j = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return j(e);
    }
    function _() {
        _ = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var n = arguments[t];
                for (var r in n) {
                    if (Object.prototype.hasOwnProperty.call(n, r)) {
                        e[r] = n[r];
                    }
                }
            }
            return e;
        };
        return _.apply(this, arguments);
    }
    function P(e, t) {
        if (e == null) return {};
        var n = C(e, t);
        var r, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                r = i[o];
                if (t.indexOf(r) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, r)) continue;
                n[r] = e[r];
            }
        }
        return n;
    }
    function C(e, t) {
        if (e == null) return {};
        var n = {};
        var r = Object.keys(e);
        var o, i;
        for (i = 0; i < r.length; i++) {
            o = r[i];
            if (t.indexOf(o) >= 0) continue;
            n[o] = e[o];
        }
        return n;
    }
    function E(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function q(e, t) {
        for (var n = 0; n < t.length; n++) {
            var r = t[n];
            r.enumerable = r.enumerable || false;
            r.configurable = true;
            if ("value" in r) r.writable = true;
            Object.defineProperty(e, r.key, r);
        }
    }
    function R(e, t, n) {
        if (t) q(e.prototype, t);
        if (n) q(e, n);
        return e;
    }
    function T(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) I(e, t);
    }
    function I(e, t) {
        I = Object.setPrototypeOf || function e(t, n) {
            t.__proto__ = n;
            return t;
        };
        return I(e, t);
    }
    function N(e) {
        var t = A();
        return function n() {
            var r = L(e), o;
            if (t) {
                var i = L(this).constructor;
                o = Reflect.construct(r, arguments, i);
            } else {
                o = r.apply(this, arguments);
            }
            return M(this, o);
        };
    }
    function M(e, t) {
        if (t && (j(t) === "object" || typeof t === "function")) {
            return t;
        }
        return D(e);
    }
    function D(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function A() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function L(e) {
        L = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return L(e);
    }
    function z(e, t, n) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: n,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = n;
        }
        return e;
    }
    var F = {
        disabled: false,
        openInNewContext: false
    };
    var B = {
        /** @private */
        allowDisabledLink: i().bool,
        children: i().node,
        disabled: i().bool,
        elementRef: i().oneOfType([ i().func, i().object ]),
        openInNewContext: i().oneOfType([ i().bool, i().string ]),
        to: i().string
    };
    var G = (0, c._)("(Opens new window)");
    /**
 * `Link` is a simple method for configuring `Button` for inline links. For more complex behaviors,
 * see the `Button` documentation.
 */    var H =  function(e) {
        T(n, e);
        var t = N(n);
        function n() {
            var e;
            E(this, n);
            for (var r = arguments.length, o = new Array(r), i = 0; i < r; i++) {
                o[i] = arguments[i];
            }
            e = t.call.apply(t, [ this ].concat(o));
            z(D(e), "component", null);
            z(D(e), "handleMount", (function(t) {
                e.component = t;
            }));
            return e;
        }
        R(n, [ {
            key: "focus",
            /**
     * Place focus on the link.
     */
            value: function e() {
                var t;
                (t = this.component) === null || t === void 0 ? void 0 : t.focus();
            }
        }, {
            key: "render",
            value: function e() {
                var t = this.props, n = t.children, o = t.openInNewContext, i = P(t, [ "children", "openInNewContext" ]);
                var a = false;
                var u;
                if (o) {
                    a = true;
                    u = typeof o === "string" ? o : G;
                }
                
                return r().createElement(k, _({
                    "data-test": "link",
                    ref: this.handleMount,
                    openInNewContext: a
                }, i), n, o &&  r().createElement(r().Fragment, null,  r().createElement(O, null),  r().createElement(l(), null, u)));
            }
        } ]);
        return n;
    }(n.Component);
    z(H, "propTypes", B);
    z(H, "defaultProps", F);
    z(H, u.legacyRefMode, true);
    /* harmony default export */    const J = H;
    // CONCATENATED MODULE: ./src/Link/index.ts
    module.exports = t;
    /******/})();
