// @splunk/react-ui/List.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var r = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(r, {
                a: r
            });
            /******/            return r;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        Item: () => /* reexport */ v,
        default: () => /* reexport */ P
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    var n =  e.n(r);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var i =  e.n(o);
    // CONCATENATED MODULE: external "styled-components"
    const l = require("styled-components");
    var a =  e.n(l);
    // CONCATENATED MODULE: external "@splunk/themes"
    const c = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/List/ListStyles.ts
    var f = a().ul.withConfig({
        displayName: "ListStyles__StyledUl",
        componentId: "sc-1yyh8l9-0"
    })([ "", ";", ";list-style-type:", ";padding-left:40px;margin-bottom:1.3em;margin-top:1em;" ], c.mixins.reset("block"), c.mixins.typography("body"), (function(e) {
        var t = e.$listStyleType;
        return t;
    }));
    var p = a().li.withConfig({
        displayName: "ListStyles__StyledItem",
        componentId: "sc-1yyh8l9-1"
    })([ "", ";list-style-type:inherit;" ], c.mixins.reset("list-item"));
    // CONCATENATED MODULE: ./src/List/Item.tsx
    function s() {
        s = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return s.apply(this, arguments);
    }
    function u(e, t) {
        if (e == null) return {};
        var r = y(e, t);
        var n, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                n = i[o];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function y(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var o, i;
        for (i = 0; i < n.length; i++) {
            o = n[i];
            if (t.indexOf(o) >= 0) continue;
            r[o] = e[o];
        }
        return r;
    }
    var d = {
        children: i().node,
        elementRef: i().oneOfType([ i().func, i().object ])
    };
    /**
 * A container for items of a `List`.
 */    function m(e) {
        var t = e.children, r = e.elementRef, o = u(e, [ "children", "elementRef" ]);
        // @docs-props-type ItemPropsBase
                
        return n().createElement(p, s({
            "data-test": "item",
            ref: r
        }, o), t);
    }
    m.propTypes = d;
    /* harmony default export */    const v = m;
    // CONCATENATED MODULE: ./src/List/List.tsx
    function b() {
        b = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return b.apply(this, arguments);
    }
    function O(e, t) {
        if (e == null) return {};
        var r = h(e, t);
        var n, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                n = i[o];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function h(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var o, i;
        for (i = 0; i < n.length; i++) {
            o = n[i];
            if (t.indexOf(o) >= 0) continue;
            r[o] = e[o];
        }
        return r;
    }
    var g = {
        children: i().node,
        elementRef: i().oneOfType([ i().func, i().object ]),
        ordered: i().bool,
        type: i().oneOf([ "disc", "decimal", "lower-alpha", "upper-alpha" ])
    };
    var j = {
        type: "disc"
    };
    function S(e) {
        var t = e.children, r = e.elementRef, o = e.type, i = e.ordered, l = O(e, [ "children", "elementRef", "type", "ordered" ]);
        // @docs-props-type ListPropsBase
        // Default prop value is "disc", we cannot determine if user or defaultProps is setting the value
                if (false) {}
        
        return n().createElement(f, b({
            as: i || o !== "disc" ? "ol" : "ul",
            "data-test": "list",
            ref: r,
            $listStyleType: i ? "decimal" : o
        }, l), t);
    }
    S.propTypes = g;
    S.defaultProps = j;
    S.Item = v;
    /* harmony default export */    const P = S;
    // CONCATENATED MODULE: ./src/List/index.ts
    module.exports = t;
    /******/})();
