// @splunk/react-ui/Markdown.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var n = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(n, {
                a: n
            });
            /******/            return n;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, n) => {
            /******/ for (var t in n) {
                /******/ if (e.o(n, t) && !e.o(r, t)) {
                    /******/ Object.defineProperty(r, t, {
                        enumerable: true,
                        get: n[t]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        MarkdownAnchorHeading: () => /* reexport */ ee,
        MarkdownBlockquote: () => /* reexport */ k,
        MarkdownCode: () => /* reexport */ R,
        MarkdownCodeBlock: () => /* reexport */ B,
        MarkdownHeading: () => /* reexport */ G,
        MarkdownImage: () => /* reexport */ ae,
        MarkdownItem: () => /* reexport */ Re,
        MarkdownLink: () => /* reexport */ ye,
        MarkdownList: () => /* reexport */ ke,
        MarkdownParagraph: () => /* reexport */ _e,
        default: () => /* reexport */ Je
    });
    // CONCATENATED MODULE: external "react"
    const n = require("react");
    var t =  e.n(n);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var i =  e.n(o);
    // CONCATENATED MODULE: external "commonmark-react-renderer"
    const a = require("commonmark-react-renderer");
    var l =  e.n(a);
    // CONCATENATED MODULE: external "commonmark"
    const c = require("commonmark");
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const u = require("@splunk/react-ui/Box");
    var f =  e.n(u);
    // CONCATENATED MODULE: external "styled-components"
    const s = require("styled-components");
    var p =  e.n(s);
    // CONCATENATED MODULE: external "@splunk/themes"
    const d = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/Markdown/MarkdownStyles.ts
    var y = p().div.withConfig({
        displayName: "MarkdownStyles__StyledCodeBlock",
        componentId: "sc-1xadih-0"
    })([ "padding:", ";margin:", " 0;background-color:", ";" ], (0, d.pick)({
        enterprise: d.variables.spacingHalf,
        prisma: (0, s.css)([ "10px ", "" ], d.variables.spacingLarge)
    }), (0, d.pick)({
        enterprise: d.variables.spacingHalf,
        prisma: d.variables.spacingLarge
    }), (0, d.pick)({
        enterprise: {
            light: d.variables.gray96,
            dark: d.variables.gray22
        },
        prisma: d.variables.backgroundColorSection
    }));
    var v = p().code.withConfig({
        displayName: "MarkdownStyles__StyledCodeInline",
        componentId: "sc-1xadih-1"
    })([ "", ";font-family:", ";background-color:", ";" ], d.mixins.reset("inline"), d.variables.monoFontFamily, (0, 
    d.pick)({
        enterprise: {
            light: d.variables.gray92,
            dark: d.variables.gray22
        },
        prisma: d.variables.neutral200
    }));
    var b = p().blockquote.withConfig({
        displayName: "MarkdownStyles__StyledHint",
        componentId: "sc-1xadih-2"
    })([ "", " ", " background-color:", ";margin-bottom:15px;padding:", ";border-left:10px solid ", ";>:last-child{margin-bottom:0;}" ], d.mixins.reset("block"), d.mixins.clearfix(), (0, 
    d.pick)({
        enterprise: {
            light: d.variables.infoColorL50,
            dark: d.variables.infoColorD50
        },
        prisma: d.variables.neutral100
    }), d.variables.spacingSmall, d.variables.infoColor);
    var g = p().img.withConfig({
        displayName: "MarkdownStyles__StyledImage",
        componentId: "sc-1xadih-3"
    })([ "max-width:100%;" ]);
    // CONCATENATED MODULE: ./src/Markdown/renderers/MarkdownBlockquote.tsx
    function O(e, r) {
        if (e == null) return {};
        var n = m(e, r);
        var t, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                t = i[o];
                if (r.indexOf(t) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, t)) continue;
                n[t] = e[t];
            }
        }
        return n;
    }
    function m(e, r) {
        if (e == null) return {};
        var n = {};
        var t = Object.keys(e);
        var o, i;
        for (i = 0; i < t.length; i++) {
            o = t[i];
            if (r.indexOf(o) >= 0) continue;
            n[o] = e[o];
        }
        return n;
    }
    var h = {
        children: i().node
    };
    function j(e) {
        var r = e.children, n = O(e, [ "children" ]);
        
        return t().createElement(b, n, r);
    }
    j.propTypes = h;
    /* harmony default export */    const k = j;
    // CONCATENATED MODULE: ./src/Markdown/renderers/MarkdownCode.tsx
    function w(e, r) {
        if (e == null) return {};
        var n = P(e, r);
        var t, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                t = i[o];
                if (r.indexOf(t) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, t)) continue;
                n[t] = e[t];
            }
        }
        return n;
    }
    function P(e, r) {
        if (e == null) return {};
        var n = {};
        var t = Object.keys(e);
        var o, i;
        for (i = 0; i < t.length; i++) {
            o = t[i];
            if (r.indexOf(o) >= 0) continue;
            n[o] = e[o];
        }
        return n;
    }
    var S = {
        literal: i().string
    };
    function x(e) {
        var r = e.literal, n = w(e, [ "literal" ]);
        
        return t().createElement(v, n, r);
    }
    x.propTypes = S;
    /* harmony default export */    const R = x;
    // CONCATENATED MODULE: external "@splunk/react-ui/Code"
    const E = require("@splunk/react-ui/Code");
    var I =  e.n(E);
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const q = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: ./src/Markdown/renderers/MarkdownCodeBlock.tsx
    function M(e, r) {
        if (e == null) return {};
        var n = C(e, r);
        var t, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                t = i[o];
                if (r.indexOf(t) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, t)) continue;
                n[t] = e[t];
            }
        }
        return n;
    }
    function C(e, r) {
        if (e == null) return {};
        var n = {};
        var t = Object.keys(e);
        var o, i;
        for (i = 0; i < t.length; i++) {
            o = t[i];
            if (r.indexOf(o) >= 0) continue;
            n[o] = e[o];
        }
        return n;
    }
    var T = {
        literal: i().string,
        language: i().string
    };
    function _(e) {
        var r = e.literal, n = e.language, o = M(e, [ "literal", "language" ]);
        
        return t().createElement(y, o,  t().createElement(I(), {
            "aria-label": (0, q._)("Code Block"),
            language: n,
            value: r,
            style: {
                margin: 0
            }
        }));
    }
    _.propTypes = T;
    /* harmony default export */    const B = _;
    // CONCATENATED MODULE: external "@splunk/react-ui/Heading"
    const L = require("@splunk/react-ui/Heading");
    var H =  e.n(L);
    // CONCATENATED MODULE: ./src/Markdown/renderers/MarkdownHeading.tsx
    function D() {
        D = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var n = arguments[r];
                for (var t in n) {
                    if (Object.prototype.hasOwnProperty.call(n, t)) {
                        e[t] = n[t];
                    }
                }
            }
            return e;
        };
        return D.apply(this, arguments);
    }
    function A(e, r) {
        if (e == null) return {};
        var n = N(e, r);
        var t, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                t = i[o];
                if (r.indexOf(t) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, t)) continue;
                n[t] = e[t];
            }
        }
        return n;
    }
    function N(e, r) {
        if (e == null) return {};
        var n = {};
        var t = Object.keys(e);
        var o, i;
        for (i = 0; i < t.length; i++) {
            o = t[i];
            if (r.indexOf(o) >= 0) continue;
            n[o] = e[o];
        }
        return n;
    }
    var F = {
        children: i().node
    };
    function z(e) {
        var r = e.level, n = e.children, o = A(e, [ "level", "children" ]);
        
        return t().createElement(H(), D({
            level: Math.min(r, 4)
        }, o), n);
    }
    z.propTypes = F;
    /* harmony default export */    const G = z;
    // CONCATENATED MODULE: external "@splunk/react-ui/Anchor"
    const J = require("@splunk/react-ui/Anchor");
    var K =  e.n(J);
    // CONCATENATED MODULE: ./src/Markdown/renderers/MarkdownAnchorHeading.tsx
    function Q() {
        Q = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var n = arguments[r];
                for (var t in n) {
                    if (Object.prototype.hasOwnProperty.call(n, t)) {
                        e[t] = n[t];
                    }
                }
            }
            return e;
        };
        return Q.apply(this, arguments);
    }
    function U(e, r) {
        if (e == null) return {};
        var n = V(e, r);
        var t, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                t = i[o];
                if (r.indexOf(t) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, t)) continue;
                n[t] = e[t];
            }
        }
        return n;
    }
    function V(e, r) {
        if (e == null) return {};
        var n = {};
        var t = Object.keys(e);
        var o, i;
        for (i = 0; i < t.length; i++) {
            o = t[i];
            if (r.indexOf(o) >= 0) continue;
            n[o] = e[o];
        }
        return n;
    }
    function W(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            W = function e(r) {
                return typeof r;
            };
        } else {
            W = function e(r) {
                return r && typeof Symbol === "function" && r.constructor === Symbol && r !== Symbol.prototype ? "symbol" : typeof r;
            };
        }
        return W(e);
    }
    var X = {
        children: i().node
    };
    // When this is used in MDX, the node isn't guaranteed to be text,
    // so this method extracts the actual text from the node.
        function Y(e) {
        var r = W(e);
        var n = "";
        if (e != null) {
            if (r === "string" || r === "number") {
                n = e.toString();
            } else if (r === "boolean") {
                n = e ? "true" : "false";
            } else if (W(e) === "object") {
                if (e instanceof Array) {
                    n = e.map(Y).join(" ");
                } else if ("props" in e) {
                    n = Y(e.props.children);
                }
            }
        }
        return n;
    }
    function Z(e) {
        return n.Children.toArray(e).map(Y).join(" ").toLowerCase().replace(/[^a-z0-9-. ]/g, "").replace(/[-. ]+/g, "-");
    }
    function $(e) {
        var r = e.level, n = e.children, o = U(e, [ "level", "children" ]);
        
        return t().createElement(G, Q({
            level: Math.min(r, 4)
        }, o),  t().createElement(K(), {
            name: Z(n)
        }, n));
    }
    $.propTypes = X;
    /* harmony default export */    const ee = $;
    // CONCATENATED MODULE: ./src/Markdown/renderers/MarkdownImage.tsx
    function re() {
        re = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var n = arguments[r];
                for (var t in n) {
                    if (Object.prototype.hasOwnProperty.call(n, t)) {
                        e[t] = n[t];
                    }
                }
            }
            return e;
        };
        return re.apply(this, arguments);
    }
    function ne(e, r) {
        if (e == null) return {};
        var n = te(e, r);
        var t, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                t = i[o];
                if (r.indexOf(t) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, t)) continue;
                n[t] = e[t];
            }
        }
        return n;
    }
    function te(e, r) {
        if (e == null) return {};
        var n = {};
        var t = Object.keys(e);
        var o, i;
        for (i = 0; i < t.length; i++) {
            o = t[i];
            if (r.indexOf(o) >= 0) continue;
            n[o] = e[o];
        }
        return n;
    }
    var oe = {
        alt: i().string,
        src: i().string,
        title: i().string
    };
    function ie(e) {
        var r = e.src, n = e.title, o = e.alt, i = ne(e, [ "src", "title", "alt" ]);
        
        return t().createElement(g, re({
            src: r,
            title: n,
            alt: o
        }, i));
    }
    ie.propTypes = oe;
    /* harmony default export */    const ae = ie;
    // CONCATENATED MODULE: external "@splunk/react-ui/Link"
    const le = require("@splunk/react-ui/Link");
    var ce =  e.n(le);
    // CONCATENATED MODULE: ./src/Markdown/renderers/MarkdownLink.tsx
    function ue() {
        ue = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var n = arguments[r];
                for (var t in n) {
                    if (Object.prototype.hasOwnProperty.call(n, t)) {
                        e[t] = n[t];
                    }
                }
            }
            return e;
        };
        return ue.apply(this, arguments);
    }
    function fe(e, r) {
        if (e == null) return {};
        var n = se(e, r);
        var t, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                t = i[o];
                if (r.indexOf(t) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, t)) continue;
                n[t] = e[t];
            }
        }
        return n;
    }
    function se(e, r) {
        if (e == null) return {};
        var n = {};
        var t = Object.keys(e);
        var o, i;
        for (i = 0; i < t.length; i++) {
            o = t[i];
            if (r.indexOf(o) >= 0) continue;
            n[o] = e[o];
        }
        return n;
    }
    var pe = {
        children: i().node,
        href: i().string,
        title: i().string
    };
    function de(e) {
        var r = e.children, n = e.href, o = fe(e, [ "children", "href" ]);
        
        return t().createElement(ce(), ue({
            to: n
        }, o), r);
    }
    de.propTypes = pe;
    /* harmony default export */    const ye = de;
    // CONCATENATED MODULE: external "@splunk/react-ui/List"
    const ve = require("@splunk/react-ui/List");
    var be =  e.n(ve);
    // CONCATENATED MODULE: ./src/Markdown/renderers/MarkdownList.tsx
    function ge() {
        ge = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var n = arguments[r];
                for (var t in n) {
                    if (Object.prototype.hasOwnProperty.call(n, t)) {
                        e[t] = n[t];
                    }
                }
            }
            return e;
        };
        return ge.apply(this, arguments);
    }
    function Oe(e, r) {
        if (e == null) return {};
        var n = me(e, r);
        var t, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                t = i[o];
                if (r.indexOf(t) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, t)) continue;
                n[t] = e[t];
            }
        }
        return n;
    }
    function me(e, r) {
        if (e == null) return {};
        var n = {};
        var t = Object.keys(e);
        var o, i;
        for (i = 0; i < t.length; i++) {
            o = t[i];
            if (r.indexOf(o) >= 0) continue;
            n[o] = e[o];
        }
        return n;
    }
    var he = {
        children: i().node,
        type: i().string
    };
    function je(e) {
        var r = e.children, n = e.type, o = Oe(e, [ "children", "type" ]);
        if (n === "ordered") {
            
            return t().createElement(be(), ge({}, o, {
                ordered: true
            }), r);
        }
        
        return t().createElement(be(), o, r);
    }
    je.propTypes = he;
    /* harmony default export */    const ke = je;
    // CONCATENATED MODULE: ./src/Markdown/renderers/MarkdownItem.tsx
    function we(e, r) {
        if (e == null) return {};
        var n = Pe(e, r);
        var t, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                t = i[o];
                if (r.indexOf(t) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, t)) continue;
                n[t] = e[t];
            }
        }
        return n;
    }
    function Pe(e, r) {
        if (e == null) return {};
        var n = {};
        var t = Object.keys(e);
        var o, i;
        for (i = 0; i < t.length; i++) {
            o = t[i];
            if (r.indexOf(o) >= 0) continue;
            n[o] = e[o];
        }
        return n;
    }
    var Se = {
        children: i().node
    };
    function xe(e) {
        var r = e.children, n = we(e, [ "children" ]);
        
        return t().createElement(ve.Item, n, r);
    }
    xe.propTypes = Se;
    /* harmony default export */    const Re = xe;
    // CONCATENATED MODULE: external "@splunk/react-ui/Paragraph"
    const Ee = require("@splunk/react-ui/Paragraph");
    var Ie =  e.n(Ee);
    // CONCATENATED MODULE: ./src/Markdown/renderers/MarkdownParagraph.tsx
    function qe(e, r) {
        if (e == null) return {};
        var n = Me(e, r);
        var t, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                t = i[o];
                if (r.indexOf(t) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, t)) continue;
                n[t] = e[t];
            }
        }
        return n;
    }
    function Me(e, r) {
        if (e == null) return {};
        var n = {};
        var t = Object.keys(e);
        var o, i;
        for (i = 0; i < t.length; i++) {
            o = t[i];
            if (r.indexOf(o) >= 0) continue;
            n[o] = e[o];
        }
        return n;
    }
    var Ce = {
        children: i().node
    };
    function Te(e) {
        var r = e.children, n = qe(e, [ "children" ]);
        
        return t().createElement(Ie(), n, r);
    }
    Te.propTypes = Ce;
    /* harmony default export */    const _e = Te;
    // CONCATENATED MODULE: ./src/Markdown/Markdown.tsx
    function Be() {
        Be = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var n = arguments[r];
                for (var t in n) {
                    if (Object.prototype.hasOwnProperty.call(n, t)) {
                        e[t] = n[t];
                    }
                }
            }
            return e;
        };
        return Be.apply(this, arguments);
    }
    function Le(e, r) {
        var n = Object.keys(e);
        if (Object.getOwnPropertySymbols) {
            var t = Object.getOwnPropertySymbols(e);
            if (r) t = t.filter((function(r) {
                return Object.getOwnPropertyDescriptor(e, r).enumerable;
            }));
            n.push.apply(n, t);
        }
        return n;
    }
    function He(e) {
        for (var r = 1; r < arguments.length; r++) {
            var n = arguments[r] != null ? arguments[r] : {};
            if (r % 2) {
                Le(Object(n), true).forEach((function(r) {
                    De(e, r, n[r]);
                }));
            } else if (Object.getOwnPropertyDescriptors) {
                Object.defineProperties(e, Object.getOwnPropertyDescriptors(n));
            } else {
                Le(Object(n)).forEach((function(r) {
                    Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(n, r));
                }));
            }
        }
        return e;
    }
    function De(e, r, n) {
        if (r in e) {
            Object.defineProperty(e, r, {
                value: n,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[r] = n;
        }
        return e;
    }
    function Ae(e, r) {
        if (e == null) return {};
        var n = Ne(e, r);
        var t, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                t = i[o];
                if (r.indexOf(t) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, t)) continue;
                n[t] = e[t];
            }
        }
        return n;
    }
    function Ne(e, r) {
        if (e == null) return {};
        var n = {};
        var t = Object.keys(e);
        var o, i;
        for (i = 0; i < t.length; i++) {
            o = t[i];
            if (r.indexOf(o) >= 0) continue;
            n[o] = e[o];
        }
        return n;
    }
    /** @public */    var Fe = {
        blockquoteRenderer: i().func,
        codeBlockRenderer: i().func,
        codeRenderer: i().func,
        elementRef: i().oneOfType([ i().func, i().object ]),
        headingRenderer: i().func,
        imageRenderer: i().func,
        itemRenderer: i().func,
        linkRenderer: i().func,
        listRenderer: i().func,
        paragraphRenderer: i().func,
        text: i().string.isRequired,
        rendererOptions: i().object
    };
    var ze = new c.Parser;
    /**
 * The `Markdown` component renders the given Markdown text as a React component.
 * The component prefers @splunk/react-ui components over plain HTML components. For example
 * links are rendered as the `@splunk/react-ui/Link` component instead of plain `<a>` tag.
 */    function Ge(e) {
        var r = e.blockquoteRenderer, n = e.codeBlockRenderer, o = e.codeRenderer, i = e.elementRef, a = e.headingRenderer, c = e.imageRenderer, u = e.itemRenderer, s = e.linkRenderer, p = e.listRenderer, d = e.paragraphRenderer, y = e.text, v = e.rendererOptions, b = Ae(e, [ "blockquoteRenderer", "codeBlockRenderer", "codeRenderer", "elementRef", "headingRenderer", "imageRenderer", "itemRenderer", "linkRenderer", "listRenderer", "paragraphRenderer", "text", "rendererOptions" ]);
        // @docs-props-type MarkdownPropsBase
        // keep this option object inside the constructor, otherwise the renderers
        // below will confuse the doc-gen
                var g = {
            escapeHtml: true,
            renderers: {
                CodeBlock: n || B,
                Code: o || R,
                Heading: a || G,
                Link: s || ye,
                List: p || ke,
                Item: u || Re,
                Paragraph: d || _e
            }
        };
 // not setting MarkdownBlockquote as the default
                if (r) {
            g.renderers.Blockquote = r;
        }
 // not setting MarkdownImage as the default
                if (c) {
            g.renderers.Image = c;
        }
        var O = new (l())(He(He({}, g), v));
        var m = ze.parse(y);
        var h = O.render(m);
        
        return t().createElement(f(), Be({
            "data-test": "markdown",
            elementRef: i
        }, b), h);
    }
    Ge.propTypes = Fe;
    /* harmony default export */    const Je = Ge;
    // CONCATENATED MODULE: ./src/Markdown/index.ts
    module.exports = r;
    /******/})();
