// @splunk/react-ui/Menu.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var t = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(t, {
                a: t
            });
            /******/            return t;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, t) => {
            /******/ for (var n in t) {
                /******/ if (e.o(t, n) && !e.o(r, n)) {
                    /******/ Object.defineProperty(r, n, {
                        enumerable: true,
                        get: t[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        Divider: () => /* reexport */ x,
        Heading: () => /* reexport */ N,
        Item: () => /* reexport */ ze,
        MenuContext: () => /* reexport */ y,
        default: () => /* reexport */ cr
    });
    // CONCATENATED MODULE: external "react"
    const t = require("react");
    var n =  e.n(t);
    // CONCATENATED MODULE: external "prop-types"
    const i = require("prop-types");
    var a =  e.n(i);
    // CONCATENATED MODULE: external "lodash/omit"
    const o = require("lodash/omit");
    var l =  e.n(o);
    // CONCATENATED MODULE: external "@splunk/react-ui/Popover"
    const s = require("@splunk/react-ui/Popover");
    // CONCATENATED MODULE: external "@splunk/ui-utils/focus"
    const c = require("@splunk/ui-utils/focus");
    // CONCATENATED MODULE: external "@splunk/ui-utils/keyboard"
    const u = require("@splunk/ui-utils/keyboard");
    // CONCATENATED MODULE: external "@splunk/react-ui/useForceUpdate"
    const p = require("@splunk/react-ui/useForceUpdate");
    var f =  e.n(p);
    // CONCATENATED MODULE: external "styled-components"
    const d = require("styled-components");
    var v =  e.n(d);
    // CONCATENATED MODULE: external "@splunk/themes"
    const b = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/Menu/DividerStyles.ts
    var m = v().div.withConfig({
        displayName: "DividerStyles__Styled",
        componentId: "l3zfh3-0"
    })([ "border-top:1px solid ", ";", "" ], (0, b.pick)({
        enterprise: b.variables.borderColor,
        prisma: b.variables.neutral200
    }), (0, b.pick)({
        prisma: {
            comfortable: (0, d.css)([ "margin:8px 0;" ]),
            compact: (0, d.css)([ "margin:6px 0;" ])
        }
    }));
    // CONCATENATED MODULE: ./src/Menu/MenuContext.tsx
    var y =  (0, t.createContext)({
        role: "menu"
    });
    // CONCATENATED MODULE: ./src/Menu/Divider.tsx
    function h() {
        h = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return h.apply(this, arguments);
    }
    var g = {};
    /**
 * A non-interactive menu item used to visually separate groups of items in the menu.
 */    function k(e) {
        var r = h({}, e);
        // @docs-props-type DividerPropsBase
                var i = (0, t.useContext)(y), a = i.role;
        
        return n().createElement(m, h({
            "aria-hidden": a === "listbox",
            "data-test": "divider",
            role: "separator"
        }, r));
    }
    k.propTypes = g;
    /* Remove the item if it is the first item after filtering */    k.filterFirst = true;
    /* Remove consecutive items with filterConsecutive = true (Dividers and Headings) */    k.filterConsecutive = true;
    /* Remove the item if it is the last item after filtering. */    k.filterLast = true;
    /* Tag the item as Divider */    k.as = "Divider";
    /* harmony default export */    const x = k;
    // CONCATENATED MODULE: external "@splunk/react-ui/Heading"
    const w = require("@splunk/react-ui/Heading");
    var S =  e.n(w);
    // CONCATENATED MODULE: external "@splunk/themes/useSplunkTheme"
    const O = require("@splunk/themes/useSplunkTheme");
    var C =  e.n(O);
    // CONCATENATED MODULE: ./src/Menu/HeadingStyles.ts
    var P = v().div.withConfig({
        displayName: "HeadingStyles__Styled",
        componentId: "mcd2ws-0"
    })([ "", " border-top:1px solid transparent;padding:", ";&:not(:first-child){", ";}" ], b.mixins.reset("block"), (0, 
    b.pick)({
        enterprise: (0, d.css)([ "", " ", " 6px" ], b.variables.spacingQuarter, b.variables.spacingHalf),
        prisma: {
            comfortable: "13px 16px 10px 16px",
            compact: "9px 16px 6px 16px"
        }
    }), (0, b.pick)({
        enterprise: (0, d.css)([ "border-top:", ";" ], b.variables.border),
        prisma: {
            comfortable: (0, d.css)([ "border-top:1px solid ", ";margin-top:6px;padding-top:19px;" ], b.variables.neutral200),
            compact: (0, d.css)([ "border-top:1px solid ", ";margin-top:6px;padding-top:15px;" ], b.variables.neutral200)
        }
    }));
    var I = v().h5.withConfig({
        displayName: "HeadingStyles__PrismaSectionTitle",
        componentId: "mcd2ws-1"
    })([ "", " font-size:", ";line-height:", ";font-weight:", ";letter-spacing:0.02em;" ], b.mixins.reset("block"), b.variables.fontSizeSmall, b.variables.lineHeight, b.variables.fontWeightBold);
    // CONCATENATED MODULE: ./src/Menu/Heading.tsx
    function j(e, r) {
        var t = Object.keys(e);
        if (Object.getOwnPropertySymbols) {
            var n = Object.getOwnPropertySymbols(e);
            if (r) n = n.filter((function(r) {
                return Object.getOwnPropertyDescriptor(e, r).enumerable;
            }));
            t.push.apply(t, n);
        }
        return t;
    }
    function E(e) {
        for (var r = 1; r < arguments.length; r++) {
            var t = arguments[r] != null ? arguments[r] : {};
            if (r % 2) {
                j(Object(t), true).forEach((function(r) {
                    _(e, r, t[r]);
                }));
            } else if (Object.getOwnPropertyDescriptors) {
                Object.defineProperties(e, Object.getOwnPropertyDescriptors(t));
            } else {
                j(Object(t)).forEach((function(r) {
                    Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r));
                }));
            }
        }
        return e;
    }
    function _(e, r, t) {
        if (r in e) {
            Object.defineProperty(e, r, {
                value: t,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[r] = t;
        }
        return e;
    }
    function T() {
        T = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return T.apply(this, arguments);
    }
    function R(e, r) {
        if (e == null) return {};
        var t = D(e, r);
        var n, i;
        if (Object.getOwnPropertySymbols) {
            var a = Object.getOwnPropertySymbols(e);
            for (i = 0; i < a.length; i++) {
                n = a[i];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function D(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var i, a;
        for (a = 0; a < n.length; a++) {
            i = n[a];
            if (r.indexOf(i) >= 0) continue;
            t[i] = e[i];
        }
        return t;
    }
    var q = {
        children: a().node,
        title: a().bool,
        outerStyle: a().object
    };
    /**
 * A non-interactive `Menu` item used to separate and label groups of `Menu` items.
 */    function K(e) {
        var r = e.children, i = e.title, a = e.outerStyle, o = R(e, [ "children", "title", "outerStyle" ]);
        var l = (0, t.useContext)(y), s = l.role;
        var c = C()(), u = c.family;
        var p = u === "prisma" ?  n().createElement(I, T({
            as: i ? "h4" : "h5",
            style: a,
            "data-test": "heading"
        }, o), r) :  n().createElement(S(), T({
            style: E(E({}, a), {}, {
                margin: 0
            }),
            level: i ? 4 : "ss",
            "data-test": "heading"
        }, o), r);
 // headings are not supported in role="listbox"
        // this causes Firefox to incorrectly read the total number
        // of items in the list if it contains a heading or divider
        // so we set them to aria-hidden to prevent that
                
        return n().createElement(P, {
            "aria-hidden": s === "listbox"
        }, p);
    }
    K.propTypes = q;
    /* Remove consecutive items with filterConsecutive = true (Dividers and Headings) */    K.filterConsecutive = true;
    /* Remove the item if it is the last item after filtering. */    K.filterLast = true;
    /* Tag the item as a Heading */    K.as = "Heading";
    /* harmony default export */    const N = K;
    // CONCATENATED MODULE: external "lodash/isString"
    const H = require("lodash/isString");
    var A =  e.n(H);
    // CONCATENATED MODULE: external "@splunk/react-icons/ChevronRight"
    const F = require("@splunk/react-icons/ChevronRight");
    var $ =  e.n(F);
    // CONCATENATED MODULE: external "@splunk/react-icons/Checkmark"
    const M = require("@splunk/react-icons/Checkmark");
    var z =  e.n(M);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Check"
    const L = require("@splunk/react-icons/enterprise/Check");
    var U =  e.n(L);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/ChevronRight"
    const W = require("@splunk/react-icons/enterprise/ChevronRight");
    var B =  e.n(W);
    // CONCATENATED MODULE: external "@splunk/react-ui/Clickable"
    const G = require("@splunk/react-ui/Clickable");
    var V =  e.n(G);
    // CONCATENATED MODULE: external "@splunk/react-ui/Switch"
    const Q = require("@splunk/react-ui/Switch");
    var Y =  e.n(Q);
    // CONCATENATED MODULE: ./src/Menu/ItemStyles.ts
    var J = (0, d.css)([ "color:", ";font-size:", ";line-height:", ";overflow:inherit;white-space:inherit;text-overflow:inherit;" ], (0, 
    b.pick)({
        prisma: b.variables.contentColorDefault,
        enterprise: b.variables.textGray
    }), b.variables.fontSizeSmall, (0, b.pick)({
        prisma: "16px"
    }));
    var X = v().span.withConfig({
        displayName: "ItemStyles__StyledItemDescriptionBottom",
        componentId: "sc-4kc053-0"
    })([ "", ";display:block;" ], J);
    var Z = v().span.withConfig({
        displayName: "ItemStyles__StyledItemDescriptionRight",
        componentId: "sc-4kc053-1"
    })([ "", ";float:right;padding-left:", ";max-width:50%;text-align:right;box-sizing:border-box;" ], J, b.variables.spacing);
    var ee = v().div.withConfig({
        displayName: "ItemStyles__StyledItemSelectedIcon",
        componentId: "sc-4kc053-2"
    })([ "position:absolute;", " color:", ";" ], (0, b.pick)({
        prisma: (0, d.css)([ "right:16px;" ]),
        enterprise: (0, d.css)([ "right:3px;left:8px;" ])
    }), (0, b.pick)({
        prisma: b.variables.contentColorActive,
        enterprise: b.variables.accentColorL10
    }));
    var re = v().span.withConfig({
        displayName: "ItemStyles__StyledItemIcon",
        componentId: "sc-4kc053-3"
    })([ "flex:0 0 auto;padding-right:", ";min-width:10px;display:inline-block;text-align:center;vertical-align:", ";", "" ], (0, 
    b.pick)({
        prisma: "8px",
        enterprise: "3px"
    }), (0, b.pick)({
        prisma: "initial",
        enterprise: "middle"
    }), (0, b.pick)({
        prisma: (0, d.css)([ "color:", ";display:inline-flex;align-items:center;min-width:20px;min-height:20px;& > svg{width:20px;height:20px;}" ], b.variables.contentColorMuted),
        enterprise: (0, d.css)([ "transform:translateY(-1px);" ])
    }));
    var te = v()(V()).withConfig({
        displayName: "ItemStyles__StyledClickable",
        componentId: "sc-4kc053-4"
    })([ "display:block;position:relative;cursor:pointer;color:", ";word-wrap:break-word;max-width:100%;width:100%;padding:", ";", " ", " ", " &:not([disabled]){&:hover{background:", ";", "}&:focus{outline:0;box-shadow:", ";", "}", " ", "}&[disabled]{color:", ";cursor:not-allowed;", ",", "{color:inherit;}", "{cursor:not-allowed;color:", ";}", "}", "" ], (0, 
    b.pick)({
        prisma: b.variables.contentColorActive,
        enterprise: {
            dark: b.variables.gray96,
            light: b.variables.gray22
        }
    }), (0, b.pick)({
        prisma: {
            comfortable: "10px 16px",
            compact: "6px 16px"
        },
        enterprise: "6px 10px"
    }), (function(e) {
        var r = e.$isSelectable, t = e.$selectableAppearance;
        return r && (0, b.pick)({
            prisma: (0, d.css)([ "padding-right:44px;", "" ], (function() {
                return t === "checkbox" && "padding-left: 42px;";
            })),
            enterprise: (0, d.css)([ "padding-right:10px;padding-left:", ";" ], (function() {
                return t === "checkbox" ? "32px" : "28px";
            }))
        });
    }), (function(e) {
        var r = e.$active, t = e.$preventFocus;
        return r && !t && (0, d.css)([ "box-shadow:", ";" ], b.variables.focusShadowInset);
    }), (function(e) {
        var r = e.$active;
        return r && (0, d.css)([ "background:", ";font-weight:", ";" ], (0, b.pick)({
            enterprise: {
                light: b.variables.gray92,
                dark: b.variables.gray22
            },
            prisma: b.variables.neutral100
        }), b.variables.fontWeightBold);
    }), (0, b.pick)({
        prisma: b.variables.interactiveColorOverlayHover,
        enterprise: b.variables.backgroundColorHover
    }), (function(e) {
        var r = e.$selected;
        return r && (0, b.pick)({
            prisma: (0, d.css)([ "background:", ";" ], b.mixins.overlayColors(b.variables.interactiveColorOverlaySelected, b.variables.interactiveColorOverlayHover))
        });
    }), b.variables.focusShadowInset, (0, b.pick)({
        prisma: (0, d.css)([ "background:", ";" ], b.variables.interactiveColorOverlayHover)
    }), (0, b.pick)({
        prisma: (0, d.css)([ "&:active{background:", ";}" ], b.variables.interactiveColorOverlayActive)
    }), (function(e) {
        var r = e.$selected;
        return r && (0, b.pick)({
            prisma: (0, d.css)([ "background:", ";&:active{background:", ";}" ], b.variables.interactiveColorOverlaySelected, b.mixins.overlayColors(b.variables.interactiveColorOverlaySelected, b.variables.interactiveColorOverlayActive))
        });
    }), (0, b.pick)({
        enterprise: {
            light: b.variables.gray45,
            dark: b.variables.contentColorDisabled
        },
        prisma: b.variables.contentColorDisabled
    }), 
    /* sc-sel */
    X, 
    /* sc-sel */
    Z, 
    /* sc-sel */
    ee, (0, b.pick)({
        prisma: b.variables.contentColorDisabled,
        enterprise: b.variables.gray80
    }), (0, b.pick)({
        prisma: (0, d.css)([ "& > * > ", "{color:", ";}" ], 
        /* sc-sel */
        re, b.variables.contentColorDisabled)
    }), (function(e) {
        var r = e.$selected;
        return r && (0, b.pick)({
            prisma: (0, d.css)([ "", "{color:", ";}" ], 
            /* sc-sel */
            re, b.variables.contentColorActive)
        });
    }));
    var ne = v()(Y()).withConfig({
        displayName: "ItemStyles__StyledSwitch",
        componentId: "sc-4kc053-5"
    })([ "position:absolute;left:", ";top:", ";" ], (0, b.pick)({
        prisma: "16px",
        enterprise: "8px"
    }), (0, b.pick)({
        prisma: "0px",
        enterprise: {
            comfortable: "-1px",
            compact: "2px"
        }
    }));
    var ie = v().span.withConfig({
        displayName: "ItemStyles__StyledInnerWrapper",
        componentId: "sc-4kc053-6"
    })([ "display:flex;align-items:flex-start;" ]);
    var ae = v().span.withConfig({
        displayName: "ItemStyles__StyledTitleAndDescriptionWrapper",
        componentId: "sc-4kc053-7"
    })([ "display:flex;flex-wrap:wrap;flex-direction:column;max-width:100%;width:100%;align-self:center;word-break:break-word;white-space:normal;", "" ], (function(e) {
        var r = e.$truncate;
        return r && (0, d.css)([ "white-space:nowrap;overflow:hidden;text-overflow:ellipsis;" ]);
    }));
    var oe = v().div.withConfig({
        displayName: "ItemStyles__StyledLabel",
        componentId: "sc-4kc053-8"
    })([ "overflow:inherit;white-space:inherit;text-overflow:inherit;max-width:100%;min-height:20px;", "" ], (function(e) {
        var r = e.$truncate;
        return r && (0, d.css)([ "display:block;clear:both;" ]);
    }));
    var le = v().span.withConfig({
        displayName: "ItemStyles__StyledMatch",
        componentId: "sc-4kc053-9"
    })([ "border-bottom:1px solid ", ";" ], (0, b.pick)({
        prisma: b.variables.contentColorActive,
        enterprise: {
            dark: b.variables.gray96,
            light: b.variables.gray45
        }
    }));
    var se = v().span.withConfig({
        displayName: "ItemStyles__StyledSubmenu",
        componentId: "sc-4kc053-10"
    })([ "float:right;padding-left:", ";color:", ";", "" ], b.variables.spacingSmall, (0, 
    b.pick)({
        prisma: "inherit",
        enterprise: b.variables.textGray
    }), (0, b.pick)({
        prisma: (0, d.css)([ "& > svg{width:20px;height:20px;}" ])
    }));
    // CONCATENATED MODULE: external "@splunk/react-icons/ArrowSquareTopRight"
    const ce = require("@splunk/react-icons/ArrowSquareTopRight");
    var ue =  e.n(ce);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/External"
    const pe = require("@splunk/react-icons/enterprise/External");
    var fe =  e.n(pe);
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const de = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: ./src/Menu/icons/External.tsx
    var ve = (0, de._)("Open externally");
    var be = (0, d.css)([ "width:0.8em;height:0.8em;vertical-align:baseline;" ]);
    var me = v()(fe()).withConfig({
        displayName: "External__StyledEnterprise",
        componentId: "sc-1turs2s-0"
    })([ "", " margin:0 0 0 3px;" ], be);
    var ye = v()(ue()).withConfig({
        displayName: "External__StyledPrisma",
        componentId: "sc-1turs2s-1"
    })([ "", " margin:0 0 0 4px;" ], be);
    function he() {
        var e = (0, b.useSplunkTheme)(), r = e.isEnterprise, t = e.isCompact;
        var i = t ? "20px" : "24px";
        return r ?  n().createElement(me, {
            screenReaderText: ve
        }) :  n().createElement(ye, {
            "aria-label": ve,
            height: i,
            width: i
        },  n().createElement("title", null, ve));
    }
    /* harmony default export */    const ge = he;
    // CONCATENATED MODULE: ./src/utils/updateReactRef.ts
    /**
 * Updates a React ref. Callback refs and object refs (from `createRef` and `useRef`) are supported.
 *
 * @param ref - The React callback or object ref. Can be `null` or `undefined`.
 * @param current - The new value of the ref.
 */
    function ke(e, r) {
        if (e) {
            if (typeof e === "function") {
                e(r);
            } else {
                // the public signature of this util uses React.Ref<T> to mirror the way React types refs.
                // the intention here is to signal "we will take care of setting 'current', not you".
                e.current = r;
 // eslint-disable-line no-param-reassign
                        }
        }
    }
    // CONCATENATED MODULE: ./src/utils/scrollIntoViewIfNeeded.ts
    // A utility that attempts to move an element into view by scrolling it's derived parent.
    var xe = function e(r) {
        if (!r) {
            return;
        }
        var t = r.offsetParent;
        if (!t) {
            return;
        }
 // Below the bottom of the container.
                if (t.scrollTop + t.clientHeight < r.offsetTop + r.clientHeight) {
            t.scrollTop = r.offsetTop + r.clientHeight - t.clientHeight;
 // Above the top of the container.
                } else if (t.scrollTop > r.offsetTop) {
            t.scrollTop = r.offsetTop;
        }
    };
    // CONCATENATED MODULE: ./src/Menu/Item.tsx
    function we(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            we = function e(r) {
                return typeof r;
            };
        } else {
            we = function e(r) {
                return r && typeof Symbol === "function" && r.constructor === Symbol && r !== Symbol.prototype ? "symbol" : typeof r;
            };
        }
        return we(e);
    }
    function Se() {
        Se = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return Se.apply(this, arguments);
    }
    function Oe(e, r) {
        if (e == null) return {};
        var t = Ce(e, r);
        var n, i;
        if (Object.getOwnPropertySymbols) {
            var a = Object.getOwnPropertySymbols(e);
            for (i = 0; i < a.length; i++) {
                n = a[i];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function Ce(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var i, a;
        for (a = 0; a < n.length; a++) {
            i = n[a];
            if (r.indexOf(i) >= 0) continue;
            t[i] = e[i];
        }
        return t;
    }
    function Pe(e, r) {
        if (!(e instanceof r)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function Ie(e, r) {
        for (var t = 0; t < r.length; t++) {
            var n = r[t];
            n.enumerable = n.enumerable || false;
            n.configurable = true;
            if ("value" in n) n.writable = true;
            Object.defineProperty(e, n.key, n);
        }
    }
    function je(e, r, t) {
        if (r) Ie(e.prototype, r);
        if (t) Ie(e, t);
        return e;
    }
    function Ee(e, r) {
        if (typeof r !== "function" && r !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(r && r.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (r) _e(e, r);
    }
    function _e(e, r) {
        _e = Object.setPrototypeOf || function e(r, t) {
            r.__proto__ = t;
            return r;
        };
        return _e(e, r);
    }
    function Te(e) {
        var r = qe();
        return function t() {
            var n = Ke(e), i;
            if (r) {
                var a = Ke(this).constructor;
                i = Reflect.construct(n, arguments, a);
            } else {
                i = n.apply(this, arguments);
            }
            return Re(this, i);
        };
    }
    function Re(e, r) {
        if (r && (we(r) === "object" || typeof r === "function")) {
            return r;
        }
        return De(e);
    }
    function De(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function qe() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function Ke(e) {
        Ke = Object.setPrototypeOf ? Object.getPrototypeOf : function e(r) {
            return r.__proto__ || Object.getPrototypeOf(r);
        };
        return Ke(e);
    }
    function Ne(e, r, t) {
        if (r in e) {
            Object.defineProperty(e, r, {
                value: t,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[r] = t;
        }
        return e;
    }
    var He = {
        active: a().bool,
        children: a().node,
        description: a().string,
        descriptionPosition: a().oneOf([ "right", "bottom" ]),
        disabled: a().bool,
        elementRef: a().oneOfType([ a().func, a().object ]),
        hasSubmenu: a().bool,
        icon: a().node,
        /** @private. The internal key of the tab */
        itemKey: a().number,
        matchRanges: a().arrayOf(a().shape({
            start: a().number.isRequired,
            end: a().number.isRequired
        })),
        onClick: a().func,
        /** @private */
        onFocus: a().func,
        openInNewContext: a().bool,
        /** @private */
        preventFocus: a().bool,
        role: a().oneOf([ "menuitem", "menuitemradio", "menuitemcheckbox", "listboxitem", "option" ]),
        selectable: a().bool,
        selectableAppearance: a().oneOf([ "checkmark", "checkbox" ]),
        selected: a().bool,
        /** @private */
        splunkTheme: a().object,
        to: a().string,
        truncate: a().bool
    };
    var Ae = {
        active: false,
        descriptionPosition: "bottom",
        disabled: false,
        hasSubmenu: false,
        openInNewContext: false,
        preventFocus: false,
        selectable: false,
        selectableAppearance: "checkmark",
        selected: false,
        truncate: false
    };
    var Fe =  function(e) {
        Ee(t, e);
        var r = Te(t);
        je(t, null, [ {
            key: "validateProps",
            // @docs-props-type ItemPropsBase
            value: function e(r) {
                if (false) {}
            }
        } ]);
        function t(e) {
            var n;
            Pe(this, t);
            n = r.call(this, e);
            Ne(De(n), "el", null);
            Ne(De(n), "handleFocus", (function(e) {
                var r = n.props, t = r.onFocus, i = r.itemKey;
                t === null || t === void 0 ? void 0 : t(e, {
                    itemKey: i
                });
            }));
            Ne(De(n), "handleMount", (function(e) {
                n.el = e;
                ke(n.props.elementRef, e);
            }));
            t.validateProps(e);
            return n;
        }
        je(t, [ {
            key: "componentDidUpdate",
            value: function e() {
                t.validateProps(this.props);
            }
        }, {
            key: "focus",
            value: function e() {
                var r;
                (r = this.el) === null || r === void 0 ? void 0 : r.focus();
            }
        }, {
            key: "scrollIntoViewIfNeeded",
            value: function e() {
                xe(this.el);
            }
        }, {
            key: "renderLabel",
            value: function e() {
                var r = this.props, t = r.children, i = r.matchRanges;
                if (!i || !A()(t)) {
                    return t;
                }
                var a = [];
 // before first match. May be empty string.
                                a.push(t.substring(0, i[0].start));
                i.forEach((function(e, r) {
                    a.push(
                    
                    // eslint-disable-next-line react/no-array-index-key
                    n().createElement(le, {
                        key: r,
                        "data-test": "match"
                    }, t.substring(e.start, e.end)));
                    if (r < i.length - 1) {
                        a.push(t.substring(e.end, i[r + 1].start));
                    } else {
                        a.push(t.substring(e.end, t.length));
                    }
                }));
                return a;
            }
        }, {
            key: "render",
            value: function e() {
                var r = this.props, t = r.active, i = r.children, a = r.hasSubmenu, o = r.selectable, s = r.selectableAppearance, c = r.selected, u = r.icon, p = r.description, f = r.disabled, d = r.onClick, v = r.preventFocus, b = r.role, m = r.splunkTheme, y = r.to, h = r.truncate, g = r.descriptionPosition, k = r.openInNewContext, x = Oe(r, [ "active", "children", "hasSubmenu", "selectable", "selectableAppearance", "selected", "icon", "description", "disabled", "onClick", "preventFocus", "role", "splunkTheme", "to", "truncate", "descriptionPosition", "openInNewContext" ]);
                var w = m.isEnterprise, S = m.isPrisma;
                var O = o || c;
                var C = {
                    nonselectable: "menuitem",
                    checkmark: "menuitemradio",
                    checkbox: "menuitemcheckbox"
                }[O ? s : "nonselectable"];
 // aria-checked updates only with checkmark appearance for now SUI-2578
                                var P = {
                    "aria-haspopup": a ? true : undefined,
                    "aria-checked": s === "checkmark" && c ? true : undefined
                };
                var I = S ? "right" : "left";
                var j = p && g === "right";
                var E = p && !j;
                var _ = c && s === "checkmark" &&  n().createElement(ee, null, S ?  n().createElement(z(), {
                    inline: true,
                    height: "16",
                    width: "16"
                }) :  n().createElement(U(), {
                    hideDefaultTooltip: true,
                    inline: true,
                    screenReaderText: null,
                    size: .85
                }));
                
                return n().createElement(te, Se({
                    $isSelectable: O,
                    $selectableAppearance: s,
                    $selected: c,
                    $active: t,
                    $preventFocus: v,
                    "data-test-selected": O ? c : null,
                    "data-test": "item",
                    "data-has-icon": !!u,
                    disabled: f,
                    onClick: d,
                    onFocus: this.handleFocus,
                    onMouseDown: function e(r) {
                        if (v) {
                            r.preventDefault();
                        }
                    },
                    role: b || C,
                    to: y,
                    title: h && A()(i) ? i : undefined,
                    openInNewContext: k
                }, P, l()(x, "onFocus"), {
                    elementRef: this.handleMount
                }), o && s === "checkbox" &&  n().createElement(ne, {
                    interactive: false,
                    selected: c,
                    selectedLabel: "Selected",
                    value: "menu-item"
                }), a &&  n().createElement(se, null, w ?  n().createElement(B(), {
                    hideDefaultTooltip: true,
                    screenReaderText: null
                }) :  n().createElement($(), null)), j &&  n().createElement(Z, {
                    "data-test": "description"
                }, p),  n().createElement(ie, null, u &&  n().createElement(re, null, u),  n().createElement(ae, {
                    $truncate: h
                }, I === "left" && _,  n().createElement(oe, {
                    $truncate: h,
                    "data-test": "label"
                }, this.renderLabel(), k &&  n().createElement(ge, null)), E &&  n().createElement(X, {
                    "data-test": "description"
                }, p), I === "right" && _)));
            }
        } ]);
        return t;
    }(t.Component);
    Ne(Fe, "propTypes", He);
    Ne(Fe, "defaultProps", Ae);
    Ne(Fe, "as", void 0);
    var $e = (0, b.withSplunkTheme)(Fe);
    var Me = $e;
    Me.propTypes = Fe.propTypes;
    Me.as = "Item";
    /* harmony default export */    const ze = Me;
    // CONCATENATED MODULE: external "@splunk/react-ui/Scroll"
    const Le = require("@splunk/react-ui/Scroll");
    var Ue =  e.n(Le);
    // CONCATENATED MODULE: ./src/Menu/MenuStyles.ts
    var We = (0, d.css)([ "border-top:1px solid ", ";" ], (0, b.pick)({
        enterprise: b.variables.gray60,
        prisma: b.variables.neutral200
    }));
    var Be = v().div.withConfig({
        displayName: "MenuStyles__Styled",
        componentId: "sc-1olffp9-0"
    })([ "", ";background-color:", ";border-radius:", ";min-width:60px;overflow:auto;position:relative;", " &:focus{outline:0;box-shadow:", ";", "}& + &{", "}" ], b.mixins.reset("block"), (0, 
    b.pick)({
        enterprise: b.variables.backgroundColor,
        prisma: b.variables.backgroundColorPopup
    }), b.variables.borderRadius, (0, b.pick)({
        prisma: (0, d.css)([ "padding:8px 0;" ])
    }), b.variables.focusShadowInset, (0, b.pick)({
        prisma: (0, d.css)([ "background:", ";" ], b.variables.interactiveColorOverlayHover)
    }), We);
    var Ge = v()(Be.withComponent(Ue())).withConfig({
        displayName: "MenuStyles__StyledScroll",
        componentId: "sc-1olffp9-1"
    })([ "", " + &,& + ", "{", "}" ], 
    /* sc-sel */
    Be, 
    /* sc-sel */
    Be, We);
    // CONCATENATED MODULE: ./src/utils/navigateList.ts
    // A utility for keyboard navigation of lists
    function Ve(e, r, t) {
        for (var n = 0; n < e.length; n += 1) {
            var i;
            var a = (n + t) % e.length;
            if (((i = e[a].current) === null || i === void 0 ? void 0 : i.props.disabled) !== true) {
                return e[a];
            }
        }
        return e[r];
    }
    function Qe(e, r, t) {
        for (var n = e.length; n > 0; n -= 1) {
            var i;
            var a = (n + t) % e.length;
            if (((i = e[a].current) === null || i === void 0 ? void 0 : i.props.disabled) !== true) {
                return e[a];
            }
        }
        return e[r];
    }
    // CONCATENATED MODULE: ./src/utils/getMenuChildNode.tsx
    // TODO(SUI-5919): fix the ref handling in the other components so this isn't needed
    /**
 * @private Get the HTML <anchor> or <button> node for a received derivative of Menu.Item
 */
    var Ye = function e(r) {
        var t, n, i, a;
        if (r.el) {
            return r.el;
 // Menu.Item
                }
        if ((t = r.c) === null || t === void 0 ? void 0 : t.el) {
            return r.c.el;
 // Multiselect.Option
                }
        if ((n = r.c) === null || n === void 0 ? void 0 : (i = n.c) === null || i === void 0 ? void 0 : i.el) {
            return r.c.c.el;
 // Select.Option
                }
        if ((a = r.item) === null || a === void 0 ? void 0 : a.el) {
            return r.item.el;
 // ComboBox.Option
                }
        return null;
    };
    // CONCATENATED MODULE: ./src/Menu/Menu.tsx
    function Je() {
        Je = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return Je.apply(this, arguments);
    }
    function Xe(e, r) {
        return nr(e) || tr(e, r) || er(e, r) || Ze();
    }
    function Ze() {
        throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function er(e, r) {
        if (!e) return;
        if (typeof e === "string") return rr(e, r);
        var t = Object.prototype.toString.call(e).slice(8, -1);
        if (t === "Object" && e.constructor) t = e.constructor.name;
        if (t === "Map" || t === "Set") return Array.from(e);
        if (t === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t)) return rr(e, r);
    }
    function rr(e, r) {
        if (r == null || r > e.length) r = e.length;
        for (var t = 0, n = new Array(r); t < r; t++) {
            n[t] = e[t];
        }
        return n;
    }
    function tr(e, r) {
        if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(e))) return;
        var t = [];
        var n = true;
        var i = false;
        var a = undefined;
        try {
            for (var o = e[Symbol.iterator](), l; !(n = (l = o.next()).done); n = true) {
                t.push(l.value);
                if (r && t.length === r) break;
            }
        } catch (e) {
            i = true;
            a = e;
        } finally {
            try {
                if (!n && o["return"] != null) o["return"]();
            } finally {
                if (i) throw a;
            }
        }
        return t;
    }
    function nr(e) {
        if (Array.isArray(e)) return e;
    }
    function ir(e, r) {
        if (e == null) return {};
        var t = ar(e, r);
        var n, i;
        if (Object.getOwnPropertySymbols) {
            var a = Object.getOwnPropertySymbols(e);
            for (i = 0; i < a.length; i++) {
                n = a[i];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function ar(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var i, a;
        for (a = 0; a < n.length; a++) {
            i = n[a];
            if (r.indexOf(i) >= 0) continue;
            t[i] = e[i];
        }
        return t;
    }
    var or = {
        children: a().node,
        /** @private */
        controlledExternally: a().bool,
        elementRef: a().oneOfType([ a().func, a().object ]),
        /** @private */
        onDownKeyPress: a().func,
        /** @private */
        onEndKeyPress: a().func,
        /** @private */
        onHomeKeyPress: a().func,
        /** @private */
        onUpKeyPress: a().func,
        retainFocus: a().bool,
        stopScrollPropagation: a().bool
    };
    var lr = {
        stopScrollPropagation: false
    };
    function sr(e) {
        var r = e.children, i = e.controlledExternally, a = e.elementRef, o = e.onDownKeyPress, p = e.onEndKeyPress, d = e.onHomeKeyPress, v = e.onUpKeyPress, b = e.retainFocus, m = e.stopScrollPropagation, h = ir(e, [ "children", "controlledExternally", "elementRef", "onDownKeyPress", "onEndKeyPress", "onHomeKeyPress", "onUpKeyPress", "retainFocus", "stopScrollPropagation" ]);
        // @docs-props-type MenuPropsBase
                var g = (0, t.useContext)(y), k = g.role, x = k === void 0 ? "menu" : k;
        var w = (0, t.useContext)(s.PopoverContext), S = w.retainFocus;
        var O = f()();
        var C = true;
        if (b != null) {
            C = b;
        } else if (S != null) {
            C = S;
        }
        var P = n().useRef(0);
        var I = [];
 // `VirtualizedResultsMenu` needs some special consideration regarding the focused element index so that on
        // keypress it won't ignore the requested focus index or element, even when it doesn't exist. `VirtualizedResultsMenu`
        // knows what it _will_ mount as a child of `Menu`, and upon mount that targeted child should receive focus.
                var j = n().useRef(null);
        var E = n().useRef(false);
 // TODO(SUI-5918): stabilize itemRefs so itemOnFocus does not churn and can access itemRefs
                var _ = (0, t.useRef)(I);
        _.current = I;
        var T = 0;
        var R;
        var D = false;
        var q = false;
        var K = false;
        var N = (0, t.useCallback)((function(e) {
            var r = e.e, t = e.data, n = e.userCallback, i = e.defaultCallback;
            var a;
            if (n) {
                var o = n(r, t);
                if (typeof o[0] === "number") {
                    // userCallback returned index of the to-render element which should capture focus on mount
                    a = {
                        current: null
                    };
                    var l = Xe(o, 2);
                    P.current = l[0];
                    E.current = l[1];
                } else {
                    // userCallback returned ref of the rendered element which should take focus
                    var s = o;
                    var c = Xe(s, 2);
                    a = c[0];
                    E.current = c[1];
                    P.current = Math.max(0, t.itemRefs.indexOf(a));
                }
            } else {
                a = i();
            }
            return a;
        }), []);
        var H = function e(r) {
            // SUI-2734: keeping this active even if controlledExternally is set
            // because Firefox + Voiceover (and possibly other a11y tools)
            // can still set focus on the Menu items and this avoids the
            // whole page scrolling when that happens
            var t = (0, u.keycode)(r.nativeEvent);
            var n = P.current;
            var i;
 // the "retainFocus" checks here mimic the prior tab key functionality
                        if (t === "down" || C && (0, c.isTabKey)(r)) {
                i = N({
                    e: r,
                    data: {
                        itemRefs: I,
                        focusedItemKey: n
                    },
                    userCallback: o,
                    defaultCallback: function e() {
                        return Ve(I, n, n + 1);
                    }
                });
            } else if (t === "up" || C && (0, c.isTabKey)(r) && r.shiftKey) {
                i = N({
                    e: r,
                    data: {
                        itemRefs: I,
                        focusedItemKey: n
                    },
                    userCallback: v,
                    defaultCallback: function e() {
                        return Qe(I, n, n - 1);
                    }
                });
            } else if (t === "home") {
                i = N({
                    e: r,
                    data: {
                        itemRefs: I,
                        focusedItemKey: n
                    },
                    userCallback: d,
                    defaultCallback: function e() {
                        return Ve(I, n, 0);
                    }
                });
            } else if (t === "end") {
                i = N({
                    e: r,
                    data: {
                        itemRefs: I,
                        focusedItemKey: n
                    },
                    userCallback: p,
                    defaultCallback: function e() {
                        return Qe(I, n, I.length - 1);
                    }
                });
            }
            if (i != null) {
                var a, l;
                (a = i.current) === null || a === void 0 ? void 0 : (l = a.focus) === null || l === void 0 ? void 0 : l.call(a);
                r.preventDefault();
                j.current = E.current ? null : i.current;
            }
        };
        var A = (0, t.useCallback)((function(e, r) {
            var t = r.itemKey;
            if (t != null) {
                var n;
                P.current = t;
                j.current = E.current ? null : (n = _.current[P.current]) === null || n === void 0 ? void 0 : n.current;
            }
        }), []);
        var F = t.Children.toArray(r).filter(t.isValidElement).reduce((function(e, r, t, a) {
            /* Filter out initial Dividers
     * Requires reduce() over filter() because a Heading may have been
     * before the Divider.
     */
            if (r.type.as === "Divider" && e.length === 0) {
                return e;
            }
 // Filter out consecutive Dividers and Headings
            // If Divider is next to Heading then it will remove both Heading and Divider unless Heading has title prop set to it
                        if ((r.type.as === "Heading" || r.type.as === "Divider") && !r.props.title && r.type.filterConsecutive && a.length > t + 1) {
                var o = a[t + 1];
                if (o.type.as === "Heading" || o.type.as === "Divider") {
                    return e;
                }
            }
 // Filter out last Dividers and Headings
                        if ((r.type.as === "Divider" || r.type.as === "Heading") && t === a.length - 1) {
                return e;
            }
 // the filterConsecutive check filters out Headings and Dividers whether styled or not
            // the typeof check prevents props from being applied to HTML elements
            // such as the zero-height scroll bottom div in <ResultsMenu>
                        if (!(r.type.as === "Divider" || r.type.as === "Heading") && typeof r.type !== "string") {
                var l =  n().createRef();
                var s = I.length;
                I.push(l);
                if (R == null && (!("disabled" in r.props) || !r.props.disabled)) {
                    R = T;
                }
                if (P.current === T && "disabled" in r.props && r.props.disabled) {
                    D = true;
                }
                if ("selected" in r.props && r.props.selected) {
                    q = true;
                    if ("disabled" in r.props && !r.props.disabled) {
                        K = true;
                    }
                }
                var c = !i && P.current === T ? undefined : -1;
                e.push( n().cloneElement(r, {
                    itemKey: T,
                    onFocus: A,
                    preventFocus: i,
                    ref: function e(t) {
                        ke(r.ref, t);
                        ke(l, t);
                        if (!t) {
                            return;
                        }
                        if (t === j.current) {
                            // When using `VirtualizedResultsMenu` the ref content changes but the
                            // count doesn't. The keyboard scrolling will break if the focus index
                            // isn't corrected during scroll
                            P.current = s;
                        } else if (E.current && s === P.current) {
                            var n;
                            (n = Ye(t)) === null || n === void 0 ? void 0 : n.focus();
                            E.current = false;
                        }
                    },
                    tabIndex: c
                }));
                T += 1;
                return e;
            }
            e.push(r);
            return e;
        }), []);
        if (R !== P.current && (!q || K) && D && R != null) {
            P.current = R;
            O();
        }
        var $ = {
            "data-test": "menu",
            onKeyDown: H,
            role: x,
            tabIndex: R == null && !i ? 0 : undefined
        };
        return m ?  n().createElement(Ge, Je({}, $, {
            elementRef: a,
            stopScrollPropagation: true
        }, l()(h, "tagName")), F) :  n().createElement(Be, Je({}, $, {
            ref: a
        }, h), F);
    }
    sr.propTypes = or;
    sr.defaultProps = lr;
    sr.Item = ze;
    sr.Divider = x;
    sr.Heading = N;
    /* harmony default export */    const cr = sr;
    // CONCATENATED MODULE: ./src/Menu/index.ts
    module.exports = r;
    /******/})();
