// @splunk/react-ui/Message.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var a = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(a, {
                a
            });
            /******/            return a;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, a) => {
            /******/ for (var i in a) {
                /******/ if (e.o(a, i) && !e.o(r, i)) {
                    /******/ Object.defineProperty(r, i, {
                        enumerable: true,
                        get: a[i]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        Link: () => /* reexport */ z,
        default: () => /* reexport */ pe
    });
    // CONCATENATED MODULE: external "react"
    const a = require("react");
    var i =  e.n(a);
    // CONCATENATED MODULE: external "prop-types"
    const n = require("prop-types");
    var s =  e.n(n);
    // CONCATENATED MODULE: external "@splunk/react-icons/CheckCircle"
    const o = require("@splunk/react-icons/CheckCircle");
    var t =  e.n(o);
    // CONCATENATED MODULE: external "@splunk/react-icons/Cross"
    const c = require("@splunk/react-icons/Cross");
    var l =  e.n(c);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Close"
    const p = require("@splunk/react-icons/enterprise/Close");
    var u =  e.n(p);
    // CONCATENATED MODULE: external "@splunk/react-icons/ExclamationCircle"
    const v = require("@splunk/react-icons/ExclamationCircle");
    var d =  e.n(v);
    // CONCATENATED MODULE: external "@splunk/react-icons/InformationCircle"
    const b = require("@splunk/react-icons/InformationCircle");
    var g =  e.n(b);
    // CONCATENATED MODULE: external "@splunk/react-icons/ExclamationTriangle"
    const f = require("@splunk/react-icons/ExclamationTriangle");
    var m =  e.n(f);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Error"
    const y = require("@splunk/react-icons/enterprise/Error");
    var k =  e.n(y);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/InfoCircle"
    const h = require("@splunk/react-icons/enterprise/InfoCircle");
    var C =  e.n(h);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Success"
    const x = require("@splunk/react-icons/enterprise/Success");
    var w =  e.n(x);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Warning"
    const O = require("@splunk/react-icons/enterprise/Warning");
    var S =  e.n(O);
    // CONCATENATED MODULE: external "@splunk/react-ui/MessageBar"
    const j = require("@splunk/react-ui/MessageBar");
    var q =  e.n(j);
    // CONCATENATED MODULE: external "@splunk/react-ui/ScreenReaderContent"
    const _ = require("@splunk/react-ui/ScreenReaderContent");
    var E =  e.n(_);
    // CONCATENATED MODULE: external "@splunk/themes"
    const P = require("@splunk/themes");
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const I = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "styled-components"
    const R = require("styled-components");
    var L =  e.n(R);
    // CONCATENATED MODULE: external "@splunk/react-ui/Link"
    const T = require("@splunk/react-ui/Link");
    var $ =  e.n(T);
    // CONCATENATED MODULE: ./src/Message/LinkStyles.ts
    var M = L()($()).withConfig({
        displayName: "LinkStyles__StyledLink",
        componentId: "w06zjv-0"
    })([ "" ]);
    // CONCATENATED MODULE: ./src/Message/Link.tsx
    function A(e, r) {
        if (e == null) return {};
        var a = N(e, r);
        var i, n;
        if (Object.getOwnPropertySymbols) {
            var s = Object.getOwnPropertySymbols(e);
            for (n = 0; n < s.length; n++) {
                i = s[n];
                if (r.indexOf(i) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, i)) continue;
                a[i] = e[i];
            }
        }
        return a;
    }
    function N(e, r) {
        if (e == null) return {};
        var a = {};
        var i = Object.keys(e);
        var n, s;
        for (s = 0; s < i.length; s++) {
            n = i[s];
            if (r.indexOf(n) >= 0) continue;
            a[n] = e[n];
        }
        return a;
    }
    var W = {
        children: s().node,
        elementRef: s().oneOfType([ s().func, s().object ]),
        openInNewContext: s().bool,
        to: s().string
    };
    /**
 * @deprecated `Message.Link` is deprecated and will be removed in a future major version. Use `Link` instead.
 *
 * A `Link` styled to work with `Message`'s `banner` appearance.
 */    function X(e) {
        var r = e.children, a = A(e, [ "children" ]);
        // @docs-props-type LinkPropsBase
                if (false) {}
        
        return i().createElement(M, a, r);
    }
    X.propTypes = W;
    /* harmony default export */    const z = X;
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const V = require("@splunk/react-ui/Box");
    var B =  e.n(V);
    // CONCATENATED MODULE: external "@splunk/react-ui/Clickable"
    const D = require("@splunk/react-ui/Clickable");
    var H =  e.n(D);
    // CONCATENATED MODULE: ./src/Message/TitleStyles.ts
    var F = L().p.withConfig({
        displayName: "TitleStyles__StyledTitle",
        componentId: "sc-6gbjha-0"
    })([ "", " ", "" ], P.mixins.reset("block"), (0, P.pick)({
        prisma: (0, R.css)([ "font-size:14px;line-height:20px;" ])
    }));
    // CONCATENATED MODULE: ./src/Message/MessageStyles.ts
    var G = L().div.withConfig({
        displayName: "MessageStyles__StyledContent",
        componentId: "eg66af-0"
    })([ "display:flex;align-items:baseline;flex:1;", ";", " ", ";" ], (0, P.pick)({
        prisma: (0, R.css)([ "padding:", " 0;color:", ";" ], P.variables.spacingMedium, P.variables.contentColorActive),
        enterprise: (0, R.css)([ "padding:12px 0;" ])
    }), (0, P.pick)({
        prisma: (0, R.css)([ "padding-left:", ";" ], P.variables.spacingXSmall),
        enterprise: (0, R.css)([ "padding-left:15px;" ])
    }), (function(e) {
        var r = e.$fillStyle;
        return r && (0, P.pick)({
            prisma: (0, R.css)([ "padding-left:", ";" ], P.variables.spacingMedium)
        });
    }));
    var J = L()(H()).withConfig({
        displayName: "MessageStyles__StyledRemove",
        componentId: "eg66af-1"
    })([ "", ";border:1px solid transparent;border-radius:", ";color:", ";", ";cursor:pointer;position:absolute;", " padding:8px;&:hover,&:focus{background:", ";border:", ";color:", ";}&:active{", ";}", "" ], P.mixins.reset("flex"), (0, 
    P.pick)({
        prisma: "50%",
        enterprise: P.variables.borderRadius
    }), (0, P.pick)({
        prisma: P.variables.contentColorMuted,
        enterprise: {
            dark: P.variables.gray96,
            light: P.variables.gray45
        }
    }), (function(e) {
        var r = e.$appearance;
        return r === "banner" && (0, P.pick)({
            prisma: (0, R.css)([ "color:", ";" ], P.variables.contentColorInverted)
        });
    }), (0, P.pick)({
        prisma: (0, R.css)([ "top:3px;right:2px;" ]),
        enterprise: (0, R.css)([ "top:8px;right:8px;" ])
    }), (0, P.pick)({
        prisma: P.variables.interactiveColorOverlayHover,
        enterprise: {
            dark: P.variables.gray30,
            light: P.variables.gray92
        }
    }), (0, P.pick)({
        prisma: (0, R.css)([ "1px solid inherit" ]),
        enterprise: (0, R.css)([ "1px solid ", "" ], P.variables.gray80)
    }), (0, P.pick)({
        enterprise: {
            dark: P.variables.gray96,
            light: P.variables.linkColor
        },
        prisma: P.variables.contentColorActive
    }), (0, P.pick)({
        prisma: (0, R.css)([ "background:", ";" ], P.variables.interactiveColorOverlayActive),
        enterprise: (0, R.css)([ "box-shadow:", ";" ], P.variables.focusShadow)
    }), (0, P.pick)({
        prisma: (0, R.css)([ "&:focus{box-shadow:0 0 0 3px ", ";}" ], P.variables.focusColor)
    }));
    var K = L().span.withConfig({
        displayName: "MessageStyles__StyledIconWrapper",
        componentId: "eg66af-2"
    })([ "position:relative;min-width:", ";text-align:center;color:", ";", " ", "" ], (0, 
    P.pick)({
        prisma: P.variables.spacingXLarge,
        enterprise: "25px"
    }), P.variables.white, (0, P.pick)({
        enterprise: (0, R.css)([ "top:-2px;" ]),
        prisma: (0, R.css)([ "top:-1.8px;" ])
    }), (0, P.pickVariant)("$type", {
        info: (0, R.css)([ "color:", ";" ], (0, P.pick)({
            prisma: P.variables.contentColorMuted,
            enterprise: P.variables.infoColor
        })),
        success: (0, R.css)([ "color:", ";" ], P.variables.accentColorPositive),
        warning: (0, R.css)([ "color:", ";" ], P.variables.accentColorWarning),
        error: (0, R.css)([ "color:", ";" ], P.variables.accentColorNegative)
    }));
    var Q = L()(B()).withConfig({
        displayName: "MessageStyles__StyledBox",
        componentId: "eg66af-3"
    })([ "", ";align-items:baseline;position:relative;border-radius:", ";margin-bottom:", ";", " word-wrap:break-word;", " ", "" ], P.mixins.reset("flex"), (0, 
    P.pick)({
        prisma: P.variables.borderRadius,
        enterprise: "5px"
    }), P.variables.spacingSmall, (0, P.pickVariant)("$hasRemoveIcon", {
        true: {
            prisma: (0, R.css)([ "padding-right:", ";" ], P.variables.spacingXXXLarge),
            enterprise: (0, R.css)([ "padding-right:40px;" ])
        },
        false: {
            prisma: (0, R.css)([ "padding-right:", ";" ], P.variables.spacingSmall),
            enterprise: (0, R.css)([ "padding-right:", ";" ], P.variables.spacingHalf)
        }
    }), (function(e) {
        var r = e.$fillStyle;
        return r && (0, P.pick)({
            enterprise: (0, R.css)([ "& > ", "{left:", ";}", "" ], 
            /* sc-sel */
            K, P.variables.spacingXSmall, (0, P.pickVariant)("$type", {
                info: (0, R.css)([ "", "" ], (0, P.pick)({
                    light: (0, R.css)([ "background-color:", ";border:1px solid ", ";" ], P.variables.infoColorL50, P.variables.infoColor),
                    dark: (0, R.css)([ "background-color:", ";" ], P.mixins.colorWithAlpha(P.variables.infoColor, .5))
                })),
                success: (0, R.css)([ "", "" ], (0, P.pick)({
                    light: (0, R.css)([ "background-color:", ";border:1px solid ", ";" ], P.variables.successColorL50, P.variables.successColor),
                    dark: (0, R.css)([ "background-color:", ";" ], P.mixins.colorWithAlpha(P.variables.successColor, .5))
                })),
                warning: (0, R.css)([ "", "" ], (0, P.pick)({
                    light: (0, R.css)([ "background-color:", ";border:1px solid ", ";" ], P.variables.warningColorL50, P.variables.warningColor),
                    dark: (0, R.css)([ "background-color:", ";" ], P.mixins.colorWithAlpha(P.variables.warningColor, .5))
                })),
                error: (0, R.css)([ "", "" ], (0, P.pick)({
                    light: (0, R.css)([ "background-color:", ";border:1px solid ", ";" ], P.variables.errorColorL50, P.variables.errorColor),
                    dark: (0, R.css)([ "background-color:", ";" ], P.mixins.colorWithAlpha(P.variables.errorColor, .5))
                }))
            })),
            prisma: (0, R.css)([ "background:linear-gradient( to right,", " ", " ", ",", " 100% );box-shadow:", ";& > ", "{color:", ";}" ], (0, 
            P.pickVariant)("$type", {
                info: (0, R.css)([ "", " 0,", " ", "," ], P.variables.contentColorActive, P.variables.contentColorActive, P.variables.spacingXLarge),
                success: (0, R.css)([ "", " 0,", " ", "," ], P.variables.accentColorPositive, P.variables.accentColorPositive, P.variables.spacingXLarge),
                warning: (0, R.css)([ "", " 0,", " ", "," ], P.variables.accentColorWarning, P.variables.accentColorWarning, P.variables.spacingXLarge),
                error: (0, R.css)([ "", " 0,", " ", "," ], P.variables.accentColorNegative, P.variables.accentColorNegative, P.variables.spacingXLarge)
            }), P.variables.backgroundColorPopup, P.variables.spacingXLarge, P.variables.backgroundColorPopup, P.variables.embossShadow, 
            /* sc-sel */
            K, P.variables.backgroundColorPopup)
        });
    }), (0, P.pick)({
        prisma: (0, R.css)([ "& ", "{color:", ";}" ], 
        /* sc-sel */
        F, (0, P.pickVariant)("$type", {
            info: P.variables.contentColorActive,
            warning: P.variables.accentColorWarning,
            error: P.variables.accentColorNegative,
            success: P.variables.accentColorPositive
        }))
    }));
    // CONCATENATED MODULE: ./src/Message/Title.tsx
    function U(e, r) {
        if (e == null) return {};
        var a = Y(e, r);
        var i, n;
        if (Object.getOwnPropertySymbols) {
            var s = Object.getOwnPropertySymbols(e);
            for (n = 0; n < s.length; n++) {
                i = s[n];
                if (r.indexOf(i) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, i)) continue;
                a[i] = e[i];
            }
        }
        return a;
    }
    function Y(e, r) {
        if (e == null) return {};
        var a = {};
        var i = Object.keys(e);
        var n, s;
        for (s = 0; s < i.length; s++) {
            n = i[s];
            if (r.indexOf(n) >= 0) continue;
            a[n] = e[n];
        }
        return a;
    }
    var Z = {
        children: s().node
    };
    /**
 * A title component for use in `Message`.
 */    function ee(e) {
        var r = e.children, a = U(e, [ "children" ]);
        // @docs-props-type TitleProps
                
        return i().createElement(F, a, r);
    }
    ee.propTypes = Z;
    /* harmony default export */    const re = ee;
    // CONCATENATED MODULE: ./src/Message/Message.tsx
    function ae() {
        ae = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var a = arguments[r];
                for (var i in a) {
                    if (Object.prototype.hasOwnProperty.call(a, i)) {
                        e[i] = a[i];
                    }
                }
            }
            return e;
        };
        return ae.apply(this, arguments);
    }
    function ie(e, r) {
        if (e == null) return {};
        var a = ne(e, r);
        var i, n;
        if (Object.getOwnPropertySymbols) {
            var s = Object.getOwnPropertySymbols(e);
            for (n = 0; n < s.length; n++) {
                i = s[n];
                if (r.indexOf(i) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, i)) continue;
                a[i] = e[i];
            }
        }
        return a;
    }
    function ne(e, r) {
        if (e == null) return {};
        var a = {};
        var i = Object.keys(e);
        var n, s;
        for (s = 0; s < i.length; s++) {
            n = i[s];
            if (r.indexOf(n) >= 0) continue;
            a[n] = e[n];
        }
        return a;
    }
    var se = {
        appearance: s().oneOf([ "default", "fill", "banner" ]),
        children: s().node,
        elementRef: s().oneOfType([ s().func, s().object ]),
        onRequestRemove: s().func,
        type: s().oneOf([ "info", "success", "warning", "error" ])
    };
    var oe = Object.freeze({
        info: (0, I._)("Info"),
        warning: (0, I._)("Warning"),
        error: (0, I._)("Alert"),
        success: (0, I._)("Success")
    });
    var te = Object.freeze({
        info: C(),
        warning: S(),
        error: k(),
        success: w()
    });
    var ce = Object.freeze({
        info: g(),
        warning: m(),
        error: d(),
        success: t()
    });
    function le(e) {
        var r = e.appearance, n = r === void 0 ? "default" : r, s = e.children, o = e.type, t = o === void 0 ? "warning" : o, c = e.onRequestRemove, p = ie(e, [ "appearance", "children", "type", "onRequestRemove" ]);
        // @docs-props-type MessagePropsBase
                var v = (0, P.useSplunkTheme)(), d = v.isEnterprise;
        if (false) {}
 // double check this can only be button
                var b = (0, a.useCallback)((function(e) {
            c === null || c === void 0 ? void 0 : c(e);
        }), [ c ]);
        if (n === "banner") {
            
            return i().createElement(q(), ae({
                type: t,
                "data-test": "banner",
                onRequestClose: b
            }, p), s);
        }
        var g = d ? te : ce;
        var f = g[t];
        var m = n === "fill";
        var y = n === "default" || n === "fill";
        
        return i().createElement(Q, ae({
            $appearance: n,
            $fillStyle: m,
            $hasRemoveIcon: !!c,
            $type: t,
            "data-test-type": t,
            "data-test": "message"
        }, p), y &&  i().createElement(K, {
            $fillStyle: m,
            $type: t
        }, d ?  i().createElement(f, {
            height: "24px",
            width: "24px",
            hideDefaultTooltip: true,
            screenReaderText: null
        }) :  i().createElement(f, {
            variant: "filled"
        }),  i().createElement(E(), null, oe[t])),  i().createElement(G, {
            "data-test": "content",
            $fillStyle: m
        }, s), c &&  i().createElement(J, {
            "data-test": "remove",
            onClick: b,
            $appearance: n
        }, d ?  i().createElement(u(), {
            height: "12px",
            hideDefaultTooltip: true,
            screenReaderText: null,
            width: "12px"
        }) :  i().createElement(l(), {
            height: "20px",
            width: "20px"
        })));
    }
    le.propTypes = se;
    le.Title = re;
    le.Link = z;
    /* harmony default export */    const pe = le;
    // CONCATENATED MODULE: ./src/Message/index.ts
    module.exports = r;
    /******/})();
