// @splunk/react-ui/MessageBar.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var a = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(a, {
                a
            });
            /******/            return a;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, a) => {
            /******/ for (var i in a) {
                /******/ if (e.o(a, i) && !e.o(r, i)) {
                    /******/ Object.defineProperty(r, i, {
                        enumerable: true,
                        get: a[i]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        default: () => /* reexport */ ie
    });
    // CONCATENATED MODULE: external "react"
    const a = require("react");
    var i =  e.n(a);
    // CONCATENATED MODULE: external "prop-types"
    const s = require("prop-types");
    var n =  e.n(s);
    // CONCATENATED MODULE: external "@splunk/react-icons/Cross"
    const t = require("@splunk/react-icons/Cross");
    var o =  e.n(t);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Close"
    const c = require("@splunk/react-icons/enterprise/Close");
    var l =  e.n(c);
    // CONCATENATED MODULE: external "@splunk/react-ui/ScreenReaderContent"
    const p = require("@splunk/react-ui/ScreenReaderContent");
    var u =  e.n(p);
    // CONCATENATED MODULE: external "@splunk/themes"
    const d = require("@splunk/themes");
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const v = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "styled-components"
    const g = require("styled-components");
    var b =  e.n(g);
    // CONCATENATED MODULE: external "@splunk/react-icons/CheckCircle"
    const h = require("@splunk/react-icons/CheckCircle");
    var y =  e.n(h);
    // CONCATENATED MODULE: external "@splunk/react-icons/ExclamationCircle"
    const m = require("@splunk/react-icons/ExclamationCircle");
    var f =  e.n(m);
    // CONCATENATED MODULE: external "@splunk/react-icons/InformationCircle"
    const k = require("@splunk/react-icons/InformationCircle");
    var x =  e.n(k);
    // CONCATENATED MODULE: external "@splunk/react-icons/ExclamationTriangle"
    const C = require("@splunk/react-icons/ExclamationTriangle");
    var w =  e.n(C);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Error"
    const S = require("@splunk/react-icons/enterprise/Error");
    var _ =  e.n(S);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/InfoCircle"
    const I = require("@splunk/react-icons/enterprise/InfoCircle");
    var z =  e.n(I);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Success"
    const q = require("@splunk/react-icons/enterprise/Success");
    var O =  e.n(q);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Warning"
    const E = require("@splunk/react-icons/enterprise/Warning");
    var B =  e.n(E);
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const M = require("@splunk/react-ui/Box");
    var j =  e.n(M);
    // CONCATENATED MODULE: external "@splunk/react-ui/Clickable"
    const N = require("@splunk/react-ui/Clickable");
    var R =  e.n(N);
    // CONCATENATED MODULE: ./src/MessageBar/MessageBarStyles.ts
    var T = b().div.withConfig({
        displayName: "MessageBarStyles__MessageBarContent",
        componentId: "sc-1uyhzza-0"
    })([ "", ";color:inherit;padding:0 8px;text-align:center;margin-right:auto;& a{display:inline-block;color:inherit;text-decoration:underline;text-underline-position:under;padding:0 3px 2px 3px;&:not([disabled]):focus{border-radius:", ";outline:0;box-shadow:", ";}}" ], d.mixins.reset("inline"), d.variables.borderRadius, (0, 
    d.pick)({
        enterprise: {
            dark: (0, g.css)([ "0 0 0 2px ", "" ], d.variables.gray80),
            light: (0, g.css)([ "0 0 0 2px ", "" ], d.variables.gray25)
        },
        prisma: {
            light: (0, d.pickVariant)("$type", {
                info: (0, g.css)([ "0 0 0 3px ", "" ], d.variables.black),
                warning: (0, g.css)([ "0 0 0 3px ", "" ], d.variables.black),
                error: (0, g.css)([ "0 0 0 3px ", "" ], d.variables.white),
                success: (0, g.css)([ "0 0 0 3px ", "" ], d.variables.white)
            }),
            dark: (0, g.css)([ "0 0 0 3px ", "" ], d.variables.black)
        }
    }));
    var W = b()(R()).withConfig({
        displayName: "MessageBarStyles__MessageBarCloseWrapper",
        componentId: "sc-1uyhzza-1"
    })([ "", ";border:1px solid transparent;border-radius:", ";color:inherit;cursor:pointer;padding:", ";", " &:hover,&:focus{background:", ";border:", ";color:", ";}&:active{", ";}", "" ], d.mixins.reset("flex"), d.variables.borderRadius, (0, 
    d.pick)({
        enterprise: d.variables.spacingXSmall,
        prisma: "3px"
    }), (0, d.pick)({
        enterprise: (0, g.css)([ "margin-top:1px;" ])
    }), (0, d.pick)({
        prisma: d.variables.interactiveColorOverlayHover,
        enterprise: {
            dark: d.variables.gray30,
            light: d.variables.gray92
        }
    }), (0, d.pick)({
        prisma: (0, g.css)([ "1px solid inherit" ]),
        enterprise: (0, g.css)([ "1px solid ", "" ], d.variables.gray80)
    }), (0, d.pick)({
        enterprise: {
            dark: d.variables.gray96,
            light: d.variables.linkColor
        }
    }), (0, d.pick)({
        prisma: (0, g.css)([ "background:", ";" ], d.variables.interactiveColorOverlayActive),
        enterprise: (0, g.css)([ "box-shadow:", ";" ], d.variables.focusShadow)
    }), (0, d.pick)({
        prisma: (0, g.css)([ "&:focus{border-radius:", ";box-shadow:", ";}" ], d.variables.borderRadius, (0, 
        d.pick)({
            light: (0, d.pickVariant)("$type", {
                info: (0, g.css)([ "0 0 0 3px ", "" ], d.variables.black),
                warning: (0, g.css)([ "0 0 0 3px ", "" ], d.variables.black),
                error: (0, g.css)([ "0 0 0 3px ", "" ], d.variables.white),
                success: (0, g.css)([ "0 0 0 3px ", "" ], d.variables.white)
            }),
            dark: (0, g.css)([ "0 0 0 3px ", "" ], d.variables.black)
        }))
    }));
    var $ = (0, g.css)([ "margin-left:auto;" ]);
    var P = b()(y()).withConfig({
        displayName: "MessageBarStyles__StyledCheckCircleIcon",
        componentId: "sc-1uyhzza-2"
    })([ "", "" ], $);
    var A = b()(f()).withConfig({
        displayName: "MessageBarStyles__StyledExclamationCircleIcon",
        componentId: "sc-1uyhzza-3"
    })([ "", "" ], $);
    var L = b()(x()).withConfig({
        displayName: "MessageBarStyles__StyledInformationCircleIcon",
        componentId: "sc-1uyhzza-4"
    })([ "", "" ], $);
    var V = b()(w()).withConfig({
        displayName: "MessageBarStyles__StyledExclamationTriangleIcon",
        componentId: "sc-1uyhzza-5"
    })([ "", "" ], $);
    var D = b()(_()).withConfig({
        displayName: "MessageBarStyles__StyledEnterpriseErrorIcon",
        componentId: "sc-1uyhzza-6"
    })([ "", " color:", ";" ], $, d.variables.errorColor);
    var H = b()(z()).withConfig({
        displayName: "MessageBarStyles__StyledEnterpriseInfoCircleIcon",
        componentId: "sc-1uyhzza-7"
    })([ "", " color:", ";" ], $, d.variables.infoColor);
    var X = b()(O()).withConfig({
        displayName: "MessageBarStyles__StyledEnterpriseSuccessIcon",
        componentId: "sc-1uyhzza-8"
    })([ "", " color:", ";" ], $, d.variables.successColor);
    var F = b()(B()).withConfig({
        displayName: "MessageBarStyles__StyledEnterpriseWarningIcon",
        componentId: "sc-1uyhzza-9"
    })([ "", " color:", ";" ], $, d.variables.warningColor);
    var G = b()(j()).withConfig({
        displayName: "MessageBarStyles__MessageBarWrapper",
        componentId: "sc-1uyhzza-10"
    })([ "", ";display:flex;align-items:center;justify-content:center;min-height:40px;margin-bottom:", ";", ";word-wrap:break-word;color:", ";", ";" ], d.mixins.reset("block"), d.variables.spacingSmall, (function(e) {
        var r = e.$hasCloseButton;
        return r ? (0, g.css)([ "", "" ], (0, d.pick)({
            prisma: (0, g.css)([ "padding:0 8px 0 32px;" ]),
            enterprise: (0, g.css)([ "padding:0 8px 0 38px;" ])
        })) : (0, g.css)([ "", "" ], (0, d.pick)({
            prisma: (0, g.css)([ "padding:0 33px 0 32px;" ]),
            enterprise: (0, g.css)([ "padding:0 26px 0 32px;" ])
        }));
    }), (0, d.pick)({
        prisma: (0, d.pickVariant)("$type", {
            info: {
                light: d.variables.contentColorActive,
                dark: d.variables.contentColorInverted
            },
            success: d.variables.contentColorInverted,
            warning: d.variables.black,
            error: d.variables.contentColorInverted
        }),
        enterprise: {
            dark: d.variables.gray96,
            light: d.variables.gray25
        }
    }), (0, d.pickVariant)("$type", {
        info: (0, g.css)([ "background-color:", ";" ], (0, d.pick)({
            prisma: {
                dark: d.variables.white,
                light: d.variables.neutral200
            },
            enterprise: {
                dark: d.mixins.colorWithAlpha(d.variables.infoColor, .5),
                light: d.variables.infoColorL50
            }
        })),
        success: (0, g.css)([ "background-color:", ";" ], (0, d.pick)({
            prisma: d.variables.accentColorPositive,
            enterprise: {
                dark: d.mixins.colorWithAlpha(d.variables.successColor, .5),
                light: d.variables.successColorL50
            }
        })),
        warning: (0, g.css)([ "background-color:", ";" ], (0, d.pick)({
            prisma: d.variables.accentColorWarning,
            enterprise: {
                dark: d.mixins.colorWithAlpha(d.variables.warningColor, .5),
                light: d.variables.warningColorL50
            }
        })),
        error: (0, g.css)([ "background-color:", ";" ], (0, d.pick)({
            prisma: d.variables.accentColorNegative,
            enterprise: {
                dark: d.mixins.colorWithAlpha(d.variables.errorColor, .5),
                light: d.variables.errorColorL50
            }
        }))
    }));
    // CONCATENATED MODULE: ./src/MessageBar/MessageBar.tsx
    function J() {
        J = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var a = arguments[r];
                for (var i in a) {
                    if (Object.prototype.hasOwnProperty.call(a, i)) {
                        e[i] = a[i];
                    }
                }
            }
            return e;
        };
        return J.apply(this, arguments);
    }
    function K(e, r) {
        if (e == null) return {};
        var a = Q(e, r);
        var i, s;
        if (Object.getOwnPropertySymbols) {
            var n = Object.getOwnPropertySymbols(e);
            for (s = 0; s < n.length; s++) {
                i = n[s];
                if (r.indexOf(i) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, i)) continue;
                a[i] = e[i];
            }
        }
        return a;
    }
    function Q(e, r) {
        if (e == null) return {};
        var a = {};
        var i = Object.keys(e);
        var s, n;
        for (n = 0; n < i.length; n++) {
            s = i[n];
            if (r.indexOf(s) >= 0) continue;
            a[s] = e[s];
        }
        return a;
    }
    var U = {
        children: n().node.isRequired,
        onRequestClose: n().func,
        type: n().oneOf([ "info", "success", "warning", "error" ]).isRequired
    };
    var Y = Object.freeze({
        info: (0, v._)("Info"),
        warning: (0, v._)("Warning"),
        error: (0, v._)("Alert"),
        success: (0, v._)("Success")
    });
    var Z = Object.freeze({
        info: H,
        warning: F,
        error: D,
        success: X
    });
    var ee = Object.freeze({
        info: L,
        warning: V,
        error: A,
        success: P
    });
    var re = "24px";
    function ae(e) {
        var r = e.children, a = e.type, s = e.onRequestClose, n = K(e, [ "children", "type", "onRequestClose" ]);
        // @docs-props-type MessageBarPropsBase
                var t = (0, d.useSplunkTheme)(), c = t.isEnterprise;
        var p = c && a === "error" ? "22px" : re;
        var g = c && a === "warning" ? "21px" : re;
        var b;
        if (c) {
            var h = Z[a];
            b =  i().createElement(h, {
                height: g,
                width: p,
                hideDefaultTooltip: true,
                screenReaderText: null,
                "aria-hidden": true,
                "data-test": "icon"
            });
        } else {
            var y = ee[a];
            b =  i().createElement(y, {
                height: re,
                width: re,
                variant: "filled",
                "aria-hidden": true,
                "data-test": "icon"
            });
        }
        
        return i().createElement(G, J({
            $type: a,
            "data-test": "message-bar",
            "data-test-type": a
        }, n, {
            role: "region",
            $hasCloseButton: Boolean(s)
        }), b,  i().createElement(u(), null, Y[a]),  i().createElement(T, {
            $type: a,
            "data-test": "content"
        }, r), s &&  i().createElement(W, {
            $type: a,
            onClick: s
        }, c ?  i().createElement(l(), {
            height: "12px",
            hideDefaultTooltip: true,
            screenReaderText: null,
            width: "12px"
        }) :  i().createElement(o(), {
            height: "19px",
            style: {
                margin: -.5
            },
            width: "19px"
        }),  i().createElement(u(), null, (0, v._)("Close"))));
    }
    ae.propTypes = U;
    /* harmony default export */    const ie = ae;
    // CONCATENATED MODULE: ./src/MessageBar/index.ts
    module.exports = r;
    /******/})();
