// @splunk/react-ui/Modal.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var r = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(r, {
                a: r
            });
            /******/            return r;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        Body: () => /* reexport */ R,
        Footer: () => /* reexport */ V,
        Header: () => /* reexport */ ye,
        default: () => /* reexport */ Le
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    var n =  e.n(r);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var i =  e.n(o);
    // CONCATENATED MODULE: external "lodash/defer"
    const a = require("lodash/defer");
    var l =  e.n(a);
    // CONCATENATED MODULE: external "lodash/keys"
    const c = require("lodash/keys");
    var s =  e.n(c);
    // CONCATENATED MODULE: external "lodash/memoize"
    const u = require("lodash/memoize");
    var p =  e.n(u);
    // CONCATENATED MODULE: external "lodash/omit"
    const f = require("lodash/omit");
    var d =  e.n(f);
    // CONCATENATED MODULE: external "@splunk/react-ui/ModalLayer"
    const y = require("@splunk/react-ui/ModalLayer");
    var b =  e.n(y);
    // CONCATENATED MODULE: external "@splunk/ui-utils/focus"
    const m = require("@splunk/ui-utils/focus");
    // CONCATENATED MODULE: external "@splunk/ui-utils/id"
    const v = require("@splunk/ui-utils/id");
    // CONCATENATED MODULE: external "@splunk/react-ui/ScrollContainerContext"
    const h = require("@splunk/react-ui/ScrollContainerContext");
    // CONCATENATED MODULE: external "styled-components"
    const g = require("styled-components");
    var x =  e.n(g);
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const O = require("@splunk/react-ui/Box");
    var S =  e.n(O);
    // CONCATENATED MODULE: external "@splunk/themes"
    const w = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/Modal/BodyStyles.ts
    var k = x()(S()).withConfig({
        displayName: "BodyStyles__StyledBox",
        componentId: "lv54z7-0"
    })([ "background-color:", ";", " flex:0 1 auto;overflow:auto;" ], (0, w.pick)({
        enterprise: w.variables.backgroundColor,
        prisma: w.variables.backgroundColorDialog
    }), (0, w.pick)({
        enterprise: {
            comfortable: (0, g.css)([ "padding:28px;" ]),
            compact: (0, g.css)([ "padding:24px;" ])
        },
        prisma: {
            comfortable: (0, g.css)([ "padding:12px 24px;&:first-child{padding-top:36px;}&:last-child{padding-bottom:36px;}" ]),
            compact: (0, g.css)([ "padding:8px 24px;&:first-child{padding-top:26px;}&:last-child{padding-bottom:26px;}" ])
        }
    }));
    // CONCATENATED MODULE: ./src/Modal/Body.tsx
    function j() {
        j = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return j.apply(this, arguments);
    }
    function C(e, t) {
        return T(e) || P(e, t) || _(e, t) || I();
    }
    function I() {
        throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function _(e, t) {
        if (!e) return;
        if (typeof e === "string") return E(e, t);
        var r = Object.prototype.toString.call(e).slice(8, -1);
        if (r === "Object" && e.constructor) r = e.constructor.name;
        if (r === "Map" || r === "Set") return Array.from(e);
        if (r === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(r)) return E(e, t);
    }
    function E(e, t) {
        if (t == null || t > e.length) t = e.length;
        for (var r = 0, n = new Array(t); r < t; r++) {
            n[r] = e[r];
        }
        return n;
    }
    function P(e, t) {
        if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(e))) return;
        var r = [];
        var n = true;
        var o = false;
        var i = undefined;
        try {
            for (var a = e[Symbol.iterator](), l; !(n = (l = a.next()).done); n = true) {
                r.push(l.value);
                if (t && r.length === t) break;
            }
        } catch (e) {
            o = true;
            i = e;
        } finally {
            try {
                if (!n && a["return"] != null) a["return"]();
            } finally {
                if (o) throw i;
            }
        }
        return r;
    }
    function T(e) {
        if (Array.isArray(e)) return e;
    }
    function M(e, t) {
        if (e == null) return {};
        var r = q(e, t);
        var n, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                n = i[o];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function q(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var o, i;
        for (i = 0; i < n.length; i++) {
            o = n[i];
            if (t.indexOf(o) >= 0) continue;
            r[o] = e[o];
        }
        return r;
    }
    var F = {
        children: i().node
    };
    /**
 * A styled container for `Modal` body content.
 */    function D(e) {
        var t = e.children, o = M(e, [ "children" ]);
        // @docs-props-type BodyPropsBase
                var i = (0, r.useState)(), a = C(i, 2), l = a[0], c = a[1];
        var s = (0, r.useCallback)((function(e) {
            c(e);
        }), []);
        
        return n().createElement(k, j({
            "data-test": "body"
        }, o, {
            elementRef: s
        }),  n().createElement(h.ScrollContainerProvider, {
            value: l
        }, t));
    }
    D.propTypes = F;
    /* harmony default export */    const R = D;
    // CONCATENATED MODULE: external "@splunk/react-ui/Divider"
    const A = require("@splunk/react-ui/Divider");
    var H =  e.n(A);
    // CONCATENATED MODULE: ./src/Modal/FooterStyles.ts
    var B = x()(S()).withConfig({
        displayName: "FooterStyles__StyledBox",
        componentId: "yszcmv-0"
    })([ "flex:0 0 auto;text-align:right;padding:", ";background-color:", ";& > button{min-width:80px;}" ], (0, 
    w.pick)({
        enterprise: w.variables.spacing,
        prisma: {
            comfortable: "24px",
            compact: "18px 24px"
        }
    }), (0, w.pick)({
        enterprise: w.variables.backgroundColor,
        prisma: w.variables.backgroundColorDialog
    }));
    // CONCATENATED MODULE: ./src/Modal/ModalContext.tsx
    var N =  (0, r.createContext)({});
    /* harmony default export */    const z = N;
    // CONCATENATED MODULE: ./src/Modal/Footer.tsx
    function L() {
        L = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return L.apply(this, arguments);
    }
    function X(e, t) {
        if (e == null) return {};
        var r = $(e, t);
        var n, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                n = i[o];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function $(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var o, i;
        for (i = 0; i < n.length; i++) {
            o = n[i];
            if (t.indexOf(o) >= 0) continue;
            r[o] = e[o];
        }
        return r;
    }
    var K = {
        children: i().node
    };
    /**
 * A styled container for `Modal` footer content.
 */    function U(e) {
        var t = e.children, o = X(e, [ "children" ]);
        // @docs-props-type FooterPropsBase
                var i = (0, r.useContext)(z), a = i.divider;
        var l = a === "both" || a === "footer";
        
        return n().createElement(n().Fragment, null, l &&  n().createElement(H(), {
            "aria-hidden": true
        }),  n().createElement(B, L({
            "data-test": "footer"
        }, o), t));
    }
    U.propTypes = K;
    /* harmony default export */    const V = U;
    // CONCATENATED MODULE: external "@splunk/react-icons/Cross"
    const W = require("@splunk/react-icons/Cross");
    var G =  e.n(W);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Close"
    const J = require("@splunk/react-icons/enterprise/Close");
    var Q =  e.n(J);
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const Y = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "@splunk/react-ui/ButtonSimple"
    const Z = require("@splunk/react-ui/ButtonSimple");
    var ee =  e.n(Z);
    // CONCATENATED MODULE: ./src/Modal/HeaderStyles.ts
    var te = x()(S()).withConfig({
        displayName: "HeaderStyles__StyledBox",
        componentId: "sc-1y722ut-0"
    })([ "", ";flex:0 0 auto;position:relative;min-height:", ";background-color:", ";padding:", ";align-items:flex-start;padding-right:", ";" ], w.mixins.reset("flex"), (0, 
    w.pick)({
        enterprise: "30px",
        prisma: {
            comfortable: "60px",
            compact: "52px"
        }
    }), (0, w.pick)({
        enterprise: w.variables.backgroundColor,
        prisma: w.variables.backgroundColorDialog
    }), (0, w.pick)({
        enterprise: "25px 28px",
        prisma: {
            comfortable: "18px 24px",
            compact: "14px 24px"
        }
    }), (function(e) {
        var t = e.$close;
        return t && (0, w.pick)({
            enterprise: "54px",
            prisma: {
                comfortable: "64px",
                compact: "56px"
            }
        });
    }));
    var re = x().div.withConfig({
        displayName: "HeaderStyles__StyledTitleWrapper",
        componentId: "sc-1y722ut-1"
    })([ "", ";flex-direction:column;align-self:center;" ], w.mixins.reset("flex"));
    var ne = x().div.withConfig({
        displayName: "HeaderStyles__StyledIcon",
        componentId: "sc-1y722ut-2"
    })([ "", ";align-items:center;justify-content:center;margin-right:", ";background-color:", ";width:", ";height:", ";flex-shrink:0;" ], w.mixins.reset("inline-flex"), w.variables.spacingLarge, w.variables.transparent, (0, 
    w.pick)({
        comfortable: "40px",
        compact: "32px"
    }), (0, w.pick)({
        comfortable: "40px",
        compact: "32px"
    }));
    var oe = x().h2.withConfig({
        displayName: "HeaderStyles__StyledTitle",
        componentId: "sc-1y722ut-3"
    })([ "", ";", " margin:0;overflow-wrap:break-word;" ], w.mixins.reset("block"), (0, 
    w.pick)({
        enterprise: (0, g.css)([ "", "" ], w.mixins.typography({
            size: 20,
            weight: "semiBold",
            color: "active",
            lineHeight: 22
        })),
        prisma: (0, g.css)([ "", "" ], w.mixins.typography({
            size: 20,
            weight: "semiBold",
            color: "active",
            lineHeight: 24
        }))
    }));
    var ie = x().p.withConfig({
        displayName: "HeaderStyles__StyledSubtitle",
        componentId: "sc-1y722ut-4"
    })([ "", ";overflow-wrap:break-word;" ], w.mixins.reset("block"));
    var ae = x().div.withConfig({
        displayName: "HeaderStyles__StyledButtonsWrapper",
        componentId: "sc-1y722ut-5"
    })([ "", ";position:absolute;top:", ";right:", ";bottom:50%;", " max-height:35px;transform-origin:bottom right;transform:rotate(-90deg) translateX(100%);" ], w.mixins.reset("block"), (0, 
    w.pick)({
        enterprise: 0,
        prisma: {
            comfortable: "-2px",
            compact: "-6px"
        }
    }), (0, w.pick)({
        enterprise: 0,
        prisma: {
            comfortable: "-2px",
            compact: "-6px"
        }
    }), (0, w.pick)({
        prisma: {
            comfortable: (0, g.css)([ "min-height:35px;" ]),
            compact: (0, g.css)([ "min-height:30px;" ])
        }
    }));
    var le = x().div.withConfig({
        displayName: "HeaderStyles__StyledClose",
        componentId: "sc-1y722ut-6"
    })([ "", ";position:absolute;right:0;top:0;transform:rotate(90deg) translate(-50%,-50%);" ], w.mixins.reset("block"));
    var ce = x()(ee()).withConfig({
        displayName: "HeaderStyles__StyledButtonSimple",
        componentId: "sc-1y722ut-7"
    })([ "float:right;width:", ";height:", ";text-align:center;" ], w.variables.inputHeight, w.variables.inputHeight);
    // CONCATENATED MODULE: ./src/Modal/Header.tsx
    function se() {
        se = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return se.apply(this, arguments);
    }
    function ue(e, t) {
        if (e == null) return {};
        var r = pe(e, t);
        var n, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                n = i[o];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function pe(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var o, i;
        for (i = 0; i < n.length; i++) {
            o = n[i];
            if (t.indexOf(o) >= 0) continue;
            r[o] = e[o];
        }
        return r;
    }
    var fe = {
        children: i().node,
        icon: i().node,
        onRequestClose: i().func,
        subtitle: i().node,
        title: i().string
    };
    /**
 * A styled container for `Modal` header content.
 */    function de(e) {
        var t = e.children, o = e.icon, i = e.onRequestClose, a = e.subtitle, l = e.title, c = ue(e, [ "children", "icon", "onRequestClose", "subtitle", "title" ]);
        // @docs-props-type StepPropsBase
                var s = (0, r.useContext)(z), u = s.titleId, p = s.subtitleId, f = s.divider;
        var d = f === "both" || f === "header";
        var y = (0, w.useSplunkTheme)(), b = y.isCompact, m = y.isEnterprise;
        var v = b ? "20px" : "24px";
        
        return n().createElement(n().Fragment, null,  n().createElement(te, se({
            $close: !!i,
            "data-test": "header"
        }, c), o &&  n().createElement(ne, null, o), l ?  n().createElement(re, null, l &&  n().createElement(oe, {
            id: u,
            "data-test": "title"
        }, l), a &&  n().createElement(ie, {
            id: p,
            "data-test": "subtitle"
        }, a)) : t, i &&  n().createElement(ae, null,  n().createElement(le, null,  n().createElement(ce, {
            appearance: "pill",
            "aria-label": (0, Y._)("Close"),
            "data-test": "close",
            onClick: i
        }, m ?  n().createElement(Q(), {
            "data-test": "cross",
            inline: false,
            screenReaderText: null,
            size: "13px"
        }) :  n().createElement(G(), {
            "data-test": "cross",
            height: v,
            width: v
        }))))), d &&  n().createElement(H(), {
            "aria-hidden": true
        }));
    }
    de.propTypes = fe;
    /* harmony default export */    const ye = de;
    // CONCATENATED MODULE: external "react-spring"
    const be = require("react-spring");
    // CONCATENATED MODULE: ./src/Modal/ModalStyles.ts
    var me = x()(be.animated.div).withConfig({
        displayName: "ModalStyles__Styled",
        componentId: "sc-5fn8ds-0"
    })([ "", ";flex-direction:column;position:fixed;left:50%;transform:translateX(-50%);z-index:", ";", "" ], w.mixins.reset("flex"), w.variables.zindexModal, (0, 
    w.pick)({
        enterprise: (0, g.css)([ "box-shadow:0 1px 5px ", ";max-height:calc(100vh - ", " * 4);max-width:calc(100vw - ", " * 4);" ], w.variables.black, w.variables.spacing, w.variables.spacing),
        prisma: (0, g.css)([ "box-shadow:", ";max-height:calc(100vh - ", " * 4);max-width:calc(100vw - ", " * 4);border-radius:", ";overflow:hidden;" ], w.variables.modalShadow, w.variables.spacingXLarge, w.variables.spacingXLarge, w.variables.borderRadius)
    }));
    // CONCATENATED MODULE: ./src/utils/updateReactRef.ts
    /**
 * Updates a React ref. Callback refs and object refs (from `createRef` and `useRef`) are supported.
 *
 * @param ref - The React callback or object ref. Can be `null` or `undefined`.
 * @param current - The new value of the ref.
 */
    function ve(e, t) {
        if (e) {
            if (typeof e === "function") {
                e(t);
            } else {
                // the public signature of this util uses React.Ref<T> to mirror the way React types refs.
                // the intention here is to signal "we will take care of setting 'current', not you".
                e.current = t;
 // eslint-disable-line no-param-reassign
                        }
        }
    }
    // CONCATENATED MODULE: ./src/Modal/Modal.tsx
    function he() {
        he = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return he.apply(this, arguments);
    }
    function ge(e) {
        return we(e) || Se(e) || Oe(e) || xe();
    }
    function xe() {
        throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function Oe(e, t) {
        if (!e) return;
        if (typeof e === "string") return ke(e, t);
        var r = Object.prototype.toString.call(e).slice(8, -1);
        if (r === "Object" && e.constructor) r = e.constructor.name;
        if (r === "Map" || r === "Set") return Array.from(e);
        if (r === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(r)) return ke(e, t);
    }
    function Se(e) {
        if (typeof Symbol !== "undefined" && Symbol.iterator in Object(e)) return Array.from(e);
    }
    function we(e) {
        if (Array.isArray(e)) return ke(e);
    }
    function ke(e, t) {
        if (t == null || t > e.length) t = e.length;
        for (var r = 0, n = new Array(t); r < t; r++) {
            n[r] = e[r];
        }
        return n;
    }
    function je(e, t) {
        var r = Object.keys(e);
        if (Object.getOwnPropertySymbols) {
            var n = Object.getOwnPropertySymbols(e);
            if (t) n = n.filter((function(t) {
                return Object.getOwnPropertyDescriptor(e, t).enumerable;
            }));
            r.push.apply(r, n);
        }
        return r;
    }
    function Ce(e) {
        for (var t = 1; t < arguments.length; t++) {
            var r = arguments[t] != null ? arguments[t] : {};
            if (t % 2) {
                je(Object(r), true).forEach((function(t) {
                    He(e, t, r[t]);
                }));
            } else if (Object.getOwnPropertyDescriptors) {
                Object.defineProperties(e, Object.getOwnPropertyDescriptors(r));
            } else {
                je(Object(r)).forEach((function(t) {
                    Object.defineProperty(e, t, Object.getOwnPropertyDescriptor(r, t));
                }));
            }
        }
        return e;
    }
    function Ie(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            Ie = function e(t) {
                return typeof t;
            };
        } else {
            Ie = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return Ie(e);
    }
    function _e(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function Ee(e, t) {
        for (var r = 0; r < t.length; r++) {
            var n = t[r];
            n.enumerable = n.enumerable || false;
            n.configurable = true;
            if ("value" in n) n.writable = true;
            Object.defineProperty(e, n.key, n);
        }
    }
    function Pe(e, t, r) {
        if (t) Ee(e.prototype, t);
        if (r) Ee(e, r);
        return e;
    }
    function Te(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) Me(e, t);
    }
    function Me(e, t) {
        Me = Object.setPrototypeOf || function e(t, r) {
            t.__proto__ = r;
            return t;
        };
        return Me(e, t);
    }
    function qe(e) {
        var t = Re();
        return function r() {
            var n = Ae(e), o;
            if (t) {
                var i = Ae(this).constructor;
                o = Reflect.construct(n, arguments, i);
            } else {
                o = n.apply(this, arguments);
            }
            return Fe(this, o);
        };
    }
    function Fe(e, t) {
        if (t && (Ie(t) === "object" || typeof t === "function")) {
            return t;
        }
        return De(e);
    }
    function De(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function Re() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function Ae(e) {
        Ae = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return Ae(e);
    }
    function He(e, t, r) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: r,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = r;
        }
        return e;
    }
    var Be = {
        children: i().node,
        closeOnClickAway: i().bool,
        divider: i().oneOf([ "header", "footer", "both", "none" ]),
        elementRef: i().oneOfType([ i().func, i().object ]),
        initialFocus: i().oneOfType([ i().object, i().oneOf([ "first", "container" ]) ]),
        onRequestClose: i().func,
        open: i().bool,
        returnFocus: i().oneOfType([ i().func, i().object ])
    };
    var Ne = {
        // TODO: Set to `false` in next major version
        closeOnClickAway: true,
        divider: "both",
        initialFocus: "first",
        open: false
    };
    var ze =  function(e) {
        Te(o, e);
        var t = qe(o);
        // @docs-props-type ModalPropsBase
        // this timeout ensures that the Modal is finished rendering before setting the focus
                function o(e) {
            var i;
            _e(this, o);
            i = t.call(this, e);
            He(De(i), "el", null);
            He(De(i), "initialFocusTimeoutId", undefined);
            He(De(i), "initialFocus", null);
            He(De(i), "headerTitleId", void 0);
            He(De(i), "headerSubtitleId", void 0);
            He(De(i), "getDefaultMotionStyle", (function() {
                return {
                    top: 0
                };
            }));
            He(De(i), "getMotionStyle", (function() {
                if (i.props.open) {
                    return {
                        top: 40
                    };
                }
                return {
                    top: 0
                };
            }));
            He(De(i), "handleFocus", (function(e) {
                var t = i.props, r = t.initialFocus, n = t.open;
                if (n) {
                    i.initialFocusTimeoutId = l()((function() {
                        var e;
 // not using the takeFocus() util here because it
                        // doesn't work for all values of "initialFocus"
                        // and it has a nested defer() call, which we want to avoid
                                                if (i.initialFocus !== r) {
                            if (r === "first") {
                                if (i.el) {
                                    /* eslint-disable prefer-destructuring */
                                    /* prefer-desturucting leads to harder-to-read syntax in this case */
                                    e = (0, m.getSortedTabbableElements)(i.el)[0];
                                    /* eslint-enable prefer-destructuring */                                }
                            } else if (r === "container") {
                                e = i.el;
                            } else if (Ie(r) === "object") {
                                e = r;
                            }
 // Typescript isn't happy with optional chaining here
                                                        if (e != null && "focus" in e) {
                                e.focus();
                            }
                            i.initialFocus = r;
                        }
                    }));
                } else {
                    i.initialFocus = null;
                    clearTimeout(i.initialFocusTimeoutId);
                    if (e) {
                        /**
           * We return focus to the invoking element here instead of in `handleRequestClose` because users can
           * close the modal by updating the value of the `open` prop without calling `onRequestClose`
           */
                        var o = i.props.returnFocus;
                        if (o != null) {
                            if (typeof o === "function") {
                                o();
                            } else {
                                l()((function() {
                                    var e, t;
                                    (e = o.current) === null || e === void 0 ? void 0 : (t = e.focus) === null || t === void 0 ? void 0 : t.call(e);
                                }));
                            }
                        }
                    }
                }
            }));
            He(De(i), "handleModalMount", (function(e) {
                i.el = e;
                ve(i.props.elementRef, e);
            }));
            He(De(i), "handleModalKeyDown", (function(e) {
                if (i.el !== null) {
                    (0, m.handleTab)(i.el, e.nativeEvent);
                }
            }));
            He(De(i), "handleRequestClose", (function(e) {
                var t, r;
                i.initialFocus = null;
                (t = (r = i.props).onRequestClose) === null || t === void 0 ? void 0 : t.call(r, e);
            }));
            He(De(i), "getProviderValue", p()((function(e) {
                var t = e.titleId, r = e.subtitleId, n = e.divider;
                return {
                    titleId: t,
                    subtitleId: r,
                    divider: n
                };
            })));
            He(De(i), "getModalStyles", p()((function(e) {
                var t = e.opacity, r = e.style, n = e.top;
                return Ce(Ce({}, r), {}, {
                    opacity: t,
                    top: n
                });
            })));
            He(De(i), "renderModal", (function(e) {
                var t = i.props, a = t.children, l = t.divider, c = t.style;
                var u = e.top, p = e.opacity;
                var f = r.Children.toArray(a).filter(r.isValidElement);
                
                return n().createElement(me, he({
                    ref: i.handleModalMount,
                    "data-test": "modal",
                    style: i.getModalStyles({
                        opacity: p,
                        style: c,
                        top: u
                    }),
                    tabIndex: -1,
                    onKeyDown: i.handleModalKeyDown,
                    role: "dialog",
                    "aria-labelledby": i.headerTitleId,
                    "aria-describedby": i.headerSubtitleId
                }, d()(i.props, [].concat(ge(s()(o.propTypes)), [ "style" ]))),  n().createElement(z.Provider, {
                    value: i.getProviderValue({
                        titleId: i.headerTitleId,
                        subtitleId: i.headerSubtitleId,
                        divider: l
                    })
                }, f));
            }));
            i.headerTitleId = (0, v.createDOMID)("titleId");
            i.headerSubtitleId = (0, v.createDOMID)("subtitleId");
            return i;
        }
        Pe(o, [ {
            key: "componentDidMount",
            value: function e() {
                this.handleFocus();
            }
        }, {
            key: "componentDidUpdate",
            value: function e(t) {
                var r = this.props.initialFocus;
                if (t.initialFocus !== r) {
                    // cancel the defer() set up in handleInitialFocus()
                    // so that focus won't jump around if "initialFocus" changes
                    clearTimeout(this.initialFocusTimeoutId);
                }
                this.handleFocus(t.open);
            }
        }, {
            key: "render",
            value: function e() {
                var t = this.props.closeOnClickAway;
                var r = t ? "interactive" : "visible";
                
                return n().createElement(b(), {
                    getDefaultMotionStyle: this.getDefaultMotionStyle,
                    getMotionStyle: this.getMotionStyle,
                    onRequestClose: this.handleRequestClose,
                    open: this.props.open,
                    renderModal: this.renderModal,
                    scrim: r
                });
            }
        } ]);
        return o;
    }(r.Component);
    He(ze, "propTypes", Be);
    He(ze, "defaultProps", Ne);
    He(ze, "Header", ye);
    He(ze, "Body", R);
    He(ze, "Footer", V);
    /* harmony default export */    const Le = ze;
    // CONCATENATED MODULE: ./src/Modal/index.ts
    module.exports = t;
    /******/})();
