// @splunk/react-ui/ModalLayer.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var n = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(n, {
                a: n
            });
            /******/            return n;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, n) => {
            /******/ for (var r in n) {
                /******/ if (e.o(n, r) && !e.o(t, r)) {
                    /******/ Object.defineProperty(t, r, {
                        enumerable: true,
                        get: n[r]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        default: () => /* reexport */ E
    });
    // CONCATENATED MODULE: external "react"
    const n = require("react");
    var r =  e.n(n);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var i =  e.n(o);
    // CONCATENATED MODULE: external "@splunk/react-ui/Animation"
    const a = require("@splunk/react-ui/Animation");
    // CONCATENATED MODULE: external "@splunk/react-ui/Layer"
    const c = require("@splunk/react-ui/Layer");
    var l =  e.n(c);
    // CONCATENATED MODULE: external "styled-components"
    const u = require("styled-components");
    var s =  e.n(u);
    // CONCATENATED MODULE: external "react-spring"
    const f = require("react-spring");
    // CONCATENATED MODULE: external "@splunk/themes"
    const p = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/ModalLayer/ModalLayerStyles.ts
    var y = (0, u.css)([ "", ";position:fixed;top:0;right:0;bottom:0;left:0;" ], p.mixins.reset("block"));
    var d = s()(f.animated.div).withConfig({
        displayName: "ModalLayerStyles__StyledClickAwayOverlay",
        componentId: "sc-1sa3n1j-0"
    })([ "", ";background-color:", ";z-index:", ";" ], y, p.variables.backgroundColorScrim, p.variables.zindexModalBackdrop);
    // CONCATENATED MODULE: ./src/ModalLayer/ModalLayer.tsx
    function v(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            v = function e(t) {
                return typeof t;
            };
        } else {
            v = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return v(e);
    }
    function b(e, t) {
        var n = Object.keys(e);
        if (Object.getOwnPropertySymbols) {
            var r = Object.getOwnPropertySymbols(e);
            if (t) r = r.filter((function(t) {
                return Object.getOwnPropertyDescriptor(e, t).enumerable;
            }));
            n.push.apply(n, r);
        }
        return n;
    }
    function m(e) {
        for (var t = 1; t < arguments.length; t++) {
            var n = arguments[t] != null ? arguments[t] : {};
            if (t % 2) {
                b(Object(n), true).forEach((function(t) {
                    R(e, t, n[t]);
                }));
            } else if (Object.getOwnPropertyDescriptors) {
                Object.defineProperties(e, Object.getOwnPropertyDescriptors(n));
            } else {
                b(Object(n)).forEach((function(t) {
                    Object.defineProperty(e, t, Object.getOwnPropertyDescriptor(n, t));
                }));
            }
        }
        return e;
    }
    function O(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function h(e, t) {
        for (var n = 0; n < t.length; n++) {
            var r = t[n];
            r.enumerable = r.enumerable || false;
            r.configurable = true;
            if ("value" in r) r.writable = true;
            Object.defineProperty(e, r.key, r);
        }
    }
    function g(e, t, n) {
        if (t) h(e.prototype, t);
        if (n) h(e, n);
        return e;
    }
    function w(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) j(e, t);
    }
    function j(e, t) {
        j = Object.setPrototypeOf || function e(t, n) {
            t.__proto__ = n;
            return t;
        };
        return j(e, t);
    }
    function S(e) {
        var t = A();
        return function n() {
            var r = _(e), o;
            if (t) {
                var i = _(this).constructor;
                o = Reflect.construct(r, arguments, i);
            } else {
                o = r.apply(this, arguments);
            }
            return P(this, o);
        };
    }
    function P(e, t) {
        if (t && (v(t) === "object" || typeof t === "function")) {
            return t;
        }
        return k(e);
    }
    function k(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function A() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function _(e) {
        _ = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return _(e);
    }
    function R(e, t, n) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: n,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = n;
        }
        return e;
    }
    var q = {
        childrenInAnimation: i().bool,
        getDefaultMotionStyle: i().func,
        getMotionStyle: i().func,
        onRequestClose: i().func,
        open: i().bool,
        renderModal: i().func.isRequired,
        scrim: i().oneOf([ "visible", "hidden", "interactive" ])
    };
    var M = {
        childrenInAnimation: false,
        open: false,
        scrim: "interactive"
    };
    var C =  function(e) {
        w(n, e);
        var t = S(n);
        function n(e) {
            var o;
            O(this, n);
            o = t.call(this, e);
            R(k(o), "allowAnimationUpdates", true);
            R(k(o), "handleClickAway", (function(e) {
                var t, n;
                (t = (n = o.props).onRequestClose) === null || t === void 0 ? void 0 : t.call(n, {
                    reason: "clickAway",
                    event: e
                });
            }));
            R(k(o), "handleAnimationEnd", (function() {
                if (o.allowAnimationUpdates) {
                    o.setState({
                        animating: false
                    });
                }
            }));
            R(k(o), "renderLayer", (function() {
                var e = o.props, t = e.getDefaultMotionStyle, n = e.getMotionStyle, i = e.renderModal, c = e.scrim;
                var l = o.props.open || o.state.animating;
                var u = l && c !== "hidden";
                
                return r().createElement(a.Spring, {
                    from: m(m({}, t === null || t === void 0 ? void 0 : t()), {}, {
                        overlayOpacity: 0,
                        opacity: 0
                    }),
                    to: m(m({}, n === null || n === void 0 ? void 0 : n()), {}, {
                        overlayOpacity: o.props.open ? 1 : 0,
                        opacity: o.props.open ? 1 : 0
                    }),
                    onRest: o.handleAnimationEnd
                }, (function(e) {
                    
                    return r().createElement("div", null, 
                    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
                    i(e), u &&  r().createElement(d, {
                        "data-test": "modal-overlay",
                        onMouseDown: c === "interactive" ? o.handleClickAway : undefined,
                        key: "clickAway",
                        style: {
                            /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
                            opacity: e.overlayOpacity
                        }
                    }));
                }));
            }));
            o.state = {
                animating: false,
                prevOpen: e.open
            };
            return o;
        }
        g(n, [ {
            key: "componentWillUnmount",
            value: function e() {
                this.allowAnimationUpdates = false;
            }
        }, {
            key: "render",
            /* eslint-enable jsx-a11y/no-static-element-interactions */
            value: function e() {
                var t = this.props.open || this.state.animating || this.props.childrenInAnimation;
                
                return r().createElement(l(), {
                    open: t,
                    closeReasons: [ "escapeKey" ],
                    onRequestClose: this.props.onRequestClose
                }, t && this.renderLayer());
            }
        } ]);
        return n;
    }(n.Component);
    R(C, "propTypes", q);
    R(C, "defaultProps", M);
    R(C, "getDerivedStateFromProps", (function(e, t) {
        if (e.open !== t.prevOpen) {
            return {
                animating: true,
                prevOpen: e.open
            };
        }
        return null;
    }));
    /* harmony default export */    const E = C;
    // CONCATENATED MODULE: ./src/ModalLayer/index.ts
    module.exports = t;
    /******/})();
