// @splunk/react-ui/Number.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var r = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(r, {
                a: r
            });
            /******/            return r;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        default: () => /* reexport */ re
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    var n =  e.n(r);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var i =  e.n(o);
    // CONCATENATED MODULE: external "lodash/has"
    const a = require("lodash/has");
    var l =  e.n(a);
    // CONCATENATED MODULE: external "lodash/isFinite"
    const s = require("lodash/isFinite");
    var u =  e.n(s);
    // CONCATENATED MODULE: external "lodash/isString"
    const p = require("lodash/isString");
    var c =  e.n(p);
    // CONCATENATED MODULE: external "lodash/omit"
    const d = require("lodash/omit");
    var f =  e.n(d);
    // CONCATENATED MODULE: external "@splunk/themes"
    const v = require("@splunk/themes");
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const m = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "@splunk/ui-utils/keyboard"
    const h = require("@splunk/ui-utils/keyboard");
    // CONCATENATED MODULE: external "@splunk/react-icons/SVGEnterprise"
    const b = require("@splunk/react-icons/SVGEnterprise");
    var y =  e.n(b);
    // CONCATENATED MODULE: ./src/Number/IncrementIcon.tsx
    function g() {
        g = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return g.apply(this, arguments);
    }
    /* eslint-disable max-len */    function k(e) {
        
        return n().createElement(y(), g({
            viewBox: "0 0 933 600"
        }, e),  n().createElement("path", {
            fill: "currentColor",
            d: "M0 466.438L466.438 0l466.44 467.466-134.59 133.56-331.85-331.848-331.85 331.85L0 466.438z"
        }));
    }
    // CONCATENATED MODULE: external "styled-components"
    const S = require("styled-components");
    var x =  e.n(S);
    // CONCATENATED MODULE: external "@splunk/react-ui/Button"
    const O = require("@splunk/react-ui/Button");
    var w =  e.n(O);
    // CONCATENATED MODULE: external "@splunk/react-ui/Clickable"
    const C = require("@splunk/react-ui/Clickable");
    var j =  e.n(C);
    // CONCATENATED MODULE: external "@splunk/react-ui/Text"
    const P = require("@splunk/react-ui/Text");
    var V =  e.n(P);
    // CONCATENATED MODULE: ./src/Number/NumberStyles.ts
    var T = x()(V()).withConfig({
        displayName: "NumberStyles__StyledText",
        componentId: "sc-2ekj9s-0"
    })([ "&[data-inline]{", "}" ], (0, v.pick)({
        enterprise: {
            comfortable: (0, S.css)([ "width:100px;" ]),
            compact: (0, S.css)([ "width:80px;" ])
        },
        prisma: (0, S.css)([ "width:100px;" ])
    }));
    var _ = x().div.withConfig({
        displayName: "NumberStyles__StyledControls",
        componentId: "sc-2ekj9s-1"
    })([ "width:", ";height:100%;", " ", "" ], (0, v.pick)({
        enterprise: {
            comfortable: "20px",
            compact: "16px"
        },
        prisma: {
            comfortable: "24px",
            compact: "24px"
        }
    }), (0, v.pick)({
        enterprise: (0, S.css)([ "transform:translateY(-1px);" ]),
        prisma: (0, S.css)([ "transform:translateX(1px);" ])
    }), (function(e) {
        var t = e.$isAppend;
        return t ? (0, S.css)([ "margin-right:8px;", "" ], (0, v.pick)({
            prisma: (0, S.css)([ "transform:translateX(-1px);" ])
        })) : (0, S.css)([ "margin-left:8px;" ]);
    }));
    var E = (0, S.css)([ "position:relative;padding:3px;font-size:inherit;min-width:0;min-height:0;height:", ";width:100%;" ], (0, 
    v.pick)({
        enterprise: "calc(50% + 1.5px)",
        prisma: "50%"
    }));
    var N = x()(w()).withConfig({
        displayName: "NumberStyles__StyledPlusButton",
        componentId: "sc-2ekj9s-2"
    })([ "", ";top:0;border-bottom-left-radius:0;border-bottom-right-radius:0;", " &:not([disabled]){&,&:hover{", ";}}" ], E, (function(e) {
        var t = e.append;
        return t && (0, v.pick)({
            enterprise: {
                light: (0, S.css)([ "border-right:1px solid ", ";" ], v.variables.gray60),
                dark: (0, S.css)([ "border-right:", ";" ], v.variables.border)
            }
        });
    }), (0, v.pick)({
        enterprise: {
            dark: (0, S.css)([ "border-color:", ";" ], v.variables.gray20)
        }
    }));
    var D = x()(w()).withConfig({
        displayName: "NumberStyles__StyledMinusButton",
        componentId: "sc-2ekj9s-3"
    })([ "", ";bottom:1px;border-top-left-radius:0;border-top-right-radius:0;", " &:not([disabled]){&,&:hover{", ";}}" ], E, (function(e) {
        var t = e.append;
        return t && (0, v.pick)({
            enterprise: {
                light: (0, S.css)([ "border-right:1px solid ", ";" ], v.variables.gray60),
                dark: (0, S.css)([ "border-right:", ";" ], v.variables.border)
            }
        });
    }), (0, v.pick)({
        enterprise: {
            dark: (0, S.css)([ "border-color:", ";" ], v.variables.gray20)
        }
    }));
    var B = x()(j()).withConfig({
        displayName: "NumberStyles__prismaButton",
        componentId: "sc-2ekj9s-4"
    })([ "", ";display:flex;align-items:center;justify-content:center;", " &:hover:not([disabled]){", "}&:active:not([disabled]){", "}" ], E, (0, 
    v.pick)({
        prisma: (0, S.css)([ "color:", ";" ], v.variables.contentColorDefault)
    }), (0, v.pick)({
        prisma: (0, S.css)([ "background-color:", ";color:", ";" ], v.variables.interactiveColorOverlayHover, v.variables.contentColorActive)
    }), (0, v.pick)({
        prisma: (0, S.css)([ "background-color:", ";color:", ";" ], v.variables.interactiveColorOverlayActive, v.variables.contentColorActive)
    }));
    var I = x()(B).withConfig({
        displayName: "NumberStyles__StyledPrismaPlusButton",
        componentId: "sc-2ekj9s-5"
    })([ "border-top-right-radius:4px;" ]);
    var q = x()(B).withConfig({
        displayName: "NumberStyles__StyledPrismaMinusButton",
        componentId: "sc-2ekj9s-6"
    })([ "border-bottom-right-radius:4px;padding-bottom:5px;" ]);
    // CONCATENATED MODULE: ./src/Number/Number.tsx
    function K(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            K = function e(t) {
                return typeof t;
            };
        } else {
            K = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return K(e);
    }
    function A() {
        A = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return A.apply(this, arguments);
    }
    function M(e, t) {
        var r = Object.keys(e);
        if (Object.getOwnPropertySymbols) {
            var n = Object.getOwnPropertySymbols(e);
            if (t) n = n.filter((function(t) {
                return Object.getOwnPropertyDescriptor(e, t).enumerable;
            }));
            r.push.apply(r, n);
        }
        return r;
    }
    function R(e) {
        for (var t = 1; t < arguments.length; t++) {
            var r = arguments[t] != null ? arguments[t] : {};
            if (t % 2) {
                M(Object(r), true).forEach((function(t) {
                    Q(e, t, r[t]);
                }));
            } else if (Object.getOwnPropertyDescriptors) {
                Object.defineProperties(e, Object.getOwnPropertyDescriptors(r));
            } else {
                M(Object(r)).forEach((function(t) {
                    Object.defineProperty(e, t, Object.getOwnPropertyDescriptor(r, t));
                }));
            }
        }
        return e;
    }
    function F(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function U(e, t) {
        for (var r = 0; r < t.length; r++) {
            var n = t[r];
            n.enumerable = n.enumerable || false;
            n.configurable = true;
            if ("value" in n) n.writable = true;
            Object.defineProperty(e, n.key, n);
        }
    }
    function z(e, t, r) {
        if (t) U(e.prototype, t);
        if (r) U(e, r);
        return e;
    }
    function X(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) $(e, t);
    }
    function $(e, t) {
        $ = Object.setPrototypeOf || function e(t, r) {
            t.__proto__ = r;
            return t;
        };
        return $(e, t);
    }
    function L(e) {
        var t = Y();
        return function r() {
            var n = J(e), o;
            if (t) {
                var i = J(this).constructor;
                o = Reflect.construct(n, arguments, i);
            } else {
                o = n.apply(this, arguments);
            }
            return G(this, o);
        };
    }
    function G(e, t) {
        if (t && (K(t) === "object" || typeof t === "function")) {
            return t;
        }
        return H(e);
    }
    function H(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function Y() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function J(e) {
        J = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return J(e);
    }
    function Q(e, t, r) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: r,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = r;
        }
        return e;
    }
    var W = {
        append: i().bool,
        children: i().node,
        defaultValue: i().number,
        describedBy: i().string,
        disabled: i().bool,
        elementRef: i().oneOfType([ i().func, i().object ]),
        inputRef: i().oneOfType([ i().func, i().object ]),
        error: i().bool,
        hideStepButtons: i().bool,
        inline: i().bool,
        inputId: i().string,
        labelledBy: i().string,
        locale: i().string,
        max: i().number,
        min: i().number,
        name: i().string,
        onBlur: i().func,
        onChange: i().func,
        onFocus: i().func,
        onKeyDown: i().func,
        onKeyUp: i().func,
        onSelect: i().func,
        placeholder: i().string,
        prepend: i().bool,
        roundTo: i().number,
        step: i().number,
        /** @private */
        splunkTheme: i().object,
        useSyntheticPlaceholder: i().bool,
        value: i().number
    };
    var Z = {
        append: false,
        disabled: false,
        error: false,
        hideStepButtons: false,
        inline: false,
        locale: "en-US",
        placeholder: "",
        prepend: false,
        roundTo: 5,
        step: 1
    };
    var ee =  function(e) {
        X(r, e);
        var t = L(r);
        z(r, null, [ {
            key: "getDecimalSeparator",
            // @docs-props-type NumberPropsBase
            value: function e(t) {
                // A more robust and expandable solution would be to use formatToParts() once IE requirements are dropped.
                var r = new Intl.NumberFormat(t.replace("_", "-")).format(1.2);
                return r.indexOf(",") > -1 ? "," : ".";
            }
        }, {
            key: "validatePrependAppend",
            value: function e(t, r) {
                if (false) {}
            }
        }, {
            key: "stringToNumber",
            value: function e(t, n) {
                var o = t;
                if (!c()(o) || o === "") {
                    return undefined;
                }
                if (r.getDecimalSeparator(n.locale) === ".") {
                    // remove leading non-numeric characters.
                    o = o.replace(/^[^\d.-]/, "");
                } else {
                    // remove leading non-numeric characters and replace comma with period
                    o = o.replace(/^[^\d,-]/, "").replace(",", ".");
                }
                var i = r.limitValue(parseFloat(o), n);
                if (u()(i)) {
                    return i;
                }
                return undefined;
            }
        }, {
            key: "numberToString",
            value: function e(t, n) {
                if (t === undefined || !u()(t)) {
                    return "";
                }
                if (r.getDecimalSeparator(n.locale) === ".") {
                    return t.toString();
                }
 // comma separated case
                                return t.toString().replace(".", ",");
            }
        }, {
            key: "limitValue",
            value: function e(t, r) {
                var n = r.min, o = r.max, i = r.roundTo;
                var a = t;
                if (!u()(t)) {
                    return a;
                }
                if (n !== undefined && u()(n)) {
                    a = Math.max(n, a);
                }
                if (o !== undefined && u()(o)) {
                    a = Math.min(o, a);
                }
 // eslint-disable-next-line no-restricted-properties
                                var l = Math.pow(10, i);
                return Math.round(a * l) / l;
            }
        } ]);
        function r(e) {
            var n;
            F(this, r);
            n = t.call(this, e);
            Q(H(n), "controlledExternally", void 0);
            Q(H(n), "text", null);
            Q(H(n), "handleInputBlur", (function(e, t) {
                var r, o;
                n.updateString();
                (r = (o = n.props).onBlur) === null || r === void 0 ? void 0 : r.call(o, e, t);
            }));
            Q(H(n), "handleInputChange", (function(e, t) {
                var r = t.value;
                n.setValueFromString(e, r);
            }));
            Q(H(n), "handleKeyDown", (function(e) {
                var t, r;
                if (!e.metaKey && !e.ctrlKey && !e.altKey && (0, h.addsCharacter)(e.nativeEvent) !== false) {
                    // only allow numbers, negative and decimals
                    if (!(0, h.isNumeric)(e.nativeEvent, {
                        locale: n.props.locale
                    })) {
                        e.preventDefault();
                    }
 // if minus is not allowed or there is already one, prevent input
                                        if ((0, h.isMinus)(e.nativeEvent) && (n.preventNegativeNumbers() || n.state.stringValue.indexOf("-") >= 0)) {
                        e.preventDefault();
                    }
 // if decimal is not allowed or there is already one, prevent input
                                        if ((0, h.isDecimal)(e.nativeEvent, {
                        locale: n.props.locale
                    }) && (n.props.roundTo <= 0 || n.state.stringValue.indexOf(n.props.locale) >= 0)) {
                        e.preventDefault();
                    }
                }
 // Arrow up and down will increment
                                if ((0, h.keycode)(e.nativeEvent) === "up") {
                    n.increment(e, n.props.step);
                    e.preventDefault();
                } else if ((0, h.keycode)(e.nativeEvent) === "down") {
                    n.increment(e, -n.props.step);
                    e.preventDefault();
                }
                (t = (r = n.props).onKeyDown) === null || t === void 0 ? void 0 : t.call(r, e);
            }));
            Q(H(n), "handleKeyUp", (function(e) {
                var t, r;
                if ((0, h.keycode)(e.nativeEvent) === "enter") {
                    n.updateString();
                }
                (t = (r = n.props).onKeyUp) === null || t === void 0 ? void 0 : t.call(r, e);
            }));
            Q(H(n), "handleIncrement", (function(e) {
                n.increment(e, n.props.step);
            }));
            Q(H(n), "handleDecrement", (function(e) {
                n.increment(e, -n.props.step);
            }));
            n.controlledExternally = l()(e, "value");
            var o = n.isControlled() ? e.value : e.defaultValue;
            n.state = {
                // value stores internal state, only in uncontrolled mode
                value: n.isControlled() ? undefined : o,
                // prevValueProp is used to keep track of changes to the value prop, only in controlled mode
                prevValueProp: n.isControlled() ? o : undefined,
                // eslint-disable-line react/no-unused-state
                // stringValue stores the current contents of the input text box - it might be out of range, invalid, ...
                stringValue: r.numberToString(o, e)
            };
            if (false) {}
            if (l()(e, "useSyntheticPlaceholder")) {
                // eslint-disable-next-line no-console
                console.warn("The 'Number' prop 'useSyntheticPlaceholder' has been marked for deprecation.");
            }
            if (false) {}
            r.validatePrependAppend(n.props.append, n.props.prepend);
            return n;
        }
        z(r, [ {
            key: "componentDidUpdate",
            value: function e(t) {
                if (false) {}
                if (false) {}
                r.validatePrependAppend(this.props.append, this.props.prepend);
            }
        }, {
            key: "getValue",
            value: function e() {
                return this.isControlled() ? this.props.value : this.state.value;
            }
        }, {
            key: "setValueFromString",
            value: function e(t, n) {
                var o = this;
                var i = r.stringToNumber(n, this.props);
                var a = this.props.name;
                this.setState({
                    value: this.isControlled() ? undefined : i,
                    stringValue: n
                }, (function() {
                    if (o.getValue() !== i || !o.isControlled()) {
                        var e, r;
                        (e = (r = o.props).onChange) === null || e === void 0 ? void 0 : e.call(r, t, {
                            value: i,
                            name: a
                        });
                    }
                }));
            }
        }, {
            key: "setValue",
            value: function e(t, n) {
                var o = this;
                var i = this.props.name;
                this.setState({
                    value: this.isControlled() ? undefined : n,
                    stringValue: r.numberToString(n, this.props)
                }, (function() {
                    if (o.getValue() !== n || !o.isControlled()) {
                        var e, r;
                        (e = (r = o.props).onChange) === null || e === void 0 ? void 0 : e.call(r, t, {
                            value: n,
                            name: i
                        });
                    }
                }));
            }
        }, {
            key: "focus",
            /**
     * Place focus on the input.
     */
            value: function e(t) {
                if (this.text) {
                    this.text.focus(t);
                }
            }
        }, {
            key: "isControlled",
            value: function e() {
                return this.controlledExternally;
            }
        }, {
            key: "updateString",
            value: function e() {
                var t = this.getValue();
                this.setState({
                    stringValue: r.numberToString(t, this.props)
                });
            }
        }, {
            key: "increment",
            value: function e(t, n) {
                var o = r.stringToNumber(this.state.stringValue, this.props);
                var i;
 // increment appropriate number
                                if (o !== undefined && u()(o)) {
                    // increment the entered value
                    i = o + n;
                } else if (this.props.defaultValue !== undefined && u()(this.props.defaultValue)) {
                    // increment the defaultValue
                    i = this.props.defaultValue + n;
                } else if (n > 0) {
                    // increment up to max or step value
                    i = this.props.max || this.props.step;
                } else {
                    // increment down to min or 0
                    i = this.props.min || 0;
 // set to min or 0
                                }
                this.setValue(t, r.limitValue(i, this.props));
            }
        }, {
            key: "preventNegativeNumbers",
            value: function e() {
                return this.props.min !== undefined && u()(this.props.min) && this.props.min >= 0;
            }
        }, {
            key: "render",
            value: function e() {
                var t = this;
                var r = this.props, o = r.append, i = r.describedBy, a = r.disabled, l = r.error, s = r.hideStepButtons, p = r.inline, c = r.value, d = c === void 0 ? this.state.value : c, v = r.max, h = r.min, b = r.prepend, y = r.splunkTheme;
                var g = f()(this.props, [ "defaultValue", "hideStepButtons", "max", "min", "roundTo", "spellCheck", "step" ]);
                var S = y.isPrisma;
                var x = S ? {} : {
                    append: o,
                    prepend: !o,
                    error: l,
                    inline: false
                };
                var O = d !== undefined && u()(d) && v !== undefined && u()(v) && d >= v;
                var w = d !== undefined && u()(d) && h !== undefined && u()(h) && d <= h;
                var C =  n().createElement(k, {
                    width: "6px",
                    height: "4px",
                    screenReaderText: (0, m._)("Increment")
                });
                var j = R({
                    "data-test": "increment",
                    disabled: a || O,
                    onClick: this.handleIncrement,
                    tabIndex: -1
                }, x);
                var P = S ?  n().createElement(I, j, C) :  n().createElement(N, j, C);
                var V =  n().createElement(k, {
                    width: "6px",
                    height: "4px",
                    screenReaderText: (0, m._)("Decrement"),
                    style: {
                        transform: "rotateX(180deg)"
                    }
                });
                var E = R({
                    "data-test": "decrement",
                    disabled: a || w,
                    onClick: this.handleDecrement,
                    tabIndex: -1
                }, x);
                var B = S ?  n().createElement(q, E, V) :  n().createElement(D, E, V);
                var K = o && !b;
                var M = (!a || S) && K && !s &&  n().createElement(_, {
                    $isAppend: K
                }, P, B);
                var F = (!a || S) && !K && !s &&  n().createElement(_, {
                    $isAppend: K
                }, P, B);
                
                return n().createElement(T, A({
                    autoCapitalize: "off",
                    autoComplete: "off",
                    autoCorrect: "off",
                    spellCheck: false,
                    "data-test": "number",
                    ref: function e(r) {
                        t.text = r;
                    }
                }, g, {
                    inline: p,
                    onChange: this.handleInputChange,
                    onBlur: this.handleInputBlur,
                    onKeyDown: this.handleKeyDown,
                    onKeyUp: this.handleKeyUp,
                    value: this.state.stringValue,
                    "aria-describedby": i,
                    error: l,
                    "data-test-value": d,
                    startAdornment: M,
                    endAdornment: F
                }));
            }
        } ]);
        return r;
    }(r.Component);
    Q(ee, "propTypes", W);
    Q(ee, "defaultProps", Z);
    Q(ee, "displayName", "Number");
    Q(ee, "getDerivedStateFromProps", (function(e, t) {
        if (l()(e, "value") && e.value !== t.prevValueProp) {
            var r = ee.stringToNumber(t.stringValue, e);
            return e.value !== r ? {
                prevValueProp: e.value,
                stringValue: ee.numberToString(e.value, e)
            } : {
                prevValueProp: e.value
            };
        }
        return null;
    }));
    var te = (0, v.withSplunkTheme)(ee);
    te.propTypes = ee.propTypes;
    /* harmony default export */    const re = te;
    // CONCATENATED MODULE: ./src/Number/index.ts
    module.exports = t;
    /******/})();
