// @splunk/react-ui/Paginator.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var r = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(r, {
                a: r
            });
            /******/            return r;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        Compact: () => /* reexport */ W,
        PageControl: () => /* reexport */ oe,
        default: () => /* reexport */ Ce
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    var n =  e.n(r);
    // CONCATENATED MODULE: external "prop-types"
    const a = require("prop-types");
    var l =  e.n(a);
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const i = require("@splunk/react-ui/Box");
    var o =  e.n(i);
    // CONCATENATED MODULE: external "@splunk/ui-utils/format"
    const c = require("@splunk/ui-utils/format");
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const s = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/ChevronLeft"
    const u = require("@splunk/react-icons/enterprise/ChevronLeft");
    var p =  e.n(u);
    // CONCATENATED MODULE: external "@splunk/react-icons/ChevronLeft"
    const f = require("@splunk/react-icons/ChevronLeft");
    var d =  e.n(f);
    // CONCATENATED MODULE: external "@splunk/themes"
    const v = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/Paginator/icons/ChevronLeft.tsx
    function g(e) {
        var t = e.className;
        var r = (0, v.useSplunkTheme)(), a = r.isCompact, l = r.isEnterprise;
        var i = a ? "20px" : "24px";
        return l ?  n().createElement(p(), {
            className: t,
            hideDefaultTooltip: true,
            screenReaderText: null
        }) :  n().createElement(d(), {
            className: t,
            height: i,
            width: i
        });
    }
    /* harmony default export */    const m = g;
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/ChevronRight"
    const b = require("@splunk/react-icons/enterprise/ChevronRight");
    var y =  e.n(b);
    // CONCATENATED MODULE: external "@splunk/react-icons/ChevronRight"
    const h = require("@splunk/react-icons/ChevronRight");
    var O =  e.n(h);
    // CONCATENATED MODULE: ./src/Paginator/icons/ChevronRight.tsx
    function k(e) {
        var t = e.className;
        var r = (0, v.useSplunkTheme)(), a = r.isCompact, l = r.isEnterprise;
        var i = a ? "20px" : "24px";
        return l ?  n().createElement(y(), {
            className: t,
            hideDefaultTooltip: true,
            screenReaderText: null
        }) :  n().createElement(O(), {
            className: t,
            height: i,
            width: i
        });
    }
    /* harmony default export */    const C = k;
    // CONCATENATED MODULE: external "styled-components"
    const x = require("styled-components");
    var P =  e.n(x);
    // CONCATENATED MODULE: external "@splunk/react-ui/ButtonSimple"
    const S = require("@splunk/react-ui/ButtonSimple");
    var w =  e.n(S);
    // CONCATENATED MODULE: ./src/Paginator/ButtonStyles.ts
    var E = P()(w()).withConfig({
        displayName: "ButtonStyles__StyledButtonSimple",
        componentId: "sc-1neztq7-0"
    })([ "flex-shrink:0;font-weight:", ";min-width:", ";min-height:", ";text-align:center;&:not([disabled],[aria-disabled='true']){", ";", ";}", "" ], v.variables.fontWeightNormal, v.variables.inputHeight, v.variables.inputHeight, (0, 
    v.pick)({
        enterprise: {
            // TODO: we can remove this pick once SUI-6160 fixes contentColorDefault in Enterprise
            // we're intentionally only overriding dark here, because light has enough contrast as-is
            // and changing only dark lets us avoid making an unnecessary change to Enterprise light in SUI 4
            dark: (0, x.css)([ "color:", ";" ], v.variables.textGray)
        },
        prisma: (0, x.css)([ "color:", ";" ], v.variables.contentColorDefault)
    }), (function(e) {
        var t = e.$current;
        return t && (0, x.css)([ "", ";", ";" ], v.mixins.typography({
            color: "active",
            weight: "bold"
        }), (0, v.pick)({
            // TODO: maintain this separation until SUI 5, where we unify on Prisma's approach
            enterprise: (0, x.css)([ "border-color:", ";" ], v.variables.accentColor),
            prisma: (0, x.css)([ "background-color:", ";" ], v.variables.interactiveColorOverlaySelected)
        }));
    }), (0, v.pick)({
        enterprise: {
            comfortable: (0, x.css)([ "padding:", ";" ], v.variables.spacingQuarter),
            compact: (0, x.css)([ "padding:3px;" ])
        },
        prisma: (0, x.css)([ "padding:", ";& + &{margin-left:8px;}" ], (0, v.pick)({
            comfortable: "5px 8px",
            compact: "4px"
        }))
    }));
    // CONCATENATED MODULE: ./src/Paginator/Button.tsx
    function _() {
        _ = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return _.apply(this, arguments);
    }
    function j(e, t) {
        if (e == null) return {};
        var r = N(e, t);
        var n, a;
        if (Object.getOwnPropertySymbols) {
            var l = Object.getOwnPropertySymbols(e);
            for (a = 0; a < l.length; a++) {
                n = l[a];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function N(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var a, l;
        for (l = 0; l < n.length; l++) {
            a = n[l];
            if (t.indexOf(a) >= 0) continue;
            r[a] = e[a];
        }
        return r;
    }
    var I = {
        children: l().node,
        disabled: l().oneOfType([ l().bool, l().oneOf([ "dimmed" ]) ]),
        onClick: l().func,
        page: l().number,
        /** @private */
        selected: l().bool
    };
    function T(e) {
        var t = e.children, a = e.disabled, l = e.onClick, i = e.page, o = e.selected, c = j(e, [ "children", "disabled", "onClick", "page", "selected" ]);
        var s = (0, r.useCallback)((function(e) {
            if (i != null) {
                l === null || l === void 0 ? void 0 : l(e, {
                    page: i
                });
            } else {
                l === null || l === void 0 ? void 0 : l(e, {});
            }
        }), [ l, i ]);
        
        return n().createElement(E, _({
            $current: o || false,
            "data-test": "page",
            "data-test-page": i,
            appearance: "pill",
            disabled: a,
            inline: false,
            onClick: s
        }, c), t);
    }
    T.propTypes = I;
    /* harmony default export */    const q = T;
    // CONCATENATED MODULE: ./src/Paginator/CompactStyles.ts
    var L = P()(o()).withConfig({
        displayName: "CompactStyles__StyledBox",
        componentId: "dwgzkj-0"
    })([ "flex:0 1 auto;flex-wrap:nowrap;margin-left:auto;" ]);
    var M = P().span.withConfig({
        displayName: "CompactStyles__StyledLabel",
        componentId: "dwgzkj-1"
    })([ "align-items:center;display:flex;flex-wrap:wrap;margin-right:", ";" ], v.variables.spacingSmall);
    // CONCATENATED MODULE: ./src/Paginator/Compact.tsx
    function R() {
        R = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return R.apply(this, arguments);
    }
    function D(e, t) {
        if (e == null) return {};
        var r = z(e, t);
        var n, a;
        if (Object.getOwnPropertySymbols) {
            var l = Object.getOwnPropertySymbols(e);
            for (a = 0; a < l.length; a++) {
                n = l[a];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function z(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var a, l;
        for (l = 0; l < n.length; l++) {
            a = n[l];
            if (t.indexOf(a) >= 0) continue;
            r[a] = e[a];
        }
        return r;
    }
    /** @public */    var G = {
        current: l().number,
        elementRef: l().oneOfType([ l().func, l().object ]),
        onChange: l().func,
        renderLabel: l().oneOfType([ l().bool, l().func ]),
        totalPages: l().number
    };
    var B = function e(t) {
        var r = t.current, n = t.totalPages;
        var a = null;
        if (r != null && n != null) {
            a = (0, c.sprintf)((0, s._)("%d of %d pages"), r, n);
        } else if (r != null) {
            a = (0, c.sprintf)((0, s._)("Page %d"), r);
        }
        return a;
    };
    function H(e) {
        var t = e.onChange, a = e.current, l = e.renderLabel, i = e.totalPages, o = D(e, [ "onChange", "current", "renderLabel", "totalPages" ]);
        // @docs-props-type PaginatorCompactPropsBase
                var c = (0, r.useCallback)((function(e, r) {
            var n = r.page;
            t === null || t === void 0 ? void 0 : t(e, {
                direction: "prev",
                page: n
            });
        }), [ t ]);
        var u = (0, r.useCallback)((function(e, r) {
            var n = r.page;
            t === null || t === void 0 ? void 0 : t(e, {
                direction: "next",
                page: n
            });
        }), [ t ]);
        var p = null;
        if (l) {
            var f = typeof l === "function" ? l : B;
            p = f({
                current: a,
                totalPages: i
            });
        }
        
        return n().createElement(L, R({
            "aria-label": (0, s._)("Pagination"),
            "data-test": "paginator",
            "data-test-current": a,
            flex: true,
            inline: true,
            tag: "nav"
        }, o), p &&  n().createElement(M, {
            "data-test": "label"
        }, p),  n().createElement(q, {
            "aria-label": (0, s._)("Go to previous page"),
            "data-test": "prev",
            disabled: a != null && a <= 1 ? "dimmed" : false,
            onClick: c,
            page: a != null ? a - 1 : undefined
        },  n().createElement(m, null)),  n().createElement(q, {
            "aria-label": (0, s._)("Go to next page"),
            "data-test": "next",
            disabled: a != null && i != null && a >= i ? "dimmed" : false,
            onClick: u,
            page: a != null ? a + 1 : undefined
        },  n().createElement(C, null)));
    }
    H.propTypes = G;
    /* harmony default export */    const W = H;
    // CONCATENATED MODULE: external "@splunk/react-ui/Select"
    const $ = require("@splunk/react-ui/Select");
    var A =  e.n($);
    // CONCATENATED MODULE: external "@splunk/react-ui/ScreenReaderContent"
    const Q = require("@splunk/react-ui/ScreenReaderContent");
    var V =  e.n(Q);
    // CONCATENATED MODULE: external "@splunk/ui-utils/id"
    const F = require("@splunk/ui-utils/id");
    // CONCATENATED MODULE: ./src/Paginator/PageSelectStyles.ts
    var J = P().span.withConfig({
        displayName: "PageSelectStyles__StyledWrapper",
        componentId: "haycz4-0"
    })([ "display:flex;" ]);
    var K = P().span.withConfig({
        displayName: "PageSelectStyles__StyledCount",
        componentId: "haycz4-1"
    })([ "align-items:center;display:flex;flex-wrap:wrap;margin:0 ", ";" ], v.variables.spacingSmall);
    var U = P()(A()).withConfig({
        displayName: "PageSelectStyles__StyledSelect",
        componentId: "haycz4-2"
    })([ "flex:0 0 auto;" ]);
    // CONCATENATED MODULE: ./src/Paginator/PageSelect.tsx
    function X(e, t) {
        if (e == null) return {};
        var r = Y(e, t);
        var n, a;
        if (Object.getOwnPropertySymbols) {
            var l = Object.getOwnPropertySymbols(e);
            for (a = 0; a < l.length; a++) {
                n = l[a];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function Y(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var a, l;
        for (l = 0; l < n.length; l++) {
            a = n[l];
            if (t.indexOf(a) >= 0) continue;
            r[a] = e[a];
        }
        return r;
    }
    /** @public */    var Z = {
        onChange: l().func,
        current: l().number,
        totalPages: l().number
    };
    function ee(e) {
        var t = e.onChange, a = e.current, l = a === void 0 ? 1 : a, i = e.totalPages, o = X(e, [ "onChange", "current", "totalPages" ]);
        // @docs-props-type PaginatorPageSelectPropsBase
                var u = (0, r.useRef)((0, F.createDOMID)("pages"));
        var p = (0, r.useRef)((0, F.createDOMID)("totalPages"));
        var f = function e(r, n) {
            var a = n.value;
            t === null || t === void 0 ? void 0 : t(r, {
                page: a
            });
        };
        var d = [];
        for (var v = 1; v <= i; v += 1) {
            d.push( n().createElement(A().Option, {
                label: "".concat(v),
                key: "".concat(v),
                value: v
            }));
        }
        var g = (0, c.sprintf)((0, s._)("of %d pages"), i);
        
        return n().createElement(J, o,  n().createElement(U, {
            value: l,
            onChange: f,
            "aria-labelledby": "".concat(u.current, " ").concat(p.current)
        }, d),  n().createElement(V(), {
            id: u.current,
            "aria-hidden": true
        }, l),  n().createElement(K, {
            id: p.current,
            "aria-hidden": true
        }, " ", g));
    }
    ee.propTypes = Z;
    /* harmony default export */    const te = ee;
    // CONCATENATED MODULE: ./src/Paginator/PageControl.tsx
    function re() {
        re = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return re.apply(this, arguments);
    }
    function ne(e, t) {
        if (e == null) return {};
        var r = ae(e, t);
        var n, a;
        if (Object.getOwnPropertySymbols) {
            var l = Object.getOwnPropertySymbols(e);
            for (a = 0; a < l.length; a++) {
                n = l[a];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function ae(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var a, l;
        for (l = 0; l < n.length; l++) {
            a = n[l];
            if (t.indexOf(a) >= 0) continue;
            r[a] = e[a];
        }
        return r;
    }
    /** @public */    var le = {
        elementRef: l().oneOfType([ l().func, l().object ]),
        onChange: l().func,
        current: l().number,
        totalPages: l().number
    };
    function ie(e) {
        var t = e.onChange, a = e.current, l = a === void 0 ? 1 : a, i = e.totalPages, o = ne(e, [ "onChange", "current", "totalPages" ]);
        // @docs-props-type PaginatorPageControlPropsBase
                var c = (0, r.useCallback)((function(e, r) {
            var n = r.page;
            // page will always be defined because we're passing "current" to Compact
            // getting TypeScript to understand this seems more trouble than it's worth
            // so we'll validate this behavior through tests
                        t === null || t === void 0 ? void 0 : t(e, {
                page: n
            });
        }), [ t ]);
        var s = (0, r.useCallback)((function(e) {
            var r = e.current, a = e.totalPages;
            if (r != null && a != null) {
                
                return n().createElement(te, {
                    current: r,
                    onChange: t,
                    totalPages: a
                });
            }
            return null;
        }), [ t ]);
        
        return n().createElement(W, re({
            current: l,
            renderLabel: s,
            onChange: c,
            totalPages: i
        }, o));
    }
    ie.propTypes = le;
    /* harmony default export */    const oe = ie;
    // CONCATENATED MODULE: external "@splunk/react-icons/DotsThreeHorizontal"
    const ce = require("@splunk/react-icons/DotsThreeHorizontal");
    var se =  e.n(ce);
    // CONCATENATED MODULE: ./src/Paginator/PageSeparatorStyles.ts
    var ue = P().span.withConfig({
        displayName: "PageSeparatorStyles__StyledEllipsisWrapper",
        componentId: "j50fqf-0"
    })([ "display:flex;align-items:center;margin:", ";color:", ";" ], (0, v.pick)({
        enterprise: "0px 11px",
        prisma: "0px 20px"
    }), v.variables.contentColorActive);
    // CONCATENATED MODULE: ./src/Paginator/PageSeparator.tsx
    function pe() {
        var e = (0, v.useSplunkTheme)(), t = e.family;
        var r = t === "enterprise" ?  n().createElement("span", {
            role: "separator"
        }, "…") :  n().createElement(se(), null);
        
        return n().createElement(ue, {
            role: "separator"
        }, r);
    }
    /* harmony default export */    const fe = pe;
    // CONCATENATED MODULE: ./src/Paginator/PaginatorStyles.ts
    var de = P()(m).withConfig({
        displayName: "PaginatorStyles__StyledChevronLeftIcon",
        componentId: "pmabsp-0"
    })([ "margin:", ";" ], (0, v.pick)({
        enterprise: "-1px 3px 0 0",
        prisma: "-2px 6px 0 0"
    }));
    var ve = P()(C).withConfig({
        displayName: "PaginatorStyles__StyledChevronRightIcon",
        componentId: "pmabsp-1"
    })([ "margin:", ";" ], (0, v.pick)({
        enterprise: "-1px 0 0 3px",
        prisma: "-2px 0 0 6px"
    }));
    var ge = P().div.withConfig({
        displayName: "PaginatorStyles__StyledPrevNext",
        componentId: "pmabsp-2"
    })([ "padding:", ";white-space:nowrap;" ], (0, v.pick)({
        enterprise: "0 3px",
        prisma: "0 8px"
    }));
    var me = P()(q).withConfig({
        displayName: "PaginatorStyles__StyledPaginatorButton",
        componentId: "pmabsp-3"
    })([ "&:not([disabled]){", ",", "{", "}}" ], 
    /* sc-sel */
    de, 
    /* sc-sel */
    ve, (0, v.pick)({
        prisma: (0, x.css)([ "color:", ";" ], v.variables.contentColorMuted)
    }));
    // CONCATENATED MODULE: ./src/Paginator/Paginator.tsx
    function be() {
        be = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return be.apply(this, arguments);
    }
    function ye(e, t) {
        if (e == null) return {};
        var r = he(e, t);
        var n, a;
        if (Object.getOwnPropertySymbols) {
            var l = Object.getOwnPropertySymbols(e);
            for (a = 0; a < l.length; a++) {
                n = l[a];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function he(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var a, l;
        for (l = 0; l < n.length; l++) {
            a = n[l];
            if (t.indexOf(a) >= 0) continue;
            r[a] = e[a];
        }
        return r;
    }
    var Oe = {
        alwaysShowLastPageLink: l().bool,
        current: l().number,
        elementRef: l().oneOfType([ l().func, l().object ]),
        numPageLinks: l().number,
        onChange: l().func,
        totalPages: l().number
    };
    function ke(e) {
        var t = e.onChange, a = e.current, l = a === void 0 ? 1 : a, i = e.alwaysShowLastPageLink, u = i === void 0 ? false : i, p = e.numPageLinks, f = p === void 0 ? 5 : p, d = e.totalPages, v = ye(e, [ "onChange", "current", "alwaysShowLastPageLink", "numPageLinks", "totalPages" ]);
        // @docs-props-type PaginatorPropsBase
                var g = (0, r.useCallback)((function(e, r) {
            var n = r.page;
            // page will always be defined because we're passing "page" to PaginatorButton
            // getting TypeScript to understand this seems more trouble than it's worth
            // so we'll validate this behavior through tests
                        t === null || t === void 0 ? void 0 : t(e, {
                page: n
            });
        }), [ t ]);
 // Can't show more links than total number of pages.
                var m = Math.min(f, d);
        var b = Math.ceil(m / 2);
        var y = Math.ceil(d - m / 2);
        var h = [];
        if (d <= 1) {
            return null;
        }
        if ((l > d || l < 1) && process.env.NODE_ENV !== "production") {}
        var O;
        var k;
        if (l <= b + 1) {
            // + 1 to avoid ellipsis between 1 and 2
            O = 1;
            k = O + Math.min(d, m) - 1;
        } else if (l > b && l < y) {
            k = Math.ceil(l + (m - 2) / 2);
            O = k - m + 2;
        } else {
            // include page #1 to avoid ellipsis between 1 and 2
            O = d - m === 1 ? 1 : d - m + 1;
            k = d;
        }
 // bump up the last page to avoid ellipsis
                k = u && k + 1 === d ? d : k;
        h.push( n().createElement(me, {
            "data-test": "prev",
            disabled: l === 1,
            onClick: g,
            page: l - 1,
            key: "prev",
            "aria-label": (0, s._)("Go to previous page")
        },  n().createElement(ge, null,  n().createElement(de, null), (0, s._)("Prev"))));
        if (l > b && d > m && O !== 1) {
            h.push( n().createElement(me, {
                key: "first",
                onClick: g,
                page: 1,
                "aria-label": (0, s._)("Go to first page")
            }, "1"));
            h.push( n().createElement(fe, {
                "data-test": "page",
                key: "prevEllipsis"
            }));
        }
        for (var C = O; C <= k; C += 1) {
            var x = C === l;
            h.push( n().createElement(me, {
                selected: x,
                onClick: g,
                page: C,
                key: C,
                "aria-current": x ? "page" : false,
                "aria-label": (0, c.sprintf)((0, s._)("Page %(pageNumber)d"), {
                    pageNumber: C
                })
            }, String(C)));
        }
        if (l <= y && d > m && k !== d) {
            h.push( n().createElement(fe, {
                "data-test": "page",
                key: "nextEllipsis"
            }));
            if (u) {
                h.push( n().createElement(me, {
                    "data-test": "last",
                    key: "last",
                    onClick: g,
                    page: d,
                    "aria-label": (0, s._)("Go to last page")
                }, String(d)));
            }
        }
        h.push( n().createElement(me, {
            "data-test": "next",
            disabled: l === d,
            onClick: g,
            page: l + 1,
            key: "next",
            "aria-label": (0, s._)("Go to next page")
        },  n().createElement(ge, null, (0, s._)("Next"),  n().createElement(ve, null))));
        
        return n().createElement(o(), be({
            "aria-label": (0, s._)("Pagination"),
            "data-test": "paginator",
            "data-test-current": l,
            flex: true,
            inline: true,
            tag: "nav"
        }, v), h);
    }
    ke.Compact = W;
    ke.PageControl = oe;
    ke.propTypes = Oe;
    /* harmony default export */    const Ce = ke;
    // CONCATENATED MODULE: ./src/Paginator/index.ts
    module.exports = t;
    /******/})();
