// @splunk/react-ui/Progress.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var t = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(t, {
                a: t
            });
            /******/            return t;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, t) => {
            /******/ for (var a in t) {
                /******/ if (e.o(t, a) && !e.o(r, a)) {
                    /******/ Object.defineProperty(r, a, {
                        enumerable: true,
                        get: t[a]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        default: () => /* reexport */ M
    });
    // CONCATENATED MODULE: external "react"
    const t = require("react");
    var a =  e.n(t);
    // CONCATENATED MODULE: external "prop-types"
    const n = require("prop-types");
    var i =  e.n(n);
    // CONCATENATED MODULE: external "lodash/isNumber"
    const o = require("lodash/isNumber");
    var s =  e.n(o);
    // CONCATENATED MODULE: external "@splunk/react-ui/AnimationToggle"
    const c = require("@splunk/react-ui/AnimationToggle");
    // CONCATENATED MODULE: external "styled-components"
    const l = require("styled-components");
    var u =  e.n(l);
    // CONCATENATED MODULE: external "tinycolor2"
    const p = require("tinycolor2");
    var f =  e.n(p);
    // CONCATENATED MODULE: external "@splunk/react-ui/Tooltip"
    const v = require("@splunk/react-ui/Tooltip");
    var b =  e.n(v);
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const d = require("@splunk/react-ui/Box");
    var m =  e.n(d);
    // CONCATENATED MODULE: external "@splunk/themes"
    const y = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/Progress/ProgressStyles.ts
    var g = u()(m()).withConfig({
        displayName: "ProgressStyles__StyledBox",
        componentId: "csowex-0"
    })([ "background-color:", ";" ], (0, y.pick)({
        enterprise: y.variables.transparent,
        prisma: y.variables.neutral200
    }));
    var h = u()(b()).withConfig({
        displayName: "ProgressStyles__StyledTooltip",
        componentId: "csowex-1"
    })([ "background-color:", ";height:", ";position:relative;overflow:hidden;padding-left:", ";", ";" ], (0, 
    y.pick)({
        enterprise: y.variables.accentColorD10,
        prisma: (0, y.pickVariant)("$type", {
            info: y.variables.interactiveColorPrimary,
            success: y.variables.accentColorPositive,
            warning: y.variables.accentColorWarning,
            error: y.variables.accentColorNegative
        })
    }), (0, y.pick)({
        enterprise: "3px",
        prisma: "4px"
    }), (0, y.pick)({
        enterprise: y.variables.spacingHalf,
        prisma: "10px"
    }), (function(e) {
        var r = e.$animated;
        return r && (0, l.css)([ "transition:width 300ms;" ]);
    }));
    var k = (0, l.keyframes)([ "from{opacity:0;}to{opacity:1;}" ]);
    var C = (0, l.keyframes)([ "from{background-position:130%;}to{background-position:-30%;}" ]);
    var w = u().div.withConfig({
        displayName: "ProgressStyles__StyledPuck",
        componentId: "csowex-2"
    })([ "height:", ";width:", ";position:absolute;right:0;top:0;background:", ";", "" ], (0, 
    y.pick)({
        enterprise: "3px",
        prisma: "4px"
    }), (0, y.pick)({
        enterprise: "300px",
        prisma: "100%"
    }), (0, y.pick)({
        enterprise: (0, l.css)([ "linear-gradient( 90deg,", ",", ",40%,", ",80%,", " )" ], y.variables.accentColorD10, y.variables.accentColorL10, y.variables.accentColorL40, y.variables.accentColorL40),
        prisma: function e(r) {
            var t = r.$animated;
            var a = (0, y.pickVariant)("$type", {
                info: y.variables.interactiveColorPrimary,
                success: y.variables.accentColorPositive,
                warning: y.variables.accentColorWarning,
                error: y.variables.accentColorNegative
            });
            if (t) {
                var n = (0, y.pickVariant)("$type", {
                    info: y.variables.statusColorInfo,
                    success: y.variables.statusColorNormal,
                    warning: y.variables.statusColorMedium,
                    error: y.variables.statusColorHigh
                });
                return (0, l.css)([ "radial-gradient(circle at center,", ",", " 30%);" ], n, a);
            }
            return a;
        }
    }), (function(e) {
        return e.$animated && (0, y.pick)({
            enterprise: (0, l.css)([ "animation-duration:1500ms;animation-name:", ";animation-iteration-count:infinite;animation-direction:alternate;" ], k),
            prisma: function r() {
                var t = (0, y.pickVariant)("$type", {
                    info: y.variables.statusColorInfo,
                    success: y.variables.statusColorNormal,
                    warning: y.variables.statusColorMedium,
                    error: y.variables.statusColorHigh
                })(e);
                return (0, l.css)([ "background-size:200% 100%;animation:", " 1500ms infinite cubic-bezier(0.33,0,0.67,1);&::after{content:'';display:block;width:60px;background:linear-gradient( to right,rgba(255,255,255,0),", " 70%,rgba(255,255,255,0.75) );height:100%;position:absolute;right:0;}" ], C, f()(t(e)).setAlpha(.75).toRgbString());
            }
        });
    }));
    // CONCATENATED MODULE: ./src/Progress/Progress.tsx
    function O() {
        O = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var a in t) {
                    if (Object.prototype.hasOwnProperty.call(t, a)) {
                        e[a] = t[a];
                    }
                }
            }
            return e;
        };
        return O.apply(this, arguments);
    }
    function S(e, r) {
        return _(e) || $(e, r) || j(e, r) || x();
    }
    function x() {
        throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function j(e, r) {
        if (!e) return;
        if (typeof e === "string") return P(e, r);
        var t = Object.prototype.toString.call(e).slice(8, -1);
        if (t === "Object" && e.constructor) t = e.constructor.name;
        if (t === "Map" || t === "Set") return Array.from(e);
        if (t === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t)) return P(e, r);
    }
    function P(e, r) {
        if (r == null || r > e.length) r = e.length;
        for (var t = 0, a = new Array(r); t < r; t++) {
            a[t] = e[t];
        }
        return a;
    }
    function $(e, r) {
        if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(e))) return;
        var t = [];
        var a = true;
        var n = false;
        var i = undefined;
        try {
            for (var o = e[Symbol.iterator](), s; !(a = (s = o.next()).done); a = true) {
                t.push(s.value);
                if (r && t.length === r) break;
            }
        } catch (e) {
            n = true;
            i = e;
        } finally {
            try {
                if (!a && o["return"] != null) o["return"]();
            } finally {
                if (n) throw i;
            }
        }
        return t;
    }
    function _(e) {
        if (Array.isArray(e)) return e;
    }
    function q(e, r) {
        if (e == null) return {};
        var t = A(e, r);
        var a, n;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (n = 0; n < i.length; n++) {
                a = i[n];
                if (r.indexOf(a) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, a)) continue;
                t[a] = e[a];
            }
        }
        return t;
    }
    function A(e, r) {
        if (e == null) return {};
        var t = {};
        var a = Object.keys(e);
        var n, i;
        for (i = 0; i < a.length; i++) {
            n = a[i];
            if (r.indexOf(n) >= 0) continue;
            t[n] = e[n];
        }
        return t;
    }
    var I = {
        elementRef: i().oneOfType([ i().func, i().object ]),
        percentage: i().number,
        tooltip: i().node,
        type: i().oneOf([ "info", "success", "warning", "error" ])
    };
    function T(e) {
        var r = e.percentage, n = e.tooltip, i = e.type, o = i === void 0 ? "info" : i, l = q(e, [ "percentage", "tooltip", "type" ]);
        // @docs-props-type ProgressPropsBase
                if (false) {}
        var u = (0, t.useState)(false), p = S(u, 2), f = p[0], v = p[1];
        var b = (0, c.useAnimationToggle)() === "on";
        var d = (0, t.useCallback)((function() {
            v(true);
        }), []);
        var m = (0, t.useCallback)((function() {
            v(false);
        }), []);
        var y = n || "".concat(r, "%");
        
        return a().createElement(g, O({
            $animated: b,
            "data-test": "progress",
            onMouseEnter: d,
            onMouseLeave: m
        }, l), s()(r) &&  a().createElement(h, {
            inline: false,
            open: f,
            content: y,
            style: {
                width: "".concat(r, "%")
            },
            role: "progressbar",
            "aria-valuenow": r,
            "aria-valuemin": 0,
            "aria-valuemax": 100,
            $animated: b,
            $type: o
        },  a().createElement(w, {
            $animated: b,
            $type: o
        })));
    }
    T.propTypes = I;
    /* harmony default export */    const M = T;
    // CONCATENATED MODULE: ./src/Progress/index.ts
    module.exports = r;
    /******/})();
