// @splunk/react-ui/RadioList.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var a = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(a, {
                a
            });
            /******/            return a;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, a) => {
            /******/ for (var o in a) {
                /******/ if (e.o(a, o) && !e.o(r, o)) {
                    /******/ Object.defineProperty(r, o, {
                        enumerable: true,
                        get: a[o]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        Option: () => /* reexport */ C,
        default: () => /* reexport */ M
    });
    // CONCATENATED MODULE: external "react"
    const a = require("react");
    var o =  e.n(a);
    // CONCATENATED MODULE: external "prop-types"
    const i = require("prop-types");
    var t =  e.n(i);
    // CONCATENATED MODULE: external "lodash/omit"
    const n = require("lodash/omit");
    var l =  e.n(n);
    // CONCATENATED MODULE: external "@splunk/ui-utils/id"
    const s = require("@splunk/ui-utils/id");
    // CONCATENATED MODULE: ./src/RadioList/RadioListContext.ts
    var c =  (0, a.createContext)({});
    c.displayName = "RadioList";
    /* harmony default export */    const d = c;
    // CONCATENATED MODULE: external "styled-components"
    const u = require("styled-components");
    var b =  e.n(u);
    // CONCATENATED MODULE: external "@splunk/themes"
    const p = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/RadioList/OptionStyles.ts
    var v = b().div.withConfig({
        displayName: "OptionStyles__StyledWrapper",
        componentId: "imaqjt-0"
    })([ "position:relative;padding:", ";" ], (0, p.pick)({
        prisma: {
            comfortable: "10px 0",
            compact: "6px 0"
        },
        enterprise: "7px 0 7px 0"
    }));
    var f = b().label.withConfig({
        displayName: "OptionStyles__StyledLabel",
        componentId: "imaqjt-1"
    })([ "", ";", ";position:relative;flex:1 1 auto;padding-left:", ";margin-left:", ";cursor:pointer;&::before,&::after{display:block;content:'';position:absolute;border-radius:50%;}&::before{box-sizing:border-box;top:0;border:", ";", "}&::after{opacity:0;width:8px;height:8px;", " background-color:", ";}", " ", "" ], p.mixins.reset("block"), (0, 
    p.pick)({
        prisma: (0, u.css)([ "color:", ";" ], p.variables.contentColorActive)
    }), (0, p.pick)({
        prisma: "10px",
        enterprise: p.variables.spacingQuarter
    }), (0, p.pick)({
        prisma: "20px",
        enterprise: "18px"
    }), (0, p.pick)({
        prisma: (0, u.css)([ "2px solid ", "" ], p.variables.contentColorMuted),
        enterprise: (0, u.css)([ "1px solid ", "" ], p.variables.gray80)
    }), (0, p.pick)({
        prisma: (0, u.css)([ "width:20px;height:20px;left:-20px;" ]),
        enterprise: (0, u.css)([ "width:18px;height:18px;left:-18px;" ])
    }), (0, p.pick)({
        prisma: (0, u.css)([ "top:6px;left:-14px;" ]),
        enterprise: (0, u.css)([ "top:5px;left:-13px;" ])
    }), (0, p.pick)({
        prisma: p.variables.white,
        enterprise: {
            dark: p.variables.gray80,
            light: p.variables.gray45
        }
    }), (0, p.pick)({
        compact: {
            enterprise: (0, u.css)([ "font-size:", ";" ], p.variables.fontSizeSmall)
        }
    }), (0, p.pick)({
        prisma: (0, u.css)([ "&:hover{&::before{box-shadow:", ";border-color:", ";background-color:", ";}" ], p.variables.hoverShadow, p.variables.contentColorDefault, p.mixins.overlayColors(p.variables.backgroundColorPage, p.variables.interactiveColorOverlayHover))
    }));
    var y = b().input.withConfig({
        displayName: "OptionStyles__StyledRadioInput",
        componentId: "imaqjt-2"
    })([ "opacity:0;position:absolute;top:10px;left:0;&:focus{& + ", "::before{box-shadow:", ";}}&:checked + ", "::after{opacity:1;}&:checked + ", "::before{", "}&:disabled,&:disabled + ", "{color:", ";cursor:not-allowed;}&:disabled + ", "::before{", ";}&:disabled:not(:checked) + ", "::before{", ";}&:disabled + ", "::after{background-color:", ";}", " ", "" ], 
    /* sc-sel */
    f, p.variables.focusShadow, 
    /* sc-sel */
    f, 
    /* sc-sel */
    f, (0, p.pick)({
        prisma: (0, u.css)([ "border-color:transparent;background-color:", ";" ], p.variables.interactiveColorPrimary),
        enterprise: {
            dark: (0, u.css)([ "border-color:", ";" ], p.variables.gray80),
            light: (0, u.css)([ "border-color:", ";" ], p.variables.gray45)
        }
    }), 
    /* sc-sel */
    f, (0, p.pick)({
        prisma: p.variables.contentColorDisabled,
        enterprise: {
            dark: p.variables.gray45,
            light: p.variables.gray80
        }
    }), 
    /* sc-sel */
    f, (0, p.pick)({
        enterprise: {
            dark: (0, u.css)([ "border-color:", ";" ], p.variables.gray45),
            light: (0, u.css)([ "border-color:", ";" ], p.variables.gray80)
        }
    }), 
    /* sc-sel */
    f, (0, p.pick)({
        prisma: (0, u.css)([ "border-color:", ";" ], p.variables.contentColorDisabled)
    }), 
    /* sc-sel */
    f, (0, p.pick)({
        prisma: p.variables.contentColorDisabled,
        enterprise: {
            dark: p.variables.gray45,
            light: p.variables.gray80
        }
    }), (function(e) {
        var r = e.$error;
        return r && (0, u.css)([ "&:not(:disabled) + ", "{color:", ";&::before{border-color:", ";}}&:checked:not(:disabled) + ", "{&::before{border-color:", ";}", "}", "" ], 
        /* sc-sel */
        f, (0, p.pick)({
            prisma: p.variables.accentColorNegative,
            enterprise: p.variables.errorColor
        }), (0, p.pick)({
            prisma: p.variables.accentColorNegative,
            enterprise: {
                dark: p.variables.errorColor,
                light: p.variables.errorColorL30
            }
        }), 
        /* sc-sel */
        f, (0, p.pick)({
            prisma: p.variables.accentColorNegative,
            enterprise: p.variables.errorColor
        }), (0, p.pick)({
            enterprise: (0, u.css)([ "&::after{background-color:", ";}" ], p.variables.errorColor)
        }), (0, p.pick)({
            prisma: (0, u.css)([ "&:not(:disabled):checked + ", "{&::before{background-color:", ";}}" ], 
            /* sc-sel */
            f, p.variables.accentColorNegative)
        }));
    }), (0, p.pick)({
        prisma: (0, u.css)([ "&:active + ", "::before{background-color:", ";}&:active + ", "::after{background-color:", ";}&:checked:active + ", "::before{border-color:transparent;background-color:", ";}&:disabled + ", ":hover::before{box-shadow:none;}&:checked:disabled + ", "{&::before{background-color:", ";}&::after{background-color:rgba(255,255,255,0.3);}}" ], 
        /* sc-sel */
        f, p.mixins.overlayColors(p.variables.backgroundColorPage, p.variables.interactiveColorOverlayActive), 
        /* sc-sel */
        f, p.mixins.overlayColors(p.variables.white, p.variables.interactiveColorOverlayActive), 
        /* sc-sel */
        f, p.mixins.overlayColors(p.variables.interactiveColorPrimary, p.variables.interactiveColorOverlayActive), 
        /* sc-sel */
        f, 
        /* sc-sel */
        f, p.variables.interactiveColorBackgroundDisabled)
    }));
    // CONCATENATED MODULE: ./src/RadioList/Option.tsx
    function g() {
        g = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var a = arguments[r];
                for (var o in a) {
                    if (Object.prototype.hasOwnProperty.call(a, o)) {
                        e[o] = a[o];
                    }
                }
            }
            return e;
        };
        return g.apply(this, arguments);
    }
    function m(e, r) {
        if (e == null) return {};
        var a = h(e, r);
        var o, i;
        if (Object.getOwnPropertySymbols) {
            var t = Object.getOwnPropertySymbols(e);
            for (i = 0; i < t.length; i++) {
                o = t[i];
                if (r.indexOf(o) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, o)) continue;
                a[o] = e[o];
            }
        }
        return a;
    }
    function h(e, r) {
        if (e == null) return {};
        var a = {};
        var o = Object.keys(e);
        var i, t;
        for (t = 0; t < o.length; t++) {
            i = o[t];
            if (r.indexOf(i) >= 0) continue;
            a[i] = e[i];
        }
        return a;
    }
    var k = {
        children: t().node,
        disabled: t().bool,
        error: t().bool,
        id: t().string,
        onChange: t().func,
        value: t().any.isRequired
    };
    function x(e) {
        var r = e.children, i = e.disabled, t = i === void 0 ? false : i, n = e.error, l = n === void 0 ? false : n, c = e.id, u = e.onChange, b = e.value, p = m(e, [ "children", "disabled", "error", "id", "onChange", "value" ]);
        // @docs-props-type OptionPropsBase
                var h = (0, a.useRef)(c || (0, s.createDOMID)("radio-option")), k = h.current;
        var x = (0, a.useContext)(d);
        var C = function e(r) {
            var a;
            r.stopPropagation();
            u === null || u === void 0 ? void 0 : u(r, {
                value: b,
                name: x.name
            });
            (a = x.onChange) === null || a === void 0 ? void 0 : a.call(x, r, {
                value: b,
                name: x.name
            });
        };
        var O = false;
        if ((x === null || x === void 0 ? void 0 : x.value) !== undefined) {
            O = x.value === b;
        }
        var w = t;
        if (x === null || x === void 0 ? void 0 : x.disabled) {
            w = x.disabled;
        }
        var S = l;
        if (x === null || x === void 0 ? void 0 : x.error) {
            S = x.error;
        }
        
        return o().createElement(v, g({
            "data-test": "option",
            "data-test-value": b
        }, p, {
            "data-checked": O
        }),  o().createElement(y, {
            id: k,
            type: "radio",
            name: x.name,
            $error: S,
            disabled: w,
            value: b.toString(),
            checked: O,
            onChange: C,
            "data-test": "input",
            required: x.required
        }),  o().createElement(f, {
            htmlFor: k,
            "data-test": "label"
        }, r));
    }
    x.propTypes = k;
    /* harmony default export */    const C = x;
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const O = require("@splunk/react-ui/Box");
    var w =  e.n(O);
    // CONCATENATED MODULE: ./src/RadioList/RadioListStyles.ts
    var S = b()(w()).withConfig({
        displayName: "RadioListStyles__StyledBox",
        componentId: "sc-1qg10hx-0"
    })([ "align-items:flex-start;flex-wrap:wrap;flex-direction:", ";", "" ], (function(e) {
        return e.direction;
    }), (function(e) {
        return e.direction === "row" && (0, u.css)([ "", "{margin-right:40px;margin-inline-end:40px;}" ], v);
    }));
    // CONCATENATED MODULE: ./src/RadioList/RadioList.tsx
    function j() {
        j = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var a = arguments[r];
                for (var o in a) {
                    if (Object.prototype.hasOwnProperty.call(a, o)) {
                        e[o] = a[o];
                    }
                }
            }
            return e;
        };
        return j.apply(this, arguments);
    }
    function q(e, r) {
        return E(e) || A(e, r) || I(e, r) || P();
    }
    function P() {
        throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function I(e, r) {
        if (!e) return;
        if (typeof e === "string") return _(e, r);
        var a = Object.prototype.toString.call(e).slice(8, -1);
        if (a === "Object" && e.constructor) a = e.constructor.name;
        if (a === "Map" || a === "Set") return Array.from(e);
        if (a === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(a)) return _(e, r);
    }
    function _(e, r) {
        if (r == null || r > e.length) r = e.length;
        for (var a = 0, o = new Array(r); a < r; a++) {
            o[a] = e[a];
        }
        return o;
    }
    function A(e, r) {
        if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(e))) return;
        var a = [];
        var o = true;
        var i = false;
        var t = undefined;
        try {
            for (var n = e[Symbol.iterator](), l; !(o = (l = n.next()).done); o = true) {
                a.push(l.value);
                if (r && a.length === r) break;
            }
        } catch (e) {
            i = true;
            t = e;
        } finally {
            try {
                if (!o && n["return"] != null) n["return"]();
            } finally {
                if (i) throw t;
            }
        }
        return a;
    }
    function E(e) {
        if (Array.isArray(e)) return e;
    }
    function B(e, r) {
        if (e == null) return {};
        var a = D(e, r);
        var o, i;
        if (Object.getOwnPropertySymbols) {
            var t = Object.getOwnPropertySymbols(e);
            for (i = 0; i < t.length; i++) {
                o = t[i];
                if (r.indexOf(o) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, o)) continue;
                a[o] = e[o];
            }
        }
        return a;
    }
    function D(e, r) {
        if (e == null) return {};
        var a = {};
        var o = Object.keys(e);
        var i, t;
        for (t = 0; t < o.length; t++) {
            i = o[t];
            if (r.indexOf(i) >= 0) continue;
            a[i] = e[i];
        }
        return a;
    }
    var N = {
        children: t().node,
        defaultValue: t().any,
        describedBy: t().string,
        direction: t().oneOf([ "row", "column" ]),
        disabled: t().bool,
        elementRef: t().oneOfType([ t().func, t().object ]),
        error: t().bool,
        labelledBy: t().string,
        name: t().string,
        onChange: t().func,
        /** @private. */
        required: t().bool,
        value: t().any
    };
    function R(e) {
        var r = e.children, i = e.defaultValue, t = e.describedBy, n = e.direction, c = n === void 0 ? "column" : n, u = e.disabled, b = u === void 0 ? false : u, p = e.error, v = p === void 0 ? false : p, f = e.labelledBy, y = e.name, g = e.onChange, m = e.required, h = e.value, k = B(e, [ "children", "defaultValue", "describedBy", "direction", "disabled", "error", "labelledBy", "name", "onChange", "required", "value" ]);
        // @docs-props-type RadioListPropsBase
        // TODO: Implement shared isControlled hook (SUI-5283).
        // See material-ui for similar pattern to handle controlled/un-controlled in functional components w/ hooks
        // https://github.com/mui-org/material-ui/blob/60d99a39836fb82f4da1477a717f642c216fb0b9/packages/material-ui/src/utils/useControlled.js
                var x = (0, a.useRef)(h !== undefined), C = x.current;
        var O = (0, a.useState)(i), w = q(O, 2), P = w[0], I = w[1];
        var _ = C ? h : P;
        var A = (0, a.useRef)(y || (0, s.createDOMID)("radio-name")), E = A.current;
        var D = function e(r, a) {
            if (!C) {
                I(a.value);
            }
            g === null || g === void 0 ? void 0 : g(r, a);
        };
 // Check correct props are passed
                if (false) {}
        if (false) {}
 // Check if changing from controlled to un-controlled
                (0, a.useEffect)((function() {
            if (false) {}
        }), [ h, C ]);
        var N = (0, a.useRef)(i), R = N.current;
        (0, a.useEffect)((function() {
            if (false) {}
        }), [ R, i, C ]);
        var M = k["aria-labelledby"];
        var T = k["aria-describedby"];
        
        return o().createElement(d.Provider, {
            value: {
                disabled: b,
                error: v,
                onChange: D,
                name: E,
                required: m,
                value: _
            }
        },  o().createElement(S, j({
            flex: true,
            direction: c,
            role: "radiogroup",
            "data-test": "radio-list",
            "data-test-value": _
        }, l()(k, [ "aria-labelledby", "aria-describedby" ]), {
            "aria-labelledby": M ? "".concat(M, " ").concat(f) : f,
            "aria-describedby": T ? "".concat(T, " ").concat(t) : t
        }), r));
    }
    R.propTypes = N;
    R.Option = C;
    /* harmony default export */    const M = R;
    // only for styled components using RadioList
        // CONCATENATED MODULE: ./src/RadioList/index.ts
    module.exports = r;
    /******/})();
