// @splunk/react-ui/RadioBar.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var t = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(t, {
                a: t
            });
            /******/            return t;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, t) => {
            /******/ for (var n in t) {
                /******/ if (e.o(t, n) && !e.o(r, n)) {
                    /******/ Object.defineProperty(r, n, {
                        enumerable: true,
                        get: t[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        Option: () => /* reexport */ _,
        default: () => /* reexport */ re
    });
    // CONCATENATED MODULE: external "react"
    const t = require("react");
    var n =  e.n(t);
    // CONCATENATED MODULE: external "prop-types"
    const a = require("prop-types");
    var l =  e.n(a);
    // CONCATENATED MODULE: external "lodash/has"
    const o = require("lodash/has");
    var i =  e.n(o);
    // CONCATENATED MODULE: external "lodash/omit"
    const u = require("lodash/omit");
    var c =  e.n(u);
    // CONCATENATED MODULE: external "@splunk/ui-utils/keyboard"
    const s = require("@splunk/ui-utils/keyboard");
    // CONCATENATED MODULE: external "@splunk/themes"
    const d = require("@splunk/themes");
    // CONCATENATED MODULE: external "@splunk/react-ui/Button"
    const f = require("@splunk/react-ui/Button");
    var p =  e.n(f);
    // CONCATENATED MODULE: external "styled-components"
    const v = require("styled-components");
    var b =  e.n(v);
    // CONCATENATED MODULE: external "@splunk/react-ui/Clickable"
    const y = require("@splunk/react-ui/Clickable");
    var h =  e.n(y);
    // CONCATENATED MODULE: ./src/RadioBar/OptionStyles.ts
    var g = b()(h()).withConfig({
        displayName: "OptionStyles__StyledClickable",
        componentId: "fyq77p-0"
    })([ "border-radius:3px;flex:1 1 auto;display:flex;justify-content:center;align-items:center;color:", ";padding:2px 12px;height:calc(", " - 6px);position:relative;gap:", ";& + &{margin-left:2px;}", " ", " &:focus{color:", ";box-shadow:", ";z-index:2;&:active{background-color:", ";}}", " ", "" ], d.variables.contentColorDefault, d.variables.inputHeight, d.variables.spacingXSmall, (function(e) {
        var r = e.$selected;
        return r && (0, v.css)([ "color:", ";background-color:", ";box-shadow:", ";" ], d.variables.contentColorActive, (0, 
        d.pick)({
            dark: d.variables.interactiveColorOverlaySelected,
            light: d.variables.white
        }), d.variables.embossShadow);
    }), (function(e) {
        var r = e.$selected, t = e.disabled;
        return r && t && (0, v.css)([ "background-color:", ";" ], d.variables.backgroundColorPage);
    }), d.variables.contentColorActive, d.variables.focusShadow, d.variables.interactiveColorOverlayActive, (function(e) {
        var r = e.$selected, t = e.disabled;
        return !r && !t && (0, v.css)([ "&:hover{color:", ";background-color:", ";}&:active{background-color:", ";}" ], d.variables.contentColorActive, d.variables.interactiveColorOverlayHover, d.variables.interactiveColorOverlayActive);
    }), (function(e) {
        var r = e.disabled;
        return r && (0, v.css)([ "color:", ";" ], d.variables.contentColorDisabled);
    }));
    var m = b().div.withConfig({
        displayName: "OptionStyles__StyledIcon",
        componentId: "fyq77p-1"
    })([ "flex:0 0 auto;margin-top:-2px;font-size:15px;" ]);
    var O = b().div.withConfig({
        displayName: "OptionStyles__StyledLabel",
        componentId: "fyq77p-2"
    })([ "", " flex:0 1 auto;", "" ], d.mixins.ellipsis, (0, d.pick)({
        enterprise: (0, v.css)([ "&:not(:first-child){margin-left:", ";}" ], d.variables.spacingXSmall)
    }));
    var x = b().div.withConfig({
        displayName: "OptionStyles__StyledAdornment",
        componentId: "fyq77p-3"
    })([ "flex:0 0 auto;", "" ], (0, d.pick)({
        enterprise: (0, v.css)([ "&:not(:first-child){margin-left:", ";}" ], d.variables.spacingXSmall)
    }));
    // CONCATENATED MODULE: ./src/RadioBar/RadioBarContext.tsx
    var k =  (0, t.createContext)({});
    k.displayName = "RadioBar";
    /* harmony default export */    const w = k;
    // CONCATENATED MODULE: ./src/RadioBar/Option.tsx
    function C() {
        C = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return C.apply(this, arguments);
    }
    function S(e, r) {
        if (e == null) return {};
        var t = j(e, r);
        var n, a;
        if (Object.getOwnPropertySymbols) {
            var l = Object.getOwnPropertySymbols(e);
            for (a = 0; a < l.length; a++) {
                n = l[a];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function j(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var a, l;
        for (l = 0; l < n.length; l++) {
            a = n[l];
            if (r.indexOf(a) >= 0) continue;
            t[a] = e[a];
        }
        return t;
    }
    var P = {
        /** @private Set by `RadioBar`. */
        append: l().bool,
        disabled: l().bool,
        startAdornment: l().node,
        endAdornment: l().node,
        icon: l().node,
        label: l().string,
        /** @private Set by `RadioBar`. */
        prepend: l().bool,
        /** @private Set by `RadioBar`. */
        selected: l().bool,
        /** The value of the `Option`. */
        value: l().any.isRequired
    };
    function E(e) {
        var r = e.disabled, a = e.icon, l = e.label, o = e.selected, i = e.startAdornment, u = e.endAdornment, c = e.value, s = S(e, [ "disabled", "icon", "label", "selected", "startAdornment", "endAdornment", "value" ]);
        // @docs-props-type OptionPropsBase
                var d = (0, t.useContext)(w), f = d.onClick, v = d.appearance, b = d.error, y = d.onKeyDown, h = d.firstValue, k = d.selectedValue;
        if (false) {}
        var j = -1;
        if (k) {
            if (c === k) {
                j = 0;
            }
        } else if (c === h) {
            j = 0;
        }
        var P = j;
        if (v === "prisma") {
            
            return n().createElement(g, C({
                "data-test": "option",
                "data-test-value": c
            }, s, {
                onClick: o ? undefined : function(e) {
                    return f === null || f === void 0 ? void 0 : f(e, {
                        value: c,
                        label: l
                    });
                },
                role: "radio",
                "aria-checked": o,
                $selected: o,
                tabIndex: P,
                disabled: r,
                onKeyDown: function e(r) {
                    return y === null || y === void 0 ? void 0 : y(r, {
                        value: c,
                        label: l
                    });
                }
            }), a &&  n().createElement(m, null, a), i &&  n().createElement(x, null, i), l &&  n().createElement(O, null, l), u &&  n().createElement(x, null, u));
        }
        
        return n().createElement(p(), C({
            "data-test": "option",
            "data-test-value": c,
            appearance: v,
            disabled: r,
            icon: a,
            selected: o,
            value: c,
            error: b
        }, s, {
            onClick: o ? undefined : function(e) {
                return f === null || f === void 0 ? void 0 : f(e, {
                    value: c,
                    label: l
                });
            },
            role: "radio",
            "aria-checked": o,
            tabIndex: P,
            onKeyDown: function e(r) {
                return y === null || y === void 0 ? void 0 : y(r, {
                    value: c,
                    label: l
                });
            }
        }), i &&  n().createElement(x, null, i), l &&  n().createElement(O, null, l), u &&  n().createElement(x, null, u));
    }
    E.propTypes = P;
    /* harmony default export */    const _ = E;
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const T = require("@splunk/react-ui/Box");
    var q =  e.n(T);
    // CONCATENATED MODULE: ./src/RadioBar/RadioBarStyles.ts
    var D = b()(q()).withConfig({
        displayName: "RadioBarStyles__StyledRadioBar",
        componentId: "gg1b79-0"
    })([ "height:", ";", " [data-inline] + &{margin-left:", ";}" ], d.variables.inputHeight, (0, 
    d.pick)({
        prisma: (0, v.css)([ "padding:2px;border-radius:", ";border:1px solid ", ";background:", ";" ], d.variables.borderRadius, (function(e) {
            var r = e.$error;
            return r ? d.variables.accentColorNegative : "transparent";
        }), (function(e) {
            var r = e.$disabled;
            return r ? d.variables.interactiveColorBackgroundDisabled : d.variables.interactiveColorBackground;
        }))
    }), (0, d.pick)({
        enterprise: d.variables.spacingHalf,
        prisma: d.variables.spacingSmall
    }));
    // CONCATENATED MODULE: ./src/RadioBar/RadioBar.tsx
    function R(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            R = function e(r) {
                return typeof r;
            };
        } else {
            R = function e(r) {
                return r && typeof Symbol === "function" && r.constructor === Symbol && r !== Symbol.prototype ? "symbol" : typeof r;
            };
        }
        return R(e);
    }
    function V() {
        V = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return V.apply(this, arguments);
    }
    function A(e, r) {
        if (e == null) return {};
        var t = B(e, r);
        var n, a;
        if (Object.getOwnPropertySymbols) {
            var l = Object.getOwnPropertySymbols(e);
            for (a = 0; a < l.length; a++) {
                n = l[a];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function B(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var a, l;
        for (l = 0; l < n.length; l++) {
            a = n[l];
            if (r.indexOf(a) >= 0) continue;
            t[a] = e[a];
        }
        return t;
    }
    function I(e, r) {
        var t = Object.keys(e);
        if (Object.getOwnPropertySymbols) {
            var n = Object.getOwnPropertySymbols(e);
            if (r) n = n.filter((function(r) {
                return Object.getOwnPropertyDescriptor(e, r).enumerable;
            }));
            t.push.apply(t, n);
        }
        return t;
    }
    function N(e) {
        for (var r = 1; r < arguments.length; r++) {
            var t = arguments[r] != null ? arguments[r] : {};
            if (r % 2) {
                I(Object(t), true).forEach((function(r) {
                    J(e, r, t[r]);
                }));
            } else if (Object.getOwnPropertyDescriptors) {
                Object.defineProperties(e, Object.getOwnPropertyDescriptors(t));
            } else {
                I(Object(t)).forEach((function(r) {
                    Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r));
                }));
            }
        }
        return e;
    }
    function $(e, r) {
        if (!(e instanceof r)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function K(e, r) {
        for (var t = 0; t < r.length; t++) {
            var n = r[t];
            n.enumerable = n.enumerable || false;
            n.configurable = true;
            if ("value" in n) n.writable = true;
            Object.defineProperty(e, n.key, n);
        }
    }
    function H(e, r, t) {
        if (r) K(e.prototype, r);
        if (t) K(e, t);
        return e;
    }
    function M(e, r) {
        if (typeof r !== "function" && r !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(r && r.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (r) X(e, r);
    }
    function X(e, r) {
        X = Object.setPrototypeOf || function e(r, t) {
            r.__proto__ = t;
            return r;
        };
        return X(e, r);
    }
    function z(e) {
        var r = F();
        return function t() {
            var n = G(e), a;
            if (r) {
                var l = G(this).constructor;
                a = Reflect.construct(n, arguments, l);
            } else {
                a = n.apply(this, arguments);
            }
            return L(this, a);
        };
    }
    function L(e, r) {
        if (r && (R(r) === "object" || typeof r === "function")) {
            return r;
        }
        return U(e);
    }
    function U(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function F() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function G(e) {
        G = Object.setPrototypeOf ? Object.getPrototypeOf : function e(r) {
            return r.__proto__ || Object.getPrototypeOf(r);
        };
        return G(e);
    }
    function J(e, r, t) {
        if (r in e) {
            Object.defineProperty(e, r, {
                value: t,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[r] = t;
        }
        return e;
    }
    /** @public */    var Q = {
        appearance: l().oneOf([ "default", "pill" ]),
        children: l().node,
        defaultValue: l().any,
        describedBy: l().string,
        elementRef: l().oneOfType([ l().func, l().object ]),
        error: l().bool,
        inline: l().bool,
        labelledBy: l().string,
        name: l().string,
        onChange: l().func,
        /** @private. */
        required: l().bool,
        /** @private */
        splunkTheme: l().object,
        value: l().any
    };
    var W = {
        appearance: "default",
        error: false,
        inline: false
    };
    /**
 * RadioBar is a form control that provides the ability to select one option out of a group.
 */    var Y =  function(e) {
        M(a, e);
        var r = z(a);
        // @docs-props-type RadioBarPropsBase
        // eslint-disable-line @typescript-eslint/no-explicit-any
                function a(e) {
            var t;
            $(this, a);
            t = r.call(this, e);
            J(U(t), "controlledExternally", void 0);
            J(U(t), "filteredValues", []);
            J(U(t), "handleClick", (function(e, r) {
                t.handleOnChange(e, r);
            }));
            J(U(t), "handleOnKeyDown", (function(e, r) {
                var n = r.value, a = r.label;
                var l = (0, s.keycode)(e.nativeEvent);
                var o = e.currentTarget;
                var i = null;
                var u = 0;
                var c = t.filteredValues.indexOf(n);
                if (l === "down" || l === "right") {
                    e.preventDefault();
                    var d = t.getNextTarget("forward", o, c), f = d.nextTargetRes, p = d.nextIndexRes;
                    i = f;
                    u = p;
                } else if (l === "up" || l === "left") {
                    e.preventDefault();
                    var v = t.getNextTarget("backward", o, c), b = v.nextTargetRes, y = v.nextIndexRes;
                    i = b;
                    u = y;
                }
                if (i) {
                    var h;
                    (h = i) === null || h === void 0 ? void 0 : h.focus();
                    var g = t.filteredValues[u];
                    t.handleOnChange(e, {
                        value: g,
                        label: a
                    });
                }
            }));
            J(U(t), "handleOnChange", (function(e, r) {
                if (t.props.value !== r.value) {
                    var n, a;
                    (n = (a = t.props).onChange) === null || n === void 0 ? void 0 : n.call(a, e, N({
                        name: t.props.name
                    }, r));
                    if (!t.isControlled()) {
                        t.setState({
                            value: r.value
                        });
                    }
                }
            }));
            t.controlledExternally = i()(e, "value");
            if (!t.isControlled()) {
                t.state = {
                    value: e.defaultValue
                };
            }
            if (false) {}
            return t;
        }
        H(a, [ {
            key: "componentDidUpdate",
            value: function e(r) {
                if (false) {}
                if (false) {}
            }
        }, {
            key: "isControlled",
            value: function e() {
                return this.controlledExternally;
            }
 // TODO: Once converted to a functional component, this function can be removed and RadioBar should use useRovingFocus hook
                }, {
            key: "getNextTarget",
            value: function e(r, t, n) {
                var a = r === "forward" ? 1 : -1;
                var l = r === "forward" ? "nextSibling" : "previousSibling";
                var o = r === "forward" ? "firstElementChild" : "lastElementChild";
                var i = r === "forward" ? 0 : this.filteredValues.length - 1;
                var u = t[l];
                var c = null;
                var s = n + a;
                while (u !== t && c === null) {
                    if (u === null) {
                        var d;
                        u = (d = t.parentElement) === null || d === void 0 ? void 0 : d[o];
                        s = i;
                    }
                    if (u.disabled) {
                        var f = u[l];
                        u = f;
                        s += a;
                    } else {
                        c = u;
                    }
                }
                return {
                    nextTargetRes: c,
                    nextIndexRes: s
                };
            }
        }, {
            key: "render",
            value: function e() {
                var r = this;
                var a = this.props, l = a.appearance, o = a.children, i = a.describedBy, u = a.error, s = a.labelledBy, d = a.splunkTheme, f = a.value, p = a.required, v = A(a, [ "appearance", "children", "describedBy", "error", "labelledBy", "splunkTheme", "value", "required" ]);
                if (false) {}
                var b = this.isControlled() ? f : this.state.value;
                var y = d.isPrisma;
                var h = true;
                this.filteredValues = [];
                var g = t.Children.toArray(o).filter(t.isValidElement).map((function(e, n, a) {
                    var l = a.length;
                    r.filteredValues.push(e.props.value);
                    if (e.props && !e.props.disabled) {
                        h = false;
                    }
                    
                    return (0, t.cloneElement)(e, {
                        selected: e.props.value === b,
                        key: e.key || n,
                        append: n < l - 1,
                        prepend: n > 0
                    });
                }));
                var m = this.filteredValues[0];
                var O = f == null || this.filteredValues.includes(b);
                if (false) {}
                var x = {
                    appearance: y ? "prisma" : l,
                    onClick: this.handleClick,
                    error: y ? false : u,
                    onKeyDown: this.handleOnKeyDown,
                    firstValue: m,
                    selectedValue: O ? b : undefined
                };
                
                return n().createElement(D, V({
                    flex: true,
                    $disabled: h,
                    $error: u,
                    "data-test": "radio-bar",
                    "data-test-value": b,
                    role: "radiogroup",
                    "aria-labelledby": s,
                    "aria-describedby": i,
                    "aria-required": p
                }, c()(v, "onChange")),  n().createElement(w.Provider, {
                    value: x
                }, g));
            }
        } ]);
        return a;
    }(t.Component);
    J(Y, "propTypes", Q);
    J(Y, "defaultProps", W);
    J(Y, "Option", _);
    var Z = (0, d.withSplunkTheme)(Y);
    var ee = Z;
    ee.propTypes = Y.propTypes;
    ee.Option = _;
    /* harmony default export */    const re = ee;
    // only for styled components using RadioBar
        // CONCATENATED MODULE: ./src/RadioBar/index.ts
    module.exports = r;
    /******/})();
