// @splunk/react-ui/Resize.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var r = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(r, {
                a: r
            });
            /******/            return r;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        default: () => /* reexport */ G
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    var n =  e.n(r);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var i =  e.n(o);
    // CONCATENATED MODULE: external "lodash/keys"
    const a = require("lodash/keys");
    var s =  e.n(a);
    // CONCATENATED MODULE: external "lodash/omit"
    const l = require("lodash/omit");
    var c =  e.n(l);
    // CONCATENATED MODULE: external "@splunk/react-ui/EventListener"
    const u = require("@splunk/react-ui/EventListener");
    var d =  e.n(u);
    // CONCATENATED MODULE: external "@splunk/react-ui/ScreenReaderContent"
    const f = require("@splunk/react-ui/ScreenReaderContent");
    var p =  e.n(f);
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const h = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "@splunk/ui-utils/keyboard"
    const v = require("@splunk/ui-utils/keyboard");
    // CONCATENATED MODULE: external "styled-components"
    const y = require("styled-components");
    var w =  e.n(y);
    // CONCATENATED MODULE: external "@splunk/themes"
    const m = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/Resize/ResizeStyles.ts
    var b = "9px";
    var g = "14px";
    var S = "3px";
    var z = "11px";
    var R = w().div.withConfig({
        displayName: "ResizeStyles__Styled",
        componentId: "sc-181mwes-0"
    })([ "", ";position:relative;overflow-x:hidden;overflow-y:hidden;" ], m.mixins.reset("block"));
    var _ = w().div.withConfig({
        displayName: "ResizeStyles__StyledBorder",
        componentId: "sc-181mwes-1"
    })([ "", ";height:100%;width:100%;border:0 solid ", ";", " ", " ", " ", "" ], m.mixins.reset("block"), (0, 
    m.pick)({
        prisma: m.variables.interactiveColorBackground,
        enterprise: {
            dark: m.variables.gray22,
            light: m.variables.white
        }
    }), (function(e) {
        var t = e.$borderSides;
        return t.n && (0, y.css)([ "border-top-width:", ";" ], b);
    }), (function(e) {
        var t = e.$borderSides;
        return t.e && (0, y.css)([ "border-right-width:", ";" ], b);
    }), (function(e) {
        var t = e.$borderSides;
        return t.s && (0, y.css)([ "border-bottom-width:", ";" ], b);
    }), (function(e) {
        var t = e.$borderSides;
        return t.w && (0, y.css)([ "border-left-width:", ";" ], b);
    }));
    var x = w().svg.withConfig({
        displayName: "ResizeStyles__StyledCornerSvg",
        componentId: "sc-181mwes-2"
    })([ "display:block;stroke:currentColor;" ]);
    var O = w().path.withConfig({
        displayName: "ResizeStyles__StyledCornerPath",
        componentId: "sc-181mwes-3"
    })([ "stroke-width:1.1px;" ]);
    var k = w().svg.withConfig({
        displayName: "ResizeStyles__StyledSideSvg",
        componentId: "sc-181mwes-4"
    })([ "display:block;fill:currentColor;" ]);
    var D = (0, y.css)([ "", ";color:", ";&:hover,&:focus{color:", ";}", "" ], m.mixins.reset("block"), (0, 
    m.pick)({
        prisma: m.variables.contentColorDefault,
        enterprise: {
            dark: m.variables.gray92,
            light: m.variables.borderColor
        }
    }), (0, m.pick)({
        prisma: m.variables.interactiveColorPrimary,
        enterprise: m.variables.linkColor
    }), (function(e) {
        var t = e.$showHandles;
        return t === "on-hover" && (0, y.css)([ "opacity:0;transition:opacity 200ms;", ":hover > div > &,&:focus{opacity:1;}" ], 
        /* sc-sel */
        R);
    }));
    var E = (0, y.css)([ "", ";width:", ";height:", ";position:absolute;background-clip:content-box;z-index:2;" ], D, z, z);
    var C = (0, y.css)([ "", ";position:absolute;z-index:1;" ], D);
    var j = w().button.withConfig({
        displayName: "ResizeStyles__StyledResize",
        componentId: "sc-181mwes-5"
    })([ "", "" ], (0, m.pickVariant)("$cardinalDirection", {
        nw: (0, y.css)([ "", ";top:0;left:0;cursor:nwse-resize;> ", "{transform:rotate(180deg);}" ], E, 
        /* sc-sel */
        x),
        n: (0, y.css)([ "", ";cursor:ns-resize;top:0;left:calc(50% - ", " / 2);padding:", " 0;" ], C, g, S),
        ne: (0, y.css)([ "", ";right:0;top:0;cursor:nesw-resize;> ", "{transform:rotate(-90deg);}" ], E, 
        /* sc-sel */
        x),
        w: (0, y.css)([ "", ";cursor:ew-resize;padding:0 ", ";left:0;top:calc(50% - ", " / 2);" ], C, S, g),
        e: (0, y.css)([ "", ";cursor:ew-resize;padding:0 ", ";right:0;top:calc(50% - ", " / 2);" ], C, S, g),
        sw: (0, y.css)([ "", ";left:0;bottom:0;cursor:nesw-resize;> ", "{transform:rotate(90deg);}" ], E, 
        /* sc-sel */
        x),
        s: (0, y.css)([ "", ";cursor:ns-resize;padding:", " 0;bottom:0;left:calc(50% - ", " / 2);" ], C, S, g),
        se: (0, y.css)([ "", ";right:0;bottom:0;cursor:nwse-resize;" ], E)
    }));
    // CONCATENATED MODULE: ./src/utils/updateReactRef.ts
    /**
 * Updates a React ref. Callback refs and object refs (from `createRef` and `useRef`) are supported.
 *
 * @param ref - The React callback or object ref. Can be `null` or `undefined`.
 * @param current - The new value of the ref.
 */
    function q(e, t) {
        if (e) {
            if (typeof e === "function") {
                e(t);
            } else {
                // the public signature of this util uses React.Ref<T> to mirror the way React types refs.
                // the intention here is to signal "we will take care of setting 'current', not you".
                e.current = t;
 // eslint-disable-line no-param-reassign
                        }
        }
    }
    // CONCATENATED MODULE: ./src/Resize/Resize.tsx
    function P(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            P = function e(t) {
                return typeof t;
            };
        } else {
            P = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return P(e);
    }
    function H() {
        H = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return H.apply(this, arguments);
    }
    function $(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function I(e, t) {
        for (var r = 0; r < t.length; r++) {
            var n = t[r];
            n.enumerable = n.enumerable || false;
            n.configurable = true;
            if ("value" in n) n.writable = true;
            Object.defineProperty(e, n.key, n);
        }
    }
    function T(e, t, r) {
        if (t) I(e.prototype, t);
        if (r) I(e, r);
        return e;
    }
    function M(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) N(e, t);
    }
    function N(e, t) {
        N = Object.setPrototypeOf || function e(t, r) {
            t.__proto__ = r;
            return t;
        };
        return N(e, t);
    }
    function B(e) {
        var t = W();
        return function r() {
            var n = X(e), o;
            if (t) {
                var i = X(this).constructor;
                o = Reflect.construct(n, arguments, i);
            } else {
                o = n.apply(this, arguments);
            }
            return K(this, o);
        };
    }
    function K(e, t) {
        if (t && (P(t) === "object" || typeof t === "function")) {
            return t;
        }
        return L(e);
    }
    function L(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function W() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function X(e) {
        X = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return X(e);
    }
    function Y(e, t, r) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: r,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = r;
        }
        return e;
    }
    var A = {
        appearance: i().oneOf([ "border", "overlay" ]),
        children: i().node,
        elementRef: i().oneOfType([ i().func, i().object ]),
        keyIncrement: i().number,
        onRequestResize: i().func.isRequired,
        resizeHandles: i().arrayOf(i().oneOf([ "nw", "n", "ne", "w", "e", "sw", "s", "se" ])).isRequired,
        showHandles: i().oneOf([ "always", "on-hover" ])
    };
    var F = {
        appearance: "overlay",
        keyIncrement: 10,
        showHandles: "always"
    };
    /**
 * Resize is a utility container with drag handles for resizing.
 */    var V =  function(e) {
        M(r, e);
        var t = B(r);
        // @docs-props-type ResizePropsBase
                function r(e) {
            var o;
            $(this, r);
            o = t.call(this, e);
            Y(L(o), "handleMount", (function(e) {
                o.setState({
                    el: e
                });
                q(o.props.elementRef, e);
            }));
            Y(L(o), "handleDragStart", (function(e, t) {
                if (e.button > 0) {
                    return;
                }
                e.preventDefault();
                e.persist();
                o.setState((function(r) {
                    var n, o;
                    return {
                        dragDirection: t,
                        startWidth: (n = r.el) === null || n === void 0 ? void 0 : n.offsetWidth,
                        startHeight: (o = r.el) === null || o === void 0 ? void 0 : o.offsetHeight,
                        startX: e.clientX,
                        startY: e.clientY
                    };
                }));
            }));
            Y(L(o), "handleDragEnd", (function() {
                o.setState({
                    dragDirection: undefined
                });
            }));
            Y(L(o), "handleDrag", (function(e) {
                var t = o.state, r = t.dragDirection, n = t.startWidth, i = n === void 0 ? 0 : n, a = t.startHeight, s = a === void 0 ? 0 : a, l = t.startX, c = l === void 0 ? 0 : l, u = t.startY, d = u === void 0 ? 0 : u;
                var f = (r === null || r === void 0 ? void 0 : r.match(/(w|e)$/)) ? e.clientX - c : 0;
                var p = (r === null || r === void 0 ? void 0 : r.match(/^(s|n)/)) ? e.clientY - d : 0;
                var h = {
                    width: (r === null || r === void 0 ? void 0 : r.match(/w$/)) ? i - f : i + f,
                    height: (r === null || r === void 0 ? void 0 : r.match(/^n/)) ? s - p : s + p
                };
                o.props.onRequestResize(e, h);
            }));
            Y(L(o), "handleKeyDown", (function(e, t) {
                var r = o.props.keyIncrement;
                var n = (0, v.keycode)(e.nativeEvent);
                var i = t.slice(-1);
                var a = {
                    e: {
                        right: 1,
                        left: -1
                    },
                    w: {
                        right: -1,
                        left: 1
                    }
                };
                var s = (i === "e" || i === "w") && (n === "right" || n === "left") ? a[i][n] : 0;
                var l = t.charAt(0);
                var c = {
                    n: {
                        down: -1,
                        up: 1
                    },
                    s: {
                        down: 1,
                        up: -1
                    }
                };
                var u = (l === "n" || l === "s") && (n === "up" || n === "down") ? c[l][n] : 0;
                if ((s || u) && o.state.el) {
                    e.preventDefault();
                    var d = {
                        width: o.state.el.offsetWidth + s * r,
                        height: o.state.el.offsetHeight + u * r
                    };
                    o.props.onRequestResize(e, d);
                }
            }));
            Y(L(o), "renderHandle", (function(e) {
                var t = o.props.showHandles;
                var i = e.length > 1;
                var a;
                if (i) {
                    a =  n().createElement(x, {
                        viewBox: "0 0 11 11",
                        width: "11px",
                        height: "11px"
                    },  n().createElement(O, {
                        d: "M0 8 L 8 0 M4 8 L 8 4"
                    }));
                } else if (e === "n" || e === "s") {
                    a =  n().createElement(k, {
                        viewBox: "0 0 14 3",
                        width: "14px",
                        height: "3px"
                    },  n().createElement("rect", {
                        width: "14",
                        height: "1",
                        x: "0",
                        y: "0"
                    }),  n().createElement("rect", {
                        width: "14",
                        height: "1",
                        x: "0",
                        y: "2"
                    }));
                } else if (e === "e" || e === "w") {
                    a =  n().createElement(k, {
                        viewBox: "0 0 3 14",
                        width: "3px",
                        height: "14px"
                    },  n().createElement("rect", {
                        width: "1",
                        height: "14",
                        x: "0",
                        y: "0"
                    }),  n().createElement("rect", {
                        width: "1",
                        height: "14",
                        x: "2",
                        y: "0"
                    }));
                }
                
                return n().createElement(j, {
                    key: e,
                    "data-test": "resize-handle-".concat(e),
                    onMouseDown: o["handleDragStart".concat(e)],
                    onKeyDown: o["handleKeyDown".concat(e)],
                    $showHandles: t,
                    $cardinalDirection: e
                }, a,  n().createElement(p(), null, r.handleLabels[e]));
            }));
            o.state = {
                el: null
            };
 // Generate handlers for each direction.
                        r.handleOrder.forEach((function(e) {
                // @ts-expect-error - dynamic function creation
                o["handleDragStart".concat(e)] = function(t) {
                    o.handleDragStart(t, e);
                };
 // @ts-expect-error - dynamic function creation
                                o["handleKeyDown".concat(e)] = function(t) {
                    o.handleKeyDown(t, e);
                };
            }));
            return o;
        }
        T(r, [ {
            key: "render",
            value: function e() {
                var t = this.props, o = t.appearance, i = t.children, a = t.resizeHandles;
                var l = this.state, u = l.dragDirection, f = l.el;
 // Handles are inserted before and after children to optimize tab order
                                var p = r.handleOrder.slice(0, 4).filter((function(e) {
                    return a.indexOf(e) > -1;
                })).map(this.renderHandle);
                var h = r.handleOrder.slice(4).filter((function(e) {
                    return a.indexOf(e) > -1;
                })).map(this.renderHandle);
                /* When appearance is border, determine which sides need margin.
       * Converts ['se', 'n'] to { e: true, s: true, n: true}
       */                var v = {};
                if (o === "border") {
                    a.join("").split("").filter((function(e, t, r) {
                        return r.indexOf(e) === t;
                    })).forEach((function(e) {
                        v[e] = true;
                    }));
                }
                
                return n().createElement(R, H({
                    "data-test": "resize",
                    ref: this.handleMount
                }, c()(this.props, s()(r.propTypes))),  n().createElement(_, {
                    $borderSides: v
                }, f && p, i, f && h, u &&  n().createElement(n().Fragment, null,  n().createElement(d(), {
                    target: window,
                    eventType: "mousemove",
                    listener: this.handleDrag
                }),  n().createElement(d(), {
                    target: window,
                    eventType: "mouseup",
                    listener: this.handleDragEnd
                }))));
            }
        } ]);
        return r;
    }(r.Component);
    Y(V, "propTypes", A);
    Y(V, "defaultProps", F);
    Y(V, "handleOrder", [ "nw", "n", "ne", "w", "e", "sw", "s", "se" ]);
    Y(V, "handleLabels", {
        nw: (0, h._)("Resize northwest"),
        n: (0, h._)("Resize north"),
        ne: (0, h._)("Resize northeast"),
        w: (0, h._)("Resize west"),
        e: (0, h._)("Resize east"),
        sw: (0, h._)("Resize southwest"),
        s: (0, h._)("Resize south"),
        se: (0, h._)("Resize southeast")
    });
    /* harmony default export */    const G = V;
    // CONCATENATED MODULE: ./src/Resize/index.ts
    module.exports = t;
    /******/})();
