// @splunk/react-ui/ScreenReaderContent.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var t = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(t, {
                a: t
            });
            /******/            return t;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, t) => {
            /******/ for (var n in t) {
                /******/ if (e.o(t, n) && !e.o(r, n)) {
                    /******/ Object.defineProperty(r, n, {
                        enumerable: true,
                        get: t[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        default: () => /* reexport */ v
    });
    // CONCATENATED MODULE: external "react"
    const t = require("react");
    var n =  e.n(t);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var a =  e.n(o);
    // CONCATENATED MODULE: external "styled-components"
    const i = require("styled-components");
    var l =  e.n(i);
    // CONCATENATED MODULE: external "@splunk/themes"
    const c = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/ScreenReaderContent/ScreenReaderContentStyles.ts
    var u = l().span.withConfig({
        displayName: "ScreenReaderContentStyles__Styled",
        componentId: "sc-1lnohwp-0"
    })([ "", ";" ], c.mixins.screenReaderContent());
    // CONCATENATED MODULE: ./src/ScreenReaderContent/ScreenReaderContent.tsx
    function f() {
        f = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return f.apply(this, arguments);
    }
    function s(e, r) {
        if (e == null) return {};
        var t = p(e, r);
        var n, o;
        if (Object.getOwnPropertySymbols) {
            var a = Object.getOwnPropertySymbols(e);
            for (o = 0; o < a.length; o++) {
                n = a[o];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function p(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var o, a;
        for (a = 0; a < n.length; a++) {
            o = n[a];
            if (r.indexOf(o) >= 0) continue;
            t[o] = e[o];
        }
        return t;
    }
    var d = {
        children: a().node.isRequired,
        elementRef: a().oneOfType([ a().func, a().object ])
    };
    /**
 * The screen reader text is used to wrap content that
 * is only accessible through screen readers.
 */    function y(e) {
        var r = e.children, t = e.elementRef, o = s(e, [ "children", "elementRef" ]);
        // @docs-props-type ScreenReaderContentPropsBase
                
        return n().createElement(u, f({
            "data-test": "screen-reader-content",
            ref: t
        }, o), r);
    }
    y.propTypes = d;
    /* harmony default export */    const v = y;
    // CONCATENATED MODULE: ./src/ScreenReaderContent/index.ts
    module.exports = r;
    /******/})();
