// @splunk/react-ui/Search.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var t = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(t, {
                a: t
            });
            /******/            return t;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, t) => {
            /******/ for (var n in t) {
                /******/ if (e.o(t, n) && !e.o(r, n)) {
                    /******/ Object.defineProperty(r, n, {
                        enumerable: true,
                        get: t[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        Divider: () => /* reexport */ g.Divider,
        Heading: () => /* reexport */ g.Heading,
        Option: () => /* reexport */ U,
        default: () => /* reexport */ fe
    });
    // CONCATENATED MODULE: external "react"
    const t = require("react");
    var n =  e.n(t);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var i =  e.n(o);
    // CONCATENATED MODULE: external "lodash/extend"
    const a = require("lodash/extend");
    var u =  e.n(a);
    // CONCATENATED MODULE: external "lodash/has"
    const l = require("lodash/has");
    var c =  e.n(l);
    // CONCATENATED MODULE: external "lodash/memoize"
    const f = require("lodash/memoize");
    var s =  e.n(f);
    // CONCATENATED MODULE: external "lodash/omit"
    const d = require("lodash/omit");
    var v =  e.n(d);
    // CONCATENATED MODULE: external "lodash/pick"
    const p = require("lodash/pick");
    var b =  e.n(p);
    // CONCATENATED MODULE: external "@splunk/ui-utils/id"
    const y = require("@splunk/ui-utils/id");
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const m = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "@splunk/ui-utils/keyboard"
    const h = require("@splunk/ui-utils/keyboard");
    // CONCATENATED MODULE: external "@splunk/react-ui/Menu"
    const g = require("@splunk/react-ui/Menu");
    // CONCATENATED MODULE: external "@splunk/react-ui/usePrevious"
    const O = require("@splunk/react-ui/usePrevious");
    var j =  e.n(O);
    // CONCATENATED MODULE: external "@splunk/react-ui/Popover"
    const w = require("@splunk/react-ui/Popover");
    var S =  e.n(w);
    // CONCATENATED MODULE: external "@splunk/react-ui/ResultsMenu"
    const P = require("@splunk/react-ui/ResultsMenu");
    var k =  e.n(P);
    // CONCATENATED MODULE: external "@splunk/react-ui/Text"
    const C = require("@splunk/react-ui/Text");
    var I =  e.n(C);
    // CONCATENATED MODULE: ./src/Search/Option.tsx
    function M(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            M = function e(r) {
                return typeof r;
            };
        } else {
            M = function e(r) {
                return r && typeof Symbol === "function" && r.constructor === Symbol && r !== Symbol.prototype ? "symbol" : typeof r;
            };
        }
        return M(e);
    }
    function R() {
        R = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return R.apply(this, arguments);
    }
    function x(e, r) {
        if (e == null) return {};
        var t = D(e, r);
        var n, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                n = i[o];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function D(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var o, i;
        for (i = 0; i < n.length; i++) {
            o = n[i];
            if (r.indexOf(o) >= 0) continue;
            t[o] = e[o];
        }
        return t;
    }
    function E(e, r) {
        if (!(e instanceof r)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function q(e, r) {
        for (var t = 0; t < r.length; t++) {
            var n = r[t];
            n.enumerable = n.enumerable || false;
            n.configurable = true;
            if ("value" in n) n.writable = true;
            Object.defineProperty(e, n.key, n);
        }
    }
    function _(e, r, t) {
        if (r) q(e.prototype, r);
        if (t) q(e, t);
        return e;
    }
    function A(e, r) {
        if (typeof r !== "function" && r !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(r && r.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (r) B(e, r);
    }
    function B(e, r) {
        B = Object.setPrototypeOf || function e(r, t) {
            r.__proto__ = t;
            return r;
        };
        return B(e, r);
    }
    function T(e) {
        var r = L();
        return function t() {
            var n = K(e), o;
            if (r) {
                var i = K(this).constructor;
                o = Reflect.construct(n, arguments, i);
            } else {
                o = n.apply(this, arguments);
            }
            return V(this, o);
        };
    }
    function V(e, r) {
        if (r && (M(r) === "object" || typeof r === "function")) {
            return r;
        }
        return H(e);
    }
    function H(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function L() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function K(e) {
        K = Object.setPrototypeOf ? Object.getPrototypeOf : function e(r) {
            return r.__proto__ || Object.getPrototypeOf(r);
        };
        return K(e);
    }
    function F(e, r, t) {
        if (r in e) {
            Object.defineProperty(e, r, {
                value: t,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[r] = t;
        }
        return e;
    }
    var W = {
        /** @private */
        active: i().bool,
        description: i().string,
        descriptionPosition: i().oneOf([ "right", "bottom" ]),
        disabled: i().bool,
        icon: i().node,
        label: i().string,
        /** @private */
        onClick: i().func,
        truncate: i().bool,
        value: i().string.isRequired
    };
    var N = {
        active: false,
        descriptionPosition: "bottom",
        disabled: false,
        truncate: false
    };
    /**
 * An option within a `Search`. This inherits from
 * [PureComponent](https://reactjs.org/docs/react-api.html#reactpurecomponent)
 * so any elements passed to it must also be pure.
 */    var z =  function(e) {
        A(t, e);
        var r = T(t);
        function t() {
            var e;
            E(this, t);
            for (var n = arguments.length, o = new Array(n), i = 0; i < n; i++) {
                o[i] = arguments[i];
            }
            e = r.call.apply(r, [ this ].concat(o));
            F(H(e), "item", null);
            F(H(e), "handleClick", (function(r) {
                var t = e.props, n = t.disabled, o = t.onClick, i = t.value;
                if (!n) {
                    o === null || o === void 0 ? void 0 : o(r, {
                        value: i
                    });
                }
            }));
            return e;
        }
        _(t, [ {
            key: "scrollIntoViewIfNeeded",
            value: function e() {
                var r;
                (r = this.item) === null || r === void 0 ? void 0 : r.scrollIntoViewIfNeeded();
            }
        }, {
            key: "render",
            value: function e() {
                var r = this;
                // eslint-disable-next-line no-unused-vars
                                var t = this.props, o = t.value, i = t.label, a = x(t, [ "value", "label" ]);
                var u = i === undefined ? o : i;
                
                return n().createElement(g.Item, R({
                    ref: function e(t) {
                        r.item = t;
                    },
                    "data-test": "search-option",
                    "data-test-value": o
                }, a, {
                    onClick: this.handleClick,
                    role: "option",
                    "aria-selected": false
                }), u);
            }
        } ]);
        return t;
    }(t.PureComponent);
    F(z, "propTypes", W);
    F(z, "defaultProps", N);
    F(z, "type", g.Item);
    /* harmony default export */    const U = z;
    // CONCATENATED MODULE: ./src/utils/updateReactRef.ts
    /**
 * Updates a React ref. Callback refs and object refs (from `createRef` and `useRef`) are supported.
 *
 * @param ref - The React callback or object ref. Can be `null` or `undefined`.
 * @param current - The new value of the ref.
 */
    function X(e, r) {
        if (e) {
            if (typeof e === "function") {
                e(r);
            } else {
                // the public signature of this util uses React.Ref<T> to mirror the way React types refs.
                // the intention here is to signal "we will take care of setting 'current', not you".
                e.current = r;
 // eslint-disable-line no-param-reassign
                        }
        }
    }
    // CONCATENATED MODULE: ./src/Search/Search.tsx
    function Y() {
        Y = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return Y.apply(this, arguments);
    }
    function $(e, r) {
        var t = Object.keys(e);
        if (Object.getOwnPropertySymbols) {
            var n = Object.getOwnPropertySymbols(e);
            if (r) n = n.filter((function(r) {
                return Object.getOwnPropertyDescriptor(e, r).enumerable;
            }));
            t.push.apply(t, n);
        }
        return t;
    }
    function G(e) {
        for (var r = 1; r < arguments.length; r++) {
            var t = arguments[r] != null ? arguments[r] : {};
            if (r % 2) {
                $(Object(t), true).forEach((function(r) {
                    J(e, r, t[r]);
                }));
            } else if (Object.getOwnPropertyDescriptors) {
                Object.defineProperties(e, Object.getOwnPropertyDescriptors(t));
            } else {
                $(Object(t)).forEach((function(r) {
                    Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r));
                }));
            }
        }
        return e;
    }
    function J(e, r, t) {
        if (r in e) {
            Object.defineProperty(e, r, {
                value: t,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[r] = t;
        }
        return e;
    }
    function Q(e, r) {
        return ne(e) || te(e, r) || ee(e, r) || Z();
    }
    function Z() {
        throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function ee(e, r) {
        if (!e) return;
        if (typeof e === "string") return re(e, r);
        var t = Object.prototype.toString.call(e).slice(8, -1);
        if (t === "Object" && e.constructor) t = e.constructor.name;
        if (t === "Map" || t === "Set") return Array.from(e);
        if (t === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t)) return re(e, r);
    }
    function re(e, r) {
        if (r == null || r > e.length) r = e.length;
        for (var t = 0, n = new Array(r); t < r; t++) {
            n[t] = e[t];
        }
        return n;
    }
    function te(e, r) {
        if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(e))) return;
        var t = [];
        var n = true;
        var o = false;
        var i = undefined;
        try {
            for (var a = e[Symbol.iterator](), u; !(n = (u = a.next()).done); n = true) {
                t.push(u.value);
                if (r && t.length === r) break;
            }
        } catch (e) {
            o = true;
            i = e;
        } finally {
            try {
                if (!n && a["return"] != null) a["return"]();
            } finally {
                if (o) throw i;
            }
        }
        return t;
    }
    function ne(e) {
        if (Array.isArray(e)) return e;
    }
    function oe(e, r) {
        if (e == null) return {};
        var t = ie(e, r);
        var n, o;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (o = 0; o < i.length; o++) {
                n = i[o];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function ie(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var o, i;
        for (i = 0; i < n.length; i++) {
            o = n[i];
            if (r.indexOf(o) >= 0) continue;
            t[o] = e[o];
        }
        return t;
    }
    var ae = {
        animateLoading: i().bool,
        children: i().node,
        defaultPlacement: i().oneOf([ "above", "below", "vertical" ]),
        defaultValue: i().string,
        describedBy: i().string,
        disabled: i().bool,
        elementRef: i().oneOfType([ i().func, i().object ]),
        error: i().bool,
        footerMessage: i().node,
        inline: i().bool,
        inputRef: i().oneOfType([ i().func, i().object ]),
        isLoadingOptions: i().bool,
        labelledBy: i().string,
        loadingMessage: i().node,
        menuStyle: i().object,
        name: i().string,
        noOptionsMessage: i().node,
        onBlur: i().func,
        onChange: i().func,
        onClose: i().func,
        onFocus: i().func,
        onKeyDown: i().func,
        onOpen: i().func,
        onScroll: i().func,
        onSelect: i().func,
        placeholder: i().string,
        value: i().string
    };
    function ue(e, r) {
        var t = r.clientX, n = r.clientY;
        var o = e.getBoundingClientRect(), i = o.top, a = o.left, u = o.bottom, l = o.right;
        return t > a && t < l && n > i && n < u;
    }
    var le = s()((function(e) {
        var r = e.anchorWidth, t = e.maxHeight, n = e.menuStyle;
        return u()({
            maxHeight: t,
            overflow: "auto",
            width: Math.max(r !== null && r !== void 0 ? r : 0, 200)
        }, n);
    }));
    function ce(e) {
        var r = e.animateLoading, o = r === void 0 ? false : r, i = e.children, a = e.defaultPlacement, u = e.defaultValue, l = e.describedBy, f = e.disabled, s = f === void 0 ? false : f, d = e.elementRef, p = e.error, g = p === void 0 ? false : p, O = e.footerMessage, w = e.inline, P = w === void 0 ? false : w, C = e.inputRef, M = e.isLoadingOptions, R = M === void 0 ? false : M, x = e.labelledBy, D = e.loadingMessage, E = e.menuStyle, q = E === void 0 ? {} : E, _ = e.name, A = e.noOptionsMessage, B = e.onChange, T = e.onClose, V = e.onFocus, H = e.onKeyDown, L = e.onOpen, K = e.placeholder, F = K === void 0 ? (0, 
        m._)("Search...") : K, W = e.value, N = oe(e, [ "animateLoading", "children", "defaultPlacement", "defaultValue", "describedBy", "disabled", "elementRef", "error", "footerMessage", "inline", "inputRef", "isLoadingOptions", "labelledBy", "loadingMessage", "menuStyle", "name", "noOptionsMessage", "onChange", "onClose", "onFocus", "onKeyDown", "onOpen", "placeholder", "value" ]);
        // @docs-props-type SearchPropsBase
                var z = !!i || R;
 // state
                var U = (0, t.useState)(0), $ = Q(U, 2), J = $[0], Z = $[1];
        var ee = (0, t.useState)(false), re = Q(ee, 2), te = re[0], ne = re[1];
        var ie = (0, t.useState)(undefined), ae = Q(ie, 2), ce = ae[0], fe = ae[1];
        var se = (0, t.useState)(u || ""), de = Q(se, 2), ve = de[0], pe = de[1];
        var be = (0, t.useState)((function() {
            return {
                popoverId: (0, y.createDOMID)("popover"),
                menuId: (0, y.createDOMID)("menu"),
                activeItemId: (0, y.createDOMID)("active-item")
            };
        })), ye = Q(be, 1), me = ye[0], he = me.popoverId, ge = me.menuId, Oe = me.activeItemId;
 // previous state
                var je = j()(J);
 // refs
                var we = (0, t.useRef)(null);
        var Se = (0, t.useRef)(null);
        var Pe = (0, t.useRef)(W !== undefined && W !== null);
        var ke = (0, t.useRef)(false);
        var Ce = (0, t.useRef)([]);
        var Ie = (0, t.useRef)(je);
        var Me = (0, t.useRef)(undefined);
        var Re = function e(r) {
            X(we, r);
            X(d, r);
        };
        var xe = function e(r) {
            X(Se, r);
            X(C, r);
        };
        var De = function e() {
            return Pe.current ? W : ve;
        };
        var Ee = function e() {
            var r = De();
            var n = t.Children.toArray(i).filter(t.isValidElement);
            var o = !ce && n.find((function(e) {
                return e.props.label !== undefined && e.props.value === r;
            }));
            if (o) {
                return o === null || o === void 0 ? void 0 : o.props.label;
            }
            return r;
        };
        var qe = function e() {
            ne(true);
            Z(-1);
            L === null || L === void 0 ? void 0 : L();
        };
        var _e = function e() {
            ne(false);
            Ie.current = null;
            T === null || T === void 0 ? void 0 : T();
        };
        var Ae = function e() {
            var r;
            (r = Se.current) === null || r === void 0 ? void 0 : r.focus();
        };
        var Be = function e(r, t) {
            /* SUI-930 On IE 11 this handler is essentially deferred after calling
     * this.textInput.focus(). this.focusCalledInternally enables the focus event to be ignored
     * when the menu closes. */
            if (ke.current) {
                ke.current = false;
            } else {
                qe();
            }
            V === null || V === void 0 ? void 0 : V(r, t);
        };
        var Te = function e(r, t) {
            var n = t.value;
            if (!Pe.current) {
                pe(n);
            }
            Z(-1);
            fe(n);
            B === null || B === void 0 ? void 0 : B(r, {
                value: n,
                name: _
            });
        };
        var Ve = function e(r, t) {
            Te(r, G({}, t));
            ke.current = true;
            Ae();
            _e();
        };
        var He = function e(r) {
            var t = Ce.current.length;
            var n = Me.current;
            var o = (0, h.keycode)(r.nativeEvent);
            if (te) {
                switch (o) {
                  case "enter":
                    {
                        if (n) {
                            Ve(r, {
                                value: n
                            });
                        }
                        break;
                    }

                  case "tab":
                    _e();
                    break;

                  case "down":
                    Z(Math.min(J + 1, t - 1));
                    break;

                  case "up":
                    Z(Math.max(J - 1, 0));
                    break;

                  default:
 // do nothing
                                }
            } else if ((0, h.addsCharacter)(r.nativeEvent) !== false || // Safari 9.0 returns undefined
            o === "enter" || o === "backspace" || o === "down" || o === "up") {
                qe();
            }
            H === null || H === void 0 ? void 0 : H(r);
        };
        var Le = function e() {
            if (document.activeElement !== Se.current) {
                ke.current = true;
                Ae();
            }
            if (!te && !s) {
                qe();
            }
        };
        var Ke = function e(r) {
            var t = r.event, n = r.reason;
            var o = n === "offScreen" || n === "escapeKey" || n === "clickAway" && we.current && !ue(we.current, t);
            if (o) {
                _e();
            }
        };
        var Fe = function e(r) {
            if (Ie.current !== J) {
                r === null || r === void 0 ? void 0 : r.scrollIntoViewIfNeeded();
            }
        };
        var We = function e(r) {
            var a = r.anchorWidth, u = r.maxHeight;
            var l = t.Children.toArray(i).filter(t.isValidElement);
 // Highlight Active
                        Me.current = undefined;
 // Only update the menu options if they are valid and the dropdown is open.
            // This is necessary to handle the outro animation of the dropdown.
                        if (z && te) {
                Ce.current = l.map((function(e, r) {
                    if (!c()(e.props, "active")) {
                        // ignore Headings and Dividers
                        return e;
                    }
                    if (r === J) {
                        if (!e.props.disabled) {
                            Me.current = e.props.value;
                        }
                        
                        return (0, t.cloneElement)(e, {
                            ref: Fe,
                            id: Oe,
                            onClick: Ve,
                            active: true
                        });
                    }
                    
                    return (0, t.cloneElement)(e, {
                        onClick: Ve
                    });
                }));
            }
            var f = le({
                anchorWidth: a,
                maxHeight: u,
                menuStyle: q
            });
            
            return n().createElement(k(), Y({
                style: f,
                controlledExternally: true,
                isLoading: R
            }, b()(N, "onScroll"), {
                noOptionsMessage: A,
                footerMessage: O,
                animateLoading: o,
                loadingMessage: D,
                menuId: ge
            }), Ce.current);
        };
        var Ne = De();
        var ze = Ee();
        var Ue = te && !!we.current && z;
        var Xe = z ? "combobox" : "textbox";
        var Ye = {
            "aria-activedescendant": Ue && J > -1 ? Oe : undefined,
            "aria-expanded": z ? te && !!we.current : undefined,
            "aria-controls": Ue ? ge : undefined
        };
        if (!(x === null || x === void 0 ? void 0 : x.length)) {
            Ye["aria-label"] = (0, m._)("Value input");
        }
        
        return n().createElement(I(), Y({
            "data-test": "search"
        }, v()(N, "onScroll"), {
            appearance: "search",
            autoCapitalize: "off",
            autoComplete: "off",
            autoCorrect: "off",
            "data-test-popover-id": he,
            "data-test-label": ze,
            "data-test-value": Ne,
            "data-test-open": Ue,
            describedBy: l,
            disabled: s,
            elementRef: Re,
            error: g,
            inline: P,
            inputRef: xe,
            labelledBy: x,
            name: _,
            onFocus: Be,
            onClick: Le,
            onChange: Te,
            onKeyDown: He,
            placeholder: F,
            spellCheck: false,
            value: ze,
            role: Xe
        }, Ye),  n().createElement(S(), {
            anchor: function() {
                return we.current;
            }(),
            autoCloseWhenOffScreen: true,
            canCoverAnchor: false,
            defaultPlacement: a,
            id: he,
            onRequestClose: Ke,
            open: Ue,
            repositionMode: "flip"
        }, We));
    }
    ce.propTypes = ae;
    ce.Option = U;
    ce.Divider = g.Divider;
    ce.Heading = g.Heading;
    /* harmony default export */    const fe = ce;
    // CONCATENATED MODULE: ./src/Search/index.ts
    module.exports = r;
    /******/})();
