// @splunk/react-ui/SidePanel.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var r = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(r, {
                a: r
            });
            /******/            return r;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        default: () => /* reexport */ K
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    var n =  e.n(r);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var i =  e.n(o);
    // CONCATENATED MODULE: external "lodash/keys"
    const a = require("lodash/keys");
    var l =  e.n(a);
    // CONCATENATED MODULE: external "lodash/omit"
    const u = require("lodash/omit");
    var c =  e.n(u);
    // CONCATENATED MODULE: external "@splunk/react-ui/ModalLayer"
    const s = require("@splunk/react-ui/ModalLayer");
    var f =  e.n(s);
    // CONCATENATED MODULE: external "styled-components"
    const p = require("styled-components");
    var y =  e.n(p);
    // CONCATENATED MODULE: external "@splunk/react-ui/TransitionOpen"
    const d = require("@splunk/react-ui/TransitionOpen");
    var b =  e.n(d);
    // CONCATENATED MODULE: external "@splunk/themes"
    const m = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/SidePanel/SidePanelStyles.ts
    var v = (0, p.css)([ "", " position:fixed;flex-direction:column;box-shadow:", ";background-color:", ";z-index:", ";" ], m.mixins.reset("flex"), (0, 
    m.pick)({
        enterprise: "0 0 12px rgba(0, 0, 0, 0.3)",
        prisma: m.variables.overlayShadow
    }), m.variables.backgroundColorSidebar, m.variables.zindexModal);
    var h = y()(b()).withConfig({
        displayName: "SidePanelStyles__StyledLeftPanelTransitionOpen",
        componentId: "mtbcuz-0"
    })([ "", ";top:0;left:0;height:100vh;" ], v);
    var g = y()(b()).withConfig({
        displayName: "SidePanelStyles__StyledRightPanelTransitionOpen",
        componentId: "mtbcuz-1"
    })([ "", ";right:0;top:0;height:100vh;" ], v);
    var O = y()(b()).withConfig({
        displayName: "SidePanelStyles__StyledTopPanelTransitionOpen",
        componentId: "mtbcuz-2"
    })([ "", ";top:0;left:0;width:100vw;" ], v);
    var S = y()(b()).withConfig({
        displayName: "SidePanelStyles__StyledBottomPanelTransitionOpen",
        componentId: "mtbcuz-3"
    })([ "", ";left:0;bottom:0;width:100vw;" ], v);
    // CONCATENATED MODULE: ./src/SidePanel/SidePanel.tsx
    function P(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            P = function e(t) {
                return typeof t;
            };
        } else {
            P = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return P(e);
    }
    function w() {
        w = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return w.apply(this, arguments);
    }
    function j(e) {
        return k(e) || _(e) || C(e) || A();
    }
    function A() {
        throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function C(e, t) {
        if (!e) return;
        if (typeof e === "string") return E(e, t);
        var r = Object.prototype.toString.call(e).slice(8, -1);
        if (r === "Object" && e.constructor) r = e.constructor.name;
        if (r === "Map" || r === "Set") return Array.from(e);
        if (r === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(r)) return E(e, t);
    }
    function _(e) {
        if (typeof Symbol !== "undefined" && Symbol.iterator in Object(e)) return Array.from(e);
    }
    function k(e) {
        if (Array.isArray(e)) return E(e);
    }
    function E(e, t) {
        if (t == null || t > e.length) t = e.length;
        for (var r = 0, n = new Array(t); r < t; r++) {
            n[r] = e[r];
        }
        return n;
    }
    function T(e, t) {
        var r = Object.keys(e);
        if (Object.getOwnPropertySymbols) {
            var n = Object.getOwnPropertySymbols(e);
            if (t) n = n.filter((function(t) {
                return Object.getOwnPropertyDescriptor(e, t).enumerable;
            }));
            r.push.apply(r, n);
        }
        return r;
    }
    function x(e) {
        for (var t = 1; t < arguments.length; t++) {
            var r = arguments[t] != null ? arguments[t] : {};
            if (t % 2) {
                T(Object(r), true).forEach((function(t) {
                    U(e, t, r[t]);
                }));
            } else if (Object.getOwnPropertyDescriptors) {
                Object.defineProperties(e, Object.getOwnPropertyDescriptors(r));
            } else {
                T(Object(r)).forEach((function(t) {
                    Object.defineProperty(e, t, Object.getOwnPropertyDescriptor(r, t));
                }));
            }
        }
        return e;
    }
    function q(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function R(e, t) {
        for (var r = 0; r < t.length; r++) {
            var n = t[r];
            n.enumerable = n.enumerable || false;
            n.configurable = true;
            if ("value" in n) n.writable = true;
            Object.defineProperty(e, n.key, n);
        }
    }
    function F(e, t, r) {
        if (t) R(e.prototype, t);
        if (r) R(e, r);
        return e;
    }
    function M(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) N(e, t);
    }
    function N(e, t) {
        N = Object.setPrototypeOf || function e(t, r) {
            t.__proto__ = r;
            return t;
        };
        return N(e, t);
    }
    function I(e) {
        var t = L();
        return function r() {
            var n = B(e), o;
            if (t) {
                var i = B(this).constructor;
                o = Reflect.construct(n, arguments, i);
            } else {
                o = n.apply(this, arguments);
            }
            return D(this, o);
        };
    }
    function D(e, t) {
        if (t && (P(t) === "object" || typeof t === "function")) {
            return t;
        }
        return z(e);
    }
    function z(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function L() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function B(e) {
        B = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return B(e);
    }
    function U(e, t, r) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: r,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = r;
        }
        return e;
    }
    var $ = {
        children: i().node,
        dockPosition: i().oneOf([ "top", "bottom", "left", "right" ]),
        innerClassName: i().string,
        innerStyle: i().object,
        onAnimationEnd: i().func,
        onRequestClose: i().func,
        open: i().bool,
        outerClassName: i().string,
        outerStyle: i().object,
        useLayerForClickAway: i().bool
    };
    var G = {
        dockPosition: "right",
        open: false,
        useLayerForClickAway: true
    };
    var H = {
        left: "slideFromLeft",
        right: "slideFromRight",
        top: "slideFromTop",
        bottom: "slideFromBottom"
    };
    var J =  function(e) {
        M(r, e);
        var t = I(r);
        // @docs-props-type SidePanelPropsBase
                function r(e) {
            var o;
            q(this, r);
            o = t.call(this, e);
            U(z(o), "handlePanelAnimationEnd", (function() {
                var e, t;
                o.setState({
                    panelAnimating: false
                });
                (e = (t = o.props).onAnimationEnd) === null || e === void 0 ? void 0 : e.call(t);
            }));
            U(z(o), "renderModal", (function() {
                var e = o.props, t = e.children, i = e.dockPosition, a = e.innerClassName, u = e.innerStyle, s = e.outerClassName, f = e.outerStyle;
                var p = {
                    left: h,
                    right: g,
                    top: O,
                    bottom: S
                }[i];
                var y = i === "left" || i === "right" ? {
                    height: "100%"
                } : {
                    width: "100%"
                };
                
                return n().createElement(p, w({
                    animateOnMount: true,
                    animation: H[i],
                    "data-test": "side-panel",
                    innerClassName: a,
                    innerStyle: x(x({}, y), u),
                    onAnimationEnd: o.handlePanelAnimationEnd,
                    open: o.props.open,
                    outerClassName: s,
                    outerStyle: f,
                    retainFocus: true,
                    role: "dialog",
                    tabIndex: -1,
                    takeFocus: true
                }, c()(o.props, [].concat(j(l()(r.propTypes)), [ "onAnimationEnd" ]))), t);
            }));
            o.state = {
                panelAnimating: false,
                prevOpen: e.open
            };
            return o;
        }
        F(r, [ {
            key: "render",
            value: function e() {
                var t = this.props, r = t.open, o = t.onRequestClose, i = t.useLayerForClickAway;
                var a = this.state.panelAnimating;
                var l = i ? "interactive" : "hidden";
                
                return n().createElement(f(), {
                    childrenInAnimation: a,
                    onRequestClose: o,
                    open: r,
                    renderModal: this.renderModal,
                    scrim: l
                });
            }
        } ]);
        return r;
    }(r.Component);
    U(J, "propTypes", $);
    U(J, "defaultProps", G);
    U(J, "getDerivedStateFromProps", (function(e, t) {
        if (e.open !== t.prevOpen) {
            return {
                panelAnimating: true,
                prevOpen: e.open
            };
        }
        return null;
    }));
    /* harmony default export */    const K = J;
    // CONCATENATED MODULE: ./src/SidePanel/index.ts
    module.exports = t;
    /******/})();
