// @splunk/react-ui/SplitButton.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var r = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(r, {
                a: r
            });
            /******/            return r;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        Item: () => /* reexport */ F,
        default: () => /* reexport */ U
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    var n =  e.n(r);
    // CONCATENATED MODULE: external "prop-types"
    const i = require("prop-types");
    var a =  e.n(i);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/ChevronDown"
    const o = require("@splunk/react-icons/enterprise/ChevronDown");
    var l =  e.n(o);
    // CONCATENATED MODULE: external "@splunk/react-ui/Dropdown"
    const c = require("@splunk/react-ui/Dropdown");
    var u =  e.n(c);
    // CONCATENATED MODULE: external "@splunk/react-ui/Menu"
    const s = require("@splunk/react-ui/Menu");
    var p =  e.n(s);
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const f = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "styled-components"
    const d = require("styled-components");
    var v =  e.n(d);
    // CONCATENATED MODULE: external "@splunk/react-ui/Button"
    const b = require("@splunk/react-ui/Button");
    var y =  e.n(b);
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const h = require("@splunk/react-ui/Box");
    var m =  e.n(h);
    // CONCATENATED MODULE: external "@splunk/themes"
    const g = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/SplitButton/SplitButtonStyles.ts
    var O = v()(y()).withConfig({
        displayName: "SplitButtonStyles__StyledButton",
        componentId: "sc-9igsyh-0"
    })([ "", "" ], (0, g.pick)({
        prisma: function e(t) {
            var r = t.appearance;
            return r === "default" && (0, d.css)([ "&:first-child{border-right:1px solid ", ";}" ], g.variables.interactiveColorBorder);
        }
    }));
    var k = v()(m()).withConfig({
        displayName: "SplitButtonStyles__StyledSeparator",
        componentId: "sc-9igsyh-1"
    })([ "width:", ";height:", ";background-color:", ";" ], (0, g.pick)({
        prisma: (0, g.pickVariant)("$appearance", {
            primary: "2px",
            destructive: "2px"
        }),
        enterprise: "1px"
    }), g.variables.inputHeight, (0, g.pickVariant)("$appearance", {
        primary: {
            enterprise: g.variables.white,
            prisma: {
                light: g.variables.white,
                dark: g.variables.black
            }
        },
        destructive: {
            enterprise: {
                light: g.variables.white,
                dark: g.variables.black
            },
            prisma: {
                light: g.variables.white,
                dark: g.variables.black
            }
        }
    }));
    var w = v()(m()).withConfig({
        displayName: "SplitButtonStyles__StyledBox",
        componentId: "sc-9igsyh-2"
    })([ "", " [data-inline] + &{margin-left:", ";}" ], (function(e) {
        var t = e.$inline;
        return !t && (0, d.css)([ "width:100%;" ]);
    }), (0, g.pick)({
        enterprise: g.variables.spacingHalf,
        prisma: g.variables.spacingSmall
    }));
    // CONCATENATED MODULE: external "lodash/omit"
    const S = require("lodash/omit");
    var j =  e.n(S);
    // CONCATENATED MODULE: ./src/SplitButton/Item.tsx
    function _(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            _ = function e(t) {
                return typeof t;
            };
        } else {
            _ = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return _(e);
    }
    function C() {
        C = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return C.apply(this, arguments);
    }
    function P(e, t) {
        if (e == null) return {};
        var r = E(e, t);
        var n, i;
        if (Object.getOwnPropertySymbols) {
            var a = Object.getOwnPropertySymbols(e);
            for (i = 0; i < a.length; i++) {
                n = a[i];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function E(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var i, a;
        for (a = 0; a < n.length; a++) {
            i = n[a];
            if (t.indexOf(i) >= 0) continue;
            r[i] = e[i];
        }
        return r;
    }
    function x(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function I(e, t) {
        for (var r = 0; r < t.length; r++) {
            var n = t[r];
            n.enumerable = n.enumerable || false;
            n.configurable = true;
            if ("value" in n) n.writable = true;
            Object.defineProperty(e, n.key, n);
        }
    }
    function q(e, t, r) {
        if (t) I(e.prototype, t);
        if (r) I(e, r);
        return e;
    }
    function R(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) M(e, t);
    }
    function M(e, t) {
        M = Object.setPrototypeOf || function e(t, r) {
            t.__proto__ = r;
            return t;
        };
        return M(e, t);
    }
    function B(e) {
        var t = V();
        return function r() {
            var n = $(e), i;
            if (t) {
                var a = $(this).constructor;
                i = Reflect.construct(n, arguments, a);
            } else {
                i = n.apply(this, arguments);
            }
            return T(this, i);
        };
    }
    function T(e, t) {
        if (t && (_(t) === "object" || typeof t === "function")) {
            return t;
        }
        return N(e);
    }
    function N(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function V() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function $(e) {
        $ = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return $(e);
    }
    function D(e, t, r) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: r,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = r;
        }
        return e;
    }
    var A = {
        appearance: a().oneOf([ "default", "primary", "destructive" ]),
        children: a().node,
        disabled: a().bool,
        isMain: a().bool,
        onClick: a().func
    };
    var H = {
        appearance: "default"
    };
    /**
 * An item within a `SplitButton`. This inherits from
 * [PureComponent](https://reactjs.org/docs/react-api.html#reactpurecomponent)
 * so any elements passed to it must also be pure.
 */    var z =  function(e) {
        R(r, e);
        var t = B(r);
        function r() {
            var e;
            x(this, r);
            for (var n = arguments.length, i = new Array(n), a = 0; a < n; a++) {
                i[a] = arguments[a];
            }
            e = t.call.apply(t, [ this ].concat(i));
            D(N(e), "item", null);
            D(N(e), "handleClick", (function(t) {
                var r = e.props, n = r.disabled, i = r.onClick;
                if (!n) {
                    i === null || i === void 0 ? void 0 : i(t);
                }
            }));
            return e;
        }
        q(r, [ {
            key: "scrollIntoViewIfNeeded",
            value: function e() {
                var t;
                (t = this.item) === null || t === void 0 ? void 0 : t.scrollIntoViewIfNeeded();
            }
            /**
     * Place focus on the item.
     */        }, {
            key: "focus",
            value: function e() {
                var t;
                (t = this.item) === null || t === void 0 ? void 0 : t.focus();
            }
        }, {
            key: "render",
            value: function e() {
                var t = this;
                var r = this.props, i = r.disabled, a = r.isMain, o = r.children, l = r.appearance, c = P(r, [ "disabled", "isMain", "children", "appearance" ]);
                if (a) {
                    
                    return n().createElement(O, C({
                        onClick: this.handleClick,
                        disabled: i,
                        append: true,
                        appearance: l
                    }, c), o);
                }
                
                return n().createElement(p().Item, C({
                    ref: function e(r) {
                        t.item = r;
                    },
                    onClick: this.handleClick,
                    disabled: i
                }, j()(c, [ "role" ])), o);
            }
        } ]);
        return r;
    }(r.PureComponent);
    D(z, "propTypes", A);
    D(z, "defaultProps", H);
    /* harmony default export */    const F = z;
    // CONCATENATED MODULE: ./src/SplitButton/SplitButton.tsx
    function G() {
        G = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return G.apply(this, arguments);
    }
    function J(e, t) {
        if (e == null) return {};
        var r = K(e, t);
        var n, i;
        if (Object.getOwnPropertySymbols) {
            var a = Object.getOwnPropertySymbols(e);
            for (i = 0; i < a.length; i++) {
                n = a[i];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function K(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var i, a;
        for (a = 0; a < n.length; a++) {
            i = n[a];
            if (t.indexOf(i) >= 0) continue;
            r[i] = e[i];
        }
        return r;
    }
    var L = {
        appearance: a().oneOf([ "default", "primary", "destructive" ]),
        children: a().node,
        disabled: a().bool,
        elementRef: a().oneOfType([ a().func, a().object ]),
        onClick: a().func,
        inline: a().bool
    };
    function Q(e) {
        var t = e.appearance, i = t === void 0 ? "default" : t, a = e.children, o = e.disabled, c = e.elementRef, s = e.inline, d = s === void 0 ? true : s, v = e.onClick, b = J(e, [ "appearance", "children", "disabled", "elementRef", "inline", "onClick" ]);
        var y = r.Children.toArray(a).filter(r.isValidElement);
        var h = 0;
        var m = 0;
        y.forEach((function(e, t) {
            var r = e.props, n = r.disabled, i = r.isMain;
            if (i) {
                h = t;
            }
            if (n) {
                m += 1;
            }
        }));
        var g =  (0, r.cloneElement)(y[h], {
            isMain: true,
            appearance: i,
            disabled: o || y[h].props.disabled
        });
 // Remove mainButton from array of dropdown items
                y.splice(h, 1);
        var S = y;
        var j =  n().createElement(O, {
            prepend: true,
            disabled: o || m === S.length,
            appearance: i,
            "data-test": "split-button-toggle",
            "aria-label": (0, f._)("Open more options"),
            icon:  n().createElement(l(), null)
        });
        
        return n().createElement(w, G({
            "aria-label": (0, f._)("Split button"),
            "data-inline": d || null,
            "data-test": "split-button-container",
            elementRef: c,
            $inline: d,
            inline: true,
            flex: true,
            role: "group",
            onClick: o ? undefined : v
        }, b), g, i !== "default" &&  n().createElement(k, {
            inline: true,
            $appearance: i
        }),  n().createElement("div", null,  n().createElement(u(), {
            toggle: j
        },  n().createElement(p(), null, S))));
    }
    Q.propTypes = L;
    Q.Item = F;
    /* harmony default export */    const U = Q;
    // CONCATENATED MODULE: ./src/SplitButton/index.ts
    module.exports = t;
    /******/})();
