// @splunk/react-ui/StaticContent.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var t = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(t, {
                a: t
            });
            /******/            return t;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, t) => {
            /******/ for (var n in t) {
                /******/ if (e.o(t, n) && !e.o(r, n)) {
                    /******/ Object.defineProperty(r, n, {
                        enumerable: true,
                        get: t[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        default: () => /* reexport */ b
    });
    // CONCATENATED MODULE: external "react"
    const t = require("react");
    var n =  e.n(t);
    // CONCATENATED MODULE: external "prop-types"
    const i = require("prop-types");
    var o =  e.n(i);
    // CONCATENATED MODULE: external "styled-components"
    const a = require("styled-components");
    var l =  e.n(a);
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const p = require("@splunk/react-ui/Box");
    var c =  e.n(p);
    // CONCATENATED MODULE: external "@splunk/themes"
    const s = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/StaticContent/StaticContentStyles.ts
    var u = l()(c()).withConfig({
        displayName: "StaticContentStyles__StyledBox",
        componentId: "sxqyrp-0"
    })([ "", ";position:relative;margin:0;font-weight:", ";word-wrap:break-word;min-height:", ";padding:", ";[data-inline] + &[data-inline]{margin-left:", ";}" ], s.mixins.reset("flex"), s.variables.fontWeightSemiBold, s.variables.inputHeight, (0, 
    s.pick)({
        enterprise: {
            comfortable: "6px 7px",
            compact: "4px 5px"
        },
        prisma: {
            comfortable: "10px 0",
            compact: "6px 0"
        }
    }), (0, s.pick)({
        enterprise: s.variables.spacingQuarter,
        prisma: s.variables.spacingLarge
    }));
    // CONCATENATED MODULE: ./src/StaticContent/StaticContent.tsx
    function f() {
        f = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return f.apply(this, arguments);
    }
    function d(e, r) {
        if (e == null) return {};
        var t = m(e, r);
        var n, i;
        if (Object.getOwnPropertySymbols) {
            var o = Object.getOwnPropertySymbols(e);
            for (i = 0; i < o.length; i++) {
                n = o[i];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function m(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var i, o;
        for (o = 0; o < n.length; o++) {
            i = n[o];
            if (r.indexOf(i) >= 0) continue;
            t[i] = e[i];
        }
        return t;
    }
    var v = {
        children: o().node,
        elementRef: o().oneOfType([ o().func, o().object ]),
        inline: o().bool,
        size: o().oneOf([ "small", "medium" ])
    };
    /**
 * This component is intended for use in a control group, either to display a static value or
 * between two controls.
 */    function y(e) {
        var r = e.children, t = e.inline, i = t === void 0 ? false : t, o = e.size, a = o === void 0 ? "medium" : o, l = d(e, [ "children", "inline", "size" ]);
        // @docs-props-type StaticContentPropsBase
                
        return n().createElement(u, f({
            "data-test": "static-content",
            "data-size": a,
            inline: i
        }, l), r);
    }
    y.propTypes = v;
    /* harmony default export */    const b = y;
    // CONCATENATED MODULE: ./src/StaticContent/index.ts
    module.exports = r;
    /******/})();
