// @splunk/react-ui/StepBar.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var r = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(r, {
                a: r
            });
            /******/            return r;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        Step: () => /* reexport */ $,
        default: () => /* reexport */ z
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    var n =  e.n(r);
    // CONCATENATED MODULE: external "prop-types"
    const i = require("prop-types");
    var o =  e.n(i);
    // CONCATENATED MODULE: external "@splunk/ui-utils/id"
    const a = require("@splunk/ui-utils/id");
    // CONCATENATED MODULE: external "@splunk/themes/useSplunkTheme"
    const l = require("@splunk/themes/useSplunkTheme");
    var s =  e.n(l);
    // CONCATENATED MODULE: external "styled-components"
    const c = require("styled-components");
    var p =  e.n(c);
    // CONCATENATED MODULE: external "@splunk/react-icons/CheckCircle"
    const d = require("@splunk/react-icons/CheckCircle");
    var u =  e.n(d);
    // CONCATENATED MODULE: external "@splunk/react-icons/ExclamationCircle"
    const v = require("@splunk/react-icons/ExclamationCircle");
    var f =  e.n(v);
    // CONCATENATED MODULE: external "@splunk/themes"
    const m = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/StepBar/StepStyles.ts
    var y = "#1A8929";
 // SUI-2439 to meet WCAG AA compliance
        var x = p().li.withConfig({
        displayName: "StepStyles__Styled",
        componentId: "sc-756fxp-0"
    })([ "", " ", ";text-align:center;padding:", ";", ";", "" ], m.mixins.reset("block"), (0, 
    m.pick)({
        enterprise: (0, c.css)([ "position:relative;flex:1 1 0;" ]),
        prisma: (0, c.css)([ "display:flex;align-items:center;flex:0 1 auto;margin:0 12px 0 12px;" ])
    }), (0, m.pick)({
        enterprise: (0, c.css)([ "25px 15px 0" ]),
        prisma: {
            comfortable: "8px 8px 8px 0",
            compact: "4px 8px 4px 0"
        }
    }), (0, m.pick)({
        prisma: (0, c.css)([ "&::before{align-items:center;box-sizing:border-box;border-radius:50%;counter-increment:", ";content:counter(", ");display:inline-flex;flex-shrink:0;font-size:12px;height:20px;width:20px;justify-content:center;line-height:20px;margin:2px 16px 2px 18px;}" ], (function(e) {
            var t = e.$idCounter;
            return t;
        }), (function(e) {
            var t = e.$idCounter;
            return t;
        }))
    }), (0, m.pickVariant)("$status", {
        error: {
            prisma: (0, c.css)([ "color:", ";&::before{visibility:hidden;position:absolute;}" ], m.variables.contentColorActive)
        },
        active: {
            prisma: (0, c.css)([ "color:", ";&::before{color:", ";background:", ";border:1px solid ", ";}" ], m.variables.contentColorActive, m.variables.contentColorInverted, m.variables.contentColorActive, m.variables.contentColorActive)
        },
        next: {
            enterprise: (0, c.css)([ "color:", ";" ], m.variables.textColor),
            prisma: (0, c.css)([ "color:", ";&::before{color:", ";border:1px solid ", ";}" ], m.variables.contentColorMuted, m.variables.contentColorMuted, m.variables.contentColorMuted)
        },
        prev: {
            prisma: (0, c.css)([ "color:", ";&::before{visibility:hidden;position:absolute;}" ], m.variables.contentColorDefault)
        }
    }));
    var b = (0, c.css)([ "height:24px;width:24px;margin:0 14px 0 16px;" ]);
    var h = p()(u()).withConfig({
        displayName: "StepStyles__StyledCheckCircleIcon",
        componentId: "sc-756fxp-1"
    })([ "", ";color:", ";" ], b, m.variables.accentColorPositive);
    var g = p()(f()).withConfig({
        displayName: "StepStyles__StyledExclamationCircleIcon",
        componentId: "sc-756fxp-2"
    })([ "", ";color:", ";" ], b, m.variables.accentColorNegative);
    var S = p().svg.withConfig({
        displayName: "StepStyles__StyledSvg",
        componentId: "sc-756fxp-3"
    })([ "position:absolute;left:0;right:0;top:0;" ]);
    var C = p().rect.withConfig({
        displayName: "StepStyles__StyledPrevOrActiveRect",
        componentId: "sc-756fxp-4"
    })([ "fill:", ";" ], y);
    var w = C.withComponent("circle");
    var O = p().rect.withConfig({
        displayName: "StepStyles__StyledGray",
        componentId: "sc-756fxp-5"
    })([ "fill:", ";" ], m.variables.gray60);
    var I = p().circle.withConfig({
        displayName: "StepStyles__StyledNext",
        componentId: "sc-756fxp-6"
    })([ "stroke:", ";stroke-width:3px;fill:", ";" ], m.variables.gray60, m.variables.backgroundColor);
    // CONCATENATED MODULE: ./src/StepBar/StepBarContext.tsx
    var k =  (0, r.createContext)({
        activeStepId: 0
    });
    k.displayName = "StepBar";
    /* harmony default export */    const E = k;
    // CONCATENATED MODULE: ./src/StepBar/Step.tsx
    function _() {
        _ = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return _.apply(this, arguments);
    }
    function j(e, t) {
        if (e == null) return {};
        var r = P(e, t);
        var n, i;
        if (Object.getOwnPropertySymbols) {
            var o = Object.getOwnPropertySymbols(e);
            for (i = 0; i < o.length; i++) {
                n = o[i];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function P(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var i, o;
        for (o = 0; o < n.length; o++) {
            i = n[o];
            if (t.indexOf(i) >= 0) continue;
            r[i] = e[i];
        }
        return r;
    }
    var N = {
        children: o().node,
        elementRef: o().oneOfType([ o().func, o().object ]),
        error: o().bool,
        stepId: o().any
    };
    function R(e) {
        var t = e.children, i = e.elementRef, o = e.error, a = o === void 0 ? false : o, l = e.stepId, c = j(e, [ "children", "elementRef", "error", "stepId" ]);
        // @docs-props-type StepPropsBase
                var p = (0, r.useContext)(k), d = p.activeStepId, u = p.idCounter, v = p.lastChildIndex;
        var f = s()(), m = f.isEnterprise;
        var y = "next";
        var b = d === l;
        var E = l === 0;
        var P = l === v;
        if (l < d) {
            y = "prev";
        }
        if (b && !a) {
            y = "active";
        }
        if (b && a) {
            y = "error";
        }
        if (m) {
            
            return n().createElement(x, _({
                $status: y,
                "data-test-status": y,
                "data-test": "step",
                "data-test-step-id": l,
                ref: i
            }, c),  n().createElement(S, {
                focusable: "false",
                width: "100%",
                height: "15px",
                viewBox: "0 0 100 15",
                preserveAspectRatio: "none",
                xmlns: "http://www.w3.org/2000/svg"
            },  n().createElement(O, {
                x: E ? "50" : "0",
                y: "6",
                width: P || E ? "50%" : "100%",
                height: "3"
            }), (y === "active" || y === "error") && !E &&  n().createElement(C, {
                x: "0",
                y: "6",
                width: "50%",
                height: "3"
            }), y === "prev" &&  n().createElement(C, {
                x: E ? "50%" : "0",
                y: "6",
                width: "100%",
                height: "3"
            })),  n().createElement(S, {
                focusable: "false",
                width: "100%",
                height: "15px",
                viewBox: "0 0 15 15",
                xmlns: "http://www.w3.org/2000/svg"
            }, y !== "next" &&  n().createElement(w, {
                cx: "50%",
                cy: "50%",
                r: "7.5"
            }), y === "next" &&  n().createElement(I, {
                cx: "50%",
                cy: "50%",
                r: "6"
            })), t);
        }
        
        return n().createElement(x, _({
            $status: y,
            "data-test-status": y,
            "data-test": "step",
            "data-test-step-id": l,
            $idCounter: u,
            ref: i
        }, c), y === "prev" &&  n().createElement(h, null), y === "error" &&  n().createElement(g, null), t);
    }
    R.propTypes = N;
    /* harmony default export */    const $ = R;
    // CONCATENATED MODULE: ./src/StepBar/StepBarStyles.ts
    var q = p().ul.withConfig({
        displayName: "StepBarStyles__Styled",
        componentId: "sc-13bf2x4-0"
    })([ "", ";", ";", ";" ], m.mixins.reset("flex"), (0, m.pick)({
        prisma: (0, c.css)([ "position:relative;display:flex;counter-reset:", ";" ], (function(e) {
            var t = e.$idCounter;
            return t;
        }))
    }), (function(e) {
        var t = e.$inline;
        return t && (0, c.css)([ "display:inline-flex;" ]);
    }));
    // CONCATENATED MODULE: ./src/StepBar/StepBar.tsx
    function A() {
        A = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return A.apply(this, arguments);
    }
    function M(e, t) {
        if (e == null) return {};
        var r = T(e, t);
        var n, i;
        if (Object.getOwnPropertySymbols) {
            var o = Object.getOwnPropertySymbols(e);
            for (i = 0; i < o.length; i++) {
                n = o[i];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function T(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var i, o;
        for (o = 0; o < n.length; o++) {
            i = n[o];
            if (t.indexOf(i) >= 0) continue;
            r[i] = e[i];
        }
        return r;
    }
    var B = {
        activeStepId: o().any.isRequired,
        children: o().node,
        elementRef: o().oneOfType([ o().func, o().object ]),
        inline: o().bool
    };
    function D(e) {
        var t = e.activeStepId, i = e.children, o = e.elementRef, l = e.inline, s = l === void 0 ? false : l, c = M(e, [ "activeStepId", "children", "elementRef", "inline" ]);
        // @docs-props-type StepBarPropsBase
                var p = (0, r.useRef)((0, a.createDOMID)("stepbar")), d = p.current;
        var u = 0;
        var v = r.Children.toArray(i).filter(r.isValidElement);
        var f = v.length - 1;
        var m = v.map((function(e, n) {
            var i = e.props.stepId || n;
            var o = e.props.error;
            if (i === t) {
                u = n;
            }
            
            return (0, r.cloneElement)(e, {
                stepId: i,
                error: o
            });
        }));
        
        return n().createElement(q, A({
            $inline: s,
            "data-test": "step-bar",
            "data-test-active-step-id": t,
            role: "progressbar",
            "aria-valuemin": 1,
            "aria-valuemax": f + 1,
            "aria-valuenow": u + 1,
            $idCounter: d,
            ref: o
        }, c),  n().createElement(E.Provider, {
            value: {
                idCounter: d,
                lastChildIndex: f,
                activeStepId: t
            }
        }, m));
    }
    D.propTypes = B;
    D.Step = $;
    /* harmony default export */    const z = D;
    // CONCATENATED MODULE: ./src/StepBar/index.ts
    module.exports = t;
    /******/})();
