// @splunk/react-ui/Switch.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var a = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(a, {
                a
            });
            /******/            return a;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, a) => {
            /******/ for (var t in a) {
                /******/ if (e.o(a, t) && !e.o(r, t)) {
                    /******/ Object.defineProperty(r, t, {
                        enumerable: true,
                        get: a[t]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        default: () => /* reexport */ X
    });
    // CONCATENATED MODULE: external "react"
    const a = require("react");
    var t =  e.n(a);
    // CONCATENATED MODULE: external "prop-types"
    const i = require("prop-types");
    var o =  e.n(i);
    // CONCATENATED MODULE: external "@splunk/react-ui/AnimationToggle"
    const n = require("@splunk/react-ui/AnimationToggle");
    var l =  e.n(n);
    // CONCATENATED MODULE: external "@splunk/react-ui/ScreenReaderContent"
    const s = require("@splunk/react-ui/ScreenReaderContent");
    var c =  e.n(s);
    // CONCATENATED MODULE: external "@splunk/ui-utils/id"
    const d = require("@splunk/ui-utils/id");
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Check"
    const p = require("@splunk/react-icons/enterprise/Check");
    var u =  e.n(p);
    // CONCATENATED MODULE: external "@splunk/themes"
    const b = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/Switch/CheckIcon.tsx
    function v() {
        var e = (0, b.useSplunkTheme)(), r = e.family;
        if (r === "enterprise") {
            
            return t().createElement(u(), {
                inline: false,
                size: "12px",
                screenReaderText: null,
                hideDefaultTooltip: true,
                style: {
                    marginTop: "1px"
                }
            });
        }
        
        return t().createElement("svg", {
            width: "12",
            height: "12",
            viewBox: "-1 -2 11 10",
            style: {
                display: "block"
            },
            xmlns: "http://www.w3.org/2000/svg"
        },  t().createElement("path", {
            d: "M8.70711 0.292893C9.09763 0.683417 9.09763 1.31658 8.70711 1.70711L4.32132 6.0929C3.73559 6.67862 2.78596 6.67869 2.20015 6.09305L0.292997 4.18646C-0.0975845 3.79599 -0.0976776 3.16282 0.292789 2.77224C0.683256 2.38166 1.31642 2.38157 1.707 2.77203L3.26061 4.32518L7.29289 0.292893C7.68342 -0.0976311 8.31658 -0.0976311 8.70711 0.292893Z",
            fill: "currentColor"
        }));
    }
    // CONCATENATED MODULE: external "styled-components"
    const f = require("styled-components");
    var h =  e.n(f);
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const g = require("@splunk/react-ui/Box");
    var m =  e.n(g);
    // CONCATENATED MODULE: external "@splunk/react-ui/Clickable"
    const y = require("@splunk/react-ui/Clickable");
    var k =  e.n(y);
    // CONCATENATED MODULE: ./src/Switch/SwitchStyles.ts
    var x = "18px";
    var C = "20px";
    var w = h()(m()).withConfig({
        displayName: "SwitchStyles__StyledBox",
        componentId: "sc-844ieu-0"
    })([ "display:inline;position:relative;color:", ";flex-shrink:0;padding:", ";[data-inline] + &{margin-left:", ";}" ], (0, 
    b.pickVariant)("$switchState", {
        error: {
            enterprise: b.variables.errorColor,
            prisma: b.variables.accentColorNegative
        },
        disabled: {
            enterprise: b.variables.textDisabledColor,
            prisma: b.variables.contentColorDisabled
        }
    }), (0, b.pick)({
        enterprise: (0, f.css)([ "calc((", " - ", ") / 2) 0" ], b.variables.inputHeight, x),
        prisma: {
            compact: "6px 0",
            comfortable: "10px 0"
        }
    }), (0, b.pick)({
        enterprise: b.variables.spacingHalf,
        prisma: b.variables.spacingSmall
    }));
    var S = h()(k()).withConfig({
        displayName: "SwitchStyles__StyledCheckboxClickable",
        componentId: "sc-844ieu-1"
    })([ "", " position:relative;width:", ";height:", ";line-height:", ";padding:2px;border-radius:2px;cursor:pointer;flex:0 0 auto;border:", ";color:", ";", " &:focus{box-shadow:", ";}", " ", " &[disabled]{cursor:not-allowed;border-color:", ";", "}" ], b.mixins.reset("inline"), x, x, x, (0, 
    b.pick)({
        enterprise: (0, b.pick)({
            light: (0, f.css)([ "1px solid ", "" ], b.variables.gray45),
            dark: (0, f.css)([ "1px solid ", "" ], b.variables.gray80)
        }),
        prisma: (0, f.css)([ "2px solid ", "" ], b.variables.contentColorMuted)
    }), (0, b.pick)({
        enterprise: (0, b.pick)({
            light: b.variables.gray45,
            dark: b.variables.gray80
        }),
        prisma: b.variables.white
    }), (0, b.pick)({
        prisma: (0, f.css)([ "top:1px;padding:1px;border-radius:3px;", " ", "" ], (function(e) {
            var r = e.$interactive;
            return r && (0, f.css)([ "&:not([disabled]):not(:focus):hover{box-shadow:0 0 0 2px ", ",0 0 0 5px ", ";}" ], b.variables.backgroundColorPage, b.variables.interactiveColorOverlayHover);
        }), (function(e) {
            var r = e.$selected;
            return !r && (0, f.css)([ "&:not([disabled]):not(:focus):hover{border-color:", ";}" ], b.variables.contentColorDefault);
        }))
    }), b.variables.focusShadow, (function(e) {
        var r = e.$selected;
        return r && (0, f.css)([ "&:not([disabled]){border-color:", ";background-color:", ";}" ], (0, 
        b.pick)({
            enterprise: {
                light: b.variables.gray45,
                dark: b.variables.gray80
            },
            prisma: b.variables.interactiveColorPrimary
        }), (0, b.pick)({
            enterprise: "#none",
            prisma: (0, f.css)([ "", ";" ], b.variables.interactiveColorPrimary)
        }));
    }), (function(e) {
        var r = e.$error;
        return r && (0, b.pick)({
            enterprise: (0, b.pick)({
                light: (0, f.css)([ "&:not([disabled]){border-color:", ";color:", ";", "}" ], b.variables.errorColorL30, b.variables.errorColor, (function(e) {
                    var r = e.$selected;
                    return r && (0, f.css)([ "border-color:", ";" ], b.variables.errorColor);
                })),
                dark: (0, f.css)([ "border-color:", ";}" ], b.variables.errorColor)
            }),
            prisma: (0, f.css)([ "&:not([disabled]){border-color:", ";", "}" ], b.variables.accentColorNegative, (function(e) {
                var r = e.$selected;
                return r && (0, f.css)([ "background-color:", ";" ], b.variables.accentColorNegative);
            }))
        });
    }), (0, b.pick)({
        enterprise: {
            light: b.variables.borderColor,
            dark: b.variables.gray45
        },
        prisma: b.variables.contentColorDisabled
    }), (function(e) {
        var r = e.$selected;
        return r && (0, b.pick)({
            prisma: (0, f.css)([ "border-color:transparent;background-color:", ";" ], b.variables.interactiveColorBackgroundDisabled)
        });
    }));
    var O = S.withComponent("span");
    var $ = h().div.withConfig({
        displayName: "SwitchStyles__StyledSome",
        componentId: "sc-844ieu-2"
    })([ "display:block;margin:", ";height:", ";width:", ";background:currentColor;border-radius:1px;" ], (0, 
    b.pick)({
        enterprise: "2px",
        prisma: "5px 2px"
    }), (0, b.pick)({
        enterprise: (0, f.css)([ "calc(", " - 10px);" ], x),
        prisma: "2px"
    }), (0, b.pick)({
        enterprise: (0, f.css)([ "calc(", " - 10px);" ], x),
        prisma: "8px"
    }));
    var _ = h().div.withConfig({
        displayName: "SwitchStyles__StyledIndicator",
        componentId: "sc-844ieu-3"
    })([ "background-color:", ";border-color:", ";box-sizing:border-box;border-radius:50%;border-width:1px;position:absolute;left:-1px;top:-1px;", ";", " ", " ", "" ], (0, 
    b.pick)({
        enterprise: {
            light: b.variables.gray98,
            dark: b.variables.gray80
        },
        prisma: b.variables.white
    }), (0, b.pick)({
        enterprise: {
            light: b.variables.gray60,
            dark: b.variables.borderColor
        },
        prisma: b.variables.transparent
    }), (function(e) {
        var r = e.$delay;
        return r && (0, f.css)([ "transition:left ", ";" ], r);
    }), (0, b.pick)({
        enterprise: (0, f.css)([ "width:", ";height:", ";border-style:solid;" ], x, x),
        prisma: (0, f.css)([ "height:16px;width:16px;margin:2px;border-style:none;box-shadow:", ";" ], b.variables.embossShadow)
    }), (function(e) {
        var r = e.$disabled;
        return r && (0, f.css)([ "border-color:", ";border-style:solid;background-color:", ";" ], (0, 
        b.pick)({
            enterprise: {
                light: b.mixins.colorWithAlpha(b.variables.borderLightColor, .8),
                dark: b.variables.gray22
            },
            prisma: b.variables.transparent
        }), (0, b.pick)({
            enterprise: {
                dark: b.variables.gray45
            },
            prisma: b.mixins.colorWithAlpha(b.variables.white, .3)
        }));
    }), (function(e) {
        var r = e.$selected;
        return r && (0, f.css)([ "left:", ";" ], (0, b.pick)({
            enterprise: (0, f.css)([ "calc(100% - ", " + 1px)" ], x),
            prisma: (0, f.css)([ "calc(100% - ", " + 1px)" ], C)
        }));
    }));
    var P = h().div.withConfig({
        displayName: "SwitchStyles__StyledToggleOutline",
        componentId: "sc-844ieu-4"
    })([ "position:absolute;border:1px solid transparent;", ";border-radius:", ";top:-1px;right:-1px;bottom:-1px;left:-1px;z-index:1;", " ", "" ], (function(e) {
        var r = e.$delay;
        return r && (0, f.css)([ "transition:border-color ", ";" ], r);
    }), (0, b.pick)({
        enterprise: (0, f.css)([ "calc(", " * 0.5)" ], x),
        prisma: "15px"
    }), (0, b.pick)({
        prisma: (0, f.css)([ "margin:-6px;" ])
    }), (function(e) {
        var r = e.$error;
        return r && (0, f.css)([ "border-color:", ";" ], (0, b.pick)({
            enterprise: b.variables.errorColor,
            prisma: b.variables.transparent
        }));
    }));
    var E = h()(k()).withConfig({
        displayName: "SwitchStyles__StyledToggleClickable",
        componentId: "sc-844ieu-5"
    })([ "position:relative;", " border-radius:", ";transition:background-color ", ";flex:0 0 auto;border:1px solid ", ";", " &:not([disabled]){&:focus{outline:0;box-shadow:", ";", "}&:hover ", "{background-color:", ";}&:hover ", "{background-color:", ";}};&[disabled]{border:1px solid ", ";background-color:", ";", " ", "}", ";", "" ], (0, 
    b.pick)({
        enterprise: (0, b.pick)({
            light: (0, f.css)([ "width:calc(", " * 2);background-color:", ";height:", ";" ], x, b.variables.gray98, x),
            dark: (0, f.css)([ "width:calc(", " * 2);background-color:", ";height:", ";" ], x, b.variables.gray45, x)
        }),
        prisma: (0, f.css)([ "width:32px;background-color:", ";height:", ";top:1px;" ], b.variables.contentColorMuted, C)
    }), x, (function(e) {
        return e.$delay;
    }), (0, b.pick)({
        enterprise: {
            light: b.variables.gray60,
            dark: b.variables.borderColor
        },
        prisma: b.variables.transparent
    }), (0, b.pick)({
        enterprise: (0, f.css)([ "box-shadow:inset 0 2px 0 rgba(0,0,0,0.1);" ])
    }), (0, b.pick)({
        enterprise: (0, f.css)([ "", ",inset 0 2px 0 rgba(0,0,0,0.1);" ], b.variables.focusShadow),
        prisma: b.variables.focusShadow
    }), (0, b.pick)({
        enterprise: (0, f.css)([ "> ", "{border-color:", ";}" ], 
        /* sc-sel */
        P, b.mixins.colorWithAlpha(b.variables.focusColor, .8))
    }), 
    /* sc-sel */
    _, (0, b.pick)({
        enterprise: (0, b.pick)({
            light: b.variables.gray96,
            dark: "#b1bcc7"
        }),
        prisma: b.variables.white
    }), 
    /* sc-sel */
    P, (0, b.pick)({
        prisma: b.variables.interactiveColorOverlayHover
    }), (0, b.pick)({
        enterprise: {
            light: b.variables.borderLightColor,
            dark: b.variables.borderColor
        },
        prisma: b.variables.transparent
    }), (0, b.pick)({
        enterprise: {
            light: b.variables.gray96,
            dark: b.variables.gray30
        },
        prisma: b.variables.interactiveColorBackgroundDisabled
    }), (0, b.pick)({
        enterprise: (0, f.css)([ "box-shadow:inset 0 2px 0 rgba(0,0,0,0.06);" ])
    }), (function(e) {
        var r = e.$selected;
        return r && (0, f.css)([ "border-color:transparent;background-color:", ";" ], (0, 
        b.pick)({
            enterprise: (0, b.pick)({
                light: b.variables.accentColorL40,
                dark: b.variables.accentColorD20
            }),
            prisma: b.mixins.colorWithAlpha(b.variables.interactiveColorPrimary, .7)
        }));
    }), (function(e) {
        var r = e.$selected;
        return r && (0, f.css)([ "border-color:", ";background-color:", ";", "" ], (0, b.pick)({
            enterprise: b.variables.accentColorL10,
            prisma: b.variables.transparent
        }), (0, b.pick)({
            enterprise: b.variables.accentColorL10,
            prisma: b.variables.interactiveColorPrimary
        }), (0, b.pick)({
            prisma: (0, f.css)([ "box-shadow:", ";" ], b.variables.embossShadow)
        }));
    }), (function(e) {
        var r = e.$error;
        return r && (0, f.css)([ "border-color:", ";", ";" ], (0, b.pick)({
            enterprise: b.variables.errorColor,
            prisma: b.variables.transparent
        }), (0, b.pick)({
            prisma: (0, f.css)([ "background-color:", ";" ], b.variables.accentColorNegative)
        }));
    }));
    var L = E.withComponent("span");
    var j = h().label.withConfig({
        displayName: "SwitchStyles__StyledLabel",
        componentId: "sc-844ieu-6"
    })([ "", " flex:1 1 auto;padding-left:", ";color:inherit;", ";" ], b.mixins.reset("inline-block"), (0, 
    b.pick)({
        enterprise: (0, f.css)([ "", ";" ], b.variables.spacingQuarter),
        prisma: "8px"
    }), (function(e) {
        var r = e.$disabled;
        return !r && (0, f.css)([ "cursor:pointer;" ]);
    }));
    // CONCATENATED MODULE: ./src/Switch/Switch.tsx
    function I(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            I = function e(r) {
                return typeof r;
            };
        } else {
            I = function e(r) {
                return r && typeof Symbol === "function" && r.constructor === Symbol && r !== Symbol.prototype ? "symbol" : typeof r;
            };
        }
        return I(e);
    }
    function D() {
        D = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var a = arguments[r];
                for (var t in a) {
                    if (Object.prototype.hasOwnProperty.call(a, t)) {
                        e[t] = a[t];
                    }
                }
            }
            return e;
        };
        return D.apply(this, arguments);
    }
    function q(e, r) {
        if (e == null) return {};
        var a = T(e, r);
        var t, i;
        if (Object.getOwnPropertySymbols) {
            var o = Object.getOwnPropertySymbols(e);
            for (i = 0; i < o.length; i++) {
                t = o[i];
                if (r.indexOf(t) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, t)) continue;
                a[t] = e[t];
            }
        }
        return a;
    }
    function T(e, r) {
        if (e == null) return {};
        var a = {};
        var t = Object.keys(e);
        var i, o;
        for (o = 0; o < t.length; o++) {
            i = t[o];
            if (r.indexOf(i) >= 0) continue;
            a[i] = e[i];
        }
        return a;
    }
    function N(e, r) {
        if (!(e instanceof r)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function R(e, r) {
        for (var a = 0; a < r.length; a++) {
            var t = r[a];
            t.enumerable = t.enumerable || false;
            t.configurable = true;
            if ("value" in t) t.writable = true;
            Object.defineProperty(e, t.key, t);
        }
    }
    function B(e, r, a) {
        if (r) R(e.prototype, r);
        if (a) R(e, a);
        return e;
    }
    function M(e, r) {
        if (typeof r !== "function" && r !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(r && r.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (r) A(e, r);
    }
    function A(e, r) {
        A = Object.setPrototypeOf || function e(r, a) {
            r.__proto__ = a;
            return r;
        };
        return A(e, r);
    }
    function H(e) {
        var r = F();
        return function a() {
            var t = Q(e), i;
            if (r) {
                var o = Q(this).constructor;
                i = Reflect.construct(t, arguments, o);
            } else {
                i = t.apply(this, arguments);
            }
            return W(this, i);
        };
    }
    function W(e, r) {
        if (r && (I(r) === "object" || typeof r === "function")) {
            return r;
        }
        return z(e);
    }
    function z(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function F() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function Q(e) {
        Q = Object.setPrototypeOf ? Object.getPrototypeOf : function e(r) {
            return r.__proto__ || Object.getPrototypeOf(r);
        };
        return Q(e);
    }
    function U(e, r, a) {
        if (r in e) {
            Object.defineProperty(e, r, {
                value: a,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[r] = a;
        }
        return e;
    }
    var V = {
        appearance: o().oneOf([ "checkbox", "toggle" ]),
        children: o().node,
        disabled: o().bool,
        elementRef: o().oneOfType([ o().func, o().object ]),
        error: o().bool,
        id: o().string,
        inline: o().bool,
        /** @private */
        interactive: o().bool,
        labelledBy: o().string,
        onClick: o().func,
        /** @private. */
        required: o().bool,
        selected: o().oneOf([ true, false, "some" ]),
        selectedLabel: o().string,
        someSelectedLabel: o().string,
        unselectedLabel: o().string,
        value: o().any
    };
    var Z = {
        appearance: "checkbox",
        disabled: false,
        error: false,
        inline: false,
        interactive: true,
        selected: false
    };
    // unclear why the manual ref is needed here (only for styled-components)
        var G = {
        checkbox: S,
        toggle: E
    };
    var J = {
        checkbox: O,
        toggle: L
    };
    /**
 * `Switch` is a basic form control with an on/off state.
 */    var K =  function(e) {
        M(a, e);
        var r = H(a);
        B(a, null, [ {
            key: "validateProps",
            // @docs-props-type SwitchPropsBase
            value: function e(r) {
                var a = r.selected, t = r.appearance;
                if (false) {}
            }
        } ]);
        function a(e) {
            var t;
            N(this, a);
            t = r.call(this, e);
            U(z(t), "clickableId", void 0);
            U(z(t), "labelId", void 0);
            U(z(t), "toggle", null);
            U(z(t), "componentDidUpdate", a.validateProps);
            U(z(t), "handleContainerClick", (function(e) {
                var r, a;
                var i = t.props, o = i.value, n = i.selected;
                (r = (a = t.props).onClick) === null || r === void 0 ? void 0 : r.call(a, e, {
                    value: o,
                    // @ts-expect-error - safe due to individual prop interfaces
                    selected: n
                });
            }));
            t.labelId = (0, d.createDOMID)("label");
            t.clickableId = (0, d.createDOMID)("clickable");
            a.validateProps(e);
            return t;
        }
        B(a, [ {
            key: "focus",
            /**
     * Place focus on the toggle.
     */
            value: function e() {
                var r;
                (r = this.toggle) === null || r === void 0 ? void 0 : r.focus();
            }
        }, {
            key: "render",
            value: function e() {
                var r = this;
                var a = this.props, i = a.appearance, o = a.children, n = a.disabled, s = a.error, d = a.id, p = a.inline, u = a.interactive, b = a.labelledBy, f = a.onClick, h = a.required, g = a.selected, m = a.selectedLabel, y = a.someSelectedLabel, k = a.unselectedLabel, x = a.value, C = q(a, [ "appearance", "children", "disabled", "error", "id", "inline", "interactive", "labelledBy", "onClick", "required", "selected", "selectedLabel", "someSelectedLabel", "unselectedLabel", "value" ]);
                var S = b;
 // consumer defined external label
                                var O = d;
 // consumer defined id
                // if has internal label defined with children....
                                if (o) {
                    S = this.labelId;
 // must use generated labelId
                                        O = d || this.clickableId;
 // must have an id
                                }
                var E = g === "some" ? "mixed" : g;
                var L = {
                    true: m,
                    false: k,
                    some: y
                };
                var I = {
                    disabled: n,
                    $interactive: u,
                    $error: s,
                    $selected: g === "some" || g,
                    "aria-labelledby": u ? S : undefined,
                    "aria-checked": u ? E : undefined,
                    "aria-invalid": s ? true : undefined,
                    "aria-required": h,
                    id: O,
                    ref: function e(a) {
                        r.toggle = a;
                    },
                    onClick: n || !u ? undefined : this.handleContainerClick,
                    "data-test": "button",
                    "data-selected": g
                };
                var T = G[i];
                var N = J[i];
                var R = function e(r) {
                    return function() {
                        var e =  t().createElement(t().Fragment, null,  t().createElement(_, {
                            $delay: r,
                            $disabled: n,
                            $error: !n && s,
                            $selected: g
                        }),  t().createElement(P, {
                            $delay: r,
                            $error: !n && s
                        }));
                        return u ?  t().createElement(T, D({
                            role: "switch"
                        }, I, {
                            $delay: r
                        }), e) :  t().createElement(N, D({}, I, {
                            $delay: r
                        }), e);
                    };
                };
                var B = function e() {
                    var r =  t().createElement(t().Fragment, null, g === true && i === "checkbox" &&  t().createElement(v, null), g === "some" && i === "checkbox" &&  t().createElement($, null));
                    return u ?  t().createElement(T, D({
                        "data-clickable": true,
                        role: i
                    }, I), r) :  t().createElement(N, I, r);
                };
                var M = n && "disabled" || s && "error" || null;
                var A = !!m || !!y || !!k;
                
                return t().createElement(w, D({
                    flex: true,
                    inline: p,
                    "data-test": "switch",
                    "data-test-selected": g,
                    "data-test-value": x,
                    "data-test-error": s ? true : undefined,
                    "data-error": s ? true : undefined,
                    "data-disabled": n ? true : undefined,
                    $switchState: M
                }, C), i === "toggle" &&  t().createElement(l(), {
                    on: R("200ms"),
                    off: R()
                }), i !== "toggle" && B(), u && A &&  t().createElement(c(), null, L["".concat(g)]), o &&  t().createElement(j, {
                    $disabled: n,
                    "data-test": "label",
                    id: S,
                    htmlFor: O,
                    "data-disabled": n || null
                }, o));
            }
        } ]);
        return a;
    }(a.Component);
    U(K, "propTypes", V);
    U(K, "defaultProps", Z);
    /* harmony default export */    const X = K;
    // CONCATENATED MODULE: ./src/Switch/index.ts
    module.exports = r;
    /******/})();
