// @splunk/react-ui/TabBar.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var r = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(r, {
                a: r
            });
            /******/            return r;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        Tab: () => /* reexport */ le,
        default: () => /* reexport */ we
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    var n =  e.n(r);
    // CONCATENATED MODULE: external "prop-types"
    const i = require("prop-types");
    var a =  e.n(i);
    // CONCATENATED MODULE: external "@splunk/ui-utils/keyboard"
    const o = require("@splunk/ui-utils/keyboard");
    // CONCATENATED MODULE: external "styled-components"
    const l = require("styled-components");
    var s =  e.n(l);
    // CONCATENATED MODULE: external "@splunk/themes"
    const c = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/TabBar/TabBarStyles.ts
    var u = s().div.withConfig({
        displayName: "TabBarStyles__Styled",
        componentId: "sc-1t85fen-0"
    })([ "", " position:relative;&::before{content:'';display:block;position:absolute;left:0;top:0;right:0;bottom:0;border:0 solid ", ";}", ";" ], c.mixins.reset("flex"), (0, 
    c.pick)({
        enterprise: c.variables.borderLightColor,
        prisma: c.variables.neutral200
    }), (0, c.pickVariant)("$layout", {
        horizontal: {
            prisma: (0, c.pickVariant)("$withUnderline", {
                /* context appearance is only supported for Prisma */
                true: (0, l.css)([ "&::before{border-bottom-width:1px;}" ])
            }),
            enterprise: (0, l.css)([ "&::before{border-bottom-width:1px;}" ])
        },
        vertical: (0, l.css)([ "display:inline-block;&::before{border-right-width:1px;}" ])
    }));
    // CONCATENATED MODULE: external "lodash/omit"
    const p = require("lodash/omit");
    var d =  e.n(p);
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const f = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "@splunk/react-icons/DotsThreeVertical"
    const v = require("@splunk/react-icons/DotsThreeVertical");
    var b =  e.n(v);
    // CONCATENATED MODULE: external "@splunk/react-ui/Dropdown"
    const h = require("@splunk/react-ui/Dropdown");
    var y =  e.n(h);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/MoreVertical"
    const m = require("@splunk/react-icons/enterprise/MoreVertical");
    var g =  e.n(m);
    // CONCATENATED MODULE: external "@splunk/react-ui/Popover"
    const x = require("@splunk/react-ui/Popover");
    var k =  e.n(x);
    // CONCATENATED MODULE: external "@splunk/react-ui/ScreenReaderContent"
    const w = require("@splunk/react-ui/ScreenReaderContent");
    var S =  e.n(w);
    // CONCATENATED MODULE: external "@splunk/ui-utils/id"
    const C = require("@splunk/ui-utils/id");
    // CONCATENATED MODULE: external "@splunk/react-ui/Clickable"
    const O = require("@splunk/react-ui/Clickable");
    var T =  e.n(O);
    // CONCATENATED MODULE: ./src/TabBar/TabStyles.ts
    var I = (0, c.pick)({
        prisma: (0, l.css)([ "color:", ";" ], c.variables.contentColorActive)
    });
    var j = (0, c.pickVariant)("$layout", {
        horizontal: {
            enterprise: (0, l.css)([ "height:3px;" ])
        },
        vertical: {
            enterprise: (0, l.css)([ "width:3px;" ])
        }
    });
    var z = s()(T()).withConfig({
        displayName: "TabStyles__StyledClickable",
        componentId: "sc-1ry8mzj-0"
    })([ "flex:0 1 auto;display:flow-root;position:relative;line-height:", ";text-align:center;white-space:nowrap;color:", ";&::after{display:block;content:attr(data-title);font-weight:", ";height:0;color:transparent;overflow:hidden;visibility:hidden;}", ";&[aria-selected='true']{cursor:default;", "}&[aria-selected='false']{box-shadow:none;font-weight:", ";&:hover:not([disabled]){", "}&[aria-haspopup='true']{", "}}&:focus{box-shadow:", ";", "}" ], (0, 
    c.pick)({
        enterprise: "24px",
        prisma: "20px"
    }), c.variables.contentColorDefault, c.variables.fontWeightBold, (0, c.pickVariant)("$layout", {
        horizontal: (0, l.css)([ "padding:", ";margin-bottom:1px;" ], (0, c.pick)({
            enterprise: (0, c.pickVariant)("$hasMenu", {
                true: "3px 4px 3px 16px",
                false: "3px 20px"
            }),
            prisma: {
                comfortable: (0, c.pickVariant)("$hasMenu", {
                    true: "0px 8px 0px 16px",
                    false: "0px 20px"
                }),
                compact: (0, c.pickVariant)("$hasMenu", {
                    true: "0px 8px 0px 12px",
                    false: "0 16px"
                })
            }
        })),
        vertical: {
            enterprise: (0, l.css)([ "width:100%;text-align:left;right:1px;padding:10px 20px;", "" ], (function(e) {
                var t = e.$icon;
                return t && (0, l.css)([ "text-align:center;" ]);
            })),
            prisma: (0, l.css)([ "width:100%;right:1px;padding:", ";" ], (0, c.pick)({
                comfortable: "0 20px",
                compact: "0 16px"
            }))
        }
    }), (0, c.pick)({
        enterprise: (0, l.css)([ "font-weight:", ";" ], c.variables.fontWeightBold),
        prisma: (0, l.css)([ "color:", ";font-weight:", ";" ], c.variables.contentColorActive, c.variables.fontWeightBold)
    }), (0, c.pick)({
        enterprise: "normal",
        prisma: c.variables.fontWeightSemiBold
    }), I, I, c.variables.focusShadowInset, (0, c.pick)({
        prisma: (0, l.css)([ "color:", ";" ], c.variables.contentColorActive)
    }));
    var M = s().div.withConfig({
        displayName: "TabStyles__StyledUnderline",
        componentId: "sc-1ry8mzj-1"
    })([ "position:absolute;", " ", ";[aria-selected='true'] > &&{background-color:", ";", ";}", ":hover:not([disabled]) > &&{", ";}[aria-selected='false'][aria-haspopup='true'] > &&{", "}" ], (0, 
    c.pick)({
        enterprise: (0, l.css)([ "background:", ";" ], c.variables.borderLightColor)
    }), (0, c.pickVariant)("$layout", {
        horizontal: (0, l.css)([ "height:0;box-sizing:border-box;width:", ";", " transition:height 0.2s;" ], (0, 
        c.pick)({
            enterprise: (0, l.css)([ "calc(100% - 20px * 2)" ]),
            prisma: {
                comfortable: (0, l.css)([ "calc(100% - 20px * 2)" ]),
                compact: (0, l.css)([ "calc(100% - 16px * 2)" ])
            }
        }), (0, c.pick)({
            enterprise: (0, l.css)([ "bottom:-1px;" ])
        })),
        vertical: {
            enterprise: (0, l.css)([ "width:0;height:calc(100% - 10px * 2);top:10px;right:-1px;transition:width 0.2s;" ]),
            prisma: (0, l.css)([ "width:1px;height:calc(100% - 10px * 2);top:10px;right:-1px;transition:width 0.2s;box-sizing:border-box;" ])
        }
    }), (0, c.pick)({
        enterprise: c.variables.accentColor,
        prisma: c.variables.contentColorActive
    }), (0, c.pickVariant)("$layout", {
        horizontal: {
            enterprise: (0, l.css)([ "height:3px;" ]),
            prisma: (0, c.pickVariant)("$withUnderline", {
                // context appearance is only supported for Prisma
                true: (0, l.css)([ "height:1px;" ])
            })
        },
        vertical: {
            enterprise: (0, l.css)([ "width:3px;" ]),
            prisma: (0, l.css)([ "width:1px;" ])
        }
    }), z, j, j);
    var _ = s().span.withConfig({
        displayName: "TabStyles__StyledMenu",
        componentId: "sc-1ry8mzj-2"
    })([ "display:inline-block;> svg{transform:translateY(-1px);}border-radius:", ";cursor:pointer;", " &:hover{", "}", "[aria-haspopup='true'] &,", ":hover &,", ":focus &{visibility:visible;}" ], c.variables.borderRadius, (0, 
    c.pickVariant)("$active", {
        false: (0, l.css)([ "visibility:hidden;" ])
    }), (0, c.pick)({
        enterprise: (0, l.css)([ "color:", ";background-color:", ";border-color:", ";" ], (0, 
        c.pick)({
            light: c.variables.linkColor,
            dark: c.variables.white
        }), c.variables.backgroundColorHover, (0, c.pick)({
            light: c.variables.gray60,
            dark: c.variables.borderColor
        })),
        prisma: (0, l.css)([ "background-color:", ";border-color:", ";box-shadow:", ";" ], c.variables.interactiveColorOverlayHover, c.variables.interactiveColorBorderHover, c.variables.hoverShadow)
    }), z, z, z);
    var E = s().span.withConfig({
        displayName: "TabStyles__StyledIcon",
        componentId: "sc-1ry8mzj-3"
    })([ "", ";" ], (0, c.pickVariant)("$iconSize", {
        inline: (0, l.css)([ "", " text-align:left;padding-right:", ";> svg{transform:translateY(-1px);}" ], (0, 
        c.pick)({
            prisma: (0, l.css)([ "display:inline-block;width:16px;height:16px;" ])
        }), (0, c.pick)({
            enterprise: "0.4em",
            prisma: "8px"
        })),
        small: (0, l.css)([ "font-size:24px;height:24px;text-align:center;display:block;padding:4px 0;" ]),
        large: (0, l.css)([ "font-size:48px;height:48px;text-align:center;display:block;padding:8px 0 0;" ])
    }));
    var P = s().div.withConfig({
        displayName: "TabStyles__StyledLabelAndMenuContainer",
        componentId: "sc-1ry8mzj-4"
    })([ "display:flex;align-items:center;justify-content:center;gap:8px;" ]);
    var $ = s().div.withConfig({
        displayName: "TabStyles__StyledLabel",
        componentId: "sc-1ry8mzj-5"
    })([ "overflow:hidden;text-overflow:ellipsis;min-width:10px;margin:", ";" ], (0, 
    c.pick)({
        enterprise: "0",
        prisma: {
            comfortable: "16px 0",
            compact: "10px 0"
        }
    }));
    var D = s().span.withConfig({
        displayName: "TabStyles__StyledCount",
        componentId: "sc-1ry8mzj-6"
    })([ "display:inline-block;border-radius:18px;color:", ";font-size:", ";line-height:10px;padding:", ";margin-inline-start:", ";", " &[disabled]{color:", ";}" ], c.variables.contentColorDefault, (0, 
    c.pick)({
        enterprise: "inherit",
        prisma: "10px"
    }), (0, c.pick)({
        enterprise: "0.4em 0.3em",
        prisma: "0.4em 0.6em"
    }), (0, c.pick)({
        enterprise: "0",
        prisma: "0.3em"
    }), (0, c.pick)({
        enterprise: (0, l.css)([ "&::before{content:'(';}&::after{content:')';}" ]),
        prisma: (0, l.css)([ "background:", ";" ], c.variables.neutral100)
    }), c.variables.contentColorDisabled);
    var R = s().div.withConfig({
        displayName: "TabStyles__StyledTooltipContent",
        componentId: "sc-1ry8mzj-7"
    })([ "padding:8px;font-size:", ";" ], c.variables.fontSizeSmall);
    // CONCATENATED MODULE: ./src/TabBar/TabBarContext.tsx
    var q = function e() {
        return undefined;
    };
    var A =  (0, r.createContext)({
        onMenuOpen: q,
        onMenuClose: q,
        isMenuOpen: false
    });
    A.displayName = "TabBar";
    /* harmony default export */    const V = A;
    // CONCATENATED MODULE: ./src/utils/getCountValue.ts
    // A utility for max count
    var K = function e(t, r) {
        if (t > r) {
            return "".concat(r, "+");
        }
        return t;
    };
    // CONCATENATED MODULE: ./src/utils/updateReactRef.ts
    /**
 * Updates a React ref. Callback refs and object refs (from `createRef` and `useRef`) are supported.
 *
 * @param ref - The React callback or object ref. Can be `null` or `undefined`.
 * @param current - The new value of the ref.
 */
    function B(e, t) {
        if (e) {
            if (typeof e === "function") {
                e(t);
            } else {
                // the public signature of this util uses React.Ref<T> to mirror the way React types refs.
                // the intention here is to signal "we will take care of setting 'current', not you".
                e.current = t;
 // eslint-disable-line no-param-reassign
                        }
        }
    }
    // CONCATENATED MODULE: ./src/TabBar/Tab.tsx
    function F(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            F = function e(t) {
                return typeof t;
            };
        } else {
            F = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return F(e);
    }
    function N() {
        N = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return N.apply(this, arguments);
    }
    function U(e, t) {
        if (e == null) return {};
        var r = W(e, t);
        var n, i;
        if (Object.getOwnPropertySymbols) {
            var a = Object.getOwnPropertySymbols(e);
            for (i = 0; i < a.length; i++) {
                n = a[i];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function W(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var i, a;
        for (a = 0; a < n.length; a++) {
            i = n[a];
            if (t.indexOf(i) >= 0) continue;
            r[i] = e[i];
        }
        return r;
    }
    function L(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function H(e, t) {
        for (var r = 0; r < t.length; r++) {
            var n = t[r];
            n.enumerable = n.enumerable || false;
            n.configurable = true;
            if ("value" in n) n.writable = true;
            Object.defineProperty(e, n.key, n);
        }
    }
    function Y(e, t, r) {
        if (t) H(e.prototype, t);
        if (r) H(e, r);
        return e;
    }
    function G(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) J(e, t);
    }
    function J(e, t) {
        J = Object.setPrototypeOf || function e(t, r) {
            t.__proto__ = r;
            return t;
        };
        return J(e, t);
    }
    function Q(e) {
        var t = ee();
        return function r() {
            var n = te(e), i;
            if (t) {
                var a = te(this).constructor;
                i = Reflect.construct(n, arguments, a);
            } else {
                i = n.apply(this, arguments);
            }
            return X(this, i);
        };
    }
    function X(e, t) {
        if (t && (F(t) === "object" || typeof t === "function")) {
            return t;
        }
        return Z(e);
    }
    function Z(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function ee() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function te(e) {
        te = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return te(e);
    }
    function re(e, t, r) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: r,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = r;
        }
        return e;
    }
    var ne = {
        /** @private. Is the tab active. */
        active: a().bool,
        ariaControls: a().string,
        count: a().number,
        disabled: a().bool,
        elementRef: a().oneOfType([ a().func, a().object ]),
        icon: a().node,
        id: a().string,
        label: a().oneOfType([ a().string, a().element ]),
        /** @private Maximum count number in tab label */
        maxCount: a().number,
        /** @private */
        splunkTheme: a().object,
        tabId: a().string,
        /** @private. The internal key of the tab */
        tabKey: a().number,
        to: a().string,
        tooltip: a().node,
        /**
   * @private
   * Accepts a Menu component that is rendered for that Tab.
   * Is not compatible with `count`.
   * The `menu` prop is only supported in horizontal layout due to both design constraints as well as keyboard support.
   */
        menu: a().node
    };
    var ie = {
        active: false
    };
    var ae =  function(e) {
        G(r, e);
        var t = Q(r);
        // @docs-props-type TabPropsBase
                function r(e) {
            var n;
            L(this, r);
            n = t.call(this, e);
            re(Z(n), "popoverId", void 0);
            re(Z(n), "ariaId", void 0);
            re(Z(n), "menuDescription", void 0);
            re(Z(n), "handleMount", (function(e) {
                n.setState({
                    anchor: e
                });
                B(n.props.elementRef, e);
            }));
            re(Z(n), "handleTooltipOpen", (function() {
                n.setState({
                    open: true
                });
            }));
            re(Z(n), "handleTooltipClose", (function() {
                n.setState({
                    open: false
                });
            }));
            re(Z(n), "handleClick", (function(e) {
                var t = n.props, r = t.active, i = t.tabId, a = t.tabKey, o = t.to;
                var l = n.context.onClick;
 // preserve the previous behaviour from 2ad8e7eaf47 to avoid firing unnecessary onChange events
                // while preventing the "to" prop from incorrectly triggering <Clickable>'s providedOnClick on the active tab
                                if (!r) {
                    l === null || l === void 0 ? void 0 : l(e, {
                        tabId: i,
                        tabKey: a
                    });
                }
                if (o == null || o != null && r) {
                    e.preventDefault();
                }
            }));
            re(Z(n), "handleFocus", (function(e) {
                n.handleTooltipOpen();
                var t = n.props, r = t.tabId, i = t.tabKey;
                var a = n.context.onFocus;
                a === null || a === void 0 ? void 0 : a(e, {
                    tabId: r,
                    tabKey: i
                });
            }));
            re(Z(n), "handleFocusMenuToggle", (function() {
                n.focus();
            }));
            re(Z(n), "handleBlur", (function() {
                n.handleTooltipClose();
            }));
            n.popoverId = (0, C.createDOMID)("popover");
            n.ariaId = (0, C.createDOMID)("aria-id");
            n.menuDescription = (0, C.createDOMID)("menu-description");
            n.state = {
                open: false,
                anchor: null
            };
            return n;
        }
        Y(r, [ {
            key: "focus",
            value: function e() {
                var t;
                (t = this.state.anchor) === null || t === void 0 ? void 0 : t.focus();
            }
        }, {
            key: "render",
            value: function e() {
                var t = this.props, r = t.active, i = t.ariaControls, a = t.count, o = t.disabled, l = t.icon, s = t.label, c = t.tabId, u = t.maxCount, p = t.splunkTheme, v = t.to, h = t.tooltip, m = t.menu, x = t.tabKey, w = U(t, [ "active", "ariaControls", "count", "disabled", "icon", "label", "tabId", "maxCount", "splunkTheme", "to", "tooltip", "menu", "tabKey" ]);
                var C = this.state, O = C.anchor, T = C.open;
                var I = this.context, j = I.appearance, q = I.disabled, A = I.iconSize, V = I.layout, B = I.width, F = I.onMenuOpen, W = I.onMenuClose, L = I.isMenuOpen, H = I.focusedTabKey;
                var Y = w.style;
                var G = q || o || false;
                var J = A || "inline";
                var Q = V || "horizontal";
                var X = B ? {
                    width: B
                } : Y;
                var Z = p.isEnterprise;
                var ee = s ? s.toString() : "";
                var te = a === 0 || a ? "".concat(a === null || a === void 0 ? void 0 : a.toString()) : "";
 // title is used for CSS styling in TabStyles. labelValue and countValue needs to be included in the title.
                                var re;
                if (Z) {
                    // countValue needs unicode space for width to not change after it's clicked.
                    var ne = te ? "(".concat(te, ")  ") : "";
                    re = "".concat(ee, " ").concat(ne);
                } else {
                    re = "".concat(ee, "    ").concat(te);
                }
                var ie = H === x && L && !!m;
                var ae =  n().createElement(_, {
                    $active: r,
                    "data-test": "menu-toggle",
                    tabIndex: -1,
                    onFocus: this.handleFocusMenuToggle
                }, Z ?  n().createElement(g(), {
                    size: 1.1
                }) :  n().createElement(b(), null));
                var oe = function e(t) {
                    // if there is no menu, or if we are not in horizontal, do not render menu container
                    if (!m || Q !== "horizontal") {
                        return t;
                    }
                    
                    return n().createElement(P, null, t,  n().createElement(y(), {
                        toggle: ae,
                        align: "center",
                        onRequestOpen: F,
                        onRequestClose: W,
                        open: ie
                    }, m));
                };
                
                // TODO: Fix the ts error: https://splunk.atlassian.net/browse/SUI-5569
                // eslint-disable-next-line
                // @ts-ignore-next-line
                return n().createElement(z, N({
                    "aria-controls": i,
                    "aria-selected": r,
                    "aria-labelledby": this.ariaId,
                    "aria-describedby": m ? this.menuDescription : undefined,
                    "aria-haspopup": ie,
                    "data-test": "tab",
                    "data-test-tab-id": c,
                    "data-test-popover-id": h ? this.popoverId : undefined,
                    $layout: Q,
                    $icon: l && J !== "inline" ? true : undefined,
                    disabled: G,
                    style: X,
                    elementRef: this.handleMount
                }, d()(w, "elementRef"), {
                    onClick: this.handleClick,
                    onFocus: this.handleFocus,
                    onMouseEnter: this.handleTooltipOpen,
                    onBlur: this.handleBlur,
                    onMouseLeave: this.handleTooltipClose,
                    role: "tab",
                    tabIndex: r ? undefined : -1,
                    "data-title": re,
                    to: v,
                    $hasMenu: !!m
                }), oe( n().createElement($, {
                    "data-test": "label",
                    $withUnderline: j === "navigation"
                }, l &&  n().createElement(E, {
                    $iconSize: J
                }, l), s, !m && (a === 0 || a) &&  n().createElement(n().Fragment, null,  n().createElement(S(), null, " "),  n().createElement(D, {
                    "data-test": "count",
                    disabled: G
                }, u ? K(a, u) : a)), !!m &&  n().createElement(S(), {
                    "aria-hidden": true,
                    id: this.menuDescription
                }, (0, f._)("Press Shift + F10 to open the tab’s menu")))),  n().createElement(M, {
                    $layout: Q,
                    $withUnderline: j === "navigation"
                }), !o && h &&  n().createElement(k(), {
                    role: "tooltip",
                    anchor: O,
                    animationConfig: {
                        tension: 400
                    },
                    appearance: "inverted",
                    closeReasons: [ "offScreen", "escapeKey" ],
                    defaultPlacement: Q === "vertical" ? "right" : "above",
                    id: this.popoverId,
                    open: !!O && T,
                    align: "center"
                },  n().createElement(R, null, h)), h &&  n().createElement(S(), {
                    "aria-hidden": "true",
                    id: this.ariaId
                }, h));
            }
        } ]);
        return r;
    }(r.Component);
    re(ae, "propTypes", ne);
    re(ae, "defaultProps", ie);
    re(ae, "contextType", V);
    var oe = (0, c.withSplunkTheme)(ae);
    oe.propTypes = ae.propTypes;
    /* harmony default export */    const le = oe;
    // CONCATENATED MODULE: ./src/utils/navigateList.ts
    // A utility for keyboard navigation of lists
    function se(e, t, r) {
        for (var n = 0; n < e.length; n += 1) {
            var i;
            var a = (n + r) % e.length;
            if (((i = e[a].current) === null || i === void 0 ? void 0 : i.props.disabled) !== true) {
                return e[a];
            }
        }
        return e[t];
    }
    function ce(e, t, r) {
        for (var n = e.length; n > 0; n -= 1) {
            var i;
            var a = (n + r) % e.length;
            if (((i = e[a].current) === null || i === void 0 ? void 0 : i.props.disabled) !== true) {
                return e[a];
            }
        }
        return e[t];
    }
    // CONCATENATED MODULE: ./src/TabBar/TabBar.tsx
    function ue() {
        ue = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return ue.apply(this, arguments);
    }
    function pe(e, t) {
        return he(e) || be(e, t) || fe(e, t) || de();
    }
    function de() {
        throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function fe(e, t) {
        if (!e) return;
        if (typeof e === "string") return ve(e, t);
        var r = Object.prototype.toString.call(e).slice(8, -1);
        if (r === "Object" && e.constructor) r = e.constructor.name;
        if (r === "Map" || r === "Set") return Array.from(e);
        if (r === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(r)) return ve(e, t);
    }
    function ve(e, t) {
        if (t == null || t > e.length) t = e.length;
        for (var r = 0, n = new Array(t); r < t; r++) {
            n[r] = e[r];
        }
        return n;
    }
    function be(e, t) {
        if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(e))) return;
        var r = [];
        var n = true;
        var i = false;
        var a = undefined;
        try {
            for (var o = e[Symbol.iterator](), l; !(n = (l = o.next()).done); n = true) {
                r.push(l.value);
                if (t && r.length === t) break;
            }
        } catch (e) {
            i = true;
            a = e;
        } finally {
            try {
                if (!n && o["return"] != null) o["return"]();
            } finally {
                if (i) throw a;
            }
        }
        return r;
    }
    function he(e) {
        if (Array.isArray(e)) return e;
    }
    function ye(e, t) {
        if (e == null) return {};
        var r = me(e, t);
        var n, i;
        if (Object.getOwnPropertySymbols) {
            var a = Object.getOwnPropertySymbols(e);
            for (i = 0; i < a.length; i++) {
                n = a[i];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function me(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var i, a;
        for (a = 0; a < n.length; a++) {
            i = n[a];
            if (t.indexOf(i) >= 0) continue;
            r[i] = e[i];
        }
        return r;
    }
    var ge = 50;
    /** @public */    var xe = {
        autoActivate: a().bool,
        activeTabId: a().string,
        appearance: a().oneOf([ "navigation", "context" ]),
        children: a().node,
        /** @private */
        disabled: a().bool,
        elementRef: a().oneOfType([ a().func, a().object ]),
        iconSize: a().oneOf([ "inline", "small", "large" ]),
        layout: a().oneOf([ "horizontal", "vertical" ]),
        onChange: a().func,
        tabWidth: a().number
    };
    function ke(e) {
        var t = e.activeTabId, i = e.autoActivate, a = e.appearance, l = a === void 0 ? "navigation" : a, s = e.children, c = e.disabled, p = c === void 0 ? false : c, d = e.elementRef, f = e.iconSize, v = f === void 0 ? "inline" : f, b = e.layout, h = b === void 0 ? "horizontal" : b, y = e.onChange, m = e.tabWidth, g = ye(e, [ "activeTabId", "autoActivate", "appearance", "children", "disabled", "elementRef", "iconSize", "layout", "onChange", "tabWidth" ]);
        // @docs-props-type TabBarPropsBase
                var x = (0, r.useState)(false), k = pe(x, 2), w = k[0], S = k[1];
        var C = (0, r.useState)(0), O = pe(C, 2), T = O[0], I = O[1];
        var j = (0, r.useRef)([]);
        var z = h === "vertical" ? "down" : "right";
        var M = h === "vertical" ? "up" : "left";
        var _ = (0, r.useCallback)((function(e, t) {
            var r = t.tabKey, n = t.tabId;
            if (r != null) {
                I(r);
            }
            if (i) {
                y === null || y === void 0 ? void 0 : y(e, {
                    selectedTabId: n
                });
            }
        }), [ i, y ]);
        var E = (0, r.useCallback)((function(e, t) {
            y === null || y === void 0 ? void 0 : y(e, {
                selectedTabId: t.tabId
            });
        }), [ y ]);
        var P = (0, r.useMemo)((function() {
            var e = 0;
 // reset tabRefs since children are changed
                        j.current = [];
            return r.Children.toArray(s).filter(r.isValidElement).map((function(i) {
                // <TabBar> allows non-<Tab> children, pass them through without any modification
                if (typeof i.type === "string" || i.type.name !== le.name) {
                    return i;
                }
                var a =  n().createRef();
                j.current.push(a);
                var o = {
                    active: t === i.props.tabId,
                    tabKey: e,
                    ref: a
                };
                e += 1;
                
                return (0, r.cloneElement)(i, o);
            }));
        }), [ t, s ]);
        var $ = m && m > ge ? m : null;
        var D = (0, r.useCallback)((function(e) {
            var t = (0, o.keycode)(e.nativeEvent);
            var r;
            if (t === z) {
                r = se(j.current, T, T + 1);
            } else if (t === M) {
                r = ce(j.current, T, T - 1);
            } else if (t === "home") {
                r = se(j.current, T, 0);
            } else if (t === "end") {
                r = ce(j.current, T, j.current.length - 1);
            } else if (h === "horizontal" && e.shiftKey && t === "f10") {
                S((function(e) {
                    if (e) {
                        var t;
                        // if we're closing, we want to shift focus to the tab
                                                var r = j.current[T];
                        r === null || r === void 0 ? void 0 : (t = r.current) === null || t === void 0 ? void 0 : t.focus();
                    }
                    return !e;
                }));
            } else if (h === "horizontal" && t === "down") {
                S(true);
            }
            if (r != null) {
                var n;
                (n = r.current) === null || n === void 0 ? void 0 : n.focus();
                e.preventDefault();
            }
        }), [ T, h, z, M ]);
        var R = (0, r.useCallback)((function(e) {
            // prevent clicking on the menu icon from changing the active tab
            e.stopPropagation();
            S(true);
        }), []);
        var q = (0, r.useCallback)((function(e) {
            var t = e.event;
            // prevent content clicks/enter key from changing the active tab
                        t === null || t === void 0 ? void 0 : t.stopPropagation();
            S(false);
        }), []);
        var A = (0, r.useMemo)((function() {
            return {
                appearance: l,
                disabled: p,
                iconSize: v,
                layout: h,
                onClick: E,
                onFocus: _,
                width: $,
                onMenuOpen: R,
                onMenuClose: q,
                isMenuOpen: w,
                focusedTabKey: T
            };
        }), [ l, p, T, v, w, h, q, R, E, _, $ ]);
        
        return n().createElement(u, ue({
            "data-tab-layout": h,
            "data-test-active-tab-id": t,
            "data-test": "tab-bar",
            role: "tablist",
            ref: d,
            $layout: h,
            $withUnderline: l === "navigation",
            onKeyDown: D
        }, g),  n().createElement(V.Provider, {
            value: A
        }, P));
    }
    ke.propTypes = xe;
    ke.Tab = le;
    /* harmony default export */    const we = ke;
    // CONCATENATED MODULE: ./src/TabBar/index.ts
    module.exports = t;
    /******/})();
