// @splunk/react-ui/TabLayout.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var n = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(n, {
                a: n
            });
            /******/            return n;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, n) => {
            /******/ for (var r in n) {
                /******/ if (e.o(n, r) && !e.o(t, r)) {
                    /******/ Object.defineProperty(t, r, {
                        enumerable: true,
                        get: n[r]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        Panel: () => /* reexport */ w,
        default: () => /* reexport */ B
    });
    // CONCATENATED MODULE: external "react"
    const n = require("react");
    var r =  e.n(n);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var a =  e.n(o);
    // CONCATENATED MODULE: external "lodash/has"
    const i = require("lodash/has");
    var l =  e.n(i);
    // CONCATENATED MODULE: external "lodash/keys"
    const c = require("lodash/keys");
    var u =  e.n(c);
    // CONCATENATED MODULE: external "lodash/omit"
    const f = require("lodash/omit");
    var s =  e.n(f);
    // CONCATENATED MODULE: external "@splunk/ui-utils/id"
    const p = require("@splunk/ui-utils/id");
    // CONCATENATED MODULE: external "@splunk/react-ui/TabBar"
    const d = require("@splunk/react-ui/TabBar");
    var y =  e.n(d);
    // CONCATENATED MODULE: external "styled-components"
    const v = require("styled-components");
    var b =  e.n(v);
    // CONCATENATED MODULE: external "@splunk/themes"
    const h = require("@splunk/themes");
    // CONCATENATED MODULE: ./src/TabLayout/TabLayoutStyles.ts
    var m = b().div.withConfig({
        displayName: "TabLayoutStyles__StyledPanel",
        componentId: "lf0quy-0"
    })([ "", "" ], h.mixins.reset("block"));
    var g = b().div.withConfig({
        displayName: "TabLayoutStyles__Styled",
        componentId: "lf0quy-1"
    })([ "text-align:center;margin-top:", ";", "" ], h.variables.spacingLarge, (function(e) {
        var t = e.$layout;
        return t === "vertical" && (0, v.css)([ "display:flex;& > [role='tablist']{flex:0 0 auto;}& > ", "{width:100%;}" ], 
        /* sc-sel */
        m);
    }));
    // CONCATENATED MODULE: ./src/TabLayout/Panel.tsx
    function O() {
        O = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var n = arguments[t];
                for (var r in n) {
                    if (Object.prototype.hasOwnProperty.call(n, r)) {
                        e[r] = n[r];
                    }
                }
            }
            return e;
        };
        return O.apply(this, arguments);
    }
    function P(e, t) {
        if (e == null) return {};
        var n = I(e, t);
        var r, o;
        if (Object.getOwnPropertySymbols) {
            var a = Object.getOwnPropertySymbols(e);
            for (o = 0; o < a.length; o++) {
                r = a[o];
                if (t.indexOf(r) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, r)) continue;
                n[r] = e[r];
            }
        }
        return n;
    }
    function I(e, t) {
        if (e == null) return {};
        var n = {};
        var r = Object.keys(e);
        var o, a;
        for (a = 0; a < r.length; a++) {
            o = r[a];
            if (t.indexOf(o) >= 0) continue;
            n[o] = e[o];
        }
        return n;
    }
    var S = {
        children: a().node,
        count: a().number,
        disabled: a().bool,
        elementRef: a().oneOfType([ a().func, a().object ]),
        icon: a().node,
        label: a().oneOfType([ a().string, a().element ]),
        panelId: a().string.isRequired,
        tooltip: a().node
    };
    function j(e) {
        var t = e.children, n = e.disabled, o = n === void 0 ? false : n, a = e.elementRef, i = e.panelId, l = P(e, [ "children", "disabled", "elementRef", "panelId" ]);
        // @docs-props-type PanelPropsBase
                
        return r().createElement(m, O({
            "data-test-panel-id": i,
            "data-test": "panel",
            ref: a,
            role: "tabpanel"
        }, s()(l, [ "icon", "label", "tooltip" ])), t);
    }
    j.propTypes = S;
    /* harmony default export */    const w = j;
    // CONCATENATED MODULE: ./src/TabLayout/TabLayout.tsx
    function T(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            T = function e(t) {
                return typeof t;
            };
        } else {
            T = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return T(e);
    }
    function _() {
        _ = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var n = arguments[t];
                for (var r in n) {
                    if (Object.prototype.hasOwnProperty.call(n, r)) {
                        e[r] = n[r];
                    }
                }
            }
            return e;
        };
        return _.apply(this, arguments);
    }
    function C(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function x(e, t) {
        for (var n = 0; n < t.length; n++) {
            var r = t[n];
            r.enumerable = r.enumerable || false;
            r.configurable = true;
            if ("value" in r) r.writable = true;
            Object.defineProperty(e, r.key, r);
        }
    }
    function E(e, t, n) {
        if (t) x(e.prototype, t);
        if (n) x(e, n);
        return e;
    }
    function q(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) R(e, t);
    }
    function R(e, t) {
        R = Object.setPrototypeOf || function e(t, n) {
            t.__proto__ = n;
            return t;
        };
        return R(e, t);
    }
    function k(e) {
        var t = D();
        return function n() {
            var r = M(e), o;
            if (t) {
                var a = M(this).constructor;
                o = Reflect.construct(r, arguments, a);
            } else {
                o = r.apply(this, arguments);
            }
            return z(this, o);
        };
    }
    function z(e, t) {
        if (t && (T(t) === "object" || typeof t === "function")) {
            return t;
        }
        return A(e);
    }
    function A(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function D() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function M(e) {
        M = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return M(e);
    }
    function L(e, t, n) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: n,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = n;
        }
        return e;
    }
    var W = {
        autoActivate: a().bool,
        activePanelId: a().string,
        appearance: a().oneOf([ "navigation", "context" ]),
        children: a().node,
        count: a().number,
        defaultActivePanelId: a().string,
        elementRef: a().oneOfType([ a().func, a().object ]),
        iconSize: a().oneOf([ "inline", "small", "large" ]),
        layout: a().oneOf([ "horizontal", "vertical" ]),
        onChange: a().func,
        tabWidth: a().number
    };
    var N = {
        appearance: "navigation",
        iconSize: "inline",
        layout: "horizontal"
    };
    /**
 * The `TabLayout` is a group of managed `Panels`. Only one panel can be open at a time.
 * `TabLayout` supports both the controlled and uncontrolled patterns.
 */    var $ =  function(e) {
        q(o, e);
        var t = k(o);
        // @docs-props-type TabLayoutPropsBase
                function o(e) {
            var n;
            C(this, o);
            n = t.call(this, e);
            L(A(n), "controlledExternally", void 0);
            L(A(n), "guid", void 0);
            L(A(n), "handleChange", (function(e, t) {
                var r, o;
                var a = t.selectedTabId;
                if (!n.isControlled()) {
                    n.setState({
                        activePanelId: a
                    });
                }
                (r = (o = n.props).onChange) === null || r === void 0 ? void 0 : r.call(o, e, {
                    activePanelId: a
                });
            }));
            n.controlledExternally = l()(e, "activePanelId");
            n.guid = (0, p.createDOMID)();
            if (!n.isControlled()) {
                n.state = {
                    activePanelId: e.defaultActivePanelId
                };
            }
            if (false) {}
            return n;
        }
        E(o, [ {
            key: "componentDidUpdate",
            value: function e(t) {
                if (false) {}
                if (false) {}
            }
        }, {
            key: "isControlled",
            value: function e() {
                return this.controlledExternally;
            }
        }, {
            key: "render",
            value: function e() {
                var t = this;
                var a = this.props, i = a.children, l = a.layout;
                var c = this.isControlled() ? this.props.activePanelId : this.state.activePanelId;
                var f;
                var p = n.Children.toArray(i).filter(n.isValidElement).map((function(e) {
                    var o = e.props;
                    var a = "".concat(t.guid, "-").concat(o.panelId);
                    var i = "".concat(t.guid, "-").concat(o.panelId, "-tab");
                    if (o.panelId === c) {
                        f =  (0, n.cloneElement)(e, {
                            "aria-labelledby": i,
                            id: a
                        });
                    }
                    
                    return r().createElement(y().Tab, {
                        count: o.count,
                        icon: o.icon,
                        key: o.panelId,
                        label: o.label,
                        tabId: o.panelId,
                        id: i,
                        ariaControls: a,
                        tooltip: o.tooltip,
                        disabled: o.disabled
                    });
                }));
                if (false) {}
                
                return r().createElement(g, _({
                    "data-test-active-panel-id": c,
                    "data-test": "tab-layout",
                    ref: this.props.elementRef,
                    $layout: l
                }, s()(this.props, u()(o.propTypes))),  r().createElement(y(), {
                    autoActivate: this.props.autoActivate,
                    activeTabId: c,
                    appearance: this.props.appearance,
                    iconSize: this.props.iconSize,
                    layout: this.props.layout,
                    onChange: this.handleChange,
                    tabWidth: this.props.tabWidth
                }, p), f);
            }
        } ]);
        return o;
    }(n.Component);
    L($, "propTypes", W);
    L($, "defaultProps", N);
    L($, "Panel", w);
    /* harmony default export */    const B = $;
    // CONCATENATED MODULE: ./src/TabLayout/index.ts
    module.exports = t;
    /******/})();
