// @splunk/react-ui/test-runner-jest.config.js
/* eslint-disable no-undef */
// eslint-disable-next-line @typescript-eslint/no-var-requires
const { getJestConfig } = require('@storybook/test-runner');

// if component is passed to storybook:test then test will run for only that component
const testMatch = process.env.STORYBOOK_TEST_COMPONENT
    ? [
          `**/${process.env.STORYBOOK_TEST_COMPONENT}.stories.[jt]s?(x)`,
          `**/${process.env.STORYBOOK_TEST_COMPONENT}.examples.[jt]s?(x)`,
      ]
    : ['**/*.stories.[jt]s?(x)', '**/*.examples.[jt]s?(x)'];

module.exports = {
    // The default configuration comes from @storybook/test-runner
    ...getJestConfig(),
    rootDir: process.cwd(),
    testMatch,
    transform: {
        '^.+\\.stories\\.[jt]sx?$': '@storybook/test-runner/playwright/transform',
        '^.+\\.examples\\.[jt]sx?$': '@storybook/test-runner/playwright/transform',
        '^.+\\.tsx?$': 'babel-jest',
    },
    // used for a11y DOM snapshots, currently disabled
    // snapshotResolver: '<rootDir>/.storybook-visual/config/snapshotResolver.js',
    reporters: [
        'default',
        'jest-junit',
        [
            'jest-stare',
            {
                resultDir: `./test-reports/${
                    process.env.STORYBOOK_TEST_A11Y === 'false' ? 'visual' : 'a11y'
                }/${process.env.STORYBOOK_THEME_FAMILY}${
                    process.env.JOB_BROWSER ? `-${process.env.JOB_BROWSER}` : ''
                }`,
                reportTitle: `${
                    process.env.STORYBOOK_TEST_A11Y === 'false' ? 'Visual' : 'A11y'
                } Test Report - ${process.env.STORYBOOK_THEME_FAMILY}${
                    process.env.JOB_BROWSER ? `-${process.env.JOB_BROWSER}` : ''
                }`,
                reportHeadline: `${
                    process.env.STORYBOOK_TEST_A11Y === 'false' ? 'Visual' : 'A11y'
                } Test Report - ${process.env.STORYBOOK_THEME_FAMILY}${
                    process.env.JOB_BROWSER ? `-${process.env.JOB_BROWSER}` : ''
                }`,
                hidePending: true,
                resultHtml:
                    process.env.STORYBOOK_TEST_A11Y === 'false'
                        ? 'visual-report.html'
                        : 'a11y-report.html',
            },
        ],
    ],
    globals: {
        __DEV__: true,
    },
};
