// @splunk/react-ui/TextArea.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var r = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(r, {
                a: r
            });
            /******/            return r;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        default: () => /* reexport */ De
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    var n =  e.n(r);
    // CONCATENATED MODULE: external "prop-types"
    const a = require("prop-types");
    var i =  e.n(a);
    // CONCATENATED MODULE: external "lodash/defer"
    const o = require("lodash/defer");
    var l =  e.n(o);
    // CONCATENATED MODULE: external "lodash/has"
    const s = require("lodash/has");
    var c =  e.n(s);
    // CONCATENATED MODULE: external "lodash/keys"
    const p = require("lodash/keys");
    var d =  e.n(p);
    // CONCATENATED MODULE: external "lodash/noop"
    const u = require("lodash/noop");
    var f =  e.n(u);
    // CONCATENATED MODULE: external "lodash/omit"
    const h = require("lodash/omit");
    var v =  e.n(h);
    // CONCATENATED MODULE: external "lodash/pickBy"
    const m = require("lodash/pickBy");
    var b =  e.n(m);
    // CONCATENATED MODULE: external "lodash/throttle"
    const y = require("lodash/throttle");
    var g =  e.n(y);
    // CONCATENATED MODULE: external "@splunk/react-ui/ControlGroup"
    const x = require("@splunk/react-ui/ControlGroup");
    // CONCATENATED MODULE: external "@splunk/react-ui/EventListener"
    const w = require("@splunk/react-ui/EventListener");
    var C =  e.n(w);
    // CONCATENATED MODULE: external "@splunk/themes"
    const k = require("@splunk/themes");
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const S = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "@splunk/ui-utils/style"
    const A = require("@splunk/ui-utils/style");
    // CONCATENATED MODULE: external "@splunk/react-icons/Cross"
    const O = require("@splunk/react-icons/Cross");
    var I =  e.n(O);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Close"
    const j = require("@splunk/react-icons/enterprise/Close");
    var E =  e.n(j);
    // CONCATENATED MODULE: external "styled-components"
    const T = require("styled-components");
    var M =  e.n(T);
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const P = require("@splunk/react-ui/Box");
    var $ =  e.n(P);
    // CONCATENATED MODULE: external "@splunk/react-ui/Button"
    const _ = require("@splunk/react-ui/Button");
    var D =  e.n(_);
    // CONCATENATED MODULE: ./src/TextArea/TextAreaStyles.ts
    var q = M()(D()).withConfig({
        displayName: "TextAreaStyles__StyledClearButton",
        componentId: "gfy8yp-0"
    })([ "display:none;visibility:hidden;border-radius:", ";cursor:pointer;font-size:0.83333em;flex-grow:0;", ";&:not([disabled]){color:", ";}" ], (0, 
    k.pick)({
        enterprise: k.variables.borderRadius,
        prisma: "50%"
    }), (0, k.pick)({
        enterprise: (0, T.css)([ "width:", ";height:", ";" ], k.variables.inputHeight, k.variables.inputHeight),
        prisma: {
            comfortable: (0, T.css)([ "width:26px;min-width:26px;min-height:26px;margin:8px;padding:0;" ]),
            compact: (0, T.css)([ "width:22px;min-width:22px;min-height:22px;margin:8px;padding:0;" ])
        }
    }), (0, k.pick)({
        enterprise: {
            light: k.variables.gray60,
            dark: k.variables.white
        },
        prisma: k.variables.contentColorMuted
    }));
    var W = M().span.withConfig({
        displayName: "TextAreaStyles__StyledSearchIconWrapper",
        componentId: "gfy8yp-1"
    })([ "", ";color:", ";pointer-events:none;padding:", ";", "" ], k.mixins.reset("inline-block"), (0, 
    k.pick)({
        enterprise: {
            light: k.variables.gray60,
            dark: k.variables.white
        },
        prisma: k.variables.contentColorMuted
    }), (0, k.pick)({
        comfortable: "0 8px",
        compact: "0 6px"
    }), (function(e) {
        var t = e.$disabled;
        return t && (0, T.css)([ "", "" ], (0, k.pick)({
            enterprise: (0, T.css)([ "color:", ";" ], k.variables.textDisabledColor),
            prisma: (0, T.css)([ "color:", ";" ], k.variables.contentColorDisabled)
        }));
    }));
    var N = "230px";
    var R = M()($()).withConfig({
        displayName: "TextAreaStyles__StyledBox",
        componentId: "gfy8yp-2"
    })([ "display:flex;justify-content:space-between;flex-grow:1;flex-shrink:1;position:relative;&[data-inline]{width:", ";flex-basis:", ";[data-inline] + &{margin-left:", ";}}&:focus-within:not([disabled]){", " ", "{visibility:visible;display:block;}}" ], N, N, k.variables.spacingSmall, (0, 
    k.pick)({
        enterprise: (0, T.css)([ "", "{display:none;}" ], 
        /* sc-sel */
        W)
    }), 
    /* sc-sel */
    q);
    var B = 12;
    var z = 10;
    var F = 0;
    var H = M().span.withConfig({
        displayName: "TextAreaStyles__StyledTextAreaWrapper",
        componentId: "gfy8yp-3"
    })([ "", " flex-grow:1;min-height:", ";position:relative;overflow:hidden;border:1px solid ", ";border-radius:", ";box-sizing:border-box;background-color:", ";box-shadow:inset 0 1px 1px rgba(0,0,0,0.075);align-items:center;padding:", ";", " ", " &:hover:not([disabled]){border-color:", ";}&:focus-within:not([disabled]){box-shadow:", ";color:", ";}", " ", " ", " ", "" ], k.mixins.reset("inline-flex"), k.variables.inputHeight, (0, 
    k.pick)({
        enterprise: {
            light: k.variables.gray60,
            dark: k.variables.gray20
        },
        prisma: k.variables.interactiveColorBorder
    }), k.variables.borderRadius, (0, k.pick)({
        enterprise: {
            light: k.variables.white,
            dark: k.variables.gray22
        },
        prisma: k.variables.transparent
    }), (0, k.pick)({
        enterprise: "0px ".concat(F, "px 0px ").concat(z, "px"),
        prisma: "0px ".concat(B, "px")
    }), (function(e) {
        var t = e.$hasEndAdornment;
        return t && (0, T.css)([ "padding-right:0;" ]);
    }), (function(e) {
        var t = e.$hasStartAdornment;
        return t && (0, T.css)([ "padding-left:0;" ]);
    }), (0, k.pick)({
        enterprise: {
            light: k.variables.gray60,
            dark: k.variables.gray20
        },
        prisma: k.variables.interactiveColorBorderHover
    }), k.variables.focusShadow, k.variables.contentColorActive, (function(e) {
        var t = e.$error;
        return t && (0, T.css)([ "&,&:hover:not([disabled]){border-color:", ";}&:focus-within:not([disabled]){box-shadow:", ";color:", ";}" ], k.variables.accentColorNegative, k.variables.focusShadow, k.variables.contentColorActive);
    }), (function(e) {
        var t = e.$append;
        return t && (0, T.css)([ "margin-right:-1px;border-top-right-radius:0;border-bottom-right-radius:0;border-right:none;" ]);
    }), (function(e) {
        var t = e.$prepend;
        return t && (0, T.css)([ "border-top-left-radius:0;border-bottom-left-radius:0;" ]);
    }), (function(e) {
        var t = e.disabled;
        return t && (0, T.css)([ "", " box-shadow:inset 0 1px 2px rgba(0,0,0,0.025);" ], (0, 
        k.pick)({
            enterprise: {
                light: (0, T.css)([ "background-color:", ";border-color:", ";" ], k.variables.gray96, k.variables.gray92),
                dark: (0, T.css)([ "background-color:", ";border-color:", ";" ], k.variables.gray22, k.variables.gray30)
            },
            prisma: (0, T.css)([ "border-color:", ";" ], k.variables.interactiveColorBorderDisabled)
        }));
    }));
    var V = M().textarea.withConfig({
        displayName: "TextAreaStyles__StyledTextArea",
        componentId: "gfy8yp-4"
    })([ "", ";outline-style:none;border:0;background:transparent;color:", ";flex-grow:1;font-family:", ";font-size:inherit;line-height:inherit;padding-top:", ";padding-bottom:", ";padding-right:", ";position:relative;resize:none;overflow:auto;white-space:pre-wrap;", " &::placeholder{color:", ";opacity:1;}", " ", "" ], k.mixins.reset("inline-flex"), k.variables.contentColorActive, k.variables.sansFontFamily, (0, 
    k.pick)({
        enterprise: {
            comfortable: k.variables.spacingQuarter,
            compact: "3px"
        },
        prisma: {
            comfortable: k.variables.spacingSmall,
            compact: "5px"
        }
    }), (0, k.pick)({
        enterprise: {
            comfortable: k.variables.spacingQuarter,
            compact: "3px"
        },
        prisma: {
            comfortable: k.variables.spacingSmall,
            compact: "5px"
        }
    }), (0, k.pick)({
        enterprise: "10px",
        prisma: "14px"
    }), (function(e) {
        var t = e.$rowsMin;
        return t && (0, T.css)([ "height:calc( ", " + (", " * ", ") + ", " );" ], k.variables.inputHeight, t - 1, k.variables.lineHeight, t > 1 ? "-3px" : "-2px");
    }), k.variables.contentColorMuted, (function(e) {
        var t = e.$error;
        return t && (0, T.css)([ "color:", ";" ], (0, k.pick)({
            enterprise: {
                light: k.variables.errorColorD10,
                dark: k.variables.errorColorL20
            },
            prisma: k.variables.contentColorActive
        }));
    }), (function(e) {
        var t = e.disabled;
        return t && (0, T.css)([ "cursor:not-allowed;color:", ";&::placeholder{color:", ";}" ], k.variables.contentColorDisabled, k.variables.contentColorDisabled);
    }));
    var K = M()(V).withConfig({
        displayName: "TextAreaStyles__StyledTextAreaShadow",
        componentId: "gfy8yp-5"
    })([ "position:absolute;overflow:hidden;left:-10000px;top:-10000px;visibility:hidden;pointer-events:none;width:", ";" ], (function(e) {
        var t = e.$startAdornmentWidth, r = e.$endAdornmentWidth;
        return (0, T.css)([ "calc(100% - (", "));" ], (0, k.pick)({
            prisma: (0, T.css)([ "", "px + ", "px" ], t || B, r || B),
            enterprise: (0, T.css)([ "", "px + ", "px" ], t || z, r || F)
        }));
    }));
    var L = (0, k.pick)({
        enterprise: {
            comfortable: "30px",
            compact: "26px"
        },
        prisma: {
            comfortable: "38px",
            compact: "30px"
        }
    });
    var G = (0, k.pick)({
        enterprise: "10px",
        prisma: "14px"
    });
    var Q = M().span.withConfig({
        displayName: "TextAreaStyles__StyledPlaceholder",
        componentId: "gfy8yp-6"
    })([ "pointer-events:none;color:", ";position:absolute;overflow:hidden;height:", ";margin-right:", ";", ";", ";", "" ], (0, 
    k.pick)({
        enterprise: k.variables.textGray,
        prisma: k.variables.contentColorMuted
    }), k.variables.lineHeight, G, (function(e) {
        var t = e.$hasStartAdornment, r = e.$startAdornmentWidth;
        return t && (r ? (0, T.css)([ "margin-left:", "px;" ], r) : (0, T.css)([ "margin-left:", ";" ], L));
    }), (function(e) {
        var t = e.$hasEndAdornment, r = e.$endAdornmentWidth;
        return t && (r ? (0, T.css)([ "max-width:calc(100% - ", " - ", "px);" ], G, r) : (0, 
        T.css)([ "max-width:calc(100% - ", " - ", ");" ], G, L));
    }), (function(e) {
        var t = e.$hasBothAdornment, r = e.$endAdornmentWidth, n = e.$startAdornmentWidth;
        return t && (0, T.css)([ "max-width:calc(100% - ", "px - ", "px);" ], n, r);
    }));
    var U = M().div.withConfig({
        displayName: "TextAreaStyles__StyledAdornmentSpacer",
        componentId: "gfy8yp-7"
    })([ "display:inline-flex;align-items:center;justify-content:center;height:", ";pointer-events:none;", ";" ], L, (function(e) {
        var t = e.$width;
        return t ? (0, T.css)([ "min-width:", "px;" ], t) : (0, T.css)([ "min-width:", ";" ], L);
    }));
    var J = M().div.withConfig({
        displayName: "TextAreaStyles__StyledAdornment",
        componentId: "gfy8yp-8"
    })([ "display:inline-flex;align-items:center;justify-content:center;position:absolute;pointer-events:none;z-index:1;height:", ";color:", ";", " ", ";", "" ], L, (0, 
    k.pick)({
        enterprise: {
            light: k.variables.gray60,
            dark: k.variables.white
        },
        prisma: k.variables.contentColorMuted
    }), (0, k.pickVariant)("$position", {
        start: (0, T.css)([ "top:1px;left:1px;" ]),
        end: (0, T.css)([ "top:1px;right:1px;" ])
    }), (function(e) {
        var t = e.disabled;
        return t && (0, T.css)([ "cursor:not-allowed;color:", ";" ], k.variables.contentColorDisabled);
    }), (0, k.pick)({
        enterprise: (0, T.css)([ "margin-left:-1px;margin-right:-1px;" ])
    }));
    // CONCATENATED MODULE: ./src/TextArea/icons/ClearButton.tsx
    function X() {
        X = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return X.apply(this, arguments);
    }
    function Y(e) {
        var t = (0, k.useSplunkTheme)(), r = t.isEnterprise, a = t.isCompact;
        var i = r ? "pill" : "secondary";
        var o = a ? "20px" : "24px";
        var l = r ?  n().createElement(E(), {
            size: 1
        }) :  n().createElement(I(), {
            width: o,
            height: o
        });
        
        return n().createElement(q, X({
            "data-test": "clear",
            icon: l,
            appearance: i,
            inline: false
        }, e));
    }
    // CONCATENATED MODULE: external "@splunk/react-icons/Magnifier"
    const Z = require("@splunk/react-icons/Magnifier");
    var ee =  e.n(Z);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Search"
    const te = require("@splunk/react-icons/enterprise/Search");
    var re =  e.n(te);
    // CONCATENATED MODULE: ./src/TextArea/icons/SearchIcon.tsx
    function ne() {
        var e = (0, k.useSplunkTheme)(), t = e.isCompact, r = e.isEnterprise;
        if (r) {
            
            return n().createElement(re(), {
                hideDefaultTooltip: true,
                inline: false,
                screenReaderText: "Search",
                size: "16px"
            });
        }
        var a = t ? "20px" : "24px";
        
        return n().createElement(ee(), {
            "aria-label": "Search",
            height: a,
            width: a
        });
    }
    // CONCATENATED MODULE: ./src/TextArea/syncHeightWithShadow.ts
    // This file was adapted from [Call-Em-All's material-ui](https://github.com/mui-org/material-ui),
    // which is MIT-licensed:
    // The MIT License (MIT)
    // Copyright (c) 2014 Call-Em-All
    // Permission is hereby granted, free of charge, to any person obtaining a copy
    // of this software and associated documentation files (the "Software"), to deal
    // in the Software without restriction, including without limitation the rights
    // to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    // copies of the Software, and to permit persons to whom the Software is
    // furnished to do so, subject to the following conditions:
    // The above copyright notice and this permission notice shall be included in all
    // copies or substantial portions of the Software.
    // THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    // IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    // FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    // AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    // LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    // SOFTWARE.
    // End of MIT license text.
    function ae(e) {
        var t = e.height, r = e.rowsMax, n = e.rowsMin, a = e.shadow;
        if (a) {
            var i = window.getComputedStyle(a);
            var o = i.getPropertyValue("line-height");
 // jsdom does not properly inherit/re-calculate the line-height property so we skip in these cases SUI-1943, SUI-2428
                        if (o === "inherit" || o === "normal") {
                return false;
            }
            var l = parseInt(o, 10);
            var s = parseInt(i.getPropertyValue("padding-top"), 10);
            var c = parseInt(i.getPropertyValue("padding-bottom"), 10);
            var p = parseInt(i.getPropertyValue("border-top-width"), 10);
            var d = parseInt(i.getPropertyValue("border-bottom-width"), 10);
            var u = a.scrollHeight + p + d + 1;
 // need one extra.
            /* eslint-disable no-restricted-globals */
            // Leverage the global `isNaN` here for IE support, rather than Number.isNaN.
            // We know newHeight is numeric because we parseInt() above, so `isNaN` should be reliable.
                        if (u === undefined || isNaN(u)) {
                return false;
            }
            /* eslint-enable no-restricted-globals */            if (r >= n) {
                var f = r * l + s + c + p + d + 1;
 // need one extra.
                                u = Math.min(f, u);
            }
            var h = n * l + s + c + p + d + 1;
 // need one extra.
                        u = Math.max(h, u);
            if (t !== u) {
                return u;
            }
        }
        return false;
    }
    // CONCATENATED MODULE: ./src/utils/updateReactRef.ts
    /**
 * Updates a React ref. Callback refs and object refs (from `createRef` and `useRef`) are supported.
 *
 * @param ref - The React callback or object ref. Can be `null` or `undefined`.
 * @param current - The new value of the ref.
 */
    function ie(e, t) {
        if (e) {
            if (typeof e === "function") {
                e(t);
            } else {
                // the public signature of this util uses React.Ref<T> to mirror the way React types refs.
                // the intention here is to signal "we will take care of setting 'current', not you".
                e.current = t;
 // eslint-disable-line no-param-reassign
                        }
        }
    }
    // CONCATENATED MODULE: ./src/TextArea/TextArea.tsx
    function oe(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            oe = function e(t) {
                return typeof t;
            };
        } else {
            oe = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return oe(e);
    }
    function le() {
        le = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return le.apply(this, arguments);
    }
    function se(e) {
        return ue(e) || de(e) || pe(e) || ce();
    }
    function ce() {
        throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function pe(e, t) {
        if (!e) return;
        if (typeof e === "string") return fe(e, t);
        var r = Object.prototype.toString.call(e).slice(8, -1);
        if (r === "Object" && e.constructor) r = e.constructor.name;
        if (r === "Map" || r === "Set") return Array.from(e);
        if (r === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(r)) return fe(e, t);
    }
    function de(e) {
        if (typeof Symbol !== "undefined" && Symbol.iterator in Object(e)) return Array.from(e);
    }
    function ue(e) {
        if (Array.isArray(e)) return fe(e);
    }
    function fe(e, t) {
        if (t == null || t > e.length) t = e.length;
        for (var r = 0, n = new Array(t); r < t; r++) {
            n[r] = e[r];
        }
        return n;
    }
    function he(e, t) {
        var r = Object.keys(e);
        if (Object.getOwnPropertySymbols) {
            var n = Object.getOwnPropertySymbols(e);
            if (t) n = n.filter((function(t) {
                return Object.getOwnPropertyDescriptor(e, t).enumerable;
            }));
            r.push.apply(r, n);
        }
        return r;
    }
    function ve(e) {
        for (var t = 1; t < arguments.length; t++) {
            var r = arguments[t] != null ? arguments[t] : {};
            if (t % 2) {
                he(Object(r), true).forEach((function(t) {
                    je(e, t, r[t]);
                }));
            } else if (Object.getOwnPropertyDescriptors) {
                Object.defineProperties(e, Object.getOwnPropertyDescriptors(r));
            } else {
                he(Object(r)).forEach((function(t) {
                    Object.defineProperty(e, t, Object.getOwnPropertyDescriptor(r, t));
                }));
            }
        }
        return e;
    }
    function me(e, t) {
        if (e == null) return {};
        var r = be(e, t);
        var n, a;
        if (Object.getOwnPropertySymbols) {
            var i = Object.getOwnPropertySymbols(e);
            for (a = 0; a < i.length; a++) {
                n = i[a];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function be(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var a, i;
        for (i = 0; i < n.length; i++) {
            a = n[i];
            if (t.indexOf(a) >= 0) continue;
            r[a] = e[a];
        }
        return r;
    }
    function ye(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function ge(e, t) {
        for (var r = 0; r < t.length; r++) {
            var n = t[r];
            n.enumerable = n.enumerable || false;
            n.configurable = true;
            if ("value" in n) n.writable = true;
            Object.defineProperty(e, n.key, n);
        }
    }
    function xe(e, t, r) {
        if (t) ge(e.prototype, t);
        if (r) ge(e, r);
        return e;
    }
    function we(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) Ce(e, t);
    }
    function Ce(e, t) {
        Ce = Object.setPrototypeOf || function e(t, r) {
            t.__proto__ = r;
            return t;
        };
        return Ce(e, t);
    }
    function ke(e) {
        var t = Oe();
        return function r() {
            var n = Ie(e), a;
            if (t) {
                var i = Ie(this).constructor;
                a = Reflect.construct(n, arguments, i);
            } else {
                a = n.apply(this, arguments);
            }
            return Se(this, a);
        };
    }
    function Se(e, t) {
        if (t && (oe(t) === "object" || typeof t === "function")) {
            return t;
        }
        return Ae(e);
    }
    function Ae(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function Oe() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function Ie(e) {
        Ie = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return Ie(e);
    }
    function je(e, t, r) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: r,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = r;
        }
        return e;
    }
    /** @public */    var Ee = {
        appearance: i().oneOf([ "default", "search" ]),
        append: i().bool,
        autoCapitalize: i().string,
        autoComplete: i().string,
        autoCorrect: i().string,
        autoFocus: i().bool,
        canClear: i().bool,
        children: i().node,
        /** @private. */
        classNamePrivate: i().string,
        defaultValue: i().string,
        describedBy: i().string,
        disabled: i().bool,
        elementRef: i().oneOfType([ i().func, i().object ]),
        endAdornment: i().node,
        error: i().bool,
        inline: i().bool,
        /** @private. */
        inputClassName: i().string,
        inputId: i().string,
        inputRef: i().oneOfType([ i().func, i().object ]),
        labelledBy: i().string,
        maxLength: i().number,
        name: i().string,
        onBlur: i().func,
        onChange: i().func,
        onFocus: i().func,
        onKeyDown: i().func,
        onSelect: i().func,
        onInputClick: i().func,
        placeholder: i().string,
        prepend: i().bool,
        /** @private. */
        required: i().bool,
        rowsMax: i().number,
        rowsMin: i().number,
        spellCheck: i().bool,
        tabIndex: i().number,
        startAdornment: i().node,
        /** @private */
        splunkTheme: i().object,
        value: i().string
    };
    var Te = {
        appearance: "default",
        append: false,
        autoFocus: false,
        canClear: false,
        disabled: false,
        error: false,
        inline: false,
        placeholder: "",
        prepend: false,
        rowsMax: 8,
        rowsMin: 2,
        tabIndex: 0
    };
    var Me = function e(t) {
        var r = t ? t.getBoundingClientRect() : undefined;
        var n = r && Math.round(r.width);
        return n;
    };
    /** Note: TextArea places role and aria props onto the input. All other props are placed on the wrapper. */    var Pe =  function(e) {
        we(r, e);
        var t = ke(r);
        xe(r, null, [ {
            key: "validateRows",
            // @docs-props-type TextAreaPropsBase
            value: function e(t) {
                var r = t.rowsMin, n = t.rowsMax;
                if (false) {}
            }
        } ]);
        function r(e) {
            var a;
            ye(this, r);
            a = t.call(this, e);
            je(Ae(a), "controlledExternally", void 0);
            je(Ae(a), "input", null);
            je(Ae(a), "shadow", null);
            je(Ae(a), "startAdornment", null);
            je(Ae(a), "endAdornment", null);
            je(Ae(a), "getAdornmentWidth", (function() {
                var e = Me(a.startAdornment);
                if (a.state.startAdornmentWidth !== e) {
                    a.setState({
                        startAdornmentWidth: e
                    });
                }
                var t = Me(a.endAdornment);
                if (a.state.endAdornmentWidth !== t) {
                    a.setState({
                        endAdornmentWidth: t
                    });
                }
            }));
            je(Ae(a), "handleResize", (function() {
                a.syncHeightWithShadow();
            }));
            je(Ae(a), "handleInputMount", (function(e) {
                a.input = e;
                ie(a.props.inputRef, e);
            }));
            je(Ae(a), "handleShadowMount", (function(e) {
                a.shadow = e;
            }));
            je(Ae(a), "handleInputChange", (function(e) {
                var t, r;
                var n = e.target.value;
                var i = a.props.name;
                if (!a.isControlled()) {
                    a.setState({
                        value: n
                    });
                }
                (t = (r = a.props).onChange) === null || t === void 0 ? void 0 : t.call(r, e, {
                    value: n,
                    name: i
                });
            }));
            je(Ae(a), "handleInputKeyDown", (function(e) {
                var t, r;
                (t = (r = a.props).onKeyDown) === null || t === void 0 ? void 0 : t.call(r, e);
            }));
            je(Ae(a), "handleInputSelect", (function(e) {
                var t, r;
                (t = (r = a.props).onSelect) === null || t === void 0 ? void 0 : t.call(r, e);
            }));
            je(Ae(a), "handleInputClick", (function(e) {
                var t, r;
                (t = (r = a.props).onInputClick) === null || t === void 0 ? void 0 : t.call(r, e);
            }));
            je(Ae(a), "handleInputFocus", (function(e) {
                var t, r;
                var n = e.target.value;
                var i = a.props.name;
                (t = (r = a.props).onFocus) === null || t === void 0 ? void 0 : t.call(r, e, {
                    value: n,
                    name: i
                });
            }));
            je(Ae(a), "handleInputBlur", (function(e) {
                var t, r;
                var n = e.target.value;
                var i = a.props.name;
                (t = (r = a.props).onBlur) === null || t === void 0 ? void 0 : t.call(r, e, {
                    value: n,
                    name: i
                });
            }));
            je(Ae(a), "handleClear", (function(e) {
                var t, r;
                e.preventDefault();
                var n = "";
                var i = a.props.name;
                if (!a.isControlled()) {
                    a.setState({
                        value: n
                    });
                }
                a.focus();
                (t = (r = a.props).onChange) === null || t === void 0 ? void 0 : t.call(r, e, {
                    value: n,
                    name: i
                });
            }));
            je(Ae(a), "syncHeightWithShadow", (function() {
                var e = ae({
                    height: a.state.height,
                    rowsMax: a.props.rowsMax,
                    rowsMin: a.props.rowsMin,
                    shadow: a.shadow
                });
                if (e !== false) {
                    a.setState({
                        height: e
                    });
                }
            }));
            je(Ae(a), "renderAdornment", (function(e, t) {
                var r = a.props.disabled;
                var i = t === "start" ? function(e) {
                    a.startAdornment = e;
                } : function(e) {
                    a.endAdornment = e;
                };
                var o = {
                    ref: i,
                    disabled: r,
                    $position: t
                };
                
                return n().createElement(J, o, e);
            }));
            je(Ae(a), "renderEndAdornment", (function() {
                var e;
                var t = a.props, r = t.appearance, i = t.endAdornment, o = t.disabled, l = t.canClear, s = t.splunkTheme;
                var c = s.isPrisma, p = s.isEnterprise;
                if (i) {
                    return a.renderAdornment(i, "end");
                }
                var d = a.context;
                var u = (e = d.labelAttrs) === null || e === void 0 ? void 0 : e.text;
                var f = u ? (0, S._)("Clear ".concat(u, " text field")) : (0, S._)("Clear text field");
                if (p && r === "search") {
                    /* Both icons are rendered, and StyledBox toggles visibility based on :focus-within */
                    return a.renderAdornment( n().createElement(n().Fragment, null,  n().createElement(Y, {
                        "aria-label": f,
                        onClick: a.handleClear
                    }),  n().createElement(W, {
                        $disabled: o
                    },  n().createElement(ne, null))), "end");
                }
                var h = !!a.getDisplayValue();
                if (h && (c && r === "search" || l)) {
                    return a.renderAdornment( n().createElement(n().Fragment, null, !o &&  n().createElement(Y, {
                        "aria-label": f,
                        onClick: a.handleClear
                    })), "end");
                }
                return undefined;
            }));
            je(Ae(a), "renderStartAdornment", (function() {
                var e = a.props, t = e.appearance, r = e.startAdornment, i = e.splunkTheme, o = e.disabled;
                var l = i.isPrisma;
                if (r) {
                    return a.renderAdornment(r, "start");
                }
                if (l && t === "search") {
                    return a.renderAdornment( n().createElement(W, {
                        $disabled: o
                    },  n().createElement(ne, null)), "start");
                }
                return undefined;
            }));
            je(Ae(a), "getDisplayValue", (function() {
                var e = a.isControlled() ? a.props.value : a.state.value;
                return e;
            }));
            a.controlledExternally = c()(e, "value");
            a.state = {
                value: e.defaultValue || ""
            };
            if (false) {}
            if (false) {}
            if (false) {}
            r.validateRows(e);
            a.handleResize = g()(a.handleResize, 100);
            return a;
        }
        xe(r, [ {
            key: "componentDidMount",
            value: function e() {
                l()(this.syncHeightWithShadow);
 // wait for styles to load
                                this.getAdornmentWidth();
            }
        }, {
            key: "componentDidUpdate",
            value: function e(t) {
                if (false) {}
                if (false) {}
                this.syncHeightWithShadow();
                r.validateRows(this.props);
                this.getAdornmentWidth();
            }
        }, {
            key: "isControlled",
            value: function e() {
                return this.controlledExternally;
            }
        }, {
            key: "blur",
            value: function e() {
                var t;
                (t = this.input) === null || t === void 0 ? void 0 : t.blur();
            }
            /**
     * Place focus on the input.
     */        }, {
            key: "focus",
            value: function e() {
                var t;
                (t = this.input) === null || t === void 0 ? void 0 : t.focus();
            }
        }, {
            key: "select",
            value: function e() {
                var t;
                (t = this.input) === null || t === void 0 ? void 0 : t.select();
            }
        }, {
            key: "render",
            value: function e() {
                var t = this.props, r = t.append, a = t.autoCapitalize, i = t.autoComplete, o = t.autoCorrect, l = t.autoFocus, s = t.children, c = t.className, p = t.classNamePrivate, u = t.disabled, h = t.describedBy, m = t.elementRef, y = t.error, g = t.inline, x = t.inputClassName, w = t.inputId, k = t.labelledBy, S = t.maxLength, O = t.name, I = t.placeholder, j = t.prepend, E = t.required, T = t.rowsMin, M = t.spellCheck, P = t.tabIndex, $ = t.title, _ = me(t, [ "append", "autoCapitalize", "autoComplete", "autoCorrect", "autoFocus", "children", "className", "classNamePrivate", "disabled", "describedBy", "elementRef", "error", "inline", "inputClassName", "inputId", "labelledBy", "maxLength", "name", "placeholder", "prepend", "required", "rowsMin", "spellCheck", "tabIndex", "title" ]);
                var D = ve(ve({}, b()(_, (function(e, t) {
                    return t === "role" || t.indexOf("aria-") === 0;
                }))), {}, {
                    "aria-describedby": h,
                    "aria-labelledby": k,
                    "aria-invalid": y || undefined
                });
                var q = v()(_, [ "inputRef", "onBlur", "onChange", "onFocus", "onKeyDown", "onSelect", "onInputClick" ].concat(se(d()(D))));
                var W = this.getDisplayValue();
                var N = {
                    className: (0, A.toClassName)(c, x),
                    $append: r || undefined,
                    $error: y,
                    $prepend: j || undefined
                };
                var B = ve(ve({}, D), {}, {
                    "data-test": "textbox",
                    autoCapitalize: a,
                    autoComplete: i,
                    autoCorrect: o,
                    autoFocus: l,
                    disabled: u,
                    id: w,
                    maxLength: S,
                    placeholder: I,
                    name: O,
                    onChange: this.handleInputChange,
                    onKeyDown: this.handleInputKeyDown,
                    onSelect: this.handleInputSelect,
                    onClick: this.handleInputClick,
                    onFocus: this.handleInputFocus,
                    onBlur: this.handleInputBlur,
                    ref: this.handleInputMount,
                    required: E,
                    spellCheck: M,
                    style: {
                        height: this.state.height
                    },
                    tabIndex: P,
                    title: $,
                    value: W,
                    $error: y
                });
                var z = this.renderStartAdornment();
                var F = this.renderEndAdornment();
                var L = !!z;
                var G = !!F;
                
                return n().createElement(R, le({
                    tabIndex: -1,
                    className: (0, A.toClassName)(c, p),
                    "data-test": "text-area",
                    "data-test-value": W,
                    elementRef: m,
                    flex: true,
                    inline: g
                }, q), z,  n().createElement(H, le({
                    $hasEndAdornment: G,
                    $hasStartAdornment: L,
                    disabled: u
                }, N), L &&  n().createElement(U, {
                    $width: this.state.startAdornmentWidth
                }),  n().createElement(K, {
                    "aria-hidden": "true",
                    onChange: f(),
                    $startAdornmentWidth: this.state.startAdornmentWidth,
                    $endAdornmentWidth: this.state.endAdornmentWidth,
                    tabIndex: -1,
                    rows: this.props.rowsMin,
                    ref: this.handleShadowMount,
                    value: I && !W ? I : W
                }),  n().createElement(V, le({
                    $rowsMin: T
                }, B, {
                    onClick: this.handleInputClick
                })),  n().createElement(C(), {
                    target: window,
                    eventType: "resize",
                    listener: this.handleResize
                }), s, G &&  n().createElement(U, {
                    $width: this.state.endAdornmentWidth
                })), F);
            }
        } ]);
        return r;
    }(r.Component);
    je(Pe, "propTypes", Ee);
    je(Pe, "defaultProps", Te);
    je(Pe, "componentType", "TextArea");
    je(Pe, "contextType", x.ControlGroupContext);
    var $e = (0, k.withSplunkTheme)(Pe);
    var _e = $e;
    _e.propTypes = Pe.propTypes;
    _e.componentType = Pe.componentType;
    /* harmony default export */    const De = _e;
    // CONCATENATED MODULE: ./src/TextArea/index.ts
    module.exports = t;
    /******/})();
