// @splunk/react-ui/Text.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var r = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(r, {
                a: r
            });
            /******/            return r;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        default: () => /* reexport */ Me
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    var n =  e.n(r);
    // CONCATENATED MODULE: external "prop-types"
    const i = require("prop-types");
    var a =  e.n(i);
    // CONCATENATED MODULE: external "lodash/has"
    const o = require("lodash/has");
    var l =  e.n(o);
    // CONCATENATED MODULE: external "lodash/keys"
    const s = require("lodash/keys");
    var c =  e.n(s);
    // CONCATENATED MODULE: external "lodash/omit"
    const p = require("lodash/omit");
    var d =  e.n(p);
    // CONCATENATED MODULE: external "lodash/pickBy"
    const u = require("lodash/pickBy");
    var f =  e.n(u);
    // CONCATENATED MODULE: external "@splunk/react-ui/ControlGroup"
    const m = require("@splunk/react-ui/ControlGroup");
    // CONCATENATED MODULE: external "@splunk/react-icons/Cross"
    const h = require("@splunk/react-icons/Cross");
    var v =  e.n(h);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Close"
    const b = require("@splunk/react-icons/enterprise/Close");
    var y =  e.n(b);
    // CONCATENATED MODULE: external "@splunk/react-icons/enterprise/Search"
    const C = require("@splunk/react-icons/enterprise/Search");
    var g =  e.n(C);
    // CONCATENATED MODULE: external "@splunk/react-icons/Magnifier"
    const x = require("@splunk/react-icons/Magnifier");
    var w =  e.n(x);
    // CONCATENATED MODULE: external "@splunk/react-ui/ScreenReaderContent"
    const k = require("@splunk/react-ui/ScreenReaderContent");
    var L =  e.n(k);
    // CONCATENATED MODULE: external "@splunk/react-ui/TextArea"
    const S = require("@splunk/react-ui/TextArea");
    var E =  e.n(S);
    // CONCATENATED MODULE: external "@splunk/themes"
    const A = require("@splunk/themes");
    // CONCATENATED MODULE: external "@splunk/ui-utils/style"
    const O = require("@splunk/ui-utils/style");
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const T = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "@splunk/react-icons/SVGEnterprise"
    const I = require("@splunk/react-icons/SVGEnterprise");
    var j =  e.n(I);
    // CONCATENATED MODULE: ./src/Text/IconOutlinedHide.tsx
    function M() {
        M = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return M.apply(this, arguments);
    }
    function R(e) {
        var t = (0, A.useSplunkTheme)(), r = t.family, i = t.density;
        if (r === "enterprise") {
            
            return n().createElement(j(), M({
                viewBox: "0 0 24 24",
                size: 1.2,
                screenReaderText: (0, T._)("Hide password")
            }, e),  n().createElement("path", {
                fillRule: "evenodd",
                clipRule: "evenodd",
                d: "M4.00691 3.89287C3.61638 3.50234 2.98322 3.50235 2.5927 3.89287C2.20217 4.2834 2.20217 4.91656 2.5927 5.30708L4.35908 7.07346C3.63043 7.75302 2.99174 8.52671 2.46085 9.37612L1.18794 11.4128C0.963393 11.7721 0.963395 12.2279 1.18794 12.5872L2.46085 14.6239C3.3213 16.0006 4.46494 17.1784 5.81576 18.0789L5.94933 18.1679C7.74121 19.3625 9.8466 20 12.0002 20C13.5116 20 14.9993 19.686 16.3715 19.0858L18.5982 21.3125C18.9887 21.703 19.6218 21.703 20.0124 21.3125C20.4029 20.922 20.4029 20.2888 20.0124 19.8983L4.00691 3.89287ZM14.8257 17.54L13.0353 15.7496C12.7056 15.8405 12.3583 15.889 11.9997 15.889C9.85196 15.889 8.11084 14.1479 8.11084 12.0001C8.11084 11.6415 8.15937 11.2943 8.25022 10.9646L5.77433 8.48871C5.15179 9.06078 4.60729 9.71541 4.15685 10.4361L3.17943 12L4.15685 13.5639C4.86685 14.6999 5.81053 15.6717 6.92516 16.4148L7.05873 16.5038C8.52208 17.4794 10.2415 18 12.0002 18C12.9658 18 13.9196 17.8431 14.8257 17.54Z",
                fill: "currentColor"
            }),  n().createElement("path", {
                d: "M10.9657 8.25019L15.7496 13.0342C15.8402 12.7049 15.8886 12.3581 15.8886 12.0001C15.8886 9.85232 14.1475 8.11121 11.9997 8.11121C11.6417 8.11121 11.2949 8.15959 10.9657 8.25019Z",
                fill: "currentColor"
            }),  n().createElement("path", {
                d: "M19.8435 13.5639C19.3932 14.2844 18.8488 14.939 18.2264 15.5109L19.6417 16.9262C20.3702 16.2467 21.0087 15.4731 21.5395 14.6239L22.8124 12.5872C23.037 12.2279 23.037 11.7721 22.8124 11.4128L21.5395 9.37612C20.6791 7.99942 19.5354 6.82164 18.1846 5.9211L18.051 5.83205C16.2591 4.63746 14.1538 4 12.0002 4C10.489 4 9.00149 4.31389 7.62941 4.91392L9.17525 6.45977C10.0811 6.15687 11.0347 6 12.0002 6C13.7589 6 15.4783 6.52059 16.9416 7.49615L17.0752 7.5852C18.1898 8.32829 19.1335 9.30013 19.8435 10.4361L20.8209 12L19.8435 13.5639Z",
                fill: "currentColor"
            }));
        }
        var a = i === "compact" ? "20" : "24";
        
        return n().createElement("svg", {
            width: a,
            height: a,
            viewBox: "0 0 24 24",
            xmlns: "http://www.w3.org/2000/svg",
            style: {
                display: "block"
            }
        },  n().createElement("title", null, (0, T._)("Hide password")),  n().createElement("path", {
            fillRule: "evenodd",
            clipRule: "evenodd",
            d: "M4.00691 3.89287C3.61638 3.50234 2.98322 3.50235 2.5927 3.89287C2.20217 4.2834 2.20217 4.91656 2.5927 5.30708L4.35908 7.07346C3.63043 7.75302 2.99174 8.52671 2.46085 9.37612L1.18794 11.4128C0.963393 11.7721 0.963395 12.2279 1.18794 12.5872L2.46085 14.6239C3.3213 16.0006 4.46494 17.1784 5.81576 18.0789L5.94933 18.1679C7.74121 19.3625 9.8466 20 12.0002 20C13.5116 20 14.9993 19.686 16.3715 19.0858L18.5982 21.3125C18.9887 21.703 19.6218 21.703 20.0124 21.3125C20.4029 20.922 20.4029 20.2888 20.0124 19.8983L4.00691 3.89287ZM14.8257 17.54L13.0353 15.7496C12.7056 15.8405 12.3583 15.889 11.9997 15.889C9.85196 15.889 8.11084 14.1479 8.11084 12.0001C8.11084 11.6415 8.15937 11.2943 8.25022 10.9646L5.77433 8.48871C5.15179 9.06078 4.60729 9.71541 4.15685 10.4361L3.17943 12L4.15685 13.5639C4.86685 14.6999 5.81053 15.6717 6.92516 16.4148L7.05873 16.5038C8.52208 17.4794 10.2415 18 12.0002 18C12.9658 18 13.9196 17.8431 14.8257 17.54Z",
            fill: "currentColor"
        }),  n().createElement("path", {
            d: "M10.9657 8.25019L15.7496 13.0342C15.8402 12.7049 15.8886 12.3581 15.8886 12.0001C15.8886 9.85232 14.1475 8.11121 11.9997 8.11121C11.6417 8.11121 11.2949 8.15959 10.9657 8.25019Z",
            fill: "currentColor"
        }),  n().createElement("path", {
            d: "M19.8435 13.5639C19.3932 14.2844 18.8488 14.939 18.2264 15.5109L19.6417 16.9262C20.3702 16.2467 21.0087 15.4731 21.5395 14.6239L22.8124 12.5872C23.037 12.2279 23.037 11.7721 22.8124 11.4128L21.5395 9.37612C20.6791 7.99942 19.5354 6.82164 18.1846 5.9211L18.051 5.83205C16.2591 4.63746 14.1538 4 12.0002 4C10.489 4 9.00149 4.31389 7.62941 4.91392L9.17525 6.45977C10.0811 6.15687 11.0347 6 12.0002 6C13.7589 6 15.4783 6.52059 16.9416 7.49615L17.0752 7.5852C18.1898 8.32829 19.1335 9.30013 19.8435 10.4361L20.8209 12L19.8435 13.5639Z",
            fill: "currentColor"
        }));
    }
    /* harmony default export */    const $ = R;
    // CONCATENATED MODULE: ./src/Text/IconOutlinedView.tsx
    function _() {
        _ = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return _.apply(this, arguments);
    }
    function P(e) {
        var t = (0, A.useSplunkTheme)(), r = t.family, i = t.density;
        if (r === "enterprise") {
            
            return n().createElement(j(), _({
                viewBox: "0 0 24 24",
                size: 1.2,
                screenReaderText: (0, T._)("View password")
            }, e),  n().createElement("path", {
                fillRule: "evenodd",
                clipRule: "evenodd",
                d: "M15.7745 11.8654C15.7745 14.0132 14.0334 15.7543 11.8857 15.7543C9.73788 15.7543 7.99677 14.0132 7.99677 11.8654C7.99677 9.71762 9.73788 7.9765 11.8857 7.9765C14.0334 7.9765 15.7745 9.71762 15.7745 11.8654ZM13.7745 11.8654C13.7745 12.9086 12.9289 13.7543 11.8857 13.7543C10.8424 13.7543 9.99677 12.9086 9.99677 11.8654C9.99677 10.8222 10.8424 9.9765 11.8857 9.9765C12.9289 9.9765 13.7745 10.8222 13.7745 11.8654Z",
                fill: "currentColor"
            }),  n().createElement("path", {
                fillRule: "evenodd",
                clipRule: "evenodd",
                d: "M11.8859 3.86536C9.73235 3.86536 7.62695 4.50282 5.83507 5.69741L5.7015 5.78646C4.35068 6.687 3.20704 7.86477 2.3466 9.24148L1.07369 11.2781C0.849136 11.6374 0.849137 12.0933 1.07369 12.4526L2.3466 14.4892C3.20704 15.8659 4.35068 17.0437 5.7015 17.9443L5.83507 18.0333C7.62695 19.2279 9.73235 19.8654 11.8859 19.8654C14.0395 19.8654 16.1449 19.2279 17.9368 18.0333L18.0703 17.9443C19.4212 17.0437 20.5648 15.8659 21.4252 14.4892L22.6982 12.4526C22.9227 12.0933 22.9227 11.6374 22.6982 11.2781L21.4252 9.24148C20.5648 7.86477 19.4212 6.687 18.0703 5.78645L17.9368 5.69741C16.1449 4.50282 14.0395 3.86536 11.8859 3.86536ZM6.94447 7.36151C8.40782 6.38594 10.1272 5.86536 11.8859 5.86536C13.6446 5.86536 15.364 6.38594 16.8274 7.36151L16.9609 7.45056C18.0756 8.19364 19.0193 9.16549 19.7293 10.3015L20.7067 11.8654L19.7293 13.4292C19.0193 14.5652 18.0756 15.5371 16.9609 16.2802L16.8274 16.3692C15.364 17.3448 13.6446 17.8654 11.8859 17.8654C10.1272 17.8654 8.40782 17.3448 6.94447 16.3692L6.8109 16.2802C5.69627 15.5371 4.75259 14.5652 4.04259 13.4292L3.06517 11.8654L4.04259 10.3015C4.75259 9.16549 5.69627 8.19364 6.8109 7.45056L6.94447 7.36151Z",
                fill: "currentColor"
            }));
        }
        var a = i === "compact" ? "20" : "24";
        
        return n().createElement("svg", {
            width: a,
            height: a,
            viewBox: "0 0 24 24",
            xmlns: "http://www.w3.org/2000/svg",
            style: {
                display: "block"
            }
        },  n().createElement("title", null, (0, T._)("View password")),  n().createElement("path", {
            fillRule: "evenodd",
            clipRule: "evenodd",
            d: "M15.7745 11.8654C15.7745 14.0132 14.0334 15.7543 11.8857 15.7543C9.73788 15.7543 7.99677 14.0132 7.99677 11.8654C7.99677 9.71762 9.73788 7.9765 11.8857 7.9765C14.0334 7.9765 15.7745 9.71762 15.7745 11.8654ZM13.7745 11.8654C13.7745 12.9086 12.9289 13.7543 11.8857 13.7543C10.8424 13.7543 9.99677 12.9086 9.99677 11.8654C9.99677 10.8222 10.8424 9.9765 11.8857 9.9765C12.9289 9.9765 13.7745 10.8222 13.7745 11.8654Z",
            fill: "currentColor"
        }),  n().createElement("path", {
            fillRule: "evenodd",
            clipRule: "evenodd",
            d: "M11.8859 3.86536C9.73235 3.86536 7.62695 4.50282 5.83507 5.69741L5.7015 5.78646C4.35068 6.687 3.20704 7.86477 2.3466 9.24148L1.07369 11.2781C0.849136 11.6374 0.849137 12.0933 1.07369 12.4526L2.3466 14.4892C3.20704 15.8659 4.35068 17.0437 5.7015 17.9443L5.83507 18.0333C7.62695 19.2279 9.73235 19.8654 11.8859 19.8654C14.0395 19.8654 16.1449 19.2279 17.9368 18.0333L18.0703 17.9443C19.4212 17.0437 20.5648 15.8659 21.4252 14.4892L22.6982 12.4526C22.9227 12.0933 22.9227 11.6374 22.6982 11.2781L21.4252 9.24148C20.5648 7.86477 19.4212 6.687 18.0703 5.78645L17.9368 5.69741C16.1449 4.50282 14.0395 3.86536 11.8859 3.86536ZM6.94447 7.36151C8.40782 6.38594 10.1272 5.86536 11.8859 5.86536C13.6446 5.86536 15.364 6.38594 16.8274 7.36151L16.9609 7.45056C18.0756 8.19364 19.0193 9.16549 19.7293 10.3015L20.7067 11.8654L19.7293 13.4292C19.0193 14.5652 18.0756 15.5371 16.9609 16.2802L16.8274 16.3692C15.364 17.3448 13.6446 17.8654 11.8859 17.8654C10.1272 17.8654 8.40782 17.3448 6.94447 16.3692L6.8109 16.2802C5.69627 15.5371 4.75259 14.5652 4.04259 13.4292L3.06517 11.8654L4.04259 10.3015C4.75259 9.16549 5.69627 8.19364 6.8109 7.45056L6.94447 7.36151Z",
            fill: "currentColor"
        }));
    }
    /* harmony default export */    const B = P;
    // CONCATENATED MODULE: external "styled-components"
    const D = require("styled-components");
    var N =  e.n(D);
    // CONCATENATED MODULE: external "@splunk/react-ui/Box"
    const q = require("@splunk/react-ui/Box");
    var V =  e.n(q);
    // CONCATENATED MODULE: external "@splunk/react-ui/Button"
    const W = require("@splunk/react-ui/Button");
    var z =  e.n(W);
    // CONCATENATED MODULE: ./src/Text/TextStyles.ts
    var Z = (0, D.css)([ "&,&[type]{cursor:not-allowed;color:", ";&::placeholder{color:", ";}}" ], (0, 
    A.pick)({
        enterprise: A.variables.textDisabledColor,
        prisma: A.variables.contentColorDisabled
    }), (0, A.pick)({
        enterprise: A.variables.textDisabledColor,
        prisma: A.variables.contentColorDisabled
    }));
    var F = (0, D.css)([ "border-radius:", ";cursor:pointer;font-size:0.83333em;flex-grow:0;", ";&:not([disabled]){color:", ";}" ], (0, 
    A.pick)({
        enterprise: A.variables.borderRadius,
        prisma: "50%"
    }), (0, A.pick)({
        enterprise: (0, D.css)([ "width:", ";height:", ";" ], A.variables.inputHeight, A.variables.inputHeight),
        prisma: {
            comfortable: (0, D.css)([ "width:26px;min-width:26px;min-height:26px;margin:8px;padding:0;" ]),
            compact: (0, D.css)([ "width:22px;min-width:22px;min-height:22px;margin:8px;padding:0;" ])
        }
    }), (0, A.pick)({
        enterprise: {
            light: A.variables.gray60,
            dark: A.variables.white
        },
        prisma: A.variables.contentColorMuted
    }));
    var H = N()(z()).withConfig({
        displayName: "TextStyles__StyledClearButton",
        componentId: "eg7n6t-0"
    })([ "display:none;visibility:hidden;", "" ], F);
    var K = N().span.withConfig({
        displayName: "TextStyles__StyledSearchIconWrapper",
        componentId: "eg7n6t-1"
    })([ "", ";color:", ";pointer-events:none;padding:", ";", "" ], A.mixins.reset("inline-block"), (0, 
    A.pick)({
        enterprise: {
            light: A.variables.gray60,
            dark: A.variables.white
        },
        prisma: A.variables.contentColorMuted
    }), (0, A.pick)({
        comfortable: "0 8px",
        compact: "0 6px"
    }), (function(e) {
        var t = e.$disabled;
        return t && (0, D.css)([ "", "" ], (0, A.pick)({
            enterprise: (0, D.css)([ "color:", ";" ], A.variables.textDisabledColor),
            prisma: (0, D.css)([ "color:", ";" ], A.variables.contentColorDisabled)
        }));
    }));
    var G = N()(V()).withConfig({
        displayName: "TextStyles__StyledBox",
        componentId: "eg7n6t-2"
    })([ "display:flex;justify-content:space-between;flex-grow:1;flex-shrink:1;position:relative;&[data-inline]{width:230px;flex-basis:230px;[data-inline] + &{margin-left:", ";}}&:focus-within:not([disabled]){", "{", "}", "{visibility:visible;display:block;}}", "" ], (0, 
    A.pick)({
        enterprise: A.variables.spacingHalf,
        prisma: A.variables.spacingSmall
    }), 
    /* sc-sel */
    K, (0, A.pick)({
        enterprise: (0, D.css)([ "display:none;" ])
    }), 
    /* sc-sel */
    H, (function(e) {
        var t = e.$isTimeInput;
        return t && (0, D.css)([ "@media screen and (min--moz-device-pixel-ratio:0){", "{display:none;}}" ], 
        /* sc-sel */
        H);
    }));
    var U = N().span.withConfig({
        displayName: "TextStyles__StyledInputWrapper",
        componentId: "eg7n6t-3"
    })([ "", ";flex-grow:1;min-height:", ";position:relative;overflow:hidden;border:1px solid ", ";border-radius:", ";box-sizing:border-box;background-color:", ";box-shadow:inset 0 1px 1px rgba(0,0,0,0.075);padding:", ";align-items:center;", " ", " &:hover:not([disabled]){border-color:", ";}&:focus-within:not([disabled]){box-shadow:", ";color:", ";}", " ", " ", " ", "" ], A.mixins.reset("inline-flex"), A.variables.inputHeight, A.variables.interactiveColorBorder, A.variables.borderRadius, (0, 
    A.pick)({
        enterprise: {
            light: A.variables.white,
            dark: A.variables.gray22
        },
        prisma: A.variables.transparent
    }), (0, A.pick)({
        enterprise: "0px 10px",
        prisma: "0px 12px"
    }), (function(e) {
        var t = e.$hasEndAdornment;
        return t && (0, D.css)([ "padding-right:0;" ]);
    }), (function(e) {
        var t = e.$hasStartAdornment;
        return t && (0, D.css)([ "padding-left:0;" ]);
    }), A.variables.interactiveColorBorderHover, A.variables.focusShadow, A.variables.contentColorActive, (function(e) {
        var t = e.$error;
        return t && (0, D.css)([ "&,&:hover:not([disabled]){border-color:", ";}&:focus-within:not([disabled]){box-shadow:", ";color:", ";}" ], A.variables.accentColorNegative, A.variables.focusShadow, A.variables.contentColorActive);
    }), (function(e) {
        var t = e.$append;
        return t && (0, D.css)([ "margin-right:-1px;border-top-right-radius:0;border-bottom-right-radius:0;border-right:none;" ]);
    }), (function(e) {
        var t = e.$prepend;
        return t && (0, D.css)([ "border-top-left-radius:0;border-bottom-left-radius:0;" ]);
    }), (function(e) {
        var t = e.disabled;
        return t && (0, D.css)([ "border-color:", ";box-shadow:inset 0 1px 2px rgba(0,0,0,0.025);", "" ], A.variables.interactiveColorBorderDisabled, (0, 
        A.pick)({
            enterprise: (0, D.css)([ "background-color:", ";" ], A.variables.interactiveColorBackgroundDisabled)
        }));
    }));
    /* Some of these need greater specificity than input[type=text] */    var J = N().input.withConfig({
        displayName: "TextStyles__StyledInput",
        componentId: "eg7n6t-4"
    })([ "", ";&,&[type]{outline-style:none;border:0;background:transparent;color:", ";flex-grow:1;font-family:", ";font-size:inherit;line-height:inherit;&[type='time']{&::-webkit-calendar-picker-indicator{display:none;}height:", ";@media screen and (min--moz-device-pixel-ratio:0){padding-right:", ";}}position:relative;&::placeholder{color:", ";opacity:1;}", " ", "}" ], A.mixins.reset("inline-flex"), (0, 
    A.pick)({
        enterprise: A.variables.textColor,
        prisma: A.variables.contentColorActive
    }), A.variables.sansFontFamily, (0, A.pick)({
        enterprise: "20px",
        prisma: {
            comfortable: "24px",
            compact: "20px"
        }
    }), (0, A.pick)({
        enterprise: A.variables.spacingHalf,
        prisma: "14px"
    }), A.variables.contentColorMuted, (function(e) {
        var t = e.$error;
        return t && (0, D.css)([ "&,&:hover{color:", ";}" ], (0, A.pick)({
            enterprise: {
                light: A.variables.errorColorD10,
                dark: A.variables.errorColorL20
            },
            prisma: A.variables.contentColorActive
        }));
    }), (function(e) {
        var t = e.disabled;
        return t && Z;
    }));
    var Q = N()(z()).withConfig({
        displayName: "TextStyles__StyledVisibilityToggle",
        componentId: "eg7n6t-5"
    })([ "", "" ], F);
    var X = N().span.withConfig({
        displayName: "TextStyles__StyledPlaceholder",
        componentId: "eg7n6t-6"
    })([ "pointer-events:none;color:", ";position:absolute;overflow:hidden;font-size:inherit;line-height:inherit;height:", ";margin-right:", ";", ";", ";", "" ], (0, 
    A.pick)({
        enterprise: A.variables.textGray,
        prisma: A.variables.contentColorMuted
    }), A.variables.lineHeight, (0, A.pick)({
        enterprise: "10px",
        prisma: "14px"
    }), (function(e) {
        var t = e.$hasStartAdornment, r = e.$startAdornmentWidth;
        return t && (r ? (0, D.css)([ "margin-left:", "px;" ], r) : (0, D.css)([ "margin-left:", ";" ], (0, 
        A.pick)({
            enterprise: {
                comfortable: "30px",
                compact: "26px"
            },
            prisma: {
                comfortable: "38px",
                compact: "30px"
            }
        })));
    }), (function(e) {
        var t = e.$hasEndAdornment, r = e.$endAdornmentWidth;
        return t && (r ? (0, D.css)([ "max-width:calc( 100% - ", " - ", "px );" ], (0, A.pick)({
            enterprise: "10px",
            prisma: "14px"
        }), r) : (0, D.css)([ "max-width:calc( 100% - ", " - ", " );" ], (0, A.pick)({
            enterprise: "10px",
            prisma: "14px"
        }), (0, A.pick)({
            enterprise: {
                comfortable: "30px",
                compact: "26px"
            },
            prisma: {
                comfortable: "38px",
                compact: "30px"
            }
        })));
    }), (function(e) {
        var t = e.$hasBothAdornment, r = e.$endAdornmentWidth, n = e.$startAdornmentWidth;
        return t && (0, D.css)([ "max-width:calc(100% - ", "px - ", "px);" ], n, r);
    }));
    var Y = (0, D.css)([ "display:inline-flex;align-items:center;justify-content:center;height:", ";pointer-events:none;" ], (0, 
    A.pick)({
        enterprise: {
            comfortable: "30px",
            compact: "26px"
        },
        prisma: {
            comfortable: "38px",
            compact: "30px"
        }
    }));
    var ee = N().div.withConfig({
        displayName: "TextStyles__StyledStartAdornmentHolder",
        componentId: "eg7n6t-7"
    })([ "", " ", ";" ], Y, (function(e) {
        var t = e.$width;
        return t ? (0, D.css)([ "min-width:", "px;" ], t) : (0, D.css)([ "min-width:", ";" ], (0, 
        A.pick)({
            enterprise: {
                comfortable: "30px",
                compact: "26px"
            },
            prisma: {
                comfortable: "38px",
                compact: "30px"
            }
        }));
    }));
    var te = N().div.withConfig({
        displayName: "TextStyles__StyledEndAdornmentHolder",
        componentId: "eg7n6t-8"
    })([ "", " ", ";" ], Y, (function(e) {
        var t = e.$width;
        return t ? (0, D.css)([ "min-width:", "px;" ], t) : (0, D.css)([ "min-width:", ";" ], (0, 
        A.pick)({
            enterprise: {
                comfortable: "30px",
                compact: "26px"
            },
            prisma: {
                comfortable: "38px",
                compact: "30px"
            }
        }));
    }));
    var re = N().div.withConfig({
        displayName: "TextStyles__StyledAdornment",
        componentId: "eg7n6t-9"
    })([ "display:inline-flex;align-items:center;justify-content:center;position:absolute;pointer-events:none;z-index:1;height:", ";color:", ";", ";", ";", "" ], (0, 
    A.pick)({
        enterprise: {
            comfortable: "30px",
            compact: "26px"
        },
        prisma: {
            comfortable: "38px",
            compact: "30px"
        }
    }), (0, A.pick)({
        enterprise: {
            light: A.variables.gray60,
            dark: A.variables.white
        },
        prisma: A.variables.contentColorMuted
    }), (function(e) {
        var t = e.$position;
        return t === "start" ? (0, D.css)([ "top:1px;left:1px;" ]) : (0, D.css)([ "top:1px;right:1px;" ]);
    }), (function(e) {
        var t = e.disabled;
        return t && Z;
    }), (0, A.pick)({
        enterprise: (0, D.css)([ "margin-left:-1px;margin-right:-1px;" ])
    }));
    // CONCATENATED MODULE: ./src/utils/updateReactRef.ts
    /**
 * Updates a React ref. Callback refs and object refs (from `createRef` and `useRef`) are supported.
 *
 * @param ref - The React callback or object ref. Can be `null` or `undefined`.
 * @param current - The new value of the ref.
 */
    function ne(e, t) {
        if (e) {
            if (typeof e === "function") {
                e(t);
            } else {
                // the public signature of this util uses React.Ref<T> to mirror the way React types refs.
                // the intention here is to signal "we will take care of setting 'current', not you".
                e.current = t;
 // eslint-disable-line no-param-reassign
                        }
        }
    }
    // CONCATENATED MODULE: ./src/Text/Text.tsx
    function ie(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            ie = function e(t) {
                return typeof t;
            };
        } else {
            ie = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return ie(e);
    }
    function ae() {
        ae = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = arguments[t];
                for (var n in r) {
                    if (Object.prototype.hasOwnProperty.call(r, n)) {
                        e[n] = r[n];
                    }
                }
            }
            return e;
        };
        return ae.apply(this, arguments);
    }
    function oe(e) {
        return pe(e) || ce(e) || se(e) || le();
    }
    function le() {
        throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function se(e, t) {
        if (!e) return;
        if (typeof e === "string") return de(e, t);
        var r = Object.prototype.toString.call(e).slice(8, -1);
        if (r === "Object" && e.constructor) r = e.constructor.name;
        if (r === "Map" || r === "Set") return Array.from(e);
        if (r === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(r)) return de(e, t);
    }
    function ce(e) {
        if (typeof Symbol !== "undefined" && Symbol.iterator in Object(e)) return Array.from(e);
    }
    function pe(e) {
        if (Array.isArray(e)) return de(e);
    }
    function de(e, t) {
        if (t == null || t > e.length) t = e.length;
        for (var r = 0, n = new Array(t); r < t; r++) {
            n[r] = e[r];
        }
        return n;
    }
    function ue(e, t) {
        var r = Object.keys(e);
        if (Object.getOwnPropertySymbols) {
            var n = Object.getOwnPropertySymbols(e);
            if (t) n = n.filter((function(t) {
                return Object.getOwnPropertyDescriptor(e, t).enumerable;
            }));
            r.push.apply(r, n);
        }
        return r;
    }
    function fe(e) {
        for (var t = 1; t < arguments.length; t++) {
            var r = arguments[t] != null ? arguments[t] : {};
            if (t % 2) {
                ue(Object(r), true).forEach((function(t) {
                    Ee(e, t, r[t]);
                }));
            } else if (Object.getOwnPropertyDescriptors) {
                Object.defineProperties(e, Object.getOwnPropertyDescriptors(r));
            } else {
                ue(Object(r)).forEach((function(t) {
                    Object.defineProperty(e, t, Object.getOwnPropertyDescriptor(r, t));
                }));
            }
        }
        return e;
    }
    function me(e, t) {
        if (e == null) return {};
        var r = he(e, t);
        var n, i;
        if (Object.getOwnPropertySymbols) {
            var a = Object.getOwnPropertySymbols(e);
            for (i = 0; i < a.length; i++) {
                n = a[i];
                if (t.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                r[n] = e[n];
            }
        }
        return r;
    }
    function he(e, t) {
        if (e == null) return {};
        var r = {};
        var n = Object.keys(e);
        var i, a;
        for (a = 0; a < n.length; a++) {
            i = n[a];
            if (t.indexOf(i) >= 0) continue;
            r[i] = e[i];
        }
        return r;
    }
    function ve(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function be(e, t) {
        for (var r = 0; r < t.length; r++) {
            var n = t[r];
            n.enumerable = n.enumerable || false;
            n.configurable = true;
            if ("value" in n) n.writable = true;
            Object.defineProperty(e, n.key, n);
        }
    }
    function ye(e, t, r) {
        if (t) be(e.prototype, t);
        if (r) be(e, r);
        return e;
    }
    function Ce(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) ge(e, t);
    }
    function ge(e, t) {
        ge = Object.setPrototypeOf || function e(t, r) {
            t.__proto__ = r;
            return t;
        };
        return ge(e, t);
    }
    function xe(e) {
        var t = Le();
        return function r() {
            var n = Se(e), i;
            if (t) {
                var a = Se(this).constructor;
                i = Reflect.construct(n, arguments, a);
            } else {
                i = n.apply(this, arguments);
            }
            return we(this, i);
        };
    }
    function we(e, t) {
        if (t && (ie(t) === "object" || typeof t === "function")) {
            return t;
        }
        return ke(e);
    }
    function ke(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function Le() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function Se(e) {
        Se = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return Se(e);
    }
    function Ee(e, t, r) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: r,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = r;
        }
        return e;
    }
    /** @public */    var Ae = {
        appearance: a().oneOf([ "default", "search" ]),
        append: a().bool,
        autoCapitalize: a().string,
        autoComplete: a().string,
        autoCorrect: a().string,
        autoFocus: a().bool,
        canClear: a().bool,
        children: a().node,
        /** @private. */
        classNamePrivate: a().string,
        defaultValue: a().string,
        describedBy: a().string,
        disabled: a().bool,
        elementRef: a().oneOfType([ a().func, a().object ]),
        endAdornment: a().node,
        error: a().bool,
        inline: a().bool,
        /** @private. */
        inputClassName: a().string,
        inputId: a().string,
        inputRef: a().oneOfType([ a().func, a().object ]),
        labelledBy: a().string,
        maxLength: a().number,
        multiline: a().bool,
        name: a().string,
        onBlur: a().func,
        onChange: a().func,
        onFocus: a().func,
        onKeyDown: a().func,
        onSelect: a().func,
        onInputClick: a().func,
        placeholder: a().string,
        prepend: a().bool,
        /** @private. */
        required: a().bool,
        rowsMax: a().number,
        rowsMin: a().number,
        spellCheck: a().bool,
        tabIndex: a().number,
        startAdornment: a().node,
        /** @private */
        splunkTheme: a().object,
        type: a().string,
        passwordVisibilityToggle: a().bool,
        useSyntheticPlaceholder: a().bool,
        value: a().string
    };
    var Oe = {
        appearance: "default",
        append: false,
        autoFocus: false,
        canClear: false,
        disabled: false,
        error: false,
        inline: false,
        multiline: false,
        placeholder: "",
        prepend: false,
        rowsMax: 8,
        rowsMin: 2,
        tabIndex: 0,
        type: "text",
        passwordVisibilityToggle: false
    };
    /** Note: Text places role and aria props onto the input. All other props are placed on the wrapper. */    var Te =  function(e) {
        Ce(r, e);
        var t = xe(r);
        ye(r, null, [ {
            key: "validateRows",
            // @docs-props-type TextPropsBase
            value: function e(t) {
                var r = t.rowsMin, n = t.rowsMax;
                if (false) {}
            }
        } ]);
        function r(e) {
            var i;
            ve(this, r);
            i = t.call(this, e);
            Ee(ke(i), "controlledExternally", void 0);
            Ee(ke(i), "input", null);
            Ee(ke(i), "startAdornment", null);
            Ee(ke(i), "endAdornment", null);
            Ee(ke(i), "getAdornmentWidth", (function() {
                var e = i.startAdornment ? i.startAdornment.getBoundingClientRect() : undefined;
                var t = e && Math.round(e.width);
                if (i.state.startAdornmentWidth !== t) {
                    i.setState({
                        startAdornmentWidth: t
                    });
                }
                var r = i.endAdornment ? i.endAdornment.getBoundingClientRect() : undefined;
                var n = r && Math.round(r.width);
                if (i.state.endAdornmentWidth !== n) {
                    i.setState({
                        endAdornmentWidth: n
                    });
                }
            }));
            Ee(ke(i), "handleInputMount", (function(e) {
                i.input = e;
                ne(i.props.inputRef, e);
            }));
            Ee(ke(i), "handleInputChange", (function(e) {
                var t, r;
                var n = e.target.value;
                var a = i.props.name;
                if (!i.isControlled()) {
                    i.setState({
                        value: n
                    });
                }
                (t = (r = i.props).onChange) === null || t === void 0 ? void 0 : t.call(r, e, {
                    value: n,
                    name: a
                });
            }));
            Ee(ke(i), "handleInputKeyDown", (function(e) {
                var t, r;
                (t = (r = i.props).onKeyDown) === null || t === void 0 ? void 0 : t.call(r, e);
            }));
            Ee(ke(i), "handleInputSelect", (function(e) {
                var t, r;
                (t = (r = i.props).onSelect) === null || t === void 0 ? void 0 : t.call(r, e);
            }));
            Ee(ke(i), "handleInputClick", (function(e) {
                var t, r;
                (t = (r = i.props).onInputClick) === null || t === void 0 ? void 0 : t.call(r, e);
            }));
            Ee(ke(i), "handleInputFocus", (function(e) {
                var t, r;
                var n = e.target.value;
                var a = i.props.name;
                (t = (r = i.props).onFocus) === null || t === void 0 ? void 0 : t.call(r, e, {
                    value: n,
                    name: a
                });
            }));
            Ee(ke(i), "handleInputBlur", (function(e) {
                var t, r;
                var n = e.target.value;
                var a = i.props.name;
                (t = (r = i.props).onBlur) === null || t === void 0 ? void 0 : t.call(r, e, {
                    value: n,
                    name: a
                });
            }));
            Ee(ke(i), "handleClear", (function(e) {
                var t, r;
                e.preventDefault();
                var n = "";
                var a = i.props.name;
                if (!i.isControlled()) {
                    i.setState({
                        value: n
                    });
                }
                i.focus();
                (t = (r = i.props).onChange) === null || t === void 0 ? void 0 : t.call(r, e, {
                    value: n,
                    name: a
                });
            }));
            Ee(ke(i), "handleVisibilityToggle", (function() {
                i.setState((function(e) {
                    return {
                        hideVisibility: !e.hideVisibility
                    };
                }));
            }));
            Ee(ke(i), "renderEndAdornment", (function() {
                var e;
                var t = i.props, r = t.appearance, a = t.endAdornment, o = t.passwordVisibilityToggle, l = t.disabled, s = t.canClear, c = t.splunkTheme;
                var p = c.isPrisma, d = c.isCompact, u = c.isEnterprise;
                var f = d ? "20px" : "24px";
                var m = {
                    ref: function e(t) {
                        i.endAdornment = t;
                    },
                    disabled: l,
                    $position: "end"
                };
                if (a) {
                    
                    return n().createElement(re, m, a);
                }
                if (o && !l) {
                    var h = i.state.hideVisibility ?  n().createElement(B, null) :  n().createElement($, null);
                    
                    return n().createElement(re, m,  n().createElement(Q, {
                        appearance: p ? "secondary" : "pill",
                        "data-test": "password-toggle",
                        inline: false,
                        onClick: i.handleVisibilityToggle,
                        type: "button",
                        icon: h
                    }));
                }
                var b = i.context;
                var C = (e = b.labelAttrs) === null || e === void 0 ? void 0 : e.text;
                var x = C ? (0, T._)("Clear ".concat(C, " text field")) : (0, T._)("Clear text field");
                var k = i.getDisplayValue();
                if (!p && r === "search") {
                    if (!k) {
                        
                        return n().createElement(re, m,  n().createElement(K, {
                            $disabled: l,
                            "data-test": "search-icon"
                        },  n().createElement(g(), {
                            screenReaderText: null,
                            hideDefaultTooltip: true,
                            size: "16px",
                            inline: false
                        })));
                    }
                    
                    return n().createElement(re, m, !l &&  n().createElement(H, {
                        appearance: p ? "secondary" : "pill",
                        "data-test": "clear",
                        inline: false,
                        icon: u ?  n().createElement(y(), {
                            size: 1,
                            screenReaderText: x
                        }) :  n().createElement(v(), {
                            "aria-label": x
                        }),
                        onClick: i.handleClear
                    }),  n().createElement(K, {
                        "data-test": "search-icon",
                        $disabled: l
                    }, u ?  n().createElement(g(), {
                        screenReaderText: null,
                        hideDefaultTooltip: true,
                        size: "16px",
                        inline: false
                    }) :  n().createElement(w(), {
                        width: f,
                        height: f
                    })));
                }
                if (!!k && (p && r === "search" || s)) {
                    
                    return n().createElement(re, m, !l &&  n().createElement(H, {
                        "data-test": "clear",
                        appearance: p ? "secondary" : "pill",
                        onClick: i.handleClear,
                        icon: u ?  n().createElement(y(), {
                            inline: false,
                            screenReaderText: x
                        }) :  n().createElement(v(), {
                            "aria-label": x,
                            height: f,
                            inline: false,
                            width: f
                        })
                    }));
                }
                return undefined;
            }));
            Ee(ke(i), "renderStartAdornment", (function() {
                var e = i.props, t = e.appearance, r = e.startAdornment, a = e.splunkTheme, o = e.disabled;
                var l = a.isPrisma, s = a.isCompact;
                var c = s ? "20px" : "24px";
                var p = {
                    ref: function e(t) {
                        i.startAdornment = t;
                    },
                    disabled: o,
                    $position: "start"
                };
                if (r) {
                    
                    return n().createElement(re, p, r);
                }
                if (l && t === "search") {
                    
                    return n().createElement(re, p,  n().createElement(K, {
                        "data-test": "search-icon",
                        $disabled: o
                    },  n().createElement(w(), {
                        "aria-label": "Search",
                        height: c,
                        width: c
                    })));
                }
                return undefined;
            }));
            Ee(ke(i), "getDisplayValue", (function() {
                var e = i.isControlled() ? i.props.value : i.state.value;
                return e;
            }));
            i.controlledExternally = l()(e, "value");
            i.state = {
                value: e.defaultValue || "",
                hideVisibility: true
            };
            if (false) {}
            if (false) {}
            if (l()(e, "useSyntheticPlaceholder")) {
                // eslint-disable-next-line no-console
                console.warn("The 'Text' prop 'useSyntheticPlaceholder' has been marked for deprecation.");
            }
            if (false) {}
            if (false) {}
            if (false) {}
            r.validateRows(e);
            return i;
        }
        ye(r, [ {
            key: "componentDidMount",
            value: function e() {
                this.getAdornmentWidth();
            }
        }, {
            key: "componentDidUpdate",
            value: function e(t) {
                if (false) {}
                if (false) {}
                r.validateRows(this.props);
                this.getAdornmentWidth();
            }
        }, {
            key: "isControlled",
            value: function e() {
                return this.controlledExternally;
            }
        }, {
            key: "blur",
            value: function e() {
                var t;
                (t = this.input) === null || t === void 0 ? void 0 : t.blur();
            }
            /**
     * Place focus on the input.
     */        }, {
            key: "focus",
            value: function e(t) {
                var r;
                (r = this.input) === null || r === void 0 ? void 0 : r.focus(t);
            }
        }, {
            key: "select",
            value: function e() {
                var t;
                (t = this.input) === null || t === void 0 ? void 0 : t.select();
            }
        }, {
            key: "render",
            value: function e() {
                var t = this.props, r = t.append, i = t.autoCapitalize, a = t.autoComplete, o = t.autoCorrect, l = t.autoFocus, s = t.children, p = t.className, u = t.classNamePrivate, m = t.disabled, h = t.describedBy, v = t.elementRef, b = t.error, y = t.inline, C = t.inputClassName, g = t.inputId, x = t.labelledBy, w = t.multiline, k = t.maxLength, S = t.name, A = t.passwordVisibilityToggle, T = t.placeholder, I = t.prepend, j = t.required, M = t.spellCheck, R = t.tabIndex, $ = t.title, _ = t.type, P = t.useSyntheticPlaceholder, B = me(t, [ "append", "autoCapitalize", "autoComplete", "autoCorrect", "autoFocus", "children", "className", "classNamePrivate", "disabled", "describedBy", "elementRef", "error", "inline", "inputClassName", "inputId", "labelledBy", "multiline", "maxLength", "name", "passwordVisibilityToggle", "placeholder", "prepend", "required", "spellCheck", "tabIndex", "title", "type", "useSyntheticPlaceholder" ]);
                var D = fe(fe({
                    role: "textbox"
                }, f()(B, (function(e, t) {
                    return t === "role" || t.indexOf("aria-") === 0;
                }))), {}, {
                    "aria-describedby": h,
                    "aria-labelledby": x,
                    "aria-invalid": b || undefined
                });
                var N = d()(B, [ "inputRef", "onBlur", "onChange", "onFocus", "onKeyDown", "onSelect", "onInputClick", "rowsMax", "rowsMin" ].concat(oe(c()(D))));
                var q = this.getDisplayValue();
                var V = {
                    className: (0, O.toClassName)(p, C),
                    $append: r || undefined,
                    $error: b,
                    $prepend: I || undefined
                };
                var W = this.state.hideVisibility ? "password" : "text";
                var z = A ? W : _;
                var Z = fe(fe({}, D), {}, {
                    "data-test": "textbox",
                    autoCapitalize: i,
                    autoComplete: a,
                    autoCorrect: o,
                    autoFocus: l,
                    id: g,
                    maxLength: k,
                    placeholder: T && !P ? T : undefined,
                    name: S,
                    onChange: this.handleInputChange,
                    onKeyDown: this.handleInputKeyDown,
                    onSelect: this.handleInputSelect,
                    onClick: this.handleInputClick,
                    onFocus: this.handleInputFocus,
                    onBlur: this.handleInputBlur,
                    ref: this.handleInputMount,
                    required: j,
                    spellCheck: M,
                    style: {
                        height: this.state.height
                    },
                    title: $,
                    tabIndex: R,
                    type: z,
                    value: q,
                    $error: b
                });
                var F = P && T && !q;
 // Do not render value in DOM when type is password, or passwordVisibilityToggle is enabled
                                var H = _ === "password" || A ? undefined : q;
 // Firefox hack (SUI-2716). Remove when fixed: https://bugzilla.mozilla.org/show_bug.cgi?id=1427173
                                var K = _ === "time";
                var Q = this.renderStartAdornment();
                var Y = this.renderEndAdornment();
                var re = !!Q;
                var ne = !!Y;
                if (w) {
                    
                    return n().createElement(E(), this.props);
                }
                
                return n().createElement(G, ae({
                    tabIndex: -1,
                    $isTimeInput: K,
                    className: (0, O.toClassName)(p, u),
                    "data-test": "text",
                    "data-test-value": H,
                    elementRef: v,
                    flex: true,
                    inline: y
                }, N), Q,  n().createElement(U, ae({
                    $hasEndAdornment: ne,
                    $hasStartAdornment: re,
                    disabled: m
                }, V), re &&  n().createElement(ee, {
                    $width: this.state.startAdornmentWidth
                }), m ?  n().createElement(J, ae({
                    className: (0, O.toClassName)(p, u),
                    "data-test": "textbox",
                    disabled: true,
                    readOnly: true,
                    type: z,
                    value: q,
                    placeholder: T && !P ? T : undefined
                }, D)) :  n().createElement(J, ae({}, Z, {
                    onClick: this.handleInputClick
                })), F &&  n().createElement(L(), null, T), F &&  n().createElement(X, {
                    $endAdornmentWidth: this.state.endAdornmentWidth,
                    $hasBothAdornment: re && ne,
                    $hasEndAdornment: ne,
                    $hasStartAdornment: re,
                    $startAdornmentWidth: this.state.startAdornmentWidth,
                    "aria-hidden": true,
                    "data-role": "placeholder"
                }, T), s, ne &&  n().createElement(te, {
                    $width: this.state.endAdornmentWidth
                })), Y);
            }
        } ]);
        return r;
    }(r.Component);
    Ee(Te, "propTypes", Ae);
    Ee(Te, "defaultProps", Oe);
    Ee(Te, "componentType", "Text");
    Ee(Te, "contextType", m.ControlGroupContext);
    var Ie = (0, A.withSplunkTheme)(Te);
    var je = Ie;
    je.propTypes = Te.propTypes;
    je.componentType = Te.componentType;
    /* harmony default export */    const Me = je;
    // CONCATENATED MODULE: ./src/Text/index.ts
    module.exports = t;
    /******/})();
