// @splunk/react-ui/Tooltip.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = t => {
            /******/ var n = t && t.__esModule ? 
            /******/ () => t["default"]
            /******/ : () => t
            /******/;
            e.d(n, {
                a: n
            });
            /******/            return n;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, n) => {
            /******/ for (var o in n) {
                /******/ if (e.o(n, o) && !e.o(t, o)) {
                    /******/ Object.defineProperty(t, o, {
                        enumerable: true,
                        get: n[o]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        default: () => /* reexport */ X
    });
    // CONCATENATED MODULE: external "react"
    const n = require("react");
    var o =  e.n(n);
    // CONCATENATED MODULE: external "prop-types"
    const r = require("prop-types");
    var a =  e.n(r);
    // CONCATENATED MODULE: external "lodash/has"
    const i = require("lodash/has");
    var l =  e.n(i);
    // CONCATENATED MODULE: external "lodash/throttle"
    const s = require("lodash/throttle");
    var u =  e.n(s);
    // CONCATENATED MODULE: external "@splunk/react-ui/Popover"
    const c = require("@splunk/react-ui/Popover");
    var p =  e.n(c);
    // CONCATENATED MODULE: external "@splunk/react-ui/ScreenReaderContent"
    const d = require("@splunk/react-ui/ScreenReaderContent");
    var f =  e.n(d);
    // CONCATENATED MODULE: external "@splunk/themes"
    const v = require("@splunk/themes");
    // CONCATENATED MODULE: external "@splunk/ui-utils/id"
    const h = require("@splunk/ui-utils/id");
    // CONCATENATED MODULE: external "@splunk/ui-utils/i18n"
    const m = require("@splunk/ui-utils/i18n");
    // CONCATENATED MODULE: external "@splunk/react-icons/SVGEnterprise"
    const y = require("@splunk/react-icons/SVGEnterprise");
    var b =  e.n(y);
    // CONCATENATED MODULE: ./src/Tooltip/InfoIcon.tsx
    function g() {
        g = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var n = arguments[t];
                for (var o in n) {
                    if (Object.prototype.hasOwnProperty.call(n, o)) {
                        e[o] = n[o];
                    }
                }
            }
            return e;
        };
        return g.apply(this, arguments);
    }
    function C(e) {
        var t = (0, v.useSplunkTheme)(), n = t.family, r = t.density;
        if (n === "enterprise") {
            
            return o().createElement(b(), g({
                screenReaderText: (0, m._)("Info Circle"),
                viewBox: "0 0 24 24"
            }, e),  o().createElement("path", {
                fill: "currentColor",
                d: "M12 20C16.4183 20 20 16.4183 20 12C20 7.58172 16.4183 4 12 4C7.58172 4 4 7.58172 4 12C4 16.4183 7.58172 20 12 20ZM12 22C17.5228 22 22 17.5228 22 12C22 6.47715 17.5228 2 12 2C6.47715 2 2 6.47715 2 12C2 17.5228 6.47715 22 12 22ZM13.2001 8.1999C13.2001 7.53716 12.6628 6.9999 12.0001 6.9999C11.3373 6.9999 10.8001 7.53716 10.8001 8.1999C10.8001 8.86264 11.3373 9.3999 12.0001 9.3999C12.6628 9.3999 13.2001 8.86264 13.2001 8.1999ZM11.0001 16.0094V11.9999C11.0001 11.4476 11.4478 10.9999 12.0001 10.9999C12.5524 10.9999 13.0001 11.4476 13.0001 11.9999V16.0094C13.0001 16.5617 12.5524 17.0094 12.0001 17.0094C11.4478 17.0094 11.0001 16.5617 11.0001 16.0094Z"
            }));
        }
        var a = r === "compact" ? "20" : "24";
        
        return o().createElement("svg", {
            width: a,
            height: a,
            viewBox: "0 0 24 24",
            xmlns: "http://www.w3.org/2000/svg",
            style: {
                display: "block"
            }
        },  o().createElement("path", {
            d: "M12 20C16.4183 20 20 16.4183 20 12C20 7.58172 16.4183 4 12 4C7.58172 4 4 7.58172 4 12C4 16.4183 7.58172 20 12 20ZM12 22C17.5228 22 22 17.5228 22 12C22 6.47715 17.5228 2 12 2C6.47715 2 2 6.47715 2 12C2 17.5228 6.47715 22 12 22ZM13.2003 8.1999C13.2003 7.53716 12.663 6.9999 12.0003 6.9999C11.3376 6.9999 10.8003 7.53716 10.8003 8.1999C10.8003 8.86264 11.3376 9.3999 12.0003 9.3999C12.663 9.3999 13.2003 8.86264 13.2003 8.1999ZM11.0005 16.0094V11.9999C11.0005 11.4476 11.4482 10.9999 12.0005 10.9999C12.5528 10.9999 13.0005 11.4476 13.0005 11.9999V16.0094C13.0005 16.5617 12.5528 17.0094 12.0005 17.0094C11.4482 17.0094 11.0005 16.5617 11.0005 16.0094Z",
            fill: "currentColor",
            fillRule: "evenodd"
        }));
    }
    /* harmony default export */    const E = C;
    // CONCATENATED MODULE: external "styled-components"
    const O = require("styled-components");
    var M =  e.n(O);
    // CONCATENATED MODULE: external "@splunk/react-ui/Clickable"
    const S = require("@splunk/react-ui/Clickable");
    var k =  e.n(S);
    // CONCATENATED MODULE: ./src/Tooltip/TooltipStyles.ts
    var R = "16px";
    var P = M().span.withConfig({
        displayName: "TooltipStyles__Styled",
        componentId: "sc-67fvgm-0"
    })([ "", ";position:relative;" ], v.mixins.reset("block"));
    var T = M().span.withConfig({
        displayName: "TooltipStyles__StyledInline",
        componentId: "sc-67fvgm-1"
    })([ "", ";position:relative;" ], v.mixins.reset("inline-block"));
    var w = M().span.withConfig({
        displayName: "TooltipStyles__StyledToggle",
        componentId: "sc-67fvgm-2"
    })([ "display:block;" ]);
    var q = M().span.withConfig({
        displayName: "TooltipStyles__StyledToggleInline",
        componentId: "sc-67fvgm-3"
    })([ "display:inline-block;" ]);
    var x = M().div.withConfig({
        displayName: "TooltipStyles__StyledContent",
        componentId: "sc-67fvgm-4"
    })([ "padding:", ";font-size:", ";line-height:", ";user-select:none;" ], (0, v.pick)({
        enterprise: v.variables.spacingHalf,
        prisma: (0, v.pickVariant)("$isLabel", {
            false: "12px 16px",
            true: "4px 8px"
        })
    }), (0, v.pick)({
        enterprise: v.variables.fontSizeSmall,
        prisma: "inherit"
    }), (0, v.pick)({
        enterprise: "18px",
        prisma: (0, v.pickVariant)("$isLabel", {
            false: "inherit",
            true: "18px"
        })
    }));
    var L = M()(k()).withConfig({
        displayName: "TooltipStyles__StyledButton",
        componentId: "sc-67fvgm-5"
    })([ "cursor:default;", " ", " &:focus{box-shadow:", ";}" ], (0, v.pick)({
        enterprise: (0, O.css)([ "display:inline-block;width:", ";height:", ";border:2px solid ", ";border-radius:", ";font-size:", ";font-weight:", ";line-height:calc(", " - 2px);text-align:center;color:", ";vertical-align:baseline;&:not([disabled]){&:hover{text-decoration:none;}}" ], R, R, v.variables.linkColor, R, v.variables.fontSizeSmall, v.variables.fontWeightSemiBold, R, v.variables.linkColor),
        prisma: (0, O.css)([ "color:", ";border-radius:50%;display:inline-block;vertical-align:middle;&:hover,&:focus{color:", ";background:", ";}" ], v.variables.contentColorMuted, v.variables.contentColorActive, v.variables.interactiveColorOverlayHover)
    }), (0, v.pick)({
        prisma: {
            comfortable: (0, O.css)([ "padding:8px;" ]),
            compact: (0, O.css)([ "padding:6px;" ])
        }
    }), v.variables.focusShadow);
    // CONCATENATED MODULE: ./src/Tooltip/Tooltip.tsx
    function _(e) {
        "@babel/helpers - typeof";
        if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
            _ = function e(t) {
                return typeof t;
            };
        } else {
            _ = function e(t) {
                return t && typeof Symbol === "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
            };
        }
        return _(e);
    }
    function I() {
        I = Object.assign || function(e) {
            for (var t = 1; t < arguments.length; t++) {
                var n = arguments[t];
                for (var o in n) {
                    if (Object.prototype.hasOwnProperty.call(n, o)) {
                        e[o] = n[o];
                    }
                }
            }
            return e;
        };
        return I.apply(this, arguments);
    }
    function A(e, t) {
        if (e == null) return {};
        var n = j(e, t);
        var o, r;
        if (Object.getOwnPropertySymbols) {
            var a = Object.getOwnPropertySymbols(e);
            for (r = 0; r < a.length; r++) {
                o = a[r];
                if (t.indexOf(o) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, o)) continue;
                n[o] = e[o];
            }
        }
        return n;
    }
    function j(e, t) {
        if (e == null) return {};
        var n = {};
        var o = Object.keys(e);
        var r, a;
        for (a = 0; a < o.length; a++) {
            r = o[a];
            if (t.indexOf(r) >= 0) continue;
            n[r] = e[r];
        }
        return n;
    }
    function H(e, t) {
        if (!(e instanceof t)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }
    function D(e, t) {
        for (var n = 0; n < t.length; n++) {
            var o = t[n];
            o.enumerable = o.enumerable || false;
            o.configurable = true;
            if ("value" in o) o.writable = true;
            Object.defineProperty(e, o.key, o);
        }
    }
    function B(e, t, n) {
        if (t) D(e.prototype, t);
        if (n) D(e, n);
        return e;
    }
    function V(e, t) {
        if (typeof t !== "function" && t !== null) {
            throw new TypeError("Super expression must either be null or a function");
        }
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                writable: true,
                configurable: true
            }
        });
        if (t) Z(e, t);
    }
    function Z(e, t) {
        Z = Object.setPrototypeOf || function e(t, n) {
            t.__proto__ = n;
            return t;
        };
        return Z(e, t);
    }
    function N(e) {
        var t = F();
        return function n() {
            var o = $(e), r;
            if (t) {
                var a = $(this).constructor;
                r = Reflect.construct(o, arguments, a);
            } else {
                r = o.apply(this, arguments);
            }
            return W(this, r);
        };
    }
    function W(e, t) {
        if (t && (_(t) === "object" || typeof t === "function")) {
            return t;
        }
        return z(e);
    }
    function z(e) {
        if (e === void 0) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }
        return e;
    }
    function F() {
        if (typeof Reflect === "undefined" || !Reflect.construct) return false;
        if (Reflect.construct.sham) return false;
        if (typeof Proxy === "function") return true;
        try {
            Date.prototype.toString.call(Reflect.construct(Date, [], (function() {})));
            return true;
        } catch (e) {
            return false;
        }
    }
    function $(e) {
        $ = Object.setPrototypeOf ? Object.getPrototypeOf : function e(t) {
            return t.__proto__ || Object.getPrototypeOf(t);
        };
        return $(e);
    }
    function U(e, t, n) {
        if (t in e) {
            Object.defineProperty(e, t, {
                value: n,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            e[t] = n;
        }
        return e;
    }
    var G = {
        /** @private */
        appearance: a().string,
        children: a().node,
        closeDelay: a().number,
        /** @private */
        closeWhen: a().oneOf([ "default", "notOnClick" ]),
        content: a().node,
        contentRelationship: a().oneOf([ "label", "description" ]),
        defaultPlacement: a().oneOf([ "above", "below", "left", "right", "theme" ]),
        elementRef: a().oneOfType([ a().func, a().object ]),
        inline: a().bool,
        onRequestClose: a().func,
        onRequestOpen: a().func,
        open: a().bool,
        openDelay: a().oneOfType([ a().oneOf([ "primary", "secondary" ]), a().number ]),
        /** @private */
        splunkTheme: a().object
    };
    var K = {
        appearance: "inverted",
        closeDelay: 300,
        closeWhen: "default",
        contentRelationship: "description",
        defaultPlacement: "theme",
        inline: true,
        openDelay: "primary"
    };
    /**
 * The Tooltip component wraps arbitrary content to be displayed when the target element is hovered
 * or focused.
 */    var J =  function(e) {
        V(r, e);
        var t = N(r);
        // @docs-props-type TooltipPropsBase
                function r(e) {
            var n;
            H(this, r);
            n = t.call(this, e);
            U(z(n), "ariaId", void 0);
            U(z(n), "controlledExternally", void 0);
            U(z(n), "popoverId", void 0);
            U(z(n), "timeout", void 0);
            U(z(n), "handleMount", (function(e) {
                // Handle mouse events with native events due to current issues with how they
                // are handled by react. See https://github.com/facebook/react/issues/4251,
                // https://github.com/facebook/react/issues/19419, and SUI-1232.
                if (e) {
                    e.addEventListener("mouseenter", n.handleMouseEnter);
                    e.addEventListener("mouseleave", n.handleMouseLeave);
                } else if (n.state.anchorEl) {
                    n.state.anchorEl.removeEventListener("mouseenter", n.handleMouseEnter);
                    n.state.anchorEl.removeEventListener("mouseleave", n.handleMouseLeave);
                }
                n.setState({
                    anchorEl: e
                });
            }));
            U(z(n), "handlePopoverMount", (function(e) {
                if (e) {
                    e.addEventListener("mouseenter", n.handleMouseEnterPopover);
                    e.addEventListener("mouseleave", n.handleMouseLeavePopover);
                } else if (n.state.popoverEl) {
                    n.state.popoverEl.removeEventListener("mouseenter", n.handleMouseEnterPopover);
                    n.state.popoverEl.removeEventListener("mouseleave", n.handleMouseLeavePopover);
                }
                n.setState({
                    popoverEl: e
                });
            }));
            U(z(n), "handlePopoverOuterMount", (function(e) {
                n.setState({
                    popoverOuterEl: e
                });
            }));
            U(z(n), "handleHitAreaMount", (function(e) {
                // Handle mouse events with native events due to current issues with how they
                // are handled by react. See https://github.com/facebook/react/issues/4251,
                // https://github.com/facebook/react/issues/19419, and SUI-1232.
                if (e) {
                    e.addEventListener("mouseenter", n.handleMouseEnterHitArea);
                    e.addEventListener("mousemove", n.handleMouseMoveHitArea);
                    e.addEventListener("mouseleave", n.handleMouseLeaveHitArea);
                } else if (n.state.hitAreaEl) {
                    n.state.hitAreaEl.removeEventListener("mouseenter", n.handleMouseEnterHitArea);
                    n.state.hitAreaEl.removeEventListener("mousemove", n.handleMouseMoveHitArea);
                    n.state.hitAreaEl.removeEventListener("mouseleave", n.handleMouseLeaveHitArea);
                }
                n.setState({
                    hitAreaEl: e
                });
            }));
            U(z(n), "handleMouseEnter", (function(e) {
                n.handleRequestOpen(e, {
                    reason: "mouseEnterToggle"
                });
            }));
            U(z(n), "handleMouseEnterPopover", (function(e) {
                n.handleRequestOpen(e, {
                    reason: "mouseEnterPopover"
                });
            }));
            U(z(n), "handleMouseEnterHitArea", (function(e) {
                n.handleRequestOpen(e, {
                    reason: "mouseEnterHitArea"
                });
            }));
            U(z(n), "handleMouseLeave", (function(e) {
                var t = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "mouseLeaveToggle";
                if (e.relatedTarget == null || e.relatedTarget !== n.state.popoverEl && e.relatedTarget !== n.state.popoverOuterEl) {
                    n.handleRequestClose(e, {
                        reason: t
                    });
                }
            }));
            U(z(n), "handleMouseLeavePopover", (function(e) {
                n.handleMouseLeave(e, "mouseLeavePopover");
            }));
            U(z(n), "handleMouseLeaveHitArea", (function(e) {
                n.handleMouseLeave(e, "mouseLeaveHitArea");
            }));
            U(z(n), "handleMouseMoveHitAreaImpl", (function(e) {
                n.handleRequestClose(e, {
                    reason: "mouseStopHitArea"
                });
            }));
            U(z(n), "handleMouseMoveHitArea", u()(n.handleMouseMoveHitAreaImpl, 10));
            U(z(n), "handleClick", (function(e) {
                if (n.props.closeWhen !== "notOnClick") {
                    n.handleRequestClose(e, {
                        reason: "toggleClick"
                    });
                }
            }));
            U(z(n), "handleFocus", (function(e) {
                n.handleRequestOpen(e, {
                    reason: "focusToggle"
                });
            }));
            U(z(n), "handleBlur", (function(e) {
                n.handleRequestClose(e, {
                    reason: "blurToggle"
                });
            }));
            U(z(n), "handlePopoverOnRequestClose", (function(e) {
                n.handleRequestClose(null, e);
            }));
            U(z(n), "handleRequestClose", (function(e, t) {
                var o = n.props.closeDelay;
                if (n.timeout !== undefined) {
                    clearTimeout(n.timeout);
                }
                n.timeout = setTimeout((function() {
                    var o, r;
                    if (!n.isControlled()) {
                        n.setState({
                            open: false
                        });
                    }
                    (o = (r = n.props).onRequestClose) === null || o === void 0 ? void 0 : o.call(r, e, t);
                }), o);
            }));
            U(z(n), "handleRequestOpen", (function(e, t) {
                var o = n.props.openDelay;
                var r = o === "primary" ? 300 : 750;
                var a = typeof o === "number" ? o : r;
                if (n.timeout !== undefined) {
                    clearTimeout(n.timeout);
                }
                n.timeout = setTimeout((function() {
                    var o = n.isOpen();
                    if (!n.isControlled()) {
                        n.setState({
                            open: true
                        });
                    }
                    if (!o) {
                        var r, a;
                        (r = (a = n.props).onRequestOpen) === null || r === void 0 ? void 0 : r.call(a, e, t);
                    }
                }), a);
            }));
            n.controlledExternally = l()(e, "open");
            n.popoverId = (0, h.createDOMID)("popover");
            n.ariaId = (0, h.createDOMID)("aria-id");
            n.state = {
                anchorEl: null,
                open: false,
                popoverEl: null,
                popoverOuterEl: null,
                hitAreaEl: null
            };
            return n;
        }
        B(r, [ {
            key: "componentDidUpdate",
            value: function e() {
                if (false) {}
            }
        }, {
            key: "componentWillUnmount",
            value: function e() {
                if (this.timeout !== undefined) {
                    clearTimeout(this.timeout);
                }
            }
        }, {
            key: "isControlled",
            value: function e() {
                return this.controlledExternally;
            }
        }, {
            key: "isOpen",
            value: function e() {
                var t = !!this.props.content;
                return !!this.state.anchorEl && this.isControlled() ? this.props.open : t && this.state.open;
            }
        }, {
            key: "render",
            value: function e() {
                var t = this.props, r = t.appearance, a = t.children, i = t.content, l = t.defaultPlacement, s = t.elementRef, u = t.inline, c = t.onRequestClose, d = t.onRequestOpen, v = t.splunkTheme, h = t.contentRelationship, y = A(t, [ "appearance", "children", "content", "defaultPlacement", "elementRef", "inline", "onRequestClose", "onRequestOpen", "splunkTheme", "contentRelationship" ]);
                var b = this.state.anchorEl;
                var g = !!i;
                var C = this.isOpen();
                var O = u ? T : P;
                var M = u ? q : w;
                var S = v.isPrisma;
                var k = S ? "below" : "above";
                var R;
                if (h === "label") {
                    R = {
                        labelledBy: this.ariaId,
                        "aria-labelledby": this.ariaId
                    };
                } else if (h === "description") {
                    R = {
                        describedBy: this.ariaId,
                        "aria-describedby": this.ariaId
                    };
                }
                
                return o().createElement(O, I({
                    "data-test": "tooltip",
                    "data-test-open": C,
                    "data-test-popover-id": this.popoverId,
                    ref: s
                }, y),  o().createElement(M, {
                    onClick: this.handleClick,
                    onFocus: this.handleFocus,
                    onBlur: this.handleBlur,
                    ref: this.handleMount,
                    "data-test": "toggle"
                }, g &&  (0, n.isValidElement)(a) ?  (0, n.cloneElement)(a, R) : a, g && !a &&  o().createElement(L, {
                    "aria-describedby": this.ariaId
                }, S ?  o().createElement(E, {
                    hideDefaultTooltip: true,
                    screenReaderText: (0, m._)("More info"),
                    style: {
                        height: "22px",
                        width: "22px"
                    }
                }) :  o().createElement(o().Fragment, null,  o().createElement("span", {
                    "aria-hidden": "true"
                }, "?"),  o().createElement(f(), null, (0, m._)("More info")))), g &&  o().createElement(f(), {
                    "aria-hidden": "true",
                    id: this.ariaId
                }, i)),  o().createElement(p(), {
                    role: "tooltip",
                    align: "center",
                    anchor: b,
                    animationConfig: {
                        tension: 400
                    },
                    appearance: r,
                    closeReasons: [ "offScreen", "escapeKey" ],
                    defaultPlacement: l === "theme" ? k : l,
                    id: this.popoverId,
                    open: C,
                    onRequestClose: this.handlePopoverOnRequestClose,
                    elementRef: this.handlePopoverMount,
                    hitAreaRef: this.handleHitAreaMount,
                    outerRef: this.handlePopoverOuterMount
                },  o().createElement(x, {
                    $isLabel: h === "label"
                }, i)));
            }
        } ]);
        return r;
    }(n.Component);
    U(J, "propTypes", G);
    U(J, "defaultProps", K);
    var Q = (0, v.withSplunkTheme)(J);
    Q.propTypes = J.propTypes;
    /* harmony default export */    const X = Q;
    // CONCATENATED MODULE: ./src/Tooltip/index.ts
    module.exports = t;
    /******/})();
