// @splunk/react-ui/Typography.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/compat get default export */
    /******/    (() => {
        /******/ // getDefaultExport function for compatibility with non-harmony modules
        /******/ e.n = r => {
            /******/ var t = r && r.__esModule ? 
            /******/ () => r["default"]
            /******/ : () => r
            /******/;
            e.d(t, {
                a: t
            });
            /******/            return t;
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (r, t) => {
            /******/ for (var n in t) {
                /******/ if (e.o(t, n) && !e.o(r, n)) {
                    /******/ Object.defineProperty(r, n, {
                        enumerable: true,
                        get: t[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, r) => Object.prototype.hasOwnProperty.call(e, r)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var r = {};
    // ESM COMPAT FLAG
        e.r(r);
    // EXPORTS
        e.d(r, {
        Typography: () => /* reexport */ v,
        default: () => /* reexport */ g
    });
    // CONCATENATED MODULE: external "react"
    const t = require("react");
    var n =  e.n(t);
    // CONCATENATED MODULE: external "prop-types"
    const o = require("prop-types");
    var a =  e.n(o);
    // CONCATENATED MODULE: external "@splunk/themes/mixins"
    const i = require("@splunk/themes/mixins");
    var l =  e.n(i);
    // CONCATENATED MODULE: external "styled-components"
    const s = require("styled-components");
    var p =  e.n(s);
    // CONCATENATED MODULE: ./src/Typography/Typography.tsx
    function y() {
        y = Object.assign || function(e) {
            for (var r = 1; r < arguments.length; r++) {
                var t = arguments[r];
                for (var n in t) {
                    if (Object.prototype.hasOwnProperty.call(t, n)) {
                        e[n] = t[n];
                    }
                }
            }
            return e;
        };
        return y.apply(this, arguments);
    }
    function f(e, r) {
        if (e == null) return {};
        var t = u(e, r);
        var n, o;
        if (Object.getOwnPropertySymbols) {
            var a = Object.getOwnPropertySymbols(e);
            for (o = 0; o < a.length; o++) {
                n = a[o];
                if (r.indexOf(n) >= 0) continue;
                if (!Object.prototype.propertyIsEnumerable.call(e, n)) continue;
                t[n] = e[n];
            }
        }
        return t;
    }
    function u(e, r) {
        if (e == null) return {};
        var t = {};
        var n = Object.keys(e);
        var o, a;
        for (a = 0; a < n.length; a++) {
            o = n[a];
            if (r.indexOf(o) >= 0) continue;
            t[o] = e[o];
        }
        return t;
    }
    var c = {
        children: a().node.isRequired,
        as: a().string.isRequired,
        elementRef: a().oneOfType([ a().func, a().object ]),
        family: a().oneOf([ "sansSerif", "monospace" ]),
        variant: a().oneOf(i.typographyVariants),
        size: a().number,
        lineHeight: a().number,
        color: a().oneOf([ "active", "default", "disabled", "inverted", "muted", "inherit" ]),
        weight: a().oneOf([ 300, 400, 500, 600, 700, 800, 900, "light", "normal", "semiBold", "bold", "extraBold", "heavy" ]),
        withReset: a().bool
    };
    var d = p().p.withConfig({
        displayName: "Typography__StyledTypography",
        componentId: "cgxs34-0"
    })([ "", "" ], (function(e) {
        var r = e.$variant, t = e.$typographyParams;
        return r ? l().typography(r, t) : l().typography(t);
    }));
    var h = {
        300: "light",
        400: "normal",
        500: "semiBold",
        600: "semiBold",
        700: "bold",
        800: "extraBold",
        900: "heavy"
    };
    /**
 * Typography renders text content with styling based on the Splunk Design System.
 */    function v(e) {
        var r = e.children, t = e.as, o = e.elementRef, a = e.variant, i = e.family, l = e.size, s = e.lineHeight, p = e.color, u = e.weight, c = e.withReset, v = f(e, [ "children", "as", "elementRef", "variant", "family", "size", "lineHeight", "color", "weight", "withReset" ]);
        // @docs-props-type TypographyPropsBase
        // TODO: Remove in next major version (SUI-5350)
                if (false) {}
        var g = {
            family: i,
            size: l,
            lineHeight: s,
            color: p,
            withReset: c,
            weight: typeof u === "number" ? h[u] : u
        };
        
        return n().createElement(d, y({
            as: t,
            ref: o,
            "data-test": "typography",
            $variant: a,
            $typographyParams: g
        }, v), r);
    }
    v.propTypes = c;
    /* harmony default export */    const g = v;
    // CONCATENATED MODULE: ./src/Typography/index.ts
    module.exports = r;
    /******/})();
