// @splunk/react-ui/useForceUpdate.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var r = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ r.d = (e, t) => {
            /******/ for (var n in t) {
                /******/ if (r.o(t, n) && !r.o(e, n)) {
                    /******/ Object.defineProperty(e, n, {
                        enumerable: true,
                        get: t[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ r.o = (r, e) => Object.prototype.hasOwnProperty.call(r, e)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ r.r = r => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(r, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(r, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var e = {};
    // ESM COMPAT FLAG
        r.r(e);
    // EXPORTS
        r.d(e, {
        default: () => /* reexport */ c
    });
    // CONCATENATED MODULE: external "react"
    const t = require("react");
    // CONCATENATED MODULE: ./src/useForceUpdate/useForceUpdate.tsx
    function n(r, e) {
        return f(r) || i(r, e) || u(r, e) || o();
    }
    function o() {
        throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
    }
    function u(r, e) {
        if (!r) return;
        if (typeof r === "string") return a(r, e);
        var t = Object.prototype.toString.call(r).slice(8, -1);
        if (t === "Object" && r.constructor) t = r.constructor.name;
        if (t === "Map" || t === "Set") return Array.from(r);
        if (t === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t)) return a(r, e);
    }
    function a(r, e) {
        if (e == null || e > r.length) e = r.length;
        for (var t = 0, n = new Array(e); t < e; t++) {
            n[t] = r[t];
        }
        return n;
    }
    function i(r, e) {
        if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(r))) return;
        var t = [];
        var n = true;
        var o = false;
        var u = undefined;
        try {
            for (var a = r[Symbol.iterator](), i; !(n = (i = a.next()).done); n = true) {
                t.push(i.value);
                if (e && t.length === e) break;
            }
        } catch (r) {
            o = true;
            u = r;
        } finally {
            try {
                if (!n && a["return"] != null) a["return"]();
            } finally {
                if (o) throw u;
            }
        }
        return t;
    }
    function f(r) {
        if (Array.isArray(r)) return r;
    }
    /**
 * This is a private component not intended for use outside @splunk/react-ui
 *
 * Triggers a rerender of a functional component. Useful as an escape hatch when a render is needed regardless of state.
 * See https://reactjs.org/docs/hooks-faq.html#is-there-something-like-forceupdate
 *
 */    function l() {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        var r = (0, t.useReducer)((function(r) {
            return r + 1;
        }), 0), e = n(r, 2), o = e[0], u = e[1];
        return u;
    }
    /* harmony default export */    const c = l;
    // CONCATENATED MODULE: ./src/useForceUpdate/index.ts
    module.exports = e;
    /******/})();
