// @splunk/react-ui/useRovingFocus.js
/******/ (() => {
    // webpackBootstrap
    /******/ "use strict";
    /******/ // The require scope
    /******/    var e = {};
    /******/
    /************************************************************************/
    /******/ /* webpack/runtime/define property getters */
    /******/    (() => {
        /******/ // define getter functions for harmony exports
        /******/ e.d = (t, r) => {
            /******/ for (var n in r) {
                /******/ if (e.o(r, n) && !e.o(t, n)) {
                    /******/ Object.defineProperty(t, n, {
                        enumerable: true,
                        get: r[n]
                    });
                    /******/                }
                /******/            }
            /******/        };
        /******/    })();
    /******/
    /******/ /* webpack/runtime/hasOwnProperty shorthand */
    /******/    (() => {
        /******/ e.o = (e, t) => Object.prototype.hasOwnProperty.call(e, t)
        /******/;
    })();
    /******/
    /******/ /* webpack/runtime/make namespace object */
    /******/    (() => {
        /******/ // define __esModule on exports
        /******/ e.r = e => {
            /******/ if (typeof Symbol !== "undefined" && Symbol.toStringTag) {
                /******/ Object.defineProperty(e, Symbol.toStringTag, {
                    value: "Module"
                });
                /******/            }
            /******/            Object.defineProperty(e, "__esModule", {
                value: true
            });
            /******/        };
        /******/    })();
    /******/
    /************************************************************************/    var t = {};
    // ESM COMPAT FLAG
        e.r(t);
    // EXPORTS
        e.d(t, {
        default: () => /* reexport */ a
    });
    // CONCATENATED MODULE: external "react"
    const r = require("react");
    // CONCATENATED MODULE: external "@splunk/ui-utils/focus"
    const n = require("@splunk/ui-utils/focus");
    // CONCATENATED MODULE: external "@splunk/ui-utils/keyboard"
    const u = require("@splunk/ui-utils/keyboard");
    // CONCATENATED MODULE: ./src/useRovingFocus/useRovingFocus.tsx
    function i(e) {
        if (e instanceof HTMLElement && e.focus) {
            e.focus();
        }
    }
    /**
 * This is a private hook not intended for use outside @splunk/react-ui
 *
 * Handles "roving focus" for a group of elements inside a containing element.
 * Elements inside the element that "rovingFocusProps" are set on will act a single tab stop
 * with arrow keys moving focus between them.
 * See https://developer.mozilla.org/en-US/docs/Web/Accessibility/Keyboard-navigable_JavaScript_widgets#managing_focus_inside_groups
 *
 * ```
 * import React from 'react';
 * import useRovingFocus from '@splunk/react-ui/useRovingFocus';
 *
 * const rovingFocusProps = useRovingFocus();
 *
 * <div {...rovingFocusProps}>
 *     <button type="button">first</button>
 *     <button type="button">second</button>
 * </div>
 * ```
 */    function o() {
        var e = (0, r.useRef)(0);
        var t = (0, r.useRef)([]);
        var o = function r(n) {
            e.current = n;
            t.current.forEach((function(t, r) {
                return r === e.current ? t.removeAttribute("tabindex") : t.setAttribute("tabindex", "-1");
            }));
        };
        var a = function r(u) {
            t.current = u != null ? (0, n.getSortedTabbableElements)(u, {
                ignoreTabIndex: true
            }) : [];
            o(e.current);
        };
        var f = function e(r) {
            var n = r.relatedTarget instanceof Element && t.current.includes(r.relatedTarget);
            if (!n) {
                o(0);
            }
        };
        var c = function e(r) {
            var n = r.target instanceof Element ? t.current.indexOf(r.target) : -1;
            if (n >= 0) {
                o(n);
            }
        };
        var l = function e(r) {
            if (r.nativeEvent.target instanceof HTMLElement) {
                var n = t.current.indexOf(r.nativeEvent.target);
                if (n >= 0) {
                    var o = (0, u.keycode)(r.nativeEvent);
                    var a = -1;
                    if (o === "left") {
                        a = n - 1;
                    } else if (o === "right") {
                        a = n + 1;
                    } else if (o === "home") {
                        a = 0;
                    } else if (o === "end") {
                        a = t.current.length - 1;
                    }
                    if (a >= 0) {
                        r.preventDefault();
                        i(t.current[a]);
                    }
                }
            }
        };
        return {
            ref: a,
            onBlur: f,
            onFocus: c,
            onKeyDown: l
        };
    }
    /* harmony default export */    const a = o;
    // CONCATENATED MODULE: ./src/useRovingFocus/index.ts
    module.exports = t;
    /******/})();
